{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "weird-cathedral",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/anndata/_core/anndata.py:21: FutureWarning: pandas.core.index is deprecated and will be removed in a future version.  The public classes are available in the top-level namespace.\n",
      "  from pandas.core.index import RangeIndex\n"
     ]
    }
   ],
   "source": [
    "import swan_vis as swan\n",
    "import pandas as pd\n",
    "import anndata\n",
    "import numpy as np\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import scipy.stats as st\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "seven-customs",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6/18/21 \n",
    "# gen report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "aggressive-maker",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding transcriptome to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Transforming to str index.\n",
      "Transforming to str index.\n",
      "AnnData expects .obs.index to contain strings, but your first indices are: Int64Index([0, 1], dtype='int64'), …\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding abundance for datasets PB65_B017, PB65_B018 to SwanGraph.\n",
      "\n",
      "Adding abundance for datasets D12 to SwanGraph.\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "db = 'data/chr11_and_Tcf3_no_gname.db'\n",
    "sg.add_transcriptome(db, include_isms=True)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance_PB65.tsv'\n",
    "sg.add_abundance(ab)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance_D12.tsv'\n",
    "sg.add_abundance(ab)\n",
    "meta = 'data/chr11_and_Tcf3_metadata.tsv'\n",
    "sg.add_metadata(meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "aging-border",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                  tpm\n",
      "ENCODE-mouseT000008451  835714.285714\n",
      "ENCODE-mouseT000008453  350000.000000\n",
      "ENCODE-mouseT000008460  250000.000000\n",
      "ENSMUST00000105344.6    250000.000000\n",
      "ENCODE-mouseT000008456  142857.142857\n",
      "ENCODE-mouseT000008457  142857.142857\n",
      "ENCODE-mouseT000008458  142857.142857\n",
      "ENCODE-mouseT000008459  142857.142857\n",
      "ENSMUST00000020637.8    142857.142857\n",
      "ENCODE-mouseT000008452  100000.000000\n",
      "ENCODE-mouseT000008454  100000.000000\n",
      "ENCODE-mouseT000008455  100000.000000\n",
      "ENCODE-mouseT000008461  100000.000000\n",
      "ENCODE-mouseT000008462  100000.000000\n",
      "ENSMUST00000020741.10   100000.000000\n",
      "                                         tname                     gid  \\\n",
      "tid                                                                      \n",
      "ENCODE-mouseT000008451  ENCODE-mouseT000008451   ENSMUSG00000020167.13   \n",
      "ENCODE-mouseT000008453  ENCODE-mouseT000008453   ENSMUSG00000020457.12   \n",
      "ENCODE-mouseT000008460  ENCODE-mouseT000008460   ENSMUSG00000020368.14   \n",
      "ENSMUST00000105344.6                  Tcf3-005   ENSMUSG00000020167.13   \n",
      "ENCODE-mouseT000008456  ENCODE-mouseT000008456  ENCODE-mouseG000002990   \n",
      "ENCODE-mouseT000008457  ENCODE-mouseT000008457   ENSMUSG00000020368.14   \n",
      "ENCODE-mouseT000008458  ENCODE-mouseT000008458   ENSMUSG00000020368.14   \n",
      "ENCODE-mouseT000008459  ENCODE-mouseT000008459   ENSMUSG00000020368.14   \n",
      "ENSMUST00000020637.8                  Canx-001   ENSMUSG00000020368.14   \n",
      "ENCODE-mouseT000008452  ENCODE-mouseT000008452   ENSMUSG00000020167.13   \n",
      "ENCODE-mouseT000008454  ENCODE-mouseT000008454  ENCODE-mouseG000002988   \n",
      "ENCODE-mouseT000008455  ENCODE-mouseT000008455  ENCODE-mouseG000002989   \n",
      "ENCODE-mouseT000008461  ENCODE-mouseT000008461   ENSMUSG00000020368.14   \n",
      "ENCODE-mouseT000008462  ENCODE-mouseT000008462   ENSMUSG00000020368.14   \n",
      "ENSMUST00000020741.10                 Drg1-001   ENSMUSG00000020457.12   \n",
      "\n",
      "                                         gname  \\\n",
      "tid                                              \n",
      "ENCODE-mouseT000008451   ENSMUSG00000020167.13   \n",
      "ENCODE-mouseT000008453   ENSMUSG00000020457.12   \n",
      "ENCODE-mouseT000008460   ENSMUSG00000020368.14   \n",
      "ENSMUST00000105344.6     ENSMUSG00000020167.13   \n",
      "ENCODE-mouseT000008456  ENCODE-mouseG000002990   \n",
      "ENCODE-mouseT000008457   ENSMUSG00000020368.14   \n",
      "ENCODE-mouseT000008458   ENSMUSG00000020368.14   \n",
      "ENCODE-mouseT000008459   ENSMUSG00000020368.14   \n",
      "ENSMUST00000020637.8     ENSMUSG00000020368.14   \n",
      "ENCODE-mouseT000008452   ENSMUSG00000020167.13   \n",
      "ENCODE-mouseT000008454  ENCODE-mouseG000002988   \n",
      "ENCODE-mouseT000008455  ENCODE-mouseG000002989   \n",
      "ENCODE-mouseT000008461   ENSMUSG00000020368.14   \n",
      "ENCODE-mouseT000008462   ENSMUSG00000020368.14   \n",
      "ENSMUST00000020741.10    ENSMUSG00000020457.12   \n",
      "\n",
      "                                                                     path  \\\n",
      "tid                                                                         \n",
      "ENCODE-mouseT000008451                                               [97]   \n",
      "ENCODE-mouseT000008453  [17, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 18, 14, 15...   \n",
      "ENCODE-mouseT000008460  [20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 3...   \n",
      "ENSMUST00000105344.6    [60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 7...   \n",
      "ENCODE-mouseT000008456                                              [101]   \n",
      "ENCODE-mouseT000008457  [20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 3...   \n",
      "ENCODE-mouseT000008458                                               [54]   \n",
      "ENCODE-mouseT000008459  [55, 56, 38, 39, 50, 51, 52, 43, 44, 45, 46, 4...   \n",
      "ENSMUST00000020637.8    [49, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 3...   \n",
      "ENCODE-mouseT000008452                               [92, 93, 94, 95, 98]   \n",
      "ENCODE-mouseT000008454                                               [99]   \n",
      "ENCODE-mouseT000008455                                              [100]   \n",
      "ENCODE-mouseT000008461  [24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 3...   \n",
      "ENCODE-mouseT000008462  [20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 3...   \n",
      "ENSMUST00000020741.10   [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,...   \n",
      "\n",
      "                                           tid     novelty  \\\n",
      "tid                                                          \n",
      "ENCODE-mouseT000008451  ENCODE-mouseT000008451     Genomic   \n",
      "ENCODE-mouseT000008453  ENCODE-mouseT000008453         NIC   \n",
      "ENCODE-mouseT000008460  ENCODE-mouseT000008460         ISM   \n",
      "ENSMUST00000105344.6      ENSMUST00000105344.6       Known   \n",
      "ENCODE-mouseT000008456  ENCODE-mouseT000008456   Antisense   \n",
      "ENCODE-mouseT000008457  ENCODE-mouseT000008457         NNC   \n",
      "ENCODE-mouseT000008458  ENCODE-mouseT000008458     Genomic   \n",
      "ENCODE-mouseT000008459  ENCODE-mouseT000008459         NNC   \n",
      "ENSMUST00000020637.8      ENSMUST00000020637.8       Known   \n",
      "ENCODE-mouseT000008452  ENCODE-mouseT000008452         ISM   \n",
      "ENCODE-mouseT000008454  ENCODE-mouseT000008454   Antisense   \n",
      "ENCODE-mouseT000008455  ENCODE-mouseT000008455  Intergenic   \n",
      "ENCODE-mouseT000008461  ENCODE-mouseT000008461         ISM   \n",
      "ENCODE-mouseT000008462  ENCODE-mouseT000008462         ISM   \n",
      "ENSMUST00000020741.10    ENSMUST00000020741.10       Known   \n",
      "\n",
      "                                                                 loc_path  \n",
      "tid                                                                        \n",
      "ENCODE-mouseT000008451                                             [5, 1]  \n",
      "ENCODE-mouseT000008453  [58, 57, 56, 55, 54, 53, 52, 51, 50, 49, 48, 4...  \n",
      "ENCODE-mouseT000008460  [102, 101, 100, 99, 98, 97, 96, 95, 94, 93, 92...  \n",
      "ENSMUST00000105344.6    [39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 2...  \n",
      "ENCODE-mouseT000008456                                           [64, 67]  \n",
      "ENCODE-mouseT000008457  [102, 101, 100, 99, 98, 97, 96, 95, 94, 93, 92...  \n",
      "ENCODE-mouseT000008458                                           [67, 66]  \n",
      "ENCODE-mouseT000008459  [86, 83, 82, 81, 80, 79, 76, 74, 73, 71, 70, 6...  \n",
      "ENSMUST00000020637.8    [103, 101, 100, 99, 98, 97, 96, 95, 94, 93, 92...  \n",
      "ENCODE-mouseT000008452                                 [7, 6, 4, 3, 2, 1]  \n",
      "ENCODE-mouseT000008454                                           [60, 61]  \n",
      "ENCODE-mouseT000008455                                           [62, 63]  \n",
      "ENCODE-mouseT000008461  [98, 97, 96, 95, 94, 93, 92, 91, 90, 89, 88, 8...  \n",
      "ENCODE-mouseT000008462  [102, 101, 100, 99, 98, 97, 96, 95, 94, 93, 92...  \n",
      "ENSMUST00000020741.10   [59, 57, 56, 55, 54, 53, 52, 51, 50, 49, 48, 4...  \n",
      "                        PB65_B017  PB65_B018            D12            sum\n",
      "tid                                                                       \n",
      "ENCODE-mouseT000008454   100000.0        0.0       0.000000  100000.000000\n",
      "ENCODE-mouseT000008459        0.0        0.0  142857.142857  142857.142857\n",
      "ENCODE-mouseT000008453   100000.0   250000.0       0.000000  350000.000000\n",
      "ENCODE-mouseT000008455   100000.0        0.0       0.000000  100000.000000\n",
      "ENCODE-mouseT000008456        0.0        0.0  142857.142857  142857.142857\n",
      "ENSMUST00000105344.6          0.0   250000.0       0.000000  250000.000000\n",
      "ENCODE-mouseT000008461   100000.0        0.0       0.000000  100000.000000\n",
      "ENCODE-mouseT000008462   100000.0        0.0       0.000000  100000.000000\n",
      "ENCODE-mouseT000008457        0.0        0.0  142857.142857  142857.142857\n",
      "ENCODE-mouseT000008451   300000.0   250000.0  285714.285714  835714.285714\n",
      "ENCODE-mouseT000008460        0.0   250000.0       0.000000  250000.000000\n",
      "ENSMUST00000020637.8          0.0        0.0  142857.142857  142857.142857\n",
      "ENSMUST00000020741.10    100000.0        0.0       0.000000  100000.000000\n",
      "ENCODE-mouseT000008452   100000.0        0.0       0.000000  100000.000000\n",
      "ENCODE-mouseT000008458        0.0        0.0  142857.142857  142857.142857\n",
      "\n",
      "Plotting transcripts for ENSMUSG00000020167.13\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/anndata/_core/anndata.py:1056: FutureWarning: is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead\n",
      "  if not is_categorical(df_full[k]):\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saving transcript path graph for ENSMUST00000105344.6 as bepis_test_ENSMUST00000105344.6_path.png\n",
      "Saving transcript path graph for ENCODE-mouseT000008452 as bepis_test_ENCODE-mouseT000008452_path.png\n",
      "Saving transcript path graph for ENCODE-mouseT000008451 as bepis_test_ENCODE-mouseT000008451_path.png\n",
      "Generating report for ENSMUSG00000020167.13\n"
     ]
    }
   ],
   "source": [
    "t_df = sg.gen_report('ENSMUSG00000020167.13', 'bepis_test',\n",
    "                     cmap='viridis',\n",
    "                     groupby='dataset',\n",
    "                     display_numbers=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "lesser-binary",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/anndata/_core/anndata.py:1056: FutureWarning: is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead\n",
      "  if not is_categorical(df_full[k]):\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                         tname                     gid  \\\n",
      "tid                                                                      \n",
      "ENCODE-mouseT000008451  ENCODE-mouseT000008451   ENSMUSG00000020167.13   \n",
      "ENCODE-mouseT000008453  ENCODE-mouseT000008453   ENSMUSG00000020457.12   \n",
      "ENCODE-mouseT000008460  ENCODE-mouseT000008460   ENSMUSG00000020368.14   \n",
      "ENSMUST00000105344.6                  Tcf3-005   ENSMUSG00000020167.13   \n",
      "ENCODE-mouseT000008456  ENCODE-mouseT000008456  ENCODE-mouseG000002990   \n",
      "ENCODE-mouseT000008457  ENCODE-mouseT000008457   ENSMUSG00000020368.14   \n",
      "ENCODE-mouseT000008458  ENCODE-mouseT000008458   ENSMUSG00000020368.14   \n",
      "ENCODE-mouseT000008459  ENCODE-mouseT000008459   ENSMUSG00000020368.14   \n",
      "ENSMUST00000020637.8                  Canx-001   ENSMUSG00000020368.14   \n",
      "ENCODE-mouseT000008452  ENCODE-mouseT000008452   ENSMUSG00000020167.13   \n",
      "ENCODE-mouseT000008454  ENCODE-mouseT000008454  ENCODE-mouseG000002988   \n",
      "ENCODE-mouseT000008455  ENCODE-mouseT000008455  ENCODE-mouseG000002989   \n",
      "ENCODE-mouseT000008461  ENCODE-mouseT000008461   ENSMUSG00000020368.14   \n",
      "ENCODE-mouseT000008462  ENCODE-mouseT000008462   ENSMUSG00000020368.14   \n",
      "ENSMUST00000020741.10                 Drg1-001   ENSMUSG00000020457.12   \n",
      "\n",
      "                                         gname  \\\n",
      "tid                                              \n",
      "ENCODE-mouseT000008451   ENSMUSG00000020167.13   \n",
      "ENCODE-mouseT000008453   ENSMUSG00000020457.12   \n",
      "ENCODE-mouseT000008460   ENSMUSG00000020368.14   \n",
      "ENSMUST00000105344.6     ENSMUSG00000020167.13   \n",
      "ENCODE-mouseT000008456  ENCODE-mouseG000002990   \n",
      "ENCODE-mouseT000008457   ENSMUSG00000020368.14   \n",
      "ENCODE-mouseT000008458   ENSMUSG00000020368.14   \n",
      "ENCODE-mouseT000008459   ENSMUSG00000020368.14   \n",
      "ENSMUST00000020637.8     ENSMUSG00000020368.14   \n",
      "ENCODE-mouseT000008452   ENSMUSG00000020167.13   \n",
      "ENCODE-mouseT000008454  ENCODE-mouseG000002988   \n",
      "ENCODE-mouseT000008455  ENCODE-mouseG000002989   \n",
      "ENCODE-mouseT000008461   ENSMUSG00000020368.14   \n",
      "ENCODE-mouseT000008462   ENSMUSG00000020368.14   \n",
      "ENSMUST00000020741.10    ENSMUSG00000020457.12   \n",
      "\n",
      "                                                                     path  \\\n",
      "tid                                                                         \n",
      "ENCODE-mouseT000008451                                               [97]   \n",
      "ENCODE-mouseT000008453  [17, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 18, 14, 15...   \n",
      "ENCODE-mouseT000008460  [20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 3...   \n",
      "ENSMUST00000105344.6    [60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 7...   \n",
      "ENCODE-mouseT000008456                                              [101]   \n",
      "ENCODE-mouseT000008457  [20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 3...   \n",
      "ENCODE-mouseT000008458                                               [54]   \n",
      "ENCODE-mouseT000008459  [55, 56, 38, 39, 50, 51, 52, 43, 44, 45, 46, 4...   \n",
      "ENSMUST00000020637.8    [49, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 3...   \n",
      "ENCODE-mouseT000008452                               [92, 93, 94, 95, 98]   \n",
      "ENCODE-mouseT000008454                                               [99]   \n",
      "ENCODE-mouseT000008455                                              [100]   \n",
      "ENCODE-mouseT000008461  [24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 3...   \n",
      "ENCODE-mouseT000008462  [20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 3...   \n",
      "ENSMUST00000020741.10   [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,...   \n",
      "\n",
      "                                           tid     novelty  \\\n",
      "tid                                                          \n",
      "ENCODE-mouseT000008451  ENCODE-mouseT000008451     Genomic   \n",
      "ENCODE-mouseT000008453  ENCODE-mouseT000008453         NIC   \n",
      "ENCODE-mouseT000008460  ENCODE-mouseT000008460         ISM   \n",
      "ENSMUST00000105344.6      ENSMUST00000105344.6       Known   \n",
      "ENCODE-mouseT000008456  ENCODE-mouseT000008456   Antisense   \n",
      "ENCODE-mouseT000008457  ENCODE-mouseT000008457         NNC   \n",
      "ENCODE-mouseT000008458  ENCODE-mouseT000008458     Genomic   \n",
      "ENCODE-mouseT000008459  ENCODE-mouseT000008459         NNC   \n",
      "ENSMUST00000020637.8      ENSMUST00000020637.8       Known   \n",
      "ENCODE-mouseT000008452  ENCODE-mouseT000008452         ISM   \n",
      "ENCODE-mouseT000008454  ENCODE-mouseT000008454   Antisense   \n",
      "ENCODE-mouseT000008455  ENCODE-mouseT000008455  Intergenic   \n",
      "ENCODE-mouseT000008461  ENCODE-mouseT000008461         ISM   \n",
      "ENCODE-mouseT000008462  ENCODE-mouseT000008462         ISM   \n",
      "ENSMUST00000020741.10    ENSMUST00000020741.10       Known   \n",
      "\n",
      "                                                                 loc_path  \n",
      "tid                                                                        \n",
      "ENCODE-mouseT000008451                                             [5, 1]  \n",
      "ENCODE-mouseT000008453  [58, 57, 56, 55, 54, 53, 52, 51, 50, 49, 48, 4...  \n",
      "ENCODE-mouseT000008460  [102, 101, 100, 99, 98, 97, 96, 95, 94, 93, 92...  \n",
      "ENSMUST00000105344.6    [39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 2...  \n",
      "ENCODE-mouseT000008456                                           [64, 67]  \n",
      "ENCODE-mouseT000008457  [102, 101, 100, 99, 98, 97, 96, 95, 94, 93, 92...  \n",
      "ENCODE-mouseT000008458                                           [67, 66]  \n",
      "ENCODE-mouseT000008459  [86, 83, 82, 81, 80, 79, 76, 74, 73, 71, 70, 6...  \n",
      "ENSMUST00000020637.8    [103, 101, 100, 99, 98, 97, 96, 95, 94, 93, 92...  \n",
      "ENCODE-mouseT000008452                                 [7, 6, 4, 3, 2, 1]  \n",
      "ENCODE-mouseT000008454                                           [60, 61]  \n",
      "ENCODE-mouseT000008455                                           [62, 63]  \n",
      "ENCODE-mouseT000008461  [98, 97, 96, 95, 94, 93, 92, 91, 90, 89, 88, 8...  \n",
      "ENCODE-mouseT000008462  [102, 101, 100, 99, 98, 97, 96, 95, 94, 93, 92...  \n",
      "ENSMUST00000020741.10   [59, 57, 56, 55, 54, 53, 52, 51, 50, 49, 48, 4...  \n",
      "                                 2      1         sum\n",
      "tid                                                  \n",
      "ENCODE-mouseT000008451   66.666667  100.0  166.666667\n",
      "ENCODE-mouseT000008453   66.666667    0.0   66.666667\n",
      "ENCODE-mouseT000008452   16.666667    0.0   16.666667\n",
      "ENCODE-mouseT000008459    0.000000   25.0   25.000000\n",
      "ENSMUST00000105344.6     16.666667    0.0   16.666667\n",
      "ENCODE-mouseT000008460   33.333333    0.0   33.333333\n",
      "ENCODE-mouseT000008456    0.000000  100.0  100.000000\n",
      "ENCODE-mouseT000008455  100.000000    0.0  100.000000\n",
      "ENCODE-mouseT000008454  100.000000    0.0  100.000000\n",
      "ENCODE-mouseT000008458    0.000000   25.0   25.000000\n",
      "ENSMUST00000020637.8      0.000000   25.0   25.000000\n",
      "ENCODE-mouseT000008461   33.333333    0.0   33.333333\n",
      "ENCODE-mouseT000008457    0.000000   25.0   25.000000\n",
      "ENCODE-mouseT000008462   33.333333    0.0   33.333333\n",
      "ENSMUST00000020741.10    33.333333    0.0   33.333333\n",
      "\n",
      "Plotting transcripts for ENSMUSG00000020167.13\n",
      "Saving transcript path graph for ENCODE-mouseT000008451 as pi_test_ENCODE-mouseT000008451_path.png\n",
      "Saving transcript path graph for ENSMUST00000105344.6 as pi_test_ENSMUST00000105344.6_path.png\n",
      "Saving transcript path graph for ENCODE-mouseT000008452 as pi_test_ENCODE-mouseT000008452_path.png\n",
      "Generating report for ENSMUSG00000020167.13\n"
     ]
    }
   ],
   "source": [
    "t_df = sg.gen_report('ENSMUSG00000020167.13', 'pi_test',\n",
    "                     cmap='magma',\n",
    "                     novelty=True,\n",
    "                     layer='pi',\n",
    "                     groupby='cluster',\n",
    "                     display_numbers=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fabulous-guide",
   "metadata": {},
   "outputs": [],
   "source": [
    "sg.plot_transcript_path('ENCODE-mouseT000008451')\n",
    "swan.save_fig('test_2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "excellent-cinema",
   "metadata": {},
   "outputs": [],
   "source": [
    "t_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "infectious-hearts",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6/17/21\n",
    "# gen_report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ceramic-triple",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding transcriptome to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Transforming to str index.\n",
      "Transforming to str index.\n",
      "AnnData expects .obs.index to contain strings, but your first indices are: Int64Index([0, 1], dtype='int64'), …\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding abundance for datasets PB65_B017, PB65_B018 to SwanGraph.\n",
      "\n",
      "Adding abundance for datasets D12 to SwanGraph.\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "db = 'data/chr11_and_Tcf3_no_gname.db'\n",
    "sg.add_transcriptome(db, include_isms=True)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance_PB65.tsv'\n",
    "sg.add_abundance(ab)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance_D12.tsv'\n",
    "sg.add_abundance(ab)\n",
    "meta = 'data/chr11_and_Tcf3_metadata.tsv'\n",
    "sg.add_metadata(meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "earned-premiere",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/anndata/_core/anndata.py:1056: FutureWarning: is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead\n",
      "  if not is_categorical(df_full[k]):\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:ylabel='tid'>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAgYAAAD4CAYAAACXOC7gAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAmCElEQVR4nO3de7xcVXn/8c83hEtQQEAuMaG/RAGVayAxUMEKUpJIrUAJP4JWYotNRWi9lHtroSoVBEVRgUaIBIoGDKAoEIxcmmLDJUDIhXCJQCGGaxMRhUaS8/SP/cycnWHPnCE5lyTn+3699mtm1t5rzdpz5px5Zq1nr6OIwMzMzAxgQF93wMzMzNYdDgzMzMyszoGBmZmZ1TkwMDMzszoHBmZmZlY3sK87YNaujud29SU0ZtaWATs+prVto92/Od3xXOsSjxiYmZlZnUcMzMzMKnTQ0dZxG9o3bAcGZmZmFV6PVW0dt6F9kG5o52NmZtYt2h0x2NA4MDAzM6uwqp/+ywAHBmZmZhU6cGBgZmZmaZUDAzMzM6vxiIGZmZnVve4cAzMzM6vxVIKZmZnVreqfcYEDAzMzsyr9cxUDBwZmZmaVVrFB/W+ktjkwMDMzq/B6ODAwMzOz5BEDMzMzq+vwiIGZmZnVeMTAzMzM6lYxoK+70CccGJiZmVXwVIKZmZnV/SE26usu9AkHBmZmZhU6PJVgZmZmNU4+NDMzs7pV0T9HDPrnWZuZmXWhA7W1dUXSTpLukLRI0kJJn83ysyX9WtLc3A4r1TlD0mJJj0oaWyofKWl+7rtIkrJ8U0nXZPk9koaV6kyU9HhuE7vqr0cMzMzMKvwhuu0jciXwDxHxgKQtgPslzcx9F0bEBeWDJe0GTAB2B94B/ELSrhGxCrgEmATcDdwMjANuAY4HlkfEzpImAOcBx0jaBjgLGAVEPveNEbG8WWc9YmBmZlahgwFtbV2JiGcj4oG8/wqwCBjSosrhwLSIWBERTwKLgdGSBgNbRsTsiAjgSuCIUp2peX86cEiOJowFZkbEsgwGZlIEE005MDAzM6uwKtTW9mbkEP8+wD1ZdJKkeZKmSNo6y4YAz5SqLcmyIXm/sXy1OhGxEngZ2LZFW005MDAzM6uwigFtbZImSZpT2iZVtSfprcB1wOci4rcU0wLvAkYAzwJfrx1aUT1alK9pnUrOMTAzM6vQ0eZVCRExGZjc6hhJG1MEBVdHxPVZ7/nS/u8BP8uHS4CdStWHAkuzfGhFebnOEkkDga2AZVl+UEOdO1v11SMGZmZmFdodMehKzvVfDiyKiG+UygeXDjsSWJD3bwQm5JUGw4FdgHsj4lngFUn7Z5vHAT8p1aldcTAeuD3zEG4FxkjaOqcqxmRZUx4xMDMzq/B69y2JfADwCWC+pLlZdiZwrKQRFEP7TwF/CxARCyVdCzxMcUXDiXlFAsAJwBXAIIqrEW7J8suBqyQtphgpmJBtLZP0ZeC+PO5LEbGsVWdVBBRm676O53b1m9XM2jJgx8fWetnCKY8d2NbfnL/e9a4NaonEHptKkLSqtGjDXEmnZ/mdkuaUjhsl6c7S49GSZuWiDo9IukzS5rnviMzefCQXeDiiVO8KSU9KekjSY5KulDSktP+prFPrz0U9de5rQ9J3s38PS3qt1N/xkraRNDMXqZhZymDtlcUwJB0i6YHsz12Sdm7o+/vy5z6+VFZ+3cs/96NVLPTRIWlUt7+QZmZrqbsWOFrf9ORUwmsRMaLJvu0lfTgibikXStoB+BEwISJm54fYUcAWknYBLgAOjYgnc95lpqQnImJeNnFKREzPep8D7pC0R0T8IfcfHBEvde9pdq+IOBHql7T8rPwaSvoacFtEnJuB1unAaeq9xTAuAQ6PiEWSPgP8E/DJ7NtG2UbV3FXV674A+Avg39b4xTIz60FeErl3nU/xodLoRGBqRMwGiML0zNw8GfjXXOyBvP0qcEpjI1nvQuA54MPtdkrSsNIoxQJJV0v6U0m/zG/Po/O4bST9OEcv7pa0V5afLenkUnsLss23SLopRzMWSDom94+U9B+S7pd0q1ZPRKlSXsBiKqsvbNEbi2EEsGXe34rObFiAv6PIuH2hi3MoGopYFBGPtnOsmVlf6K7kw/VNT57RIK0+lXBMad9sYIWkgxvq7AHc36S93Sv2zcnyZh4A3lN6fEepP59vUmdn4FvAXln3Y8CBFIHJmXnMvwAPRsReWXZliz5A8cG6NCL2jog9gBkqLl35NjA+IkYCU4Bzumhnh8xKJW+3z/LeWgzjU8DNkpZQJNKcC5BTNkcCl1b0OYCfZ/BTeW2vmdm6qCPU1rah6aupBICvUIwanNZme+KNizJUlTXuL2tnKuHJiJgPIGkhxdB9SJoPDMtjDqSY4iAibpe0raStWrQ5H7hA0nkU0wP/KWkPikBoZk77b0SxwMWa6K3FMD4PHBYR90g6BfgGRbDwTeC0iFiV51J2QEQslbQ9xbk+EhGzWp5NSQYTkwAu+dr2TPpEq5fZzKz7vN59/ythvdJnYyARcTuwGbB/qXghMLJJlYUU895l+1JcztHMPhRrUldS8R+vaiMIn87iFaVDOkqPO+gMpJp9eK5k9dd0M4CIeIzivOYDX5X0z9nGwogYkdueETGmxbkAPF+bbsjb2rD92iyGgd64GMYb2pK0HbB3RNSW8bwGeH/eHwVMk/QUxfWzFysTQyNiad6+ANwAjO7iHFcTEZMjYlREjHJQYGa9aRVqa9vQ9PXkyDnAqaXH3wEmStqvViDpLyXtSJF4eEYtez5vz6RzCUlKdSTp74HBwIxmTx4Rz5Q+mKuGwZuZBXw8n+sg4KVc3vIpimAFSfsCw/P+O4BXI+Lf8zz2BR4FtpP0x3nMxpJaTYvA6gtYTGT1hS16ejGM5cBWknbNOoeSQVdEDI+IYRExjCJf4TMR8ePMrdgiz+8t2VZtAQ8zs3VaRwxoa9vQ9OQ4ySB1LuQAMCMiTi8fEBE3S3qx9Pj5zJC/IIeeOyg+hK+PiOcknQb8NOfnXwdOjYjyc5wv6YvA5hRZ+AeXrkiAIsegtkjEvIg4bg3P7Wzg+5LmAa/S+QF7HXBcnvd9wGNZvmf2rSP7fUJE/EHFZX0X5TTEQIoh+YUtnvdc4FpJxwNPA0dD7y2GIelvgOvyPJYDf93F67QDcENOLwwEfhARM7KtIylyLLYDbpI0NyLGNm3JzKyXbYijAe3wAke23vACR2bWru5Y4OichR9p62/OP+7+sw0qguifmRVmZmZd6MYlkdcrDgzMzMwq9NcFjhwYmJmZVdgQ1yhohwMDMzOzChviqobtcGBgZmZWwSMGZmZmVtfhEQMzMzOreb3DgYGZmZmlDXFVw3Y4MDAzM6vQX1c+dGBgZmZWwcmHZmZmVuepBDMzM6vr8FSCmZmZ1bze4f+VYGZmZsk5BmZmZlbnqQQzMzOr84iBmZmZ1fmqBDMzM6tb6cDAzMzMajyVYGZmZnX9NTDon+MkZmZmXegItbV1RdJOku6QtEjSQkmfzfJtJM2U9Hjebl2qc4akxZIelTS2VD5S0vzcd5EkZfmmkq7J8nskDSvVmZjP8bikiV3114GBmZlZhe4KDICVwD9ExHuB/YETJe0GnA7cFhG7ALflY3LfBGB3YBxwsaTaakuXAJOAXXIbl+XHA8sjYmfgQuC8bGsb4CxgP2A0cFY5AKniwMDMzKxCB2pr60pEPBsRD+T9V4BFwBDgcGBqHjYVOCLvHw5Mi4gVEfEksBgYLWkwsGVEzI6IAK5sqFNrazpwSI4mjAVmRsSyiFgOzKQzmKjkHAMzM7MKKzu6/7tzDvHvA9wD7BARz0IRPEjaPg8bAtxdqrYky17P+43ltTrPZFsrJb0MbFsur6hTySMGZmZmFdqdSpA0SdKc0japqj1JbwWuAz4XEb9t8dRVwxDRonxN61TyiIGZmVmFdq9KiIjJwORWx0jamCIouDoirs/i5yUNztGCwcALWb4E2KlUfSiwNMuHVpSX6yyRNBDYCliW5Qc11LmzVV89YmBmZlYhQm1tXcm5/suBRRHxjdKuG4HaVQITgZ+UyifklQbDKZIM781ph1ck7Z9tHtdQp9bWeOD2zEO4FRgjaetMOhyTZU15xMDMzKxCN/4TpQOATwDzJc3NsjOBc4FrJR0PPA0cDRARCyVdCzxMcUXDiRGxKuudAFwBDAJuyQ2KwOMqSYspRgomZFvLJH0ZuC+P+1JELGvVWRUBhdm6r+O5Xf1mNbO2DNjxsbX+VD/wF6e29Tfnrj/92ga1EpJHDMzMzCqs6oGrEtYHDgxsvfEnJ1Ym+lo/NuiGe/q6C7aOmtmx9m20kz+wIXJgYGZmVqG//q8EBwZmZmYV+msKngMDMzOzCt14VcJ6xYGBmZlZBScfmpmZWZ2nEszMzKzOVyWYmZlZnQMDMzMzq/PlimZmZlbnHAMzMzOr6/BVCWZmZlbTTwcMHBiYmZlVcfKhmZmZdeqnQwYODMzMzCp4xMDMzMzqOjocGJiZmVmNRwzMzMysxusYmJmZWScHBmZmZlbj5EMzMzPr5BEDMzMzqwlflWBmZmadHBiYmZlZjacSzMzMrM6BgZmZmdX5qgQzMzOr8QJHZmZm1qmfXpUwoK87YGZmti5StLe11ZY0RdILkhaUys6W9GtJc3M7rLTvDEmLJT0qaWypfKSk+bnvIknK8k0lXZPl90gaVqozUdLjuU3sqq8ODMzMzKpEm1t7rgDGVZRfGBEjcrsZQNJuwARg96xzsaSN8vhLgEnALrnV2jweWB4ROwMXAudlW9sAZwH7AaOBsyRt3aqjDgzMzMyqhNrb2mkqYhawrM1nPhyYFhErIuJJYDEwWtJgYMuImB0RAVwJHFGqMzXvTwcOydGEscDMiFgWEcuBmVQHKHUODMzMzKp074hBMydJmpdTDbVv8kOAZ0rHLMmyIXm/sXy1OhGxEngZ2LZFW005MDAzM6vS0d4maZKkOaVtUpvPcAnwLmAE8Czw9SyvGoaIFuVrWqdS06sSJO3bqmJEPNBqv5mZ2Xqt/WmCycDkN918xPO1+5K+B/wsHy4BdiodOhRYmuVDK8rLdZZIGghsRTF1sQQ4qKHOna361WrE4Ou5fRe4h+Kkv5f3L2rV6NqStKqUpTlX0ulZfqekOaXjRkm6M+9vLunqzNZcIOkuSW/NfSHpqlK9gZJelPSzfHy2pJMb+vCUpLfn/X+UtDCHe+ZK2k/SDXl/saSXS319v6ThmRX6eGaJbpLtKLNIF2db+5aeb1xmny6unW+WbyNpZrY1s5w00iJr9RxJz0j6XcM5tcpaLb/mN5bKL5f0UPZ3eu01Le1/X9Yd31C+kaQHa69xk5/zQfl8CyX9R7PjzMz6QndelVDZfpEzUHMkULti4UZgQv7NHk6RZHhvRDwLvCJp/8wfOA74SalO7YqD8cDtmYdwKzBG0tb5+TEmy5pqOmIQEQdnx6cBkyJifj7eAzi5Wb1u8lpEjGiyb3tJH46IWxrKPws8HxF7Akh6N/B67vs9sIekQRHxGnAo8Ot2OiLpj4GPAPtGxIoMFjaJiCNz/0HAyRHxkVKdaykyTadJupQiW/QS4MN0ZpLul2X7Zbbpd7NfS4D7JN0YEQ8DpwO3RcS5GTCcDpym1bNW3wH8QtKuEbEK+CnwHeDxhtOpZ61KmkCRtXpM7mv2mn8+In6b5/UN4CTg3Hy8UbZR9Sb7LLAI2LLJ6/o24GJgXEQ8LWn7quPMzPpMNy5wJOmHFN/c3y5pCcWVAgdJGpHP9BTwtwARsTA/Rx4GVgIn5t92gBMornAYBNySG8DlwFWSFlOMFEzItpZJ+jJwXx73pYhomQTZzgJH76kFBfkkC/JE+sr5wD/R+WLUDAb+u/YgIh5t2H8L8GcU2ZrHAj8EPtDG8w0GXoqIFdnuS60OzijuQ8DHsmgqcDZFEHA4cGVGcXdLeltGjMOAxRHxRLYxLY99OG8PKrV1J3AapaxV4Ml8M4wGZkfE3dlOY/cOz76Qr8N3JCn7U6kUFIjijVg+9u+A64D3NbwGQyle63OALzRp+mPA9RHxdD7PC836YGa2vouIYyuKL29x/DkUf0Mby+cAe1SU/y9wdJO2pgBT2u1rO8mHiyRdlsO+H1QxD7Ko3SdYQ4O0+lTCMaV9s4EVkg5uqDOF4pv0bElfkbRLw/5pFEMzmwF7UUyJtOPnwE6SHpN0saQPdnH8tsBvMisUmmSNNuxrlTW6Qw4fkbe1b9ZvOtOU5lmrAJupSJq5W9IR5UqSvg88B7wH+HaWDaEY+rq04nm+CZxKkZrTzK7A1iqmh+6XdFzVQSol9Tz35N1dnJ6ZWffp6amEdVU7gcFfAQsphoY/R/Et9q96sE+Qw9ql7ZqG/V+hGDWoi4i5wDspRhS2oRiOf29p/zyKb+bHAjc3tNfsRxsR8TtgJMWCEi8C10j6ZIu+90rWaA/U+aOIGEXxTf6bkt5VPyDiryimKxbROfXwTeC00vBW8QTSR4AXIuL+LvoykOJ1/TOK62y/KGnXN3QuYnJEjIqIUTsO37+LJs3MulGH2ts2MF1OJeTwxIW5rRMi4vacM9m/ofx3wPXA9ZI6gMNYfXTjRuACiqH5bUvl/0MxZVC2BfCbbHcVxRD+nZLmUyR4XNGkey8Bb5M0ML+VV2WN1tT2bdKkHOB5SYMj4tmcdqgNuTdrq5VmWatExNK8fUJFQuc+wK9qFSNilaRrgFOA7wOjgGk5XfF24DBJKylyJz6qYmnPzYAtJf17RPxlRV9eiojfA7+XNAvYG3isi3MwM+sdG+BoQDuajhhk4gMqsvznNW6918WmzqEYrgZA0gGZcYmKqwB2o5RzkKZQJF7MbyifRfFhtkXW/wvgofwwfHfDtMSIinbrcr7+DoqsUCiCiHLW6HEq7A+8nNMD9wG7qLiaYROKpJEbS3UmNmnrDVmrzfpV0VY9azWzVTfNc387cADwcPZz5ywX8OfAI3mewyNiWEQMo8hX+ExE/DgizoiIoVk+IZ+jMSggz+MDKq4Q2ZwioOjpKSozs7b116mEViMGn83bRRTfEmsEfK3HelQYJGlu6fGMiDi9fEBE3CzpxVLRu4BL8gNsAHATRWJcuc4S4FuNTxYR8yR9B7hLUlB8K/9U7n4r8O3Mol9JsTRlV4tXnEbxbforwIN0JpjcTDGKsRh4lZySiYiVkk6iyO7fCJgSEQuzzrnAtZKOB54mk0taZa1K+hrFlMDmmf16WUScTZOsVeC9wL/lKMsA4NyIeFjSAGCqpC0pfu4PUWTErhFJn86+XxoRiyTNAOZR5CJcFhELWjZgZtabNsAP/XaoRUJ6cYD0QETs21A2LyL26tGemTU48KgL+umvqTUz6IZ2c4itv5nZ8aO1nvx/1wXfaOtvzq9O/sIGlWjQauXDE4DPAO9smDrYAvhlT3fMzMysL22I0wTtaDWV8AOKa/+/SrGoTs0rXS2OYGZmtt7bAK84aEerlQ9fprjOvWpRBjMzsw2aRwzMzMyskwMDMzMzq/GIgZmZmXVyYGBmZmY1avXfXjZg7fyvBDMzM+snPGJgZmZWxVMJZmZmVuPkQzMzM+vkwMDMzMzqHBiYmZlZTX+9KsGBgZmZWQXnGJiZmVknBwZmZmZW58DAzMzMajyVYGZmZp0cGJiZmVmNr0owMzOzTh4xMDMzsxrnGJiZmVknBwZmZmZW58DAzMzMavrrVMKAvu6AmZnZukjR3tZWW9IUSS9IWlAq20bSTEmP5+3WpX1nSFos6VFJY0vlIyXNz30XSVKWbyrpmiy/R9KwUp2J+RyPS5rYVV8dGJiZmVWJNrf2XAGMayg7HbgtInYBbsvHSNoNmADsnnUulrRR1rkEmATsklutzeOB5RGxM3AhcF62tQ1wFrAfMBo4qxyAVHFgYGZmVqUbA4OImAUsayg+HJia96cCR5TKp0XEioh4ElgMjJY0GNgyImZHRABXNtSptTUdOCRHE8YCMyNiWUQsB2byxgBlNQ4MzMzMKrQ7lSBpkqQ5pW1Sm0+xQ0Q8C5C322f5EOCZ0nFLsmxI3m8sX61ORKwEXga2bdFWU04+NDMzq9L+aMBkYHI3PrOa9KZZ+ZrWqeQRAzMzswrqaG9bC8/n9AB5+0KWLwF2Kh03FFia5UMrylerI2kgsBXF1EWztppyYGBmZlahO69KaOJGoHaVwETgJ6XyCXmlwXCKJMN7c7rhFUn7Z/7AcQ11am2NB27PPIRbgTGSts6kwzFZ1pSnEszMzKp04zoGkn4IHAS8XdISiisFzgWulXQ88DRwNEBELJR0LfAwsBI4MSJWZVMnUFzhMAi4JTeAy4GrJC2mGCmYkG0tk/Rl4L487ksR0ZgEuRoHBmZmZlW6MTCIiGOb7DqkyfHnAOdUlM8B9qgo/18ysKjYNwWY0m5fHRiYmZlV6K8rHzowMDMzq6CO/hkZODAwMzOr0j/jAgcGZmZmVTyVYGZmZp0cGJiZmVmNRwzMzMyskwMDMzMzq1nL5Y7XWw4MzMzMKngqwczMzDpF/4wMHBiYmZlV8IiBmZmZdXJgYGZmZjVOPjQzM7M6BwZmZmbWycmHZmZmVuPkQzMzM+vkwMDMzMxqPGJgZmZmderon5GBAwMzM7Mq/TMucGBgZmZWxVMJZmZm1slTCWZmZlbXP+MCBwZmZmZVPJVgZmZmdb4qwczMzDr1z7jAgYGZmVkV+X8lmJmZWZ3/u6KZmZnV9NcRgwF93QEzM7N1UrS5tUHSU5LmS5oraU6WbSNppqTH83br0vFnSFos6VFJY0vlI7OdxZIukqQs31TSNVl+j6Rha3raPRYYSFqVL0BtOz3L76y9KPl4lKQ7S49HS5qVL8Yjki6TtHnuO0LSvCyfL+mIUr0rJD0p6SFJj0m6UtKQ0v7yD2WupIt66tzXhqTvZv8elvRaqb/je+tNJGliPsfjkiaWyg+R9ED25y5JOzf0/X35cx9fKnvDL0OWn58/x3mSbpD0tu5+Lc3M1oY6oq3tTTg4IkZExKh8fDpwW0TsAtyWj5G0GzAB2B0YB1wsaaOscwkwCdglt3FZfjywPCJ2Bi4EzlvT8+7JEYPX8gWobeeW9m0v6cONFSTtAPwIOC0i3g28F5gBbCFpb+AC4PCIeA/wUeACSXuVmjglIvYG3g08CNwhaZPS/oNL/fn7bj3bbhIRJ0bECOAw4Fel/k6nF95EkrYBzgL2A0YDZ5UCkEuAj2f/fgD8U63f+XznAbdWnFbjLwPATGCPiNgLeAw4482+VmZmPSqivW3NHQ5MzftTgSNK5dMiYkVEPAksBkZLGgxsGRGzIyKAKxvq1NqaDhxS+yL4ZvXVVML5lD5USk4EpkbEbIAoTI+I54GTgX/NF4m8/SpwSmMjWe9C4DngDQFIM5KGlUYpFki6WtKfSvplfnsencdtI+nH+W337lpwIulsSSeX2luQbb5F0k05mrFA0jG5f6Sk/5B0v6Rb84feSm+8icYCMyNiWUQsp/gArwUTAWyZ97cClpb69nfAdcALXZxD0VDEzyNiZT68GxjaTj0zs96ijjY3aZKkOaVtUkVzAfw8/97X9u8QEc8C5O32WT4EeKZUd0mWDcn7jeWr1cm/rS8D267JefdkYDBIq08lHFPaNxtYIenghjp7APc3aW/3in1zsryZB4D3lB7fUerP55vU2Rn4FrBX1v0YcCBFYHJmHvMvwIP5bfdMig/cVsYBSyNi74jYA5ghaWPg28D4iBgJTAHO6aKd3ngTNWsL4FPAzZKWAJ8AzgXIKZsjgUsr+lz1y9Dor4Fbmp61mVlfaHPEICImR8So0ja5orUDImJfii+rJ0r6kxbPXPVNP1qUt6rzpvXkVQmv5ZBzM1+hGDU4rc32xBtPsqqscX/ZwRHxUhfP82REzAeQtJBi6D4kzQeG5TEHAkcBRMTtkraVtFWLNudTTHucB/wsIv5T0h4UgdDMHO3ZCHi2i741051volZ1Pg8cFhH3SDoF+AZFsPBNiumfVRUjVwdExFJJ21Oc6yMRMavecekfgZXA1ZUnVgQTkwDetc9R7Dh8/6rDzMy6XzdelBARS/P2BUk3UEzVPi9pcEQ8myO8tRHXJcBOpepDKUZol7D66GqtvFxniaSBFKO6y9akr312VUJE3A5sBpT/0i8ERjapshAY1VC2L/Bwi6fZB1jUbKeknUojCJ/O4hWlQzpKjzvoDKSafXiuZPXXdDOAiHiM4rzmA1+V9M/ZxsJSDsGeETGmxblAvomy7931JqLhTVTZlqTtgL0j4p4svwZ4f94fBUyT9BQwniLH4Yg89/ovA1D7ZSCfdyLwEYq8hcpfwXIk7qDAzHqTOjra2rpsp5hO3qJ2HxgDLABuBGoJ3hOBn+T9G4EJmSQ+nCI/7N4cKX5F0v459XtcQ51aW+OB25v9Xe1KX1+ueA5waunxd4CJkvarFUj6S0k7UiQenqHMns/bM4GvNzaqwt8DgymSFytFxDOlD+aqYfBmZgEfz+c6CHgpIn4LPEURrCBpX2B43n8H8GpE/Huex77Ao8B2kv44j9lYUqtpEeidN9GtwBhJW2fS4ZgsWw5sJWnXrHMoGXRFxPCIGBYRwyjyFT4TET9u8cuApHEUo0UfjYhXuzhvM7Pe19Hm1rUdgLskPQTcC9wUETMopmMPlfQ4xd/UcwEiYiFwLcUX3xnAiRGxKts6AbiMIpfsV3ROw14ObCtpMfAFMjl9TfTkVMIgSXNLj2dExGodjYibJb1Yevy8pAkUw+7bU7zks4DrI+I5SacBP835+deBUyOi/BznS/oisDlFQtvBEfGH0v47JNVe3HkRcdwantvZwPclzQNepfMD9jrguDzv+yiy7QH2zL51ZL9PiIg/qLis76KchhhIMSS/sMXzngtcK+l44GngaCjeRJJqb6KVvPFNdAUwiOINVH4TXZVvomUUVzUQEcskfTn7D/CliFgGIOlvgOvyPJZT5Aa0sgNwQ04vDAR+kL8MUASBm9I5lXJ3RHy6shUzsz7QXQscRcQTwN4V5f8DHNKkzjlU5J1FxByKaejG8v8lPxPWltZwpMGs1x141AV+s9pqBt1wT9cHWb80s+NHa3SpXtnYUWe39Tfn1jlnr/VzrUu8JLKZmVmVfvrF2YGBmZlZFf8TJTMzM6tp54qDDZEDAzMzsyqeSjAzM7M6BwZmZmZW1z9nEhwYmJmZVemudQzWNw4MzMzMqjgwMDMzs7pV/XMuwYGBmZlZFY8YmJmZWZ0DAzMzM6vrcGBgZmZmNeEcAzMzM6tx8qGZmZnVOcfAzMzM6hwYmJmZWZ0DAzMzM6vzv102MzOzOo8YmJmZWZ2vSjAzM7Oa8DoGZmZmVueVD83MzKzOOQZmZmZW56sSzMzMrM4jBmZmZlYTq1b1dRf6hAMDMzOzKk4+NDMzs7p+ernigL7ugJmZ2booOqKtrR2Sxkl6VNJiSaf3cNfXigMDMzOzKtHR3tYFSRsB3wU+DOwGHCtptx7u/RrzVIKZmVmFbkw+HA0sjognACRNAw4HHu6uJ+hOin56OYbZ+kzSpIiY3Nf9sHWL3xd9Q9IkYFKpaHL55yBpPDAuIj6Vjz8B7BcRJ/VuT9vjqQSz9dOkrg+xfsjviz4QEZMjYlRpawzOVFWtN/q2JhwYmJmZ9awlwE6lx0OBpX3Uly45MDAzM+tZ9wG7SBouaRNgAnBjH/epKScfmq2fPI9sVfy+WAdFxEpJJwG3AhsBUyJiYR93qyknH5qZmVmdpxLMzMyszoGBmZmZ1TkwMOshknaQ9ANJT0i6X9JsSUdKOkhSSDq+dOw+WXZyPj5a0kJJHZJGlY47NNuan7cfKu07NsvnSZoh6e1ZfqGkubk9Juk3pTpfy+dZJOkiSVWXVVkfkfS70v3DJD0u6Y8kfVLSi6Wf66dKx/2RpJ/nz/RhScOy/OpckneBpCmSNu6DU7L1gAMDsx6QH7A/BmZFxDsjYiRFJvLQPGQ+cEypygTgodLjBcBfALMamn4J+POI2BOYCFyVzzcQ+BZwcETsBcwDTgKIiM9HxIiIGAF8G7g+67wfOADYC9gDeB/wwbU9d+t+kg6h+NmNi4ins/ia2s81Ii4rHX4lcH5EvJdixb0Xsvxq4D3AnsAg4FOYVXBgYNYzPgT8ISIurRVExH9HxLfz4dPAZjmqIGAccEvp2EUR8WhjoxHxYETUrn9emG1sSrGAioC3ZHtbUn2d9LHAD2vNAZsBmwCbAhsDz6/pCVvPkPQB4HvAn0XEr7o4djdgYETMBIiI30XEq3n/5kjAvXQGqWarcWBg1jN2Bx7o4pjpwNHA+/PYFW/yOY4CHoyIFRHxOnACxUjEUop/1HJ5+WBJ/w8YDtwOEBGzgTuAZ3O7NSIWvck+WM/aFPgJcEREPNKw76icNpouqbZ4zq7AbyRdL+lBSefnP/CpyymETwAzerz3tl5yYGDWCyR9V9JDku4rFV9LERiUv8W3297uwHnA3+bjjSkCg32Ad1BMJZzRUG0CMD0iVmWdnYH3UnxzHAJ8SNKfvMlTs571OvBfwPEN5T8FhuW00S+AqVk+EPgAcDLF1NA7gU821L2YYorrP3uoz7aec2Bg1jMWAvvWHkTEicAhwHalsuco/vAfCtzWbsOShgI3AMeVhpZHZJu/yqHiaylGIsomsHoAciRwdw43/45iKmP/dvthvaID+P/A+ySdWSuMiP+JiNoI0/eAkXl/CcUo0hMRsZIiz6X+PpR0FsV78Au90HdbTzkwMOsZt1PM/59QKtu84rh/Bk6rfYvviqS3ATcBZ0TEL0u7fg3sJqkWeBwKLCrVezewNTC7VOdp4IOSBuaIwwfLdWzdkDkCHwE+XruSRdLg0iEfpfPndh+wdel98CHyX/vmlQtjgWMjoqM3+m7rJy+JbNYDIiIkHQFcKOlU4EXg98BpDcf9V1V9SUdSZKFvB9wkaW5EjKW40mBn4IuSvpiHj4mIpZL+BZgl6XXgv1l9CPlYYFqsvtTpdIoPjvkUiYgzIuKna3Ha1kMiYpmkcRQ/35eA/SV9FFgJLCN/1hGxKi95vS2TUO+nGFEAuJTifTE7r0q9PiK+1LtnYusDL4lsZmZmdZ5KMDMzszoHBmZmZlbnwMDMzMzqHBiYmZlZnQMDMzMzq3NgYGZmZnUODMzMzKzu/wD1hQ7EdEdY1QAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "t_df = sg.gen_report('ENSMUSG00000020167.13', 'test', groupby='sample')\n",
    "sns.heatmap(data=t_df, cmap='viridis')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "minute-brisbane",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GM12878</th>\n",
       "      <th>K562</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008451</th>\n",
       "      <td>18.124219</td>\n",
       "      <td>18.124219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000105344.6</th>\n",
       "      <td>16.124234</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008452</th>\n",
       "      <td>16.124234</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          GM12878       K562\n",
       "tid                                         \n",
       "ENCODE-mouseT000008451  18.124219  18.124219\n",
       "ENSMUST00000105344.6    16.124234   0.000000\n",
       "ENCODE-mouseT000008452  16.124234   0.000000"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cooked-georgia",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                              GM12878           K562\n",
      "tid                                                 \n",
      "ENCODE-mouseT000008451  285714.285714  285714.285714\n",
      "ENSMUST00000105344.6     71428.571429       0.000000\n",
      "ENCODE-mouseT000008452   71428.571429       0.000000\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "tid\n",
       "ENCODE-mouseT000008451    True\n",
       "ENSMUST00000105344.6      True\n",
       "ENCODE-mouseT000008452    True\n",
       "dtype: bool"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(t_df)\n",
    "t_df.any(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "golden-mercy",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0sAAACACAYAAADag/KcAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAgPklEQVR4nO3debgkVZnn8e+vCgFlX10ailJBXEBAQRGXKexWUWZsaURxQMUNGqcdGVtnFLdCpbVFHW1tF1REWRQVkRbEDcQVh1UUl2ERik2UrdgLpOrtP05c77lJRGbkEhmZ9/4+zxPPjcx845wTkedmxpsRcUIRgZmZmZmZmc21qO0GmJmZmZmZTSInS2ZmZmZmZiWcLJmZmZmZmZVwsmRmZmZmZlbCyZKZmZmZmVkJJ0tmZmZmZmYl1mqzcmlRwMzQ5ep36YGfV/nTfZf5wGcrllfV8nXWuaLd+aOKdaiM77f8Ls/XWaZ6Pfuru72Y6lf7ewe7xdRoR1X3qlF+7fiqblUZ3jtqrOvQZ/vrxo2qvuqY8ls4jHLb9fuJOSemRtDckBrrM7K6upltx6i2pSrWbU7MiLbX0P1ZNdpa3ahMth2H+Get839U+Z51LWDwmHrr32/MUB2s3iJD1THOZQfpxE3UUSOm7w+JIWK6xTWynn22dZjyh97JKI+54IJLvxsRe3UrvSmtJkvpQzE1Yc5O15x/lNmDX3NjFpXHzHl+cenzqowvf77est3iytuxqFZ9ddZhNqayTKpisvKpaBvlMQCLKpfP6ovydi+qKHdR1XxlOfl8lshE75g58RUxD4hT+fL5lsn76qKK+Lyb13l+0Zx/kYp665Q553OoYz0r4yqer9WOiuf7jKlsQ9U2qohfNHeV5+z81lp+iLZWb6PyndE5n06V5ZfvXHau59y6o/T5uXWUb5c6MZXL1mpD1fr0V9cDX6uzfNX70F+MqupSVTnl26JOfLc2aVHv5VW1Pot6x2iYmEWUz1fEpEZV1ZHHVJXbO6be81X/iBUxFf9gqvrH6+jEqvUBWj6vGjHV/9zDlLOod0xnQlRnmVrlVsX0+XytcmrEqOL52m0qLysq21GnTYMvG7WW7baeNbZNxfxai/fcnJb4NDwzMzMzM7MSTpbMzMzMzMxKOFkyMzMzMzMr4WTJzMzMzMyshJMlMzMzMzOzEk6WzMzMzMzMSjhZMjMzMzMzK+FkyczMzMzMrISTJTMzMzMzsxJOlszMzMzMzEo4WTIzMzMzMyvhZMnMzMzMzKyEkyUzMzMzM7MSTpbMzMzMzMxKOFkyMzMzMzMr4WTJzMzMzMyshJMlMzMzMzOzEk6WzMzMzMzMSjhZMjMzMzMzK+FkyczMzMzMrISTJTMzMzMzsxJOlszMzMzMzEo4WTIzMzMzMyvhZMnMzMzMzKyEkyUzMzMzM7MSTpbMzMzMzMxKOFkyMzMzMzMr4WTJzMzMzMyshJMlMzMzMzOzEk6WzMzMzMzMSjhZMjMzMzMzK+FkyczMzMzMrISTJTMzMzMzsxJOlszMzMzMzEo4WTIzMzMzMyvhZMnMzMzMzKyEkyUzMzMzM7MSTpbMzMzMzMxKOFkyMzMzMzMrsVbL9f8G7l8FEPmzUR5cJ8asi82Bm9puhC0o7nM2Tu5vNk7ubzZO67ZVcdvJ0qqI2LXlNtgCIel89zcbJ/c5Gyf3Nxsn9zcbJ0nnt1W3T8MzMzMzMzMr4WTJzMzMzMysRNvJ0tEt128Li/ubjZv7nI2T+5uNk/ubjVNr/U0RHinBzMzMzMysU9tHlszMzMzMzCaSkyUzMzMzM7MSfSdLSl4q6TRJ10q6V9IfJZ0p6bWSRj4cuaT1JL1J0s8k/VnSKkkrJJ0k6Xmjrs8mxzj7m6Qlkg6RdIKkSyTdLuk+STdK+qmk90haMqr6bDK18RlX0oZNJN0gKbJpadP12vi11d8kLZa0n6SvSLpc0p3FZ95lks6Q9L8lbddE3daelvbh9pZ0YtG37iy+V2+S9HNJ/yLp0aOu09pXfMbsIOkgSR+XdI6ku7PvtOUN1TvynKGva5YkbQJ8HXh2l7ALgX0i4upBGlRS5y5FnY/qEnYi8KqIuG8UddpkGGd/k/RN4IWAeoTeC7wjIj40TH02mdr4jKtoxxeAgzqefmREXNVUnTZ+bfW34nv188AuPUI/FhGHjapea9e4+5ukzYGvAct6hN4HvDMiPjhsnTY5JJ0M/EOXkCMiYvmI62wkZ6j9C4KktYFTgWcWT11DGpnicmAr4NXA44AnAWdIelpE3F63/Io6twHOAB5aPHUucDzpjtE7AgcDmwH/HVgDvHyY+mxytNDfdmA2UToP+CFwKXB7Ud++wNOBdYCjJK0TEUcOUZ9NmDY+4yra8RxSorSGtBPR2l3LrTlt9TdJe5C+VzcsnjoP+BZwZfF466LeFwxbl02Ocfe34gjVGcDMTWtXAccBvwRuJfWz/wY8A1gb+FdJd0bEJwet0ybO4o7HtwA3A40csW40Z4iIWhPwRiCK6QJgk47X1wW+k8UcVbfsLnWekpX3eWBRx+tLgBVZzN7D1ulpMqZx9zfgEuATwPZdYg7L6vsL8Ji2t5On0U1tfMaVtOEhwB+K8v8NuCqrb2nb28jTSN/rNr5TH0racQjgHmD/LrGLgYe3vZ08jWZq4Tv15VlZVwPbVMS9Lou7EVir7W3laTQTcDjwfuDFpDMjIP0QOPN+Lx9xfY3lDHUbsBbw56LwNcATKuK2BO4s4lYBmw2x0jtlK7QCWLci7gVZ3Hltdw5Pw08t9bdNasZ9PetvR7S9rTyNZmqjz1WU/5Gi7GuBDXCyNC+ntvob8JWsP7207e3gaTxTS9+px2d97fU9Ys/PYndse3t5am5qKllqOmeoO8DDs4EtivkzI+I3ZUER8efiwxjS6Up/X7P8Mi/N5o+OiFUVcWeQDiMD7OoLBeeFsfe3iLi1ZujXsvkdB63PJk4bn3FzSHoK6ddfgDdExB2jKtsmztj7W3GKyn7Fwx9HxEmDlmVTp43Pty2z+ct6xF6aza83RJ22cDWaM9RNlp6bzX+nR2z++l41yx+4zkip4nezpzw63vRro7/Vle/APngM9dl4tNrnJD0I+BzpM/nUiDhlFOXaxGqjv72S2e/8zw1Rjk2fNvrbn7L5XteozLy+mrmJk1ldjeYMdZOlHbL5C3rEnl+xXG2SFpEuNAS4H7i46Tptooy1v/Upr2PFGOqz8Wi7z72VdKTyTuCfRlSmTa42+tuzsvmzJG0m6QhJv86GDf9NMcTvY4aoxyZPG/3t1Gz+rcWRzQeQ9FpmB4E4LiJuGaJOW4DGkTPUHQ0v/+C8qkfstaRfBxYD20lSkcn1YyvShc4A10XE/T3i851Wf8hPv3H3t1qK0X1elT11ehP1WCta63OSHge8vXj49oi4dtCybGq00d9mdkhvA7YBTgYe1hHz+GI6VNLbIuKoAeqxydNGfzuZdMH9PqSR734v6UvMHQ3vhaTR8Chi3zBAPWaN5wx1k6WNs/mbugVGxP2Sbgc2Kcpfj/RraT9q11e4uWJZm04bZ/Pj6G91vRl4bDH/a5wszScbZ/Nj63PFL2KfI10fcD5pREab/zbO5hvvb5LWATYqHq4GTivK+wNwDOkc/k1I16jsRdpR/qCk+yLiY/3UZRNp42x+LJ9vERGS9gPeQ0qCNiAN3dzpQuBdwLeb+qHT5r2Ns/lGcoa6p+Gtn81XXTSVuyeb36BmHW3WZ5Nl4t5/SXsC7y0e3g8cEhFrmqjLWtFWn3s9sAdpB/Zg96kFY9z9bZNsftPi8XdIo6IdGREnRcSnI+L5wD9msR+UtNUA9dlkaeXzLSJWA0cBR5Bu6F7mSaTTkJ82aD224DXev+smS23yLw3WKkmPJY2CN3Mk9vCIOKfFJtk8IGkJ6R4UAB+NiIvabI/Na53f9XcAB5aNGBURnyGdQgXpZqGHNtw2m6ck7UU67e9DwDmki+k3Jh1J3450H567SKfinSlpZKOL2oLVSM5QN1nKD8HWuZt8PkrYIMPf5vXVGXFs2Ppssoy7v1WS9Ejg+6S7PkPaqfV5/PNPG33uU6RfxFYA7x6wDJtO4+5vnct8IyJuLo1Mjs7m/26A+myyjP3zrUiUTied/vl14G8j4nsRcVtE3BcRl0fE+0nDmt9btOs4SZ3X0Zn10njOUDdZWpnNb1YVBH+9CH7D4uH9pF8N+lW7vpKYlVVBNjVWZvPj6G9VZW8NnEW6eBDg0xHxv0ZVvk2Uldl8431O0gGkm+MB/I+IGFm/tamwMpsfx2fcncWyMy7sEZ+PmOZ7F06/ldn8uL5TP0zax1wDvLHqFOOIOBc4tni4AemmpWb9WJnNN5Iz1B3g4VLgkcX8UroPmbwV6eJQgMsGvGDvWuBu0ugWW0laq8foFvmQlB6jf/qNu789gKRHkBKlpcVTx5CuL7H5adx97rXF3xuAXSTtUhG3UTb/T5JWFvOfjYg/lcTbdBhrfysutr+M2eF1b+uxSP76RpVRNi3G2t+KMzIeXzz8bURc32ORHwCHFPNP6bc+W/AazxnqJkuXMHvjpl2BH3WJ3TWbv6Rm+XNExBpJvwOeTGrjTnS/N8DQddpEGWt/61ScBnAWsG3x1AnA6zxSz7w27j6n4u/DmB04pJd/zuZPY+5NH226tPEZ9ytmk6UNuwUyN0HqlVjZ5Bt3f3tENn97jfi8j603YJ22QI0jZ6h7Gl4/d7vN7/jc607RQ9cpSR2vf7cq1qZGG/0NAElbAGcC2xdPnQS80qOUzXut9TlbkNrob2dk80/uEZu/7rM1pt+4+1ueIG1dIz7/pb/btXRmVRrNGeomSz8Ebizm/07SEyoasSWwf/FwFXPv4Nyvr2bzh0iquijx+cweATg/Iq4Yok6bDG30NyRtSjodYOb0gW+QRoxaPUy5NhXG2uciYllEqNfE3NNlHpm99stB6rWJ0cZn3KmkU1UA9pHU7dz+/H44/kFg+o27v13O7BDOW0vao0f8/tn8+QPWaQtbozlDrWSpOPfvyOKhgC9Jyu/bQNGwLzJ7CPUTVaPtSDpWUhTT8oo6Lwa+WTxcAnyiuIFjXs4S0ohSM0rLsunSRn+TtBHwPeCJxVP/Aexf407QNg+00eds4WrpO3Ul6aJ7SKfhHVe2QyHpYGDf4uFdzP2OtSk07v4WEfcwN9H6YrG/VlbW4cDfFg/vZe5Or9lE5Ax1r1miqGBf4Jmkm4hdLOkzpF8QtgJew+z50L8F3tdH2VUOI92o7KFF+TtIOo50mHZH0gWBM7+OnRARp4+gTpsM4+5vZzB76sn1pOuU9k5HbCvdHRHfG7JemxxtfMbZwtVGf/sA6ZfVXYu/l0g6pqhzE+Dvi+dnHBIRNz6gFJtG4+5vhwPPId0EeVtSXzse+AXppqBLgP2Ap2bLHBER1w5Zr02IYqCP13Q8/cRs/tnF6Iu5k4e45+BhNJUzRETtifRheibppk9V0wXAkh7lHJvFL+8RuwtwRY86TwDW7mddPE3+NM7+1qOOqumqtreRp+ntczXbc1VWztK2t4+n0U4tfaduQTotq1udd5Ou1Wx9G3ma3v5GSsourfFd+hfg7W1vH08j72/LBtivOmjQ/lbENpIz1L1mCYCIuJV0g7r9STcbux64jzQq01mk85yfGhFX91NujzovImWi/0y6A/RNpEO11wBfA54fEQdExH2jqtMmQxv9zRY29zkbp5a+U28k3Qj0JaRTpa4p6rwNuAh4P7BtRHxxVHXaZBh3f4uIC0n7b68ETiH9+HM36f5NN5OOMn0A2D4ijqwoxqy2pnIGFZmYmZmZmZmZZfo6smRmZmZmZrZQOFkyMzMzMzMr4WTJzMzMzMyshJMlMzMzMzOzEk6WzMzMzMzMSjhZMjMzMzMzK+FkyczMzMzMrISTJTMzMzMzsxJOlszMzMzMzEo4WTIzMzMzMyvhZMnMzMzMzKyEkyUzW7AkLZUUxXRs2+2ZZJIWSzpU0o8l3Sxpdbbtdu6jnKnb5qNad1uYJO1T9JVVkv6mwXpeXtSzUtKWTdVjttCs1XYDzOYLSdHl5TuBPwEXAt8ATo6Iv4ylYQuIpKXAQcXDsyPi7NYaM49IWgScCuzddlvGbSGvuw1P0rrAR4qHR0fEdQ1WdyLwDuAxwPuB1zRYl9mC4WTJbDzWL6ZHA/sBv5b04oi4tN1mzTtLgXdnj89upxnzzouYTRZWAP8OXAXMJPxXjr9JY/MiFu662/BeT/pcWgV8oMmKImK1pPcBXwIOkvShiPhdk3WaLQROlsyasU/H402APYADgAcDOwLfl7RLRNwy7sZZEhFXAWq7HVPgBdn8/hHxi0ELmsJtPrJ1t4VF0oOBtxYPj42I68dQ7YnAe4FtSD8c7T+GOs3mNV+zZNaAiPhmx/SFiHgdsBtwUxG2BPg/7bXSrLats/mLWmtFOxbyuttwXgFsUcx/aRwVRsRq4ITi4YslLRlHvWbzmZMlszGKiN8Ah2dPvbittpj1YZ2ZmYi4t82GtGAhr7sN59Di7xURcc4Y6z2++LsYOHiM9ZrNS06WzMbv9Gz+UZIeMvNA0rJslK3lxXPbS/qopN9Jur147aDOQiU9TtLHJF0i6TZJ90haIemrkjpPC+xctqzeHSUdLemKoqwbJf1A0svqrugwbep3exQDbPwwW/zd2bJ/nTrKrz0yWwPbd4mkD0v6vaS7ihGsfi7p9ZJGeor0IG3P2wz8l+z5zm26vM+21NrmSiPQvVzStyRdozSS2D3F/IVF3/yH/P9nFOs97LqPs893iX+MpH+XdJmkuyVdX2zHPUrq21vSadk2XiHpk5Ie1qutNdZleda2ZcPGjqhPrC/pMEnfL7bLvZJukXSepPdI2qJq2X5I2hHYqXh4Yo/YV5X0rX6ml+TlFdcp/bJ4eICkaTrt1WzyRIQnT55GMAExM/WIe1AeCzwie21Z9vxy0mkcd3fEB3BQR5lHAPeXxOXT2cBmFW3qrPflpAuSq8o6DVi3x3oO1aZ+t0ePekrfG9KF1zOvHdvUupSsx17ArV3K+h6wzoj65UBt72hzt2l5n+3puc2BzYFza9b/olG/Z4Ou+7D9ZJDPgJL4fYG7KupeA7wq+xw6pks7/whsO2TfW56Vt2yY2BH1ieeTRiXttuztwAtH8H/3zqzMPXvEHl1zvaqmrUvK/Ej2+s6j+Czx5GmhTh7gwWz8On+5vL0i7unA24HVwOeBn5ESmO2BG2aCJL2f2YuIVwNfAc4C7iENJPFq4KGkX8jPkrR7RNzTpX27MXuq4DHAj4tydyMNRbseaXSw46k4jbCBNvXaHveTBtXYgXRxM8BJRb1DaWBddgbeQhrk4DPAOcC9wK7AP5K273OKdX1Xi22/hNmBSt4HPKGY7zw68vth2ljhs6T+BnA58GXg0qLdG5Le82cBTy1beATvWd/r3kKfv6Ek/klFG+4DPgqcTzqDZC/gZaQ+91lJPwHeCLwK+BXpf3lF0b6DSf9HDwOOBZ7Ro43jMmyf2Jf0mbCYtD1PA84kbccNgD2Blxbzp0h6TkScNUR7n1P8XUN6H7r5IvBtZvvaGtKoqWtKYr8MrAtcAbwZWB0R15TE5QORPI/ZI01m1q+2szVPnubLRPZLX4+412WxV3a8toy5vxj+EXh8l7KeRvpCDdK9nJ5VErMpcF5W5lElMZ313g7sXhK3HXBdFrdvU20acHvk8ctrvGdLs/hjx7h9VwDblcQ9hTQkdQC3MMTRpRG/D2fX6ds129Vrm29J2pmNom3rdSlrG2Cbpta77rpPUJ8PUiKxpCTu8CzmgqK9nwQWdcQ9hJRAzcQ+ZYj3enlWzrJBY0fQJ7YGbiuWvwHYrWLZ3YCVRdw1wIMGXO/FzB7du6SP5W4olvl9xetLsm10fI+ytslivzHoe+jJk6fwNUtm4yTpscCR2VNf77HIIRHx2y6vzxyhAHhLRPy4MyDS0OQvJp3KA3CopI171PuWKBkiOSIuY+6NDt88xjZB7+0xak2ty4HFtuws61zSr9+Qhpt/yiCNLjT5PjTpUcxeT3tiRNxVFRgRKyJiRcfTbaz3JPX5AyPi6pLnPwzcUcw/iXT07H9GxJyjFxFxN3PvB/S8Pupuyij6xIbF/H4RcV7FsucBbyoebkU6ujNoe2eum/r/dRaQtCXpyB7AxRVhO2fzv+pWXrENZo5cPrFOG8ysnJMlswZIelHHdJCko0m/6M6chnc98MEuxawAvtWljnWYvQfMzaTTdEoVX5xfLh6uBzy3S723Al/oUtZ3gJmdt93zC8EbbBP02B6j1uC6XBQRP+nyen7qz+NrNPUBGn4fmnZ3Nv+EyqgSbaz3hPX5C8p+5Cjqvpe5p4N9JiLuryjnp9n8QH1wxIbpEyLd3w7g3B7/e5B+rJjZLoP+L2yTzde9j95O2XxVIrRLjZjcrcXfrT3Ig9ngnCyZNeOUjukLpNPvZn5t/B3w3Ii4sUsZP42I6PL6TswOa3x2RNzXo03fy+ZLz+sv/KRGWfkO/W7ZfFNtgt7bY9SaWpdeNzW9LpvfpEdslSbfh6b9hvRDAsBrJH1e0u6S6nxftbHek9Tn/1+P1/+UzZ9bM27QPjhKw/SJJ5BOgQS4peSHrDkT6VqjO4v4xw3Y3k2z+brJUn70p86RpaqY3M3F37VJybmZDcADPJiNx13An0k3tTwF+Fr0vmfLdT1ef3g2f2mNNuQxD6+MStc89JLHPGIMbYLe22PUmlqXm7q8Bmmwhxnr1qi3TJPvQ6MiYrWkQ4CTSTt5ry6mlZLOIR31+G5EXFCyeBvrPUl9/uYer+d9qzI2Iu7NDkQM2gdHZsg+sTSb36uY6ho0UVwnm7+jMmqufo4s3RQRf6xRZj540IOZTQLNrA8+smTWgIhQx7R+RDwqIvaNiONrJEowe755lQ2y+cpz+DP5F+UGlVFzT3mpkte3/hjaBL23x6g1tS5lI1yNWpPvQ+Mi4jTS9VrfJA14AbAxaejnI4HzJf1aUueObxvrPUl9vp++NY5+ODJD9ImNhqh27QGXyz/fN6yMmmvmyNLKsmvOiuvbZk7vq3MKHsxd93F/fprNG06WzKZX/otlnVMs8qSm26+dlTd0rKgv3/lrqk1tmOZ1mea2AxARF0fEPsBmpB3i9wI/YnZHeQfg25IOyBZrY72nfltPiJ77IwP2ifzzaXnJD1ndpqUDrkt+6t2mlVEFSQ9i9pS/qtPr8uuV6pyCl9d9H/USeTMr4WTJbHrlp2FsVyM+j7m+Mgq2rVFWHpOX1VSb2jDN6zLNbZ8jIu6IiO9ExLsiYhnp1LX/W7ws4COSFheP21jvebOtG5AfYel1lGbzuoX22SfyUxn7GhxiCFdl8z2TJVKiNLN9qo4a7ZzN1z2yNFP31WO+3tNsXnGyZDa9LmZ2Z2RZ8etkN/nITt0u7n6GpF47Nntm8/kwvE21qa781KJhR39qe12GMc1t7yoibo6INzE7stuWzCYgbaz3vN3WI7Aym39EVVBh4IFFevSJi5i9due5ksYx0MGVzB7J2b5GfJ3BHfoaCU/SUmavN6ubXJlZCSdLZlOquO7p9OLh5sBBVbGStgZeVjy8i7kjcnXaFHhll7Key+wvtOdExA1jaFNd+Sk3Q+0UTcC6DGya296Hq7L5taCd9V4g23pQ+f2hnl0VJOnppHs/DeuqbH6mT6wGTiie24h0c95GFXXODDbxWEm9rluqM7jDTEK1mrnbtUqefPYaJdHMunCyZDbdjmL2aMqHi52OOSRtQrr57Uzy8KmIWNmj3A9J2q3zSUmPBo7JnvrwGNtUx5XZ/Ch2vtpcl2FNZdslPU/SGyVVXpgvaVvSEM+QEuQrspfbWO+p3NZj8Atmjy7tL+nJnQHFZ8rx3QoZQZ/4l6wdb5P05m7DjkvaQtI7JA1zM9fvF38XAbv2iJ2pZw3pZsFlHlP8vToiVtWoP0+Wvlsj3swqeOhwsykWEb+Q9K/A20gja/1I0pdJ90G6h3TB82uZvTP8r4B39Sj226Sdjp9J+iLwE9KvmbsBr2H2AvVvRMTJY2pTLRFxq6SLSKes7Cnp08CZZBfSFzfVrVtea+syrClu+8OBjwIflPRD0q/ifyCN0rg5qR++hNmk46MR8deRvtpY7yne1o0qhiD/OPBO4EHA2cX/5Pmk4bWfBryCdMrsfwAvrChq2D5xraT9izrWJiW3B0s6mXTPu7tJo9ZtB+wOPBNYDJw9xOqfQhqAAmAZc+9N12nmyNJlebtnSHowaehvgPUlrVNjRNWZU6WvjIi6A0KYWZmI8OTJ0wgmIGamIcpYlpWzvI/l3kO663x0mc4GNqtTL3AgsKpLWacD6zbZpkG3B2mUrMp6O2KXZq8dO67t20QfaPh9OHvYvl13m5N2nru1dWZaQ9qBXtTUeve77m30+T771rFZ7NIesX9t75Dv9zqk0w2rtsdtwH8lfe7MPLesoT6xO+mIU52y7gB2HHLdLyrKurxLzJZZnSdVxIh03dVM3DXAD7qU+bgs9r3D/s968rTQJ5+GZzYPRMS7SKdyfJx0PvsdpIvOryXdyHHfiFgWEb1uWjlT3vGkX2s/R/oFdxVpONyzgAMiYu/ocSrIqNtUV0ScATwdOJF0Wt7Q9xdpa11GYQrbfhzpl/Y3AaeSboB8F+no5m3AL4FPAE+OiMMiovR+QW2s9xRu68ZFOgLyAuBQ4Oeknf5VpPf134CdI91DqZtR9YlfkAZcOBD4Kunz4U5SgnsL6YjXZ4GXAg+LiF8PtNKzPln8fbSkPSpidsrmS69XiogA3kDqR/eTjo51O1p0YPF3NWl9zGwISv+DZraQSVoG/LB4eERELG+tMWZm84CkdYEVpKNHR0fEIWOoczEpmVxKOlK1f9N1ms13PrJkZmZmNmLF0fcPFA9fIanX8Omj8DJSorQGOGIM9ZnNe06WzMzMzJrxKdLpfuuSBgBpTHFU6R3Fw2Mj4ndN1me2UDhZMjMzM2tAcXTpTcXD10n6mwarexnpmqzbaDgxM1tInCyZmZmZNSQivhkRioh1I+K6Bus5vqhn44j4c1P1mC00TpbMzMzMzMxKeDQ8MzMzMzOzEj6yZGZmZmZmVsLJkpmZmZmZWQknS2ZmZmZmZiWcLJmZmZmZmZVwsmRmZmZmZlbCyZKZmZmZmVkJJ0tmZmZmZmYlnCyZmZmZmZmV+E8kSYiX/ru+rQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1008x108 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib as mpl\n",
    "\n",
    "t_df = sg.gen_report('ENSMUSG00000020167.13', 'test', layer='pi', groupby='sample')\n",
    "\n",
    "# create a colorbar between 0 and 1\n",
    "plt.rcParams.update({'font.size': 30})\n",
    "fig, ax = plt.subplots(figsize=(14, 1.5))\n",
    "fig.subplots_adjust(bottom=0.5)\n",
    "fig.patch.set_visible(False)\n",
    "ax.patch.set_visible(False)\n",
    "\n",
    "cmap = 'magma'\n",
    "cmap = plt.get_cmap(cmap)\n",
    "# try:\n",
    "#     cmap = plt.get_cmap(cmap)\n",
    "# except:\n",
    "#     raise ValueError('Colormap {} not found'.format(cmap))\n",
    "\n",
    "norm = mpl.colors.Normalize(vmin=0, vmax=1)\n",
    "cb = mpl.colorbar.ColorbarBase(ax,\n",
    "                    cmap=cmap,\n",
    "                    norm=norm,\n",
    "                    orientation='horizontal')\n",
    "cb.set_label('Proportion of isoform use (' +'$\\pi$'+')')\n",
    "# plt.savefig(gid_prefix+'_colorbar_scale.png', format='png', dpi=200)\n",
    "plt.show()\n",
    "# plt.clf()\n",
    "# plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "vulnerable-flooring",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAzoAAAB8CAYAAAChWMVYAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAf7klEQVR4nO3debgsRZ3m8e97F3aVTQRFRAVEWRRUQARlEwHRBm1BFPFiq0jbM21r60y39njt1nFGH5exaVFo4QBuoA1uyKIiKGiDXDYBZe2LoCCLIJflsv7mj4iy8hSVVVlVWafOyfN+nqeeysqKjIiTWXGyfpURkYoIzMzMzMzMmmTBpCtgZmZmZmZWNwc6ZmZmZmbWOA50zMzMzMyscRzomJmZmZlZ4zjQMTMzMzOzxnGgY2ZmZmZmjbNokoWvrw3jYR7unUhqL5amKX3RN8/yNAPmOVQ9KrxRqehB05fVYZQ6Txej1Gng8gYra9qE6qPUbZjtC5+9ShO7j3Q8u+tZ7piPVWnZY/+MlKvt8zCGbcfyGRlTXjHBYzhaPoW9PMuOf8VPwEingLINVLXsVvpKBbfzHP30171+1U7vJduWvhhw29I8B9u2Z3kVPjPldRosz2p/c4VyS/7+ofIatE4V/rbayirLZ0x/f6XP3qBlTWuH3XMatNzqeXX/3r/siofOioh9emTf1UQDnYd5mB21Z3qh9sUlLSju4f7rmba+sIMWLOi6ntL1hW0r5P+EllOWb9lyab4Dll0hz5iWhq5pypZLtx1mm7I0KklTum1ZHSik6V6HsnJL83zCe4XlBWV16p9vbekXlKwfMJ/qdaqyvso+rqus4fOEHoHOuMuegb+taDz1iAppBqtDaZ5FpXlGaboq5VXKt+wUUJJm+jEpy2fQ9Uyn4he67l/uqpQx/V9G/zwXlKYZLP2Cki+kCzqO57RtyvIqzbckTZVlxrHt433T99pmYWkZ3fNdWCXNtDzL8inbtrCekvXF+j/h7+z+3vT1/etUqR5034/T8+9ebtk+WlhWz5I6pHRV8ur++ZlWp9I0Zce/rA6F9dPSqCRNcb26pn/Ce9O2WVCSZkHXNAs3um59huCua2ZmZmZm1jgOdMzMzMzMrHEc6JiZmZmZWeM40DEzMzMzs8ZxoGNmZmZmZo3jQMfMzMzMzBrHgY6ZmZmZmTWOAx0zMzMzM2scBzpmZmZmZtY4DnTMzMzMzKxxHOiYmZmZmVnjONAxMzMzM7PGcaBjZmZmZmaN40DHzMzMzMwax4GOmZmZmZk1jgMdMzMzMzNrHAc6ZmZmZmbWOA50zMzMzMyscRzomJmZmZlZ4zjQMTMzMzOzxnGgY2ZmZmZmjeNAx8zMzMzMGseBjpmZmZmZNY4DHTMzMzMzaxwHOmZmZmZm1jgOdMzMzMzMrHEc6JiZmZmZWeM40DEzMzMzs8ZxoGNmZmZmZo3jQMfMzMzMzBrHgY6ZmZmZmTWOAx0zMzMzM2scBzpmZmZmZtY4DnTMzMzMzKxxHOiYmZmZmVnjONAxMzMzM7PGcaBjZmZmZmaN40DHzMzMzMwax4GOmZmZmZk1jiJicoVLVwIrJ1YBG9T6wJ2TroRV5uM1d/hYzS0+XnOLj9fc4uM1d8zksbozIvYZdKNF46jJAFZGxEsmXAerSNLFPl5zh4/X3OFjNbf4eM0tPl5zi4/X3DEXjpW7rpmZmZmZWeM40DEzMzMzs8aZdKBzzITLt8H4eM0tPl5zh4/V3OLjNbf4eM0tPl5zx6w/VhOdjMDMzMzMzGwcJn1Fx8zMzMzMrHYOdMzMzMzMrHEGDnSUHCzp+5JukfSQpFsl/VjSOyTVPmW1pDUlvU/SBZJul7RS0k2STpb06rrLawJJT5F0kKSjJV0o6S5Jj0i6W9Llkr4g6aU1lhcDPKbqKrcpJJ07wP5bXmO5blsDkrR0wM/7yJ97t6/uJC2UtLWkJZL+VdIvJD1Q2BdLh8hzn/z5vym3h9tz+/g7SWuO4W9YLOmd+Rx6az6n3iLpe/lcq7rLnJS6jpek1SW9VtJnJZ2fj9HDku6V9GtJx0vas+a6Lx+gDZ5bZ9mTUOOxmhrk/1fNf4Pb1gDHK287zLnt3BHrPta2NVBQImkd4FvAHh1vbZgfewBHSjowIn47aGVKytwul/mcjrc2yY+DJH0NODwiHq6jzLlO0geBfwZW7fL22vmxLelYfQU4IiIemLEK2qzgtjXjbpx0BRroFOD1dWQkaVXgeOCQjreemh87A++R9PqIuKKmMjcFTgW263jrGfmxP/AOSW+MiHvqKHPCRj5ekt4CfBFYq8vbi4Et82OJpDOBwyLijlHKnKdqa1uT4LY1o2b1ua1yoCNpFeA7wK551c2k2RauBzYG3g48H9geOEPSyyLi3lEqJ+lZwBnA0/Kqi4CvkO7Cug3wLmA94M3A48BbRymvQbagHeTcCPwIuIy039YB9gTeACwEDgU2kLRvRDxeQ9lXAR/uk6aWILjBDuzz/shBqdvWSL5Bak/9PAWYyssBnFBD2W5f0y3seP1H4C5g8yHyOgE4OC/fRTq//Yp05+9DgR2A5wJnStoxIm4eqsaZpLVJbXDLvOrXwHHALcBmpDb4TGAv4FRJe0fEo6OUOQvUcbyeTTvIuRX4IfBL4HZgTdJ3lEOA1YB9gB/l7yN1/Zh3B+nY9DJTd4ofpzrbVssRpOM0Vm5bwHDH6xz6f/+A1BvsK8Dq+fXxA5TRy3jaVkRUegB/SzpZB7AMWKfj/dWAMwtpPlU17x5lnlbI78vAgo73NwFuKqR5zahlNuEBHAt8H3gleWa9Lml2BVYU9t3hI5bZyufcSf/9c/EBnNvahzNUntvW+Pfxuwv774cj5uX21X2//CPwCeAvgWfndUsK+2tpxXz+orDNTcAmHe8vIH1RaqX5Zg11/2whvzOA1TreXxe4pJDmPZPe37PheJEC/fNJv8gvLEnzAuD3hXw/WkPdl+e8lk96P86VY5W3mSpss+kM1d1ta8jjVbGsfQr5XldDfmNtW1UrsYgUhQfp192tStJtANyX060E1hvhD39hx4lntZJ0+xXS/XLSH7bZ8KAjCO2R7m8K++68Ecv0F7HR9t+5rX04A2W5bc3MMb2osP/ePGJebl/V99UwX8YuLWyzX0ma1Zke/G89Qh03AB7K+dwHbFCSbut8zg3gNkq+2M/lx6DHa4Dz2/7F/3M11HOsX8bmwmPItjVV2GbTGaij29YIx6tivqcU8v3HGvIba9uqOhnBHqQ+ygA/joiruiWKiNtJ3TogdZ36i4r5d3NwYfmYiFhZku4MUvc5gJdIeu4IZTZCRNxdMek3C8vbjKMuNiu5bY2ZpK2A1mQf95D6itssJGlz4EX55XUR8YNu6SLiQdLV8paDRij2AGCVvPz1fO7sVuaVpO4kkLqZvnKEMhthgPPbGcD9eXkTSU8eU5VsdjkAt62xkbQu8Lr88jHq6ZI9VlUDnb0Ly2f2SVt8f5/BqjN4mZHCwbMKqzxTVHUrCsurl6aypnHbGr+3F5a/3iOYtMkrfq7PKk2VzOj5reYy55WIeIzp4xl9jpsf3LbG6y20x4CfHRG/m2Rlqqga6GxdWF7WJ+3FJdtVJmkBaWIDgEeBy8dd5jxV3Fc31ZTn8/J0n3fl6T7/IOmnkj4iacOaymg0SafnqTAfzvvxsjxd5ItqyNtta8yUptg/tLDquBqzd/uq3yDnt8tIv2ICvGCE6Wln9Jw6H0nagHZPlAdIA53rsJ6kH6k9pfUdki6S9H991buvYyX9Nk/zfI+kqyUdK+kVNZbhtjVehxeW6zy3wZjaVtVAZ4vC8vI+aW+hfSLYfMgTwcbAGnn5d9F/Nozil/QtSlNZp+LsFqfXlOeGwMtJg/0Wk/rL7gosBZZL+tuaymmy/Uj7cTFpP76QNJ7qUknHSRrll0m3rfHbn/S5B/hVRFzcK/GA3L7qV/n8lttL6xfMNUnT1A4k/9jQOmk/Rjpn9uI2OJzi+e3MqGdWUUgzvu1JCqIWk2bleynwQeAaSZ+Q1DkDliV7kWY7W4U0K+XzgXcA5yndm3HdUTJ32xovSS+kPV33ncB3ay5iLG2r6vTSaxeWe07tFhGPSrqXNI3xItLJ4L4B61W5vOyukm2thKSdaUfmK4HP1ZDtDcDZwBWkY7I6sBVpFpDnkC53fk7SuhHxkRrKa5q7SF1nlpFmDBKwKemL8845zeGk/ub7VAhSulm7sOy2NR7j+sXL7Ws81i4sV20TmxS27fdlqtNatM+991Rox26DA5L0HOAfCqv+T01Z/57U3eky4A+kL+ybke5fsg1pit//CWxEGghuyQrSNOAXkW5N8hjpR7e9aXc1ew0p4Hl5DH9rEret8Sp2yf5q1Ht/vfG1rYozIjxMe4aFRRXS/66QfqMhZmDYubD9+RXSb15If804Zm1o0oP0q/AthX329zXk+Yoe7y0E/qVQXgC7THo/zKYH8DJgcY/3DyQNrG3tvw8NWY7b1niP49OAR/L+eghYv6Z83b6q76slhf2wtEL6awvpN6uQ/oJC+pcNUb+nF7a/pUL6xYX0D016/076eFXIb01Sl9xWnkfVVM9d6JiGv+P9I0ndgVvlvmXS+3Y2HCvgxcCaPd7flfTFtpXvsSPUz21rxOPVI69VSD8EtfLbtsZ6jrVtVe26Nkkx6Qo0iaQ1STd+bXW5OB349Kj5RsRPe7z3WET8E9NnLPrQqGU2SUT8IiIe6fH+acA7C6s+kO/kPlKxI25vT/RW2r8ofi8iarlxoNvXjJnpNlGlPLfTinK3lq8B2+ZVlwB/X0feEXF+9Oj+FhFHA/+rsKrfjX3nhYhYFhH393j/Z6Rf7Vuf88MlDdwltFvWNaWx5HWkm4gDLIuIK+rKeNxtq2qgU+x6tlqF9MUxBCtKU1Urr8p4hFHLmxckrUbqU7lDXnUBcHDkkHkGLKX9j2X3EceazDsR8TXgmvzyKaSxGoNy2xqvcQ7U7Gcpbl/DmOk2MWh5axSW3QZL5PEZU7Snvr0G2DdmdsbDzwB/ystb5i501kdEXEDqlgvpCvWwM3y6bY1PsdvaTJ/bYIS2VTXQuaewvF5ZIvjzjEOt+eofpT2P/SAql9clzT1lieYzSauQ7uWxR151EenGeMMcn6FExO9J3UQgjSfYdKbKbpBzC8vPG2L7ewrLbls1krQj6Y7skPob95uquFZuX0O7p7A8E23iPtK5EWDtCoNr3Qb7yJMefYn2bIc3AHtGyT1UxiUHVf9ZWDXM/+j56tzC8rD7zW1rDCQ9nfZYqpWkq6YzapS2VTXQubawvGmftBuTInJIN18b5mrBLbTnv984B0+9PKuwfG1pqnlK0mLSzUH3zasuBfaJ4Qf8jaI4+G+dCZQ/1426/9y2xqf4i9cJke7jMdPcvgZX+fyW20urW839tGdgqyx30bghv1xIOmf24jbY31Gk2bsgzaS1R0zu/h5ug8MZeb+5bY3N22h/rz8tIu6ZUD2G+oxUDXSuLCy/pE/a4vtXlqbqIX9Yf51fLiJNrzvWMpsqn5i/Tvty/q+AV0X1u0vXzb+gjGak/ee2NR65m9jBhVXHT6gqbl+DG+T89iLaJ/yrR+j2O6Pn1CaT9Dngr/PLW0hBzm8nVyO3wSHVtd/ctuq3pLA8iW5rLUN9RqoGOoPcHb14d9l+d6Uducx8yXqQO1vPG/my7UnAG/Kqq4G9IuKu8q3GWp+NaM9V/zD13aR0PnllYXnYX6Dctur3BtK4KYCfRcR1M10Bt6+hzdrzW81lNo6kTwKte0fdSgpybpxgfVYFdiqs8lWC6uo4t4HbVq0k7UL7vHIT8OMJ1WPotlU10PkJ7bsK7yVpq5KKbAC8Kb9cSZrda1inFJaPyAPpu9mXNNc2wMURcUNJunklD8w8jvbxuJYJ9FnusJR0bxiA82ZyfFATSDoE2DK/XAGcP2RWblv1m+QkBC1LcfsaWA5KL80vN5e0b7d0uZ0UZz48pVu6ir5NCkYBDsnnzm5lbkV7XOVtwHkjlNkokj4GfCC//AMpyJnxHxg6vI/2Dx7XRcT1k6zMXJHv69cKSh5ntB/Vvo3bVp2K57apGZy8qtPQbatSoBPppksfzy8FnChpWv+4fBI4gTSHPaS567teOZA0JSnyY2lJmZeTPrCQbs52VP7yXsxnE+Dowqquec03hYGZh+VV1wO7R8RtQ+TV91jlu9Vu0u29/P5CSf/M9DtVf7ws/Xwj6b/ngey90hwA/Hth1ae7zSbktjXzJG0K7J5friCNh6u6rdvX7PDRwvLRnfs7t49/o32j0G9FRNeuLpKWFo7pVLc0EXEH8IX8ci1gqvMHh3yOPYl28PqxCY37mnUkfZj2FOp3kH7E+82QefU9XpI+JOn5ffJ5N+l+Vi3zvg1KOkzSq/J3krI0uwCn0f6cnxgRN5ekdduaQfl2JAfll0Ga1XCQ7WdF2+o3ELnoaFL3jF2B7YHLJX2J9CV6Y+CvgFZlrwY+NkhFSryXdCPFp+X8t5Z0EmlA0jbAEbT77H01Ik6vocwm+DjtgZmPAJ8Hdujxv6bl7Ih4oF+iLo4E/oekn5OmrL6WNA3g6qRZqA4i3bn9z/WLCP960rYH8P8kXUO6LHwV6TMu0uDo15Ju9NnyE+ATI5b5Xty26rKE9gnzlDFcSXH7KiHp2aTPb9G2heU9uky48R8RcWlxRUR8R9LJpHFWzwKW5fPblaR2cBjtaflvJf26OKqPkrrObEm6enqJpH8nTXCwGakNPjOnPRc4poYyJ6qO4yXpnUz/0nMU6Urc5n2KP3+E+1q9EfiYpEtJv/z/Grib6XdvL/4dXwFOHLKsWaGmtrU9qWvhzZLOIo0RvgN4jPS9ce/8aP3/vAr4uxqq77aVDPy/sMMbScEiwDkRsXy0WpaWMd62NeDdS9chfRGLHo9lwCZ98pkqpO95t1ZgO9IsGr3K/CqwyiB/S5MfpIbba3+VPTYd5liRBoVVyf9+4G8mvX9m24N0daXK/nucdKVujR55uW3N7LET8F+F/bXzgNu7fY22/3cb4v/ckpK8ViVN3NJr2+uBF/ap09JC+qk+aTcl3dSyV5k/BNae9L6eLcero80M8tht2OMFXFaxjEdIQdiiSe/rWXKsPjfAtqcC67ltTe54dcnzp4W0bx6iTrOibQ1yRYeIuFvSXqRfEN9K+qK0Pin6ugr4BnB8pK5utYiISyVtS4rA/xLYHHgScDtpTu3jIsKDyCbr1aSbV76MNGjtqaRfQR8jXSG4ghQgT0XEHydVyVns/cD3SftvW2ADUrtaRPqSey1pPM7xEVHb4Fa3rVrsQXtK4msi4udjKMPtawZExEOkPv0nkKYK34nUFlcA15G6JB4TNV6xi4jludvqEtJ4yq1IPyjeSRo3dBLpKmHUVaYN5a3AK0htcCvS/+f1SN3//0jqxXIe6X/mpKa1no0+RfrxeyfS1Z2nkfbdaqSr0v8F/JzUXe2SOgt22xqNpM1IPbggHatTx1TU2NuWfIzNzMzMzKxpqs66ZmZmZmZmNmc40DEzMzMzs8ZxoGNmZmZmZo3jQMfMzMzMzBrHgY6ZmZmZmTWOAx0zMzMzM2scBzpmZmZmZtY4DnTMzMzMzKxxHOiYmZmZmVnjONAxM5tBkjaVFPkxNen6TJqk7SU9lvfHiyddn9lC0i55nzwiaatJ18fMbC5yoGNmZn1JeoqkgyQdLelCSXflL+F3S7pc0hckvXSIrP+NdC76XkQsy2XtVggGR33sVvgblvdId7+kmyR9R9I7Ja3RY1+c27HtrwbYjwf2qmNLRJwPnAMsAj5fNX8zM2tzoGNmZj1J+iDwB+Bk4N3ADsC6pC/hawPbAkcCF0k6qVeQ0JHv64Gd8sul9dZ6YGsAmwCvA44Brpa0U+9N/mzrAa5GLRmgTh/Jz3tIevUA25mZGekkZWZm1ssWwKp5+UbgR8BlwJ3AOsCewBuAhcChwAaS9o2Ix/vkuzQ/nxkRlxTWXwkc2GO7NwEH5+WTgW/0SHtlyfojgNsLr58EvAQ4jBS8PQs4Q9JLI+L6Hvk/SjqXLgGW9UiHpKcC+3ZsVyoizpd0PrAL8FHgrF7pzcxsOgc6ZmbWTwCnA58CfhoR0fH+MZJ2BX4ArAXsDbwNOL4sw3yFYpv88sRphUXcCXy7x7YvKrz8TUSUpu3h7IhY3rHuJEmfBH4CbE4KeD5OO6jq5kxgf+AQSe+PiId7pD0UWAw8TgpaXlOhnieSAp0dJe0cET+vsI2ZmeGua2Zm1t8HI2L/iDivS5ADQET8DPiHwqolffI8Mj/fB3xn9CrWIyJ+B7ynsOq1klYtSw9M5ef1gNf2yf5t+fnHwC0Vq/RN4KG8/NcVtzEzMxzomJnNSpIWSXqHpB9I+r2kh/IEABdL+pikjSrms0DS4ZJ+IulOSQ9Iul7SFyW9IKdZ2mdg/N0Vq/3NwvI2ZYkkrQfsl1+eFhEPVMx/ppwDPJiXVwee2yPtBcB1efltZYkkbQe8ML+cqlqRiLgHOCO/PFDSWlW3NTOb7xzomJnNMpK2II0tOZY0pmMjYBXSBAAvBj4EXCfpsD75PJnUDes4YDfSVYfWF/cjgGWSDqmx6isKy6v3SLcfqQsXuX6zSkQ8BhSDu7X7bHJCft5X0gYlaZbk5z8Bpw1YpdY+WgPYa8BtzczmLQc6ZmaziKSNgfOB5+VV1wMfJg3AP5L2r/trAlOS3lKSj0hfqF+RV90DfJo0TuRtpCBqAenqwnY1VX/rwvJNPdK9qrB8UU1l10bSQtIkCy339tnkRNK4m0Wk/duZ32LgzfnlKRHxYGeaPv6zsOzZ18zMKvJkBGZms8uxwFPz8reAQyPiocL7X5S0BPgyKVA5WtI5EXFrRz5vB/bIyzcAu0VEcVzIiZK+SJpB7XU11f1dheXTe6TbMT+vAH5dU9l12oP2FamVpGCzVETcLOkc0tWWJcBnOpLsD6yfl6eGqM9lpHE6q9Led2Zm1oev6JiZzRKStgX2yS+XA4d1BDkARMQUcHR++SSmD55veW9h+dCOIKeVzyXA+4avcZuknYHD88uVwOdK0q1KmtEM4LoKU1DPqDz26ajCqu9HxMoKm07l520kbd/x3pL8fO0ws6blmdxuzC+3kuRzt5lZBf5naWY2e7y+sPyvfbo4fZI07XPndkh6Lu1uZBdGRLHrU6evAncNWtGO8jYETqF9TvmniLi5JPkzAeXlP45S7oj2lnRA4XGopM8CV5HuGwTpitM/VczvVNpd3Ja0VuYxO62JF05geK0xQ6uQxmyZmVkf7rpmZjZ77FBYPrtXwoj4raTfAM8HtpT05IhofdF+SSFpz8H+EfGIpAsYsvuapDVJ00M/I686nTQWqMy6heVJBjpf6vP+74BDIuI3VTKLiAclnQK8g/Y9dR4hjdlZRBrDc2KvPPooBqPr5PqZmVkPvqJjZjZ7FH+pv640Vdu1+VnAhoX1Ty8s30h/VdI8gaTVgO/SDtAuAA4uu9dOVrwnzYrSVDPvQdK9bU4n3a/m+fneQIOYys/rk8blQOHeOd26Dw6gOCFCrxntzMws8xUdM7PZ40n5+dFuY3O6uK/LtpBmZGupco+a+yukmUbSKqTuWq0JDy4C9ouIfnkV/64nD1pujZ4dEcvrzDAiLpB0HWkM0hJJNwHb5renRsz+KYXlQWdtMzObl3xFx8xs9mhd4ViUA4l+ijePLF4dKQYba1TIZ83+SdrydMnfJN3jB+BSYJ9C17leit3V1i1NNXe1xuHsB3wgL9/L4PfO6TRbuvyZmc0ZDnTMzGaP4hTRm5ememKaAG4rrP99Yfk5FfKpkgYASYuAr9Me0/Mr4FURcXf5VtP8ljReBZoZ6BTvqfOmvO7kIe6d06m1rx5m+ufEzMxKONAxM5s9ijfPfFVpKkDSM4Et88vfdFxNubiwvHuffBYDO1epXL6R5knAG/Kqq4G9IqLyrG15quTW+KPNmjZVcp5t7pyO1VOj5Jmv7rWC0Sv7jIEyM7OsUScYM7M57tTC8n/Lg/3LfID2//D/KL4RETeQpkkG2FHSTj3yeQvtm1mWygHJcbSvUlwL7BkRt/fbtosL8/OTgBcMsf1sdxTpb7wQ+N4w987psB1pWmlo7zszM+vDgY6Z2SwREVcAZ+SXzwGO7zZWR9Jbad8kdAXwhS7Zfbaw/BVJz+hMIGk74DP96iVJpOmYD8urrgd2j4jbyrfq6YeF5R1KU81REfGdiNgpP4aatrvDjoXls2rIz8xsXvCsa2Zms8u7gEuAp5Kunmwv6QRScLE2aWzMawrpj4yIbmM2jiNdrdkdeC5wpaQv57wXAbvQnvr4u7TH3DzOE32cdH8YgEeAzwM7pPinp7Mjotusbz8gjTVZBdgt19XKtbof3g/8aJIVMTObSxzomJnNIhFxi6RdSMHH84AtSIFGpwdIQc5XS/IJSQcA3wd2JQVJ7+9IthJ4O7AV7UCn271timN4FpMCnSqeDSzvUrc/SvoBcABwoKQ1SgKieU/S2rRntzutwvTdZmaWueuamdksExHXAtsA7yR1VbqNdCXlbtIVmf8NbB4RJ/bJ517SFZO/As7L268EbgCOAV4cEV8H1itsNlNTF7e6261FCnisuzfSvsnq0ZOsiJnZXCNP3mJmNr9JWgZsD/wJWGemZvWSdDnphppnR8SrZ6LMuUbST0lX5C6MiF6TSpiZWQdf0TEzm8ckvYwU5ACcO8NTFy/Nz3tLevEMljsnSHo5KcgB+Mgk62JmNhc50DEzayhJW0tar8f7LyDd/LPlS+OvVVtEnAb8Ir9cOpNlzxEfzc/nRIRnWzMzG5C7rpmZNZSkDwMfAn4M/Jw0McAjwAbAK4ADSZMLAHwrIt44gTpuD/yS9MPbSyPi4j6bzAv5as75wKPAiyLiqj6bmJlZBwc6ZmYNlQOdf6mQ9GRgSUSsHHOVzMzMZowDHTOzhpK0AfBa4NXA1sD6pGmmHwRuBS4ApiLiZ5Oqo5mZ2bg40DEzMzMzs8bxZARmZmZmZtY4DnTMzMzMzKxxHOiYmZmZmVnjONAxMzMzM7PGcaBjZmZmZmaN40DHzMzMzMwa5/8DpR0qWq3BVrQAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1008x108 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "t_df = sg.gen_report('ENSMUSG00000020167.13', 'test', layer='tpm', groupby='sample')\n",
    "t_df = np.log2(t_df+1)\n",
    "\n",
    "max_val = t_df.max().max()\n",
    "min_val = t_df.min().min()\n",
    "\n",
    "plt.rcParams.update({'font.size': 30})\n",
    "fig, ax = plt.subplots(figsize=(14, 1.5))\n",
    "fig.subplots_adjust(bottom=0.5)\n",
    "fig.patch.set_visible(False)\n",
    "ax.patch.set_visible(False)\n",
    "\n",
    "cmap = 'viridis'\n",
    "try:\n",
    "    cmap = plt.get_cmap(cmap)\n",
    "except:\n",
    "    raise ValueError('Colormap {} not found'.format(cmap))\n",
    "\n",
    "norm = mpl.colors.Normalize(vmin=min_val, vmax=max_val)\n",
    "\n",
    "cb = mpl.colorbar.ColorbarBase(ax,\n",
    "                    cmap=cmap,\n",
    "                    norm=norm,\n",
    "                    orientation='horizontal')\n",
    "cb.set_label('log2(TPM)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "conceptual-institute",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                          GM12878      K562\n",
      "tid                                        \n",
      "ENCODE-mouseT000008451  17.253506  18.34661\n",
      "ENSMUST00000105344.6    15.931592   0.00000\n",
      "ENCODE-mouseT000008452  14.931615   0.00000\n",
      "18.346610396681648\n",
      "0.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/anndata/_core/anndata.py:1056: FutureWarning: is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead\n",
      "  if not is_categorical(df_full[k]):\n"
     ]
    }
   ],
   "source": [
    "t_df = sg.gen_report('ENSMUSG00000020167.13', 'test', groupby='sample')\n",
    "\n",
    "# transforming data to plot gene report heatmap\n",
    "t_df = np.log2(t_df+1)\n",
    "print(t_df)\n",
    "\n",
    "max_val = t_df.max().max()\n",
    "min_val = t_df.min().min()\n",
    "print(max_val)\n",
    "print(min_val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "wanted-coast",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:ylabel='tid'>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAfYAAAD4CAYAAAAelsR4AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAlmElEQVR4nO3deZwdVZ338c8XgmyiggKyaXAIKLsQA4o+EpEtw7AoDGEcQYXJiOiMzoMCOiqjMqKoKKIyERBwkEU2UUIgIzJRJyB7FtYICDEIIoggPEC6v88fdRoq17rdN53udPed7/v1qte999SpU6dud/Krs9Rp2SYiIiK6w0ojXYGIiIgYOgnsERERXSSBPSIioosksEdERHSRBPaIiIguMm6kKxDRqbse3DCPcMRSJqzy0pGuQoxSK736bi1vGb2/27yj/3OG4lxDKS32iIiILpIWe0RERINeejvKN9payAnsERERDZ53T0f5RlsgHW31iYiIGBU6bbGPNgnsERERDXrG6JLrCewRERENeklgj4iI6Bo9QxjYJZ0J7AM8YnvrknYBsEXJ8grgj7a3bzj2fuBJoAdYYntif+dKYI+IiGgwxC32s4BTgXP6Emwf3Pde0leBJ/o5frLtRzs5UQJ7REREg+eHcIzd9mxJ45v2SRLwt8A7huJco+3xu4iIiFGhB3e0SZom6cbaNm0ZT/U24GHb97TZb+BqSTd1UnZa7BEREQ16Omyw254OTF+OUx0CnNfP/l1sL5a0HjBL0p22Z7fLnBZ7REREg94Ot+UhaRzwLuCCdnlsLy6vjwCXApP6KzOBPSIiokEP6mhbTu8E7rS9qGmnpDUlrdX3HtgDmN9fgQnsERERDZ63Oto6Iek8YA6whaRFkg4vu6bS0g0vaUNJM8rH9YFfSLoN+BVwhe2Z/Z0rY+wRERENhqA1/gLbh7RJf19D2mJgSnl/L7DdspwrgT0iIqJBb4et8dEmgT0iIqLBULbYV6QE9oiIiAY9Y3QaWgJ7REREg3TFR0REdJHnvPJIV2FQEtgjIiIa9KYrPiIiontk8lxEREQX6XFa7BEREV2jNy32iIiI7vGcx2aIHJu1joiIGGaZPBcREdFFevIce0RERPfIynMRERFdpDez4iMiIrpHWuwRERFd5PksKRsREdE9xuoCNcNWa0k9km6tbceW9Gsl3VjLN1HStbXPkyTNlnSXpDslnS5pjbJvf0lzS/o8SfvXjjtL0n2SbpN0t6RzJG1U239/OaavPqcM17UvD0nfKvW7XdIztfoeKGkdSbMk3VNe164dd5ykheV727OWvmO57oWSTpGkkr6qpAtK+vWSxteOOayc4x5Jh9XSd5N0c6nPLyRt1lL3N5Wf+4G1tPr3Xv+5HyRpgaReSROH/IuMiFhOvaijbbQZzhb7M7a3b7NvPUl7276ynihpfeCHwFTbc0oQejewlqQJwFeA3W3fJ2lTYJake23PLUV83PZF5biPAj+TtLXt58r+ybYfHdrLHFq2jwIogfYn9e9Q0peBn9o+sdwoHQscI2lLYCqwFbAh8F+SNrfdA3wHmAZcB8wA9gKuBA4HHre9maSpwJeAgyWtA3wWmAgYuEnS5bYfL2XtZ/sOSR8C/hV4X6nbyqWMqxouq+l7nw+8C/iPQX9ZERHDKC32ZXMSVVBodRRwtu05AK5cZPth4Gjg323fV/bdB3wR+HhrIeW4k4HfAXt3WilJ42u9BPMlnSvpnZJ+WVqvk0q+dSRdVnoPrpO0bUk/XtLRtfLmlzLXlHRF6U2YL+ngsn9HSf8t6SZJV0naYIAq7gecXd6fDexfSz/f9rPle1kITCrlvcz2HNsGzmk5pq+si4Ddyg3RnsAs24+VYD6L6mYAqkD/svL+5cDiWt0+AlwMPDLANVQF2XfYvquTvBERI6GHlTraRpvhrNHqWror/uDavjnAs5ImtxyzNXBTm/K2ath3Y0lv52bg9bXPP6vV52NtjtkM+AawbTn274C3Ut1YfLLk+TfgFtvblrRz+qkDVIFxse3tbG8NzJS0CvBN4EDbOwJnAicMUM76th8CKK/rlfSNgAdr+RaVtI3K+9b0pY6xvQR4AnhlP2UBHAHMkLQIeC9wIkAZ8jgAOK2hzgauLjcv0wa4voiIUaPX6mgbbYYzsD9je/vadkHL/i/Q3GpvR1RBYqC01v11k2v1ObnNMffZnme7F1hA1fVtYB4wvuR5K/B9ANvXAK+U9PJ+6jEPeKekL0l6m+0ngC2obmRmSbqV6rvYuJ8y+tP0m+V+0gd7zMeAKbY3Br4HfK2kfx04pnT9t9rF9g5UPSdHSfo/jVfQhqRpkm6UdOMF5z69LIdGRCyX5z2uo60Tks6U9Iik+bW04yX9ttbgnNLm2L3K/KmFZRi2XyPWh1AC4mrAzrXkBcCObQ5ZQDXuW7cDcHs/p3kjcEe7nZI2qX2hHyzJz9ay9NY+9/LinIR2wW8JS3+nqwHYvpvquuYBX5T0mVLGgtqNxja29+jnWgAe7uuuL6993d6LgE1q+Tam6iZfxNI3C33pSx0jaRxV1/pj7cqStC6wne3rS/oFwFvK+4nA+ZLuBw4Evq0ysdH24vL6CHApMGmAa1yK7em2J9qeePB71liWQyMilksP6mjr0Fm8OKxZd3ItDsxo3VnmL32LqnG0JXBImVfV1kgPDpwAfKL2+VTgMEk79SVI+ntJr6aaOHdc3+zt8vpJ4KutharyT8AGwMx2J7f9YO0LbepGbmc28J5yrl2BR23/Cbif6mYDSTsAm5b3GwJP2/7Pch07AHcB60p6c8mziqT+hhUALgf6ZqkfBvyolj5V1Uz3TYEJwK9Kd/2TknYu4+eHthzTV9aBwDWlZ+IqYA9Ja6uadb9HSXsceLmkzcsxu1Nummxvanu87fFU4/Ufsn1ZmVuwVrm+NUtZL9ytRkSMZr1eqaOtE7ZnUzWeltUkYKHte8tE8POp5ki1NZyz4lcvXcx9ZtpeqgvB9gxJv699frjM0P6KpPWoWsmzgUts/07SMcCPy/j088AnbNfPcZKkTwNrUM0Cn1ybEQ/VGHtfd/Fc24cO8tqOB74naS7wNC8GyIuBQ8t13wDcXdK3KXXrLfU+0vZzqh4LO6V044+j6tJe0M95TwQulHQ48ABwEIDtBZIupOq9WAIcVesWP5LqTnF1qtnwfU8inAF8X9JCql+2qaWsxyR9vtQf4HO2HwOQ9A/AxeU6Hgc+MMD3tD5waXVPwTjgB7ZnlrIOoJpjsC5whaRbbe/ZtqSIiBWs09Z4mT9Un0M03fb0Dk/zYUmHUs0Z+79l0nJd07ynneiHqkZaxOh314Mb5pc1ljJhlZeOdBVilFrp1Xcv96y2Exbs09H/OZ/a6icdnUsvPsa8dfm8PvAo1VDu54ENbH+g5ZiDgD1tH1E+vxeYZPsj7c6TleciIiIaDPeSsuVRbgAkfRf4SUO2dnOo2hrpMfaIiIhRqccrdbQNVsvaJQfQPAfpBmCCpE0lvYRq2PTy/spNiz0iIqLBUD6jLuk8YFfgVWUtkM8Cu0ranqor/n7gH0veDYHTbU+xvUTSh6kmMa8MnGm7v7lYCewRERFNhnJVOduHNCSf0SbvYmBK7fMMqiXBO5LAHhER0WA0rirXiQT2iIiIBr1jdBpaAntERESD53sT2CMiIrpGp6vKjTYJ7BEREQ2WYR34USWBPSIiokEmz0VERHSRdMVHRER0kd50xUdERHSP53uHd6344ZLAHhER0SBj7BEREV0kXfERERFdJC32iIiILpJZ8REREV1kSQJ7RERE90hXfERERBdJYI+IiOgiCewRERFdJIE9IiKii+Q59oiIiC6ypDez4iMiIrrGUHbFSzoT2Ad4xPbWJe0k4G+A54BfA++3/ceGY+8HngR6gCW2J/Z3rrF5OxIRETHMeq2Otg6dBezVkjYL2Nr2tsDdwHH9HD/Z9vYDBXVIYI+IiGhkq6Ots7I8G3isJe1q20vKx+uAjYei3gnsERERDXpRR9sQ+QBwZZt9Bq6WdJOkaQMVlDH2iIiIBp12s5dgWw+4021P7/Q8kj4FLAHObZNlF9uLJa0HzJJ0Z+kBaJTAHhER0aCnw1nxJYh3HMjrJB1GNaluN9tuU/7i8vqIpEuBSUACe4x9tz670UhXIUaZD792/EhXIUapWb3LX0an4+eDJWkv4Bjg7bafbpNnTWAl20+W93sAn+uv3IyxR0RENBjKWfGSzgPmAFtIWiTpcOBUYC2q7vVbJZ1W8m4oaUY5dH3gF5JuA34FXGF7Zn/nSos9IiKiQXPH+GDL8iENyWe0ybsYmFLe3wtstyznSmCPiIhokCVlIyIiukink+dGmwT2iIiIBkPZFb8iJbBHREQ0GO5Z8cMlgT0iIqJBAntEREQXGcq/7rYiJbBHREQ0yBh7REREF+nNrPiIiIjuMUYb7AnsERERTTJ5LiIiopuM0SZ7AntERESDtNgjIiK6SG9vAntERET3SIs9IiKie+Q59oiIiG6SwB4REdE9MnkuIiKim6TFHhER0T2cWfERERHdJIE9IiKie4zRrvix+adrIiIihps73Dog6UxJj0iaX0tbR9IsSfeU17XbHLuXpLskLZR07EDnSmCPiIhoYnW2deYsYK+WtGOBn9qeAPy0fF6KpJWBbwF7A1sCh0jasr8TJbBHREQ0sDvbOivLs4HHWpL3A84u788G9m84dBKw0Pa9tp8Dzi/HtZXAHhER0aRXHW2Spkm6sbZN6/AM69t+CKC8rteQZyPgwdrnRSWtrUyei4iIaKDOW+PTgenDVY2mU/Z3QFrsERERTYZw8lwbD0vaAKC8PtKQZxGwSe3zxsDi/gpNYI+IiGgytJPnmlwOHFbeHwb8qCHPDcAESZtKegkwtRzXVgJ7REREk6F93O08YA6whaRFkg4HTgR2l3QPsHv5jKQNJc0AsL0E+DBwFXAHcKHtBf2dK2PsERERTXqHrijbh7TZtVtD3sXAlNrnGcCMTs/VNrBL2qG/A23f3OlJIiIixpwx+tfd+uuK/2rZvgVcTzXj77vl/SnDWSlJPZJurW3HlvRrJd1YyzdR0rXl/RqSzpU0T9J8Sb+Q9NKyz5K+XztunKTfS/pJ+Xy8pKNb6nC/pFeV95+StEDS3FKfnSRdWt4vlPREra5vKWMh15fVhC4o4yKocko5Zm795qndykL9rUwk6biS/y5Je9bST5D0oKSnWq5p1VKfhaV+49t855fX0s+QdFup70V932lt/5vKsQe2pK8s6Za+77jNz3nXcr4Fkv67Xb6IiJEgd7aNNm0Du+3JticDvwF2sD3R9o7AG4GFw1yvZ2xvX9tOrO1bT9LeDcf8M/Cw7W1sbw0cDjxf9v0Z2FrS6uXz7sBvO6mIpDcD+1B9B9sC7wQetH2A7e2BI4Cf1+r6P8CXgJPLakKPl7pAtXLQhLJNA75TztHfykKNKxOV/VOBrahWM/p2KQfgx1SLGrQ6HHjc9mbAyaWeferf+b619I/Z3q5c+wNUYz19383KpYyrGs71z1TjQY0kvQL4NrCv7a2Ag9rljYgYEcM/K35YdDJ57vW25/V9sD0f2H7YajSwk4B/bUjfgFqwtn2X7Wdr+68E/rq8PwQ4r8PzbQA82leW7UfL+EcjSQLeAVxUkuqrCe0HnOPKdcAryiMO/a0s1G5lov2A820/a/s+qputSaWO1/UtetCiXtZFwG6lvm3Z/lPtulZn6V/jjwAX0/KIhqSNqb7r0/sp+u+AS2w/UM7T9JhHREQso04C+x2STi/dpm+X9F36aYkNkdVbuuIPru2bAzwraXLLMWcCx0iaI+kLkia07D8fmCppNWBbqiGFTlwNbCLpbknflvT2AfK/EvhjmckIS68S1G4Fof5WFmq3MtEyr0ZUP6bU74lSX4DVVK2YdJ2k/esHSfoe8Dvg9cA3S9pGwAHAaQ3n+TrwCfqferI5sHYZXrlJ0qFNmVRb0em/zv/9AJcXETF0uq4rvub9wAKqrtWPAreXtOHU2hV/Qcv+L9DSard9K/A6qhb9OsANkt5Q2z8XGE/VWm+dXdjuR2PbTwE7UnWd/x64QNL7+ql7f6sEtdu3zCsLDcMxr7E9kaol/XVJf/VCBvv9wIZUN3R9N1lfB46x3bPUCaR9gEds3zRAXcZRfa9/DewJfFrS5n9ROXt6GQaa+M6p6w5QZETEEOpwSdnRZsDH3Wz/P6rx2JOHvzqdsX2NpM8DO7ekPwVcAlwiqZfqcYF678LlwFeAXXmxpQrwB6ou97q1gD+WcnuAa4FrJc2jWkjgrDbVe5Sqi31caRXXVwlqt4LQS9qkQ1mZyPZDWnplomVejah2zCJJ44CXU/4oQd/wgu17VU1IfCPw674DbfdIugD4OPA9YCJwfunJfxUwRdISYCdgX0lTgNWAl0n6T9t/31CXR23/GfizpNnAdsDdA1xDRMSKMQpb451o22KXdGF5nVdmRC+1rbgqtnUCVXcvAJJ26ZsxXmahb0k18a/uTOBz9TkDxWyqYLRWOf5dwG0lmG3R0q2/fUO5L7Bt4GdA3yzx+mpClwOHltnxOwNPlO71/lYWarcy0eVUQwurStqUakLer9rVq6GsA4FrbFvS2pJWLdf+KmAX4PZSz81KuoC/Ae4s17mp7fG2x1ON13/I9mW2j7O9cUmfWs7RGtQp1/E2VU8orEF1QzDcQzwRER0bq13x/bXY/7m83kHVSusj4MvDVqPK6pJurX2eaXupv1Nre4ak+qDrXwHfKQFoJeAKqold9WMWAd9oPZntuZJOBX4hyVSt4iPK7pcC3yyzuJdQTVIb6C/3HEPVmv0CcAtwRkmfQdWLsBB4mjKkYXuJpL6VhVYGzqytLHQicKGqVYoeoMwet72g3HzdXup1VF+3uKQvU3WpryFpEXC67eNLPb4vaSFVS31qOccbgP8ovRwrASfavl3SSsDZkl5G9XO/DThygGtvS9IHS91Ps32HpJnAXKqx+NPLxMyIiNFhFAbtTsgD/DFZSTfb3qElbW55/Clihblg4ZvG6D+zGC6nbz5+pKsQo9Ss3h8u9+D3X33lax39n/Pro/9lVA2097fy3JHAh4DXtXS9rwX8crgrFhERMZJGYzd7J/rriv8B1bPfX6QsilI8afuxYa1VRETESBuFM9470Taw236C6jnndgvXR0REdK1ubLFHRET875XAHhER0T3SYo+IiOgmCewRERHdQ/39tYtRrJO14iMiImKMSIs9IiKiSbriIyIiukcmz0VERHSTMRrYM8YeERHRxB1uAyh/JfTW2vYnSR9tybOrpCdqeT4z2GqnxR4REdFgqGbF276L6k9+I2ll4LfApQ1Zf257n+U9XwJ7REREg2EaY98N+LXt3wxL6aQrPiIiolmHXfGSpkm6sbZN66fUqcB5bfa9WdJtkq6UtNVgq50We0RERJMOW+y2pwPTB8on6SXAvsBxDbtvBl5r+ylJU4DLgAmdVrUuLfaIiIgGcmfbMtgbuNn2w607bP/J9lPl/QxgFUmvGky9E9gjIiKaDNGs+JpDaNMNL+nVklTeT6KKz38YTLXTFR8REdFgKNeKl7QGsDvwj7W0DwLYPg04EDhS0hLgGWCq7UFN30tgj4iIaDKEs+JtPw28siXttNr7U4FTh+JcCewRERENsqRsREREN0lgj4iI6CIJ7BEREd0jXfERERFdJIE9IiKimySwR0REdJEE9oiIiO6RrviIiIhuksAeERHRPYZySdkVKYE9xoyFz64/0lWIiP9F0hUfERHRTRLYIyIiukgCe0RERPdIV3xEREQXUe/YjOwJ7BEREU3GZlxPYI+IiGiSrviIiIhuksAeERHRPdJij4iI6CYJ7BEREd1jKJeUlXQ/8CTQAyyxPbFlv4BvAFOAp4H32b55MOdKYI+IiGgwDF3xk20/2mbf3sCEsu0EfKe8LrOVBle3iIiILmd3tg2N/YBzXLkOeIWkDQZTUAJ7REREA7mzrUMGrpZ0k6RpDfs3Ah6sfV5U0pZZuuIjIiKadBi0S6CuB+vptqe3ZNvF9mJJ6wGzJN1pe3a9mMHXYGkJ7BEREQ06nTxXgnhrIG/Ns7i8PiLpUmASUA/si4BNap83BhYvQ3VfkK74iIiIBurtbBuwHGlNSWv1vQf2AOa3ZLscOFSVnYEnbD80mHqnxR4REdFk6CbGrQ9cWj3RxjjgB7ZnSvpgdRqfBsygetRtIdXjbu8f7MkS2CMiIhoM1eNutu8FtmtIP6323sBRQ3G+BPaIiIgmWXkuIiKie2St+IiIiC6i3rEZ2RPYIyIimozNuJ7AHhER0SRd8REREd0kXfERERFdZGzG9QT2iIiIJumKj4iI6CKZFR8REdFNxmZcT2CPiIhooqFbK36FSmCPiIho0uGfbR1tEtgjIiIapMUeERHRTcZmXGel4SpYUo+kW2vbsSX9Wkk31vJNlHRt7fMkSbMl3SXpTkmnS1qj7Ntf0tySPk/S/rXjzpJ0n6TbJN0t6RxJG9X231+O6avPKcN17ctD0rdK/W6X9EytvgdKWkfSLEn3lNe1a8cdJ2lh+d72rKXvWK57oaRTVP4gsKRVJV1Q0q+XNL52zGHlHPdIOqyWvpukm0t9fiFps5a6v6n83A+spdW/9/rP/aTyc5wr6VJJrxjq7zIiYnmo1x1to82wBXbgGdvb17YTa/vWk7R36wGS1gd+CBxjewvgDcBMYC1J2wFfAfaz/XpgX+ArkratFfFx29sBWwC3AD+T9JLa/sm1+vzTkF7tELF9lO3tgSnAr2v1vQg4Fvip7QnAT8tnJG0JTAW2AvYCvi1p5VLkd4BpwISy7VXSDwcet70ZcDLwpVLWOsBngZ2AScBnazcQ3wHeU+r3A+Bf++pdzvcl4KqGy+r73ifW0mYBW9veFrgbOG5Zv6uIiGFld7aNMsMZ2PtzErWgUHMUcLbtOVD94XnbF9l+GDga+Hfb95V99wFfBD7eWkg57mTgd8Bf3EC0I2l8rZdgvqRzJb1T0i9L63VSybeOpMtKa/O6vpsLScdLOrpW3vxS5pqSrii9CfMlHVz27yjpvyXdJOkqSRsMUMX9gLPL+7OB/Wvp59t+tnwvC4FJpbyX2Z5j28A5Lcf0lXURsFtpze8JzLL9mO3HqQJw382AgZeV9y8HFtfq9hHgYuCRAa6hKsi+2vaS8vE6YONOjouIWFHU29k22gxnYF+9pSv+4Nq+OcCzkia3HLM1cFOb8rZq2HdjSW/nZuD1tc8/q9XnY22O2Qz4BrBtOfbvgLdS3Vh8suT5N+CW0tr8JFXA7M9ewGLb29neGpgpaRXgm8CBtncEzgROGKCc9W0/BFBe1yvpGwEP1vItKmkblfet6UsdUwLsE8Ar+ykL4AhghqRFwHuBEwHKkMcBwGkNdTZwdbl5mdbmuj4AXNn2qiMiRsIYbbEP5+S5Z0qXbTtfoGq1H9NheeIvpzI0pbXur5ts+9EBznOf7XkAkhZQdX1b0jxgfMnzVuDdALavkfRKSS/vp8x5VMMGXwJ+YvvnkramupGZVYa9VwYeGqBu7bReJ1TfS7v0wR7zMWCK7eslfRz4GlWw/zrV8ElPuZa6XWwvlrQe1bXeaXv2CxWXPgUsAc5tvLDqZmAawH6f3YY3HfSapmwREUNv9MXsjoxUVzy2rwFWA3auJS8AdmxzyAJgYkvaDsDt/ZzmjcAd7XZK2qTWgv9gSX62lqW39rmXF2+E2gW/JSz9na4GYPtuquuaB3xR0mdKGQtqY+jb2N6jn2sBeLivu7689nV7LwI2qeXbmKqbfBFLd3H3pS91jKRxVF3rj7UrS9K6wHa2ry/pFwBvKe8nAudLuh84kGqMf/9y7YvL6yPApVTj9pTzHgbsQzVu3/hPyPZ02xNtT0xQj4gVSb29HW2jzYgF9uIE4BO1z6cCh0naqS9B0t9LejXVxLnj+mZvl9dPAl9tLVSVfwI2oJp818j2g7XA2tSN3M5s4D3lXLsCj9r+E3A/1c0GknYANi3vNwSetv2f5Tp2AO4C1pX05pJnFUn9DSsAXA70zVI/DPhRLX1qmem+KdUkuV+V7vonJe1cxs8PbTmmr6wDgWtKcL0K2EPS2mXS3B4l7XHg5ZI2L8fsTrlpsr2p7fG2x1ON13/I9mVlbsFa5frWLGXNL5/3ouqt2df20wNcd0TEitfb4TbKDGdX/OqSbq19nmn72HoG2zMk/b72+WFJU6m6rdej+spmA5fY/p2kY4Afl/Hp54FP2K6f4yRJnwbWoJqQNdn2c7X9P5PUU97PtX3oIK/teOB7kuYCT/NigLwYOLRc9w1Us70Btil16y31PtL2c6oeCzuldOOPo+rSXtDPeU8ELpR0OPAAcBCA7QWSLqTqvVgCHGW77zqPBM4CVqcax+4byz4D+L6khVQt9amlrMckfb7UH+Bzth8DkPQPwMXlOh6nGhvvz/rApaV7fhzwA9t9N1qnAqvy4lDEdbY/2FhKRMQIGKoFaiRtQjUX69VUcW267W+05NmVquF1X0m6xPbnBnW+Nj2gEaPOCQv2yS9rLOXabVYf6SrEKDWr94dNQ6bLZM+Jx3f0f85VNx7f77nK0OkGtm8uvZg3Afvbvr2WZ1fgaNv7DL7Glaw8FxER0WSIGr5lWLTviaYnJd1B9bRRf3PEBm2kx9gjIiJGpw7H2CVNk3RjbWv3aG/f/LA3Atc37H5zWe/kyg7mXLWVFntERESDTme8254OTB+wPOmlVHOxPlomXNfdDLzW9lOSpgCXUU2EXmZpsUdERDQZwgVqyqTvi4FzbV/yl6fyn2w/Vd7PAFaR9KrBVDuBPSIioskQBfbyuPEZwB22v9Ymz6tLPlQtX74S8IfBVDtd8REREU2G7hn1XaiW4Z5Xewz8k8BrAMo6KgcCR0paAjwDTG23cNdAEtgjIiIaDNVz7LZ/QfOKpfU8p1Kt77HcEtgjIiKajNF1XhLYIyIimvSMwvViO5DAHhER0SQt9oiIiC6SwB4REdFFehPYIyIiuoczxh4REdE9MnkuIiKii2SMPSIioosksEdERHSRBPaIiIgu0uGfbR1tEtgjIiKapMUeERHRRTIrPiIions4z7FHRER0kaw8FxER0UUyxh4REdFFMis+IiKii6TFHhER0T3c0zPSVRiUBPaIiIgmmTwXERHRRcbo424rjXQFIiIiRiP3uqOtE5L2knSXpIWSjm3YL0mnlP1zJe0w2HonsEdERDRxb2fbACStDHwL2BvYEjhE0pYt2fYGJpRtGvCdwVY7gT0iIqKBe3o62jowCVho+17bzwHnA/u15NkPOMeV64BXSNpgMPXOGHuMGZ/a6ica6TqMFpKm2Z4+0vUYaZ8am0Ogwya/F0NrVu8PO/o/R9I0qlZ2n+ktP4eNgAdrnxcBO7UU05RnI+ChjitcpMUeMTZNGzhL/C+U34sRYHu67Ym1rfXmqukGoXVwvpM8HUlgj4iIGF6LgE1qnzcGFg8iT0cS2CMiIobXDcAESZtKegkwFbi8Jc/lwKFldvzOwBO2l7kbHjLGHjFWZRw1muT3YhSyvUTSh4GrgJWBM20vkPTBsv80YAYwBVgIPA28f7Dnk8foWrgRERHxl9IVHxER0UUS2CMiIrpIAnvEMJG0vqQfSLpX0k2S5kg6QNKukizp8FreN5a0o8vngyQtkNQraWIt3+6lrHnl9R21fYeU9LmSZkp6VUk/WdKtZbtb0h9rx3y5nOeOspxl1goYRSQ9VXs/RdI9kl4j6X2Sfl/7uR5Ry/caSVeXn+ntksaX9HPLkqbzJZ0paZURuKRYARLYI4ZBCZCXAbNtv872jlQzYTcuWeYBB9cOmQrcVvs8H3gXMLul6EeBv7G9DXAY8P1yvnHAN4DJtrcF5gIfBrD9Mdvb294e+CZwSTnmLcAuwLbA1sCbgLcv77XH0JO0G9XPbi/bD5TkC/p+rrZPr2U/BzjJ9huoVjx7pKSfC7we2AZYHTiC6EoJ7BHD4x3Ac2W2KwC2f2P7m+XjA8BqpVUvYC/gylreO2zf1Vqo7Vts9z3buqCUsSrV4hYC1izlvYzmZ2APAc7rKw5YDXgJsCqwCvDwYC84hoektwHfBf7a9q8HyLslMM72LADbT9l+uryfUZYrNfArXrzJjC6TwB4xPLYCbh4gz0XAQcBbSt5nl/Ec7wZusf2s7eeBI6l6AhZT/aGJM+qZJb0W2BS4BsD2HOBnVEtWPgRcZfuOZaxDDK9VgR8B+9u+s2Xfu8uwy0WS+hY22Rz4o6RLJN0i6aTyB0heULrg3wvMHPbax4hIYI9YASR9S9Jtkm6oJV9IFdjrrehOy9sK+BLwj+XzKlSB/Y3AhlRd8ce1HDYVuMh2TzlmM+ANVC23jYB3SPo/y3hpMbyeB/4HOLwl/cfA+DLs8l/A2SV9HPA24GiqoZXXAe9rOfbbVENEPx+mOscIS2CPGB4LgBf+nrLto4DdgHVrab+j+o97d+CnnRYsaWPgUuDQWtfs9qXMX5eu1gupegLqprL0DcQBwHWlu/YpqqGAnTutR6wQvcDfAm+S9Mm+RNt/sN3Xw/NdYMfyfhFVL869tpdQzfN44fdQ0mepfgf/ZQXUPUZIAnvE8LiGavz7yFraGg35PgMc09eKHoikVwBXAMfZ/mVt12+BLSX13TjsDtxRO24LYG1gTu2YB4C3SxpXWvxvrx8To0MZI98HeE/fkxQtf85zX178ud0ArF37PXgHcHs55ghgT+AQu4M/Ih5jVpaUjRgGti1pf+BkSZ8Afg/8GTimJd//NB0v6QCqWdDrAldIutX2nlQz3TcDPi3p0yX7HrYXS/o3YLak54HfsHQX7CHA+V56qcmLqP7jn0c1kW6m7R8vx2XHMLH9mKS9qH6+jwI7S9oXWAI8RvlZ2+4pj0z+tEyivImqRQ9wGtXvxZzyVOMltj+3Yq8kVoQsKRsREdFF0hUfERHRRRLYIyIiukgCe0RERBdJYI+IiOgiCewRERFdJIE9IiKiiySwR0REdJH/D/fO1K8cG7Z7AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.heatmap(data=t_df, cmap='viridis')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "material-jesus",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:ylabel='tid'>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAfMAAAD8CAYAAABjLk0qAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAjBElEQVR4nO3deZxddX3/8debBNkEJCo0LP0FS1DZZImIYiuIrFoWhRKqJVJoKqJVf2UJttaVioJVUdFGQIJigQYUVARTlqIWkD0LYUmBQpqwGcT1B8zM+/fH+Q6cjOfOXCYzmTnyfvI4j3vP95zv93zPvUM+57ucc2WbiIiIaK81xroCERERsWoSzCMiIlouwTwiIqLlEswjIiJaLsE8IiKi5RLMIyIiWi7BPCIiYpRJOkfSo5IW1tImSZon6d7yulFt28mSlki6W9K+Q5WfYB4RETH6zgX2G5A2C7jK9lTgqrKOpG2A6cC2Jc+ZkiYMVniCeURExCizfR2wYkDyQcCc8n4OcHAt/QLbT9m+H1gC7DpY+RNHrqoRo+uXRx2ZxxXGStY7+6/HugoxTk1YYw+tahm9fdd29W/OxAl7/i0ws5Y02/bsLrJuYns5gO3lkjYu6ZsBN9T2W1rSOtehm4pGREREsxK4uwne3Wq6EBn0wiLBPCIioklfX3f7DX/A+hFJk0urfDLwaElfCmxR229zYNnoVCEiIuIPWU9Pd8vwXQbMKO9nAJfW0qdLWkvSlsBU4GeDFZSWeURERJMR/FVRSf8G7AG8TNJS4KPAqcBFko4GHgQOqw7rRZIuAu4EeoDjbPcOVn6CeURERJNuu9m7YPuIDpv26rD/KcAp3ZafYB4REdFkBIP5aEswj4iIaJJgHhER0XIJ5hEREe2m3lWaqb5aJZhHREQ0Scs8IiKi5fra8wTpBPOIiIgmaZlHRES0XIJ5REREy2UCXERERMtlzDwiIqLl0s0eERHRcgnmERER7aYE84iIiJYbwZ9AHW0J5hEREU16Mps9IiKi3TKbPSIiouUyZh4REdFyCeYREREtlwlwERERLdfTO9Y16FqCeURERJN0s0dERLRcgnlERETL5da0iIiIlnNa5hEREe2WlnlERETLZTZ7REREy7VoAtwao1WwpF5Jt9eWWSX9Wkk31/abJuna2vqukq6TdLekuySdJWndsu1gSfNL+gJJB9fynSvpfkl3SLpH0nmSNqttf6Dk6a/PGaN17qtC0ldK/e6U9LtafQ+VNEnSPEn3lteNavlOlrSkfG771tJ3Kee9RNIZklTS15J0YUm/UdKUWp4Z5Rj3SppRS99L0q2lPj+RtNWAur+2fO+H1tLqn3v9ez9M0iJJfZKmjfgHGRGxqvrc3TIOjGbL/He2d+ywbWNJ+9v+YT1R0ibAvwPTbV9fAs87gPUlTQVOB/a2fb+kLYF5ku6zPb8UcYLtuSXfB4FrJG1n++myfU/bj4/saY4s28cBlOD6/fpnKOmzwFW2Ty0XR7OAkyRtA0wHtgU2Bf5D0ta2e4GvAjOBG4DLgf2AHwJHA0/Y3krSdOAzwOGSJgEfBaYBBm6RdJntJ0pZB9leLOm9wD8C7y51m1DKuLLhtJo+94XA24F/HfaHFRExmlo0AW7UWuZDOI0qEAx0HDDH9vUArsy1/QhwPPDPtu8v2+4HPg2cMLCQku/zwMPA/t1WStKUWm/AQknnS3qLpJ+WVuquZb9Jkr5beglukLRDSf+YpONr5S0sZa4n6Qel12ChpMPL9l0k/aekWyRdKWnyEFU8CJhT3s8BDq6lX2D7qfK5LAF2LeVtYPt62wbOG5Cnv6y5wF7lImhfYJ7tFSWAz6O6AIAquG9Q3m8ILKvV7f3AxcCjQ5xDVZC92Pbd3ewbETEmWtQyH81gvo5W7mY/vLbteuApSXsOyLMdcEuH8rZt2HZzSe/kVuBVtfVravX5UIc8WwFfBHYoef8SeCPVxcSHyz4fB26zvUNJO2+QOkAVDJfZfo3t7YArJK0JfAk41PYuwDnAKUOUs4nt5QDldeOSvhnwUG2/pSVts/J+YPpKeWz3AE8CLx2kLIBjgMslLQX+CjgVoAxnHAJ8raHOBn5ULlhmDnF+ERHjR4uC+Vh1swN8iqp1flKX5YkqMAyVNnB7XTfd7PfbXgAgaRFVt7YlLQCmlH3eSNX9j+2rJb1U0oaDlLkAOF3SZ6i6zn8saTuqi5d5ZRh7ArB8iLp1MvA8ofpcOqUPN8+HgANs3yjpBOBfqAL8F4CTbPeWc6nb3fYySRtTnetdtq8b9GxqygXATIAvvP51HPXKrbvNGhGxalo0m32sutmxfTWwNrBbLXkRsEuHLIuoxnHrdgbuHOQwOwGLO22UtEWtpf6ekvxUbZe+2nofz138dAp4Paz8ma4NYPseqvNaAHxa0j+VMhbZ3rEs29veZ5BzAXikvyu+vPZ3aS8FtqjttzlVF/jS8n5g+kp5JE2k6jZf0aksSS8HXmP7xpJ+IfCG8n4acIGkB4BDgTNVJifaXlZeHwW+A+w6xDmuxPZs29NsT0sgj4jVqkUt8zEL5sUpwIm19S8DMyS9rj9B0rsk/RHV5LeT+2ddl9cPA58bWKgqfwdMBq7odHDbD9WCaVMXcSfXAe8sx9oDeNz2L4EHqC4wkLQzsGV5vynwW9vfKuexM3A38HJJry/7rClpsCEDgMuA/tnlM4BLa+nTVc1Q3xKYCvysdMX/StJuZTz8yAF5+ss6FLi6jKtfCewjaSNVs+X3KWlPABtK6o+oe1MulGxvaXuK7SlU4+/vtf3dMldg/XJ+65WyFg5xjhER40NfX3fLODCa3ezrSLq9tn6F7Vn1HWxfLumx2vojZWb16aVbto8qcF5i+2FJJwHfK+PNzwAn2q4f4zRJHwHWpZq9vWdtJjtUY+b9/SbzbR85zHP7GPANSfOB3/JcULwYOLKc903APSV9+1K3vlLvY20/reoWrjNKF/1Equ7qRYMc91TgIklHAw8ChwHYXiTpIqpeih7guDKTHeBY4FxgHapZ7P13EJwNfFPSEqoW+fRS1gpJnyz1B/iE7RUAkv4GuLicxxPAXw/xOW0CfKd0vU8Evm37ilLWIVRzBl4O/EDS7bb37VhSRMTq1qLfM5dbVNl4YfvlUUfmjzVWst7ZQ11PxgvVhDX2aBoOfV58wd939W+Opn9uyGOVSdfHUA3JLgCOomp4Xkg1H+sB4C/KXUTP21h3s0dERIxPPb3dLUMod/z8HTCt3NE0gao3dBbVJOupwFVlfVgSzCMiIpqM7AS4iVTDzxOpWuTL6PzskOctwTwiIqJJl8Fc0kxJN9eWlZ6pYft/qSY/P0h1C/KTtn9E52eHPG/5oZWIiIgmXba6bc8GZnfaXu4MOojqDqdfAP8u6V0jUMNnJZhHREQ0cJfBvIuZdm+heiDZYwCSLqF6TscjkibbXj7g2SHPW7rZIyIimtjdLUN7ENhN0rrlmR97UT2no9OzQ563tMwjIiKa9IzMA2HKI7DnUv1eSA9wG1W3/ItpeHbIcCSYR0RENBnBR7Xa/ijVz0vXPUXVSl9lCeYRERFNxslz17uRYB4REdGg2wlw40GCeURERJME84iIiJZLMI+IiGi53vHx86bdSDCPiIho4PbE8gTziIiIRulmj4iIaLkE84iIiHZLN3tERETb9aRlHhER0Wp5aExERETbpZs9IiKi5drTME8wj4iIaJJu9oiIiLZLN3tERES7uWesa9C9BPOIiIgGuc88IiKi7RLMIyIi2s3tmf+WYB4REdEk3ewRERFtl2AeERHRbn29Y12D7iWYR2vsfukaY12FGGcWnrv3WFchxin7mVUvpE+rXsZqkmAeERHRIGPmERERLWenZR4REdFqaZlHRES0XF9vWuYRERGt5kyAi4iIaLc8AS4iIqLlMgEuIiKi5dLNHhER0XLpZo+IiGi53t72PHUywTwiIqJBm1rm7bnsiIiIWI1sdbV0Q9JLJM2VdJekxZJeL2mSpHmS7i2vGw23rgnmERERDUYymANfBK6w/SrgNcBiYBZwle2pwFVlfVgSzCMiIhr0WV0tQ5G0AfBnwNkAtp+2/QvgIGBO2W0OcPBw65pgHhER0aCvV10tkmZKurm2zBxQ1CuAx4BvSLpN0lmS1gM2sb0coLxuPNy6ZgJcREREg25a3QC2ZwOzB9llIrAz8H7bN0r6IqvQpd4kLfOIiIgGIzhmvhRYavvGsj6XKrg/ImkyQHl9dLh1TTCPiIhoMFJj5rYfBh6S9MqStBdwJ3AZMKOkzQAuHW5d080eERHRYISfzf5+4HxJLwLuA46ialBfJOlo4EHgsOEWnmAeERHRoG8Ey7J9OzCtYdNeI1F+gnlERESD3r72jEQnmEdERDTIT6BGRES0XF+Lns2eYB4REdEgLfOIiIiW6yPBPCIiotXa9BOoCeYRERENepzZ7BEREa2WlnlERETLdftDK+NBgnlEREQDZwJcREREu+U+84iIiJbr/UOYACdp58Ey2r515KsTERExPrSpZT7YZcfnyvIV4EZgNvD18v6M0ayUpF5Jt9eWWSX9Wkk31/abJuna8n5dSedLWiBpoaSfSHpx2WZJ36zlmyjpMUnfL+sfk3T8gDo8IOll5f0/SFokaX6pz+skfae8XyLpyVpd3yBpS0k3SrpX0oXlJ+9Q5YySZ379gknSfpLuLttm1dInSZpXyponaaPatpPL/ndL2reWfoqkhyT9esA5rVXqs6TUb0qHz/yyWvrZku4o9Z3b/5nWtr+25D10QPoESbf1f8Ydvuc9yvEWSfrPTvtFRIwFo66W8aBjMLe9p+09gf8BdrY9zfYuwE7AklGu1+9s71hbTq1t21jS/g15PgA8Ynt729sBRwPPlG2/AbaTtE5Z3xv4324qIun1wNuoPoMdgLcAD9k+xPaOwDHAj2t1/S/gM8DnbU8Fnih1AdgfmFqWmcBXyzEmUF007Q9sAxwhaZuSZxZwVSnrqrJO2T4d2BbYDzizlAPwPWDXhtM5GnjC9lbA50s9+9U/8wNr6R+y/Zpy7g8C76t9NhNKGVc2HOsDwOKG9P68LwHOBA60vS2r8Du+ERGjoc/dLeNBNwMCr7K9oH/F9kJgx1Gr0dBOA/6xIX0ytQBt+27bT9W2/xB4a3l/BPBvXR5vMvB4f1m2H7e9rNPOkgS8GZhbkuYAB5f3BwHnuXID8BJJk6kC7xLb99l+Grig7NufZ06Hsi6w/ZTt+6kusHYtdbzB9vKG6tXLmgvsVerbke1f1s5rHaD+p/t+4GLg0QGfweZUn/VZgxT9l8Alth8sx3l0kH0jIla7PqurZTzoJpgvlnRW6RJ9k6SvM0iLa4SsM6Cb/fDatuuBpyTtOSDPOcBJkq6X9ClJUwdsvwCYLmltYAeq4YJu/AjYQtI9ks6U9KYh9n8p8AvbPWV9KbBZeb8Z8FBt3/5tndIBNukPzOV14yHKGsyzeUr9niz1BVhb0s2SbpB0cD2TpG8ADwOvAr5U0jYDDgG+1nCcLwAnAn2D1GVrYKMydHKLpCObdpI0s9Tr5hVP3TPE6UVEjBx3uYwH3QTzo4BFVN2mHwTuLGmjaWA3+4UDtn+KAa1z27cDr6BquU8CbpL06tr2+cAUqlb55QPK6/R92PavgV2ousUfAy6U9O5B6t50meYhtg2WZzjHGU6eP7Y9jarF/AVJf/LsDvZRwKZUF3H9F1ZfAE6y3bvSAaS3AY/avmWIukyk+lzfCuwLfETS1r9XOXt2GeKZNmmt39scETFqeqyulvFgyFvTbP8/qvHVz49+dbpj+2pJnwR2G5D+a+AS4BJJfcABrNyLcBlwOrAHz7VIAX5O1Z1etz7wi1JuL3AtcK2kBcAM4NwO1Xucqvt8Ymn9bg70d8svBbao7du/7UUd0gEekTTZ9vLSJd/fHd2prMH051kqaSKwIbCinOOy8nqfqkmFOwH/3Z/Rdq+kC4ETgG8A04ALSi/9y4ADJPUArwMOlHQAsDawgaRv2X5XQ10et/0b4DeSrgNeA6T5HRHjQpt+ArVjy1zSReV1QZnJvNKy+qrY0SlUXbkASNq9f6Z3mT2+DdXkvbpzgE/U5wAU11EFoPVL/rcDd5QA9soBXfY7NpT7LNsGrgH6Z3fPAC4t7y8Djiyz2ncDnixd5zcBU1XNgn8R1cS2y2p5ZnQoa3qZob4l1aS6n3WqV0NZhwJX27akjSStVc79ZcDuwJ2lnluVdAF/DtxVznNL21NsT6Eaf3+v7e/aPtn25iV9ejnGwEBOOY8/VXVnwbpUFwGjPXwTEdG1vi6X8WCwlvkHyutiqtZYPwGfHbUaVdaRdHtt/Qrbs+o72L5c0mO1pD8BvlqCzhrAD6gmZ9XzLAW+OPBgtudL+jLwE0mmav0eUza/GPhSmX3dQzXRbOYQ9T+JqtX6KeA24OySfjlVb8ES4LeU4QrbPZLeRzUrfAJwju1FJc+pwEWSjqaaTX5YybOoXHDdWep1XH+Xt6TPUnWXrytpKXCW7Y+VenxT0hKqFvn0coxXA/9aejPWAE61faekNYA5kjag+t7vAI4d4tw7kvSeUvev2V4s6QpgPtX/D2eVyZUREePCeJmp3g15iJ+FkXSr7Z0HpM0vtypFrDbbT3p3i/7XitVh4RPnj3UVYpyyn1nlPvJLXvuRrv7NeftNnxzz/vjBngB3LPBe4BUDutXXB3462hWLiIgYS21qmQ/Wzf5tqnuzP015UEnxK9srRrVWERERY6y3RRPgOgZz209S3Yd8xOqrTkRExPjwh9Iyj4iIeMEaL89d70aCeURERIO0zCMiIlpuvNxD3o0E84iIiAZ/EBPgIiIiXsjSzR4REdFyLYrlCeYRERFNxstvlXcjwTwiIqJBWuYREREtlzHziIiIluttUTDv+HvmERERL2R9qKulW5ImSLpN0vfL+iRJ8yTdW143Gm5dE8wjIiIa2N0tz8MHgMW19VnAVbanAlex8o+aPS8J5hEREQ36uly6IWlz4K3AWbXkg4A55f0c4ODh1jXBPCIiokGfu1skzZR0c22Z2VDcF4ATWTn+b2J7OUB53Xi4dc0EuIiIiAbd9qDbng3M7rRd0tuAR23fImmPEaja70kwj4iIaNA7cr+0sjtwoKQDgLWBDSR9C3hE0mTbyyVNBh4d7gHSzR4REdFgpMbMbZ9se3PbU4DpwNW23wVcBswou80ALh1uXdMyj4iIaLAaHhpzKnCRpKOBB4HDhltQgnlERESD0Yjltq8Fri3vfw7sNRLlJphHREQ0yONcIyIiWq43v5oWERHRbmmZR0REtFyLYnmCeURERJO0zCMiIlrOLWqbJ5hHREQ0SMs8IiKi5XoTzCMiItrtef5W+ZhKMI+IiGgwcr+zMvoSzCMiIhpkzDwiIqLl0s0eERHRcpkAFxER0XIZM4+IiGg5t6ifPcE8IiKiQSbARUREtFyLYnmCeURERJO0zCMiIlquN2PmERER7ZaWeURERMu1qGGeYB4REdGkr0VT4BLMIyIiGqRlHhER0XJ5AlxERETL9bZoBlyCeURERIOMmUdERLRcxswjIiJaLi3ziIiIlkvLPCIiouV63J757AnmERERDZxu9oiIiHZrT7s8wTwiIqJRJsBFRES0nFs0A26Nsa5ARETEeNSHu1qGImkLSddIWixpkaQPlPRJkuZJure8bjTcuiaYR0RENOjt8r8u9AB/b/vVwG7AcZK2AWYBV9meClxV1oclwTwiIqLBSLXMbS+3fWt5/ytgMbAZcBAwp+w2Bzh4uHVNMI+IiGjQbTCXNFPSzbVlZqcyJU0BdgJuBDaxvRyqgA9sPNy6ZgJcREREg74ub06zPRuYPdR+kl4MXAx80PYvJa1aBWsSzCMiIhpYI3enuaQ1qQL5+bYvKcmPSJpse7mkycCjwy0/3ewRERENRnA2u4CzgcW2/6W26TJgRnk/A7h0uHUdtWAuqVfS7bVlVkm/VtLNtf2mSbq2tr6rpOsk3S3pLklnSVq3bDtY0vySvkDSwbV850q6X9Idku6RdJ6kzWrbHyh5+utzxmid+6qQ9JVSvzsl/a5W30MHu41B0smSlpTPbd9a+i7lvJdIOqP8USFpLUkXlvQbyzhOf54Z5Rj3SppRS99L0q2lPj+RtNWAur+2fO+H1tLqn3v9ez+tfI/zJX1H0ktG+rOMiFgVvfR0tXRhd+CvgDfX/k0/ADgV2FvSvcDeZX1YRrOb/Xe2d+ywbWNJ+9v+YT1R0ibAvwPTbV9fAs87gPUlTQVOB/a2fb+kLYF5ku6zPb8UcYLtuSXfB4FrJG1n++myfU/bj4/saY4s28fBs5Mkvl//DCV9luo2hlPLxdEs4KRyi8N0YFtgU+A/JG1tuxf4KjATuAG4HNgP+CFwNPCE7a0kTQc+AxwuaRLwUWAaYOAWSZfZfqKUdZDtxZLeC/wj8O5StwmljCsbTqvpc58HnGy7R9JngJOBk4b7uUVEjLS+Eepmt/0ToNMA+V4jcYyx6mY/jSoQDHQcMMf29QCuzLX9CHA88M+27y/b7gc+DZwwsJCS7/PAw8D+3VZK0pRab8BCSedLeoukn5ZW6q5lv0mSvltalTdI2qGkf0zS8bXyFpYy15P0g9JrsFDS4WX7LpL+U9Itkq5UNWYymE63MRwEXGD7qfK5LAF2LeVtYPt6V48yOm9Anv6y5gJ7lYugfYF5tleUAD6P6gIAquC+QXm/IbCsVrf3U40HdTXmY/tHtvsvaW8ANu8mX0TE6tLX5X/jwWgG83W0cjf74bVt1wNPSdpzQJ7tgFs6lLdtw7abS3ontwKvqq1fU6vPhzrk2Qr4IrBDyfuXwBupLiY+XPb5OHCb7R1K2nmD1AGqYLjM9mtsbwdcoWoyxJeAQ23vApwDnDJEOZ1uY9gMeKi239KStll5PzB9pTwlqD4JvHSQsgCOAS6XtJSqy+hUgDKccQjwtYY6G/hRuWDpdLvGX1P1FkREjBttCuZj1c0O8Cmq1nm3XauC35tp0JQ2cHtdN93s99teACBpEVW3tiUtAKaUfd5I1f2P7aslvVTShoOUuQA4vXQnf9/2jyVtR3XxMq8MY08Alg9Rt06aum88SPpw83wIOMD2jZJOAP6FKsB/ATjJdq9+/1aL3W0vk7Qx1bneZfu6Zysu/QPV05HObzyx6gJgJsCm676eSWtt3bRbRMSI8zgJ1N0Ys9nstq8G1qZ6tF2/RcAuHbIsohrHrdsZuHOQw+xE9aSdRqqel9vfUn9PSX6qtktfbb2P5y5+OgW8Hlb+TNcGsH0P1XktAD4t6Z9KGYts71iW7W3vM8i5QLmNodS9fhvDUmCL2n6bU3WBL2Xl7uv+9JXySJpI1W2+olNZkl4OvMb2jSX9QuAN5f004AJJDwCHAmeqTE60vay8Pgp8B9i1v+Ayue5twDvd4RcNbM+2Pc32tATyiFidenmmq2U8GOtb004BTqytfxmYIel1/QmS3iXpj6gmv53cP+u6vH4Y+NzAQlX5O2AycEWng9t+qBZMm7qIO7kOeGc51h7A47Z/CTxAdYGBpJ2BLcv7TYHf2v5WOY+dgbuBl0t6fdlnTUmDDRlA59sYLgOmq5qhviUwFfhZ6Yr/laTdynj4kQPy9Jd1KHB1CahXAvtI2kjVbPl9StoTwIaS+iPq3pQLJdtb2p5iewrV+Pt7bX+3zBVYv5zfeqWshWV9P6pemQNt/3aI846IWO361NfVMh6MZjf7OpJur61fYXulh8jbvlzSY7X1R8rM6tNLt2wfVeC8xPbDkk4CvlfGm58BTrRdP8Zpkj4CrEs1qWrP2kx2qMbM+5+KP9/2kcM8t48B35A0H/gtzwXFi4Ejy3nfBNxT0rcvdesr9T7W9tOqbuE6o3TRT6Tqrl40yHFPBS6SdDTwIHAYgO1Fki6i6qXoAY4rM9kBjgXOBdahGpfuH5s+G/impCVULfLppawVkj5Z6g/wCdsrACT9DXBxOY8nqMa6B7MJ8J3S9T4R+Lbt/ourLwNr8dwwww2239NYSkTEGOjr7kdUxgW16fda44Vt+0nvzh9rrGThE41TLSKwn1nlZ6VOfemhXf2bc+/P547cc1mHKY9zjYiIaNDn9rTME8wjIiIatGk2e4J5REREg/EyU70bCeYREREN0jKPiIhoOWfMPCIiot3Gy6Nau5FgHhER0cAtus88wTwiIqKBnZZ5REREq/U6s9kjIiJaLbPZIyIiWi7d7BERES2XCXAREREtl5Z5REREy/W6Z6yr0LUE84iIiAZpmUdERLRcHucaERHRcrk1LSIiouXSzR4REdFyCeYREREt15fZ7BEREe2WlnlERETrJZhHRES0WlrmERERLZdb0yIiIlouLfOIiIiWs58Z6yp0LcE8IiKiUVrmERER7daibvY1xroCERER45G7/K8bkvaTdLekJZJmjXRdE8wjIiIa9XW5DE7SBOArwP7ANsARkrYZyZommEdERDSw+7paurArsMT2fbafBi4ADhrJumbMPFpjwYpzNdZ1GC8kzbQ9e6zrMfbOHesKjCv5uxhZ9jNd/ZsjaSYws5Y0e8D3sBnwUG19KfC6Va/hc9Iyj2inmUPvEi9A+bsYA7Zn255WWwZeUDVdFHQ32N6lBPOIiIjRtRTYora+ObBsJA+QYB4RETG6bgKmStpS0ouA6cBlI3mAjJlHtFPGRaNJ/i7GIds9kt4HXAlMAM6xvWgkjyF7RLvtIyIiYjVLN3tERETLJZhHRES0XIJ5xCiRtImkb0u6T9Itkq6XdIikPSRZ0tG1fXcqaceX9cMkLZLUJ2labb+9S1kLyuuba9uOKOnzJV0h6WUl/fOSbi/LPZJ+Ucvz2XKcxZLOkJR7+ccRSb+uvT9A0r2S/ljSuyU9Vvtej6nt98eSflS+0zslTSnp55fHiS6UdI6kNcfglGKUJJhHjIISFL8LXGf7FbZ3oZrBunnZZQFweC3LdOCO2vpC4O3AdQOKfhz4c9vbAzOAb5bjTQS+COxpewdgPvA+ANsfsr2j7R2BLwGXlDxvAHYHdgC2A14LvGlVzz1GnqS9qL67/Ww/WJIv7P9ebZ9V2/084DTbr6Z68tijJf184FXA9sA6wDHEH4wE84jR8Wbgadtf60+w/T+2v1RWHwTWLq13AfsBP6ztu9j23QMLtX2b7f77UxeVMtaieiiFgPVKeRvQfB/rEcC/9RcHrA28CFgLWBN4ZLgnHKND0p8CXwfeavu/h9h3G2Ci7XkAtn9t+7fl/eUugJ/x3IVl/AFIMI8YHdsCtw6xz1zgMOANZd+nnucx3gHcZvsp288Ax1K1+JdR/ZjD2fWdJf0fYEvgagDb1wPXAMvLcqXtxc+zDjG61gIuBQ62fdeAbe8oQypzJfU/kGRr4BeSLpF0m6TTyo98PKt0r/8VcMWo1z5WmwTziNVA0lck3SHpplryRVTBvN5a7ra8bYHPAH9b1tekCuY7AZtSdbOfPCDbdGCu7d6SZyvg1VQttM2AN0v6s+d5ajG6ngH+Czh6QPr3gCllSOU/gDklfSLwp8DxVMMmrwDePSDvmVTDPz8epTrHGEgwjxgdi4Cd+1dsHwfsBby8lvYw1T/WewNXdVuwpM2B7wBH1rpddyxl/nfpRr2IqsVfN52VLxoOAW4oXbG/purm363besRq0Qf8BfBaSR/uT7T9c9v9PTlfB3Yp75dS9dbcZ7uHat7Gs3+Hkj5K9Tf4f1dD3WM1SjCPGB1XU41nH1tLW7dhv38CTupvLQ9F0kuAHwAn2/5pbdP/AttI6r9Y2BtYXMv3SmAj4PpangeBN0maWFr2b6rnifGhjHm/DXhn/x0QkibXdjmQ5763m4CNan8HbwbuLHmOAfYFjnCXv9sZ7ZHHuUaMAtuWdDDweUknAo8BvwFOGrDffzXll3QI1ezllwM/kHS77X2pZqhvBXxE0kfK7vvYXibp48B1kp4B/oeVu1ePAC7wyo98nEv1j/0CqslwV9j+3iqcdowS2ysk7Uf1/T4O7CbpQKAHWEH5rm33ltsbryoTIW+harkDfI3q7+L6cgfiJbY/sXrPJEZLHucaERHRculmj4iIaLkE84iIiJZLMI+IiGi5BPOIiIiWSzCPiIhouQTziIiIlkswj4iIaLn/D2mXEP4ka9YrAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "t_df = sg.gen_report('ENSMUSG00000020167.13', 'test', groupby='sample', layer='pi')\n",
    "sns.heatmap(data=t_df, cmap='magma')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "laden-thailand",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6/17/21\n",
    "# order_transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "sharing-lemon",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ENCODE-mouseT000008451', 'ENCODE-mouseT000008452', 'ENCODE-mouseT000008453', 'ENCODE-mouseT000008454', 'ENCODE-mouseT000008455']\n",
      "['ENSMUST00000105344.6', 'ENSMUST00000020741.10', 'ENSMUST00000020637.8', 'ENCODE-mouseT000008462', 'ENCODE-mouseT000008461']\n",
      "False\n",
      "\n",
      "['ENCODE-mouseT000008451', 'ENCODE-mouseT000008452', 'ENCODE-mouseT000008453', 'ENCODE-mouseT000008454', 'ENCODE-mouseT000008455']\n",
      "[[300000.         100000.         100000.         100000.\n",
      "  100000.              0.              0.              0.\n",
      "       0.              0.         100000.         100000.\n",
      "       0.         100000.              0.        ]\n",
      " [250000.              0.         250000.              0.\n",
      "       0.              0.              0.              0.\n",
      "       0.         250000.              0.              0.\n",
      "       0.              0.         250000.        ]\n",
      " [285714.28571429      0.              0.              0.\n",
      "       0.         142857.14285714 142857.14285714 142857.14285714\n",
      "  142857.14285714      0.              0.              0.\n",
      "  142857.14285714      0.              0.        ]]\n",
      "\n",
      "['ENSMUST00000105344.6', 'ENSMUST00000020741.10', 'ENSMUST00000020637.8', 'ENCODE-mouseT000008462', 'ENCODE-mouseT000008461']\n",
      "[[     0.         100000.              0.         100000.\n",
      "  100000.              0.              0.              0.\n",
      "       0.              0.         100000.         100000.\n",
      "  100000.         100000.         300000.        ]\n",
      " [250000.              0.              0.              0.\n",
      "       0.         250000.              0.              0.\n",
      "       0.              0.              0.              0.\n",
      "  250000.              0.         250000.        ]\n",
      " [     0.              0.         142857.14285714      0.\n",
      "       0.              0.         142857.14285714 142857.14285714\n",
      "  142857.14285714 142857.14285714      0.              0.\n",
      "       0.              0.         285714.28571429]]\n"
     ]
    }
   ],
   "source": [
    "# how can I order adata?\n",
    "tids_before = sg.t_df.index.tolist()\n",
    "tids = sg.t_df.sort_index(ascending=False).index.tolist()\n",
    "print(tids_before[:5])\n",
    "print(tids[:5])\n",
    "print(tids_before == tids)\n",
    "\n",
    "print()\n",
    "print(sg.adata.var.index.tolist()[:5])\n",
    "print(sg.adata.X[:5, :])\n",
    "print()\n",
    "sg.adata = sg.adata[:, tids]\n",
    "print(sg.adata.var.index.tolist()[:5])\n",
    "print(sg.adata.X[:5, :])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "intensive-trunk",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3, 15)\n",
      "(15,)\n",
      "[250000.         100000.         142857.14285714 100000.\n",
      " 100000.         250000.         142857.14285714 142857.14285714\n",
      " 142857.14285714 142857.14285714 100000.         100000.\n",
      " 350000.         100000.         835714.28571429]\n"
     ]
    }
   ],
   "source": [
    "print(np.shape(sg.adata.X))\n",
    "data = sg.adata.layers['tpm'].sum(axis=0)\n",
    "print(np.shape(data))\n",
    "print(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "contrary-speech",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008451</th>\n",
       "      <td>835714.285714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008453</th>\n",
       "      <td>350000.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000105344.6</th>\n",
       "      <td>250000.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008460</th>\n",
       "      <td>250000.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000020637.8</th>\n",
       "      <td>142857.142857</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  tpm\n",
       "ENCODE-mouseT000008451  835714.285714\n",
       "ENCODE-mouseT000008453  350000.000000\n",
       "ENSMUST00000105344.6    250000.000000\n",
       "ENCODE-mouseT000008460  250000.000000\n",
       "ENSMUST00000020637.8    142857.142857"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols = sg.adata.var.index.tolist()\n",
    "temp = pd.DataFrame(index=['tpm'], data=[data], columns=cols).transpose()\n",
    "temp.head()\n",
    "\n",
    "temp.sort_values(by='tpm', ascending=False, inplace=True)\n",
    "temp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "timely-fellowship",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6/17/21\n",
    "# coloring for the report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "vital-caribbean",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding transcriptome to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Transforming to str index.\n",
      "Transforming to str index.\n",
      "AnnData expects .obs.index to contain strings, but your first indices are: Int64Index([0, 1], dtype='int64'), …\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding abundance for datasets PB65_B017, PB65_B018 to SwanGraph.\n",
      "\n",
      "Adding abundance for datasets D12 to SwanGraph.\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "db = 'data/chr11_and_Tcf3_no_gname.db'\n",
    "sg.add_transcriptome(db, include_isms=True)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance_PB65.tsv'\n",
    "sg.add_abundance(ab)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance_D12.tsv'\n",
    "sg.add_abundance(ab)\n",
    "meta = 'data/chr11_and_Tcf3_metadata.tsv'\n",
    "sg.add_metadata(meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "smart-superior",
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_col = 'sample'\n",
    "cmap = {'GM12878': 'plum', 'K562': 'goldenrod'}\n",
    "sg.set_metadata_colors(obs_col, cmap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "freelance-compression",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['plum', 'goldenrod']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.adata.uns['sample_colors']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "strange-pharmacy",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "      <th>cluster</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PB65_B017</th>\n",
       "      <td>PB65_B017</td>\n",
       "      <td>2</td>\n",
       "      <td>GM12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PB65_B018</th>\n",
       "      <td>PB65_B018</td>\n",
       "      <td>2</td>\n",
       "      <td>GM12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D12</th>\n",
       "      <td>D12</td>\n",
       "      <td>1</td>\n",
       "      <td>K562</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             dataset cluster   sample\n",
       "index                                \n",
       "PB65_B017  PB65_B017       2  GM12878\n",
       "PB65_B018  PB65_B018       2  GM12878\n",
       "D12              D12       1     K562"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "sustained-marijuana",
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_col = 'sample'\n",
    "\n",
    "# set sample colors\n",
    "gm_blue = 'plum'\n",
    "k562_red = 'goldenrod'\n",
    "\n",
    "obs_order = list(sg.adata.obs_names)\n",
    "\n",
    "# sample type\n",
    "sample_order = sg.adata.obs.loc[obs_order, 'sample']\n",
    "sample_order = sample_order.unique().tolist()\n",
    "c_dict = {'GM12878': gm_blue,\n",
    "          'K562': k562_red}\n",
    "\n",
    "sample_colors = [c_dict[s] for s in sample_order]\n",
    "sg.adata.uns['{}_colors'.format(obs_col)] = sample_colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "integral-minneapolis",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "wrapped-helicopter",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/anndata/_core/anndata.py:1154: FutureWarning: is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead\n",
      "  if is_string_dtype(df[key]) and not is_categorical(df[key])\n",
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/anndata/_core/anndata.py:1172: ImplicitModificationWarning: Initializing view as actual.\n",
      "  \"Initializing view as actual.\", ImplicitModificationWarning\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'cluster' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'sample' as categorical\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWsAAAHPCAYAAACVy/uAAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAwkklEQVR4nO3debxkRX338c+XYRnZZFMyiggqRjF5RESi4IMsGkhQISKKiqLyQg3GaKIi5okaNTFExWBMRDGKQ0AFUQJxYxlWB2RzQ4XIIiIyioALgggz83v+qDNOz3Vm+tyu012nbn/fr9d5dfe5/euqrnvmN+dW1amjiMDMzPptndIVMDOz4Zyszcwq4GRtZlYBJ2szswo4WZuZVcDJ2sysAuuWrkDfbHf0F1vPZbz5mP0FMH+z+bHxgo1HLvPuex40cuwGP7pn5Fgra/lmG40c+7htf9ZhTWbvmp8/pGj5o1p6110s+/U9Kl2PUThZd2DjBRuz/8Lnjhx/8WVPGDn2MX/ztZFjrax79/6TkWMv+Y+PdliT2Xv0qa8pWv6objv2uNJVGJm7QczMKuBkbWZWASdrM7MKOFmbmVXAydrMrAJO1mZmFXCyNjOrgJO1mVkFnKzNzCrgZG1mVgEnazOzCjhZm5lVwMnazKwCTtZmZhVwsjYzq4CTtZlZBZyszcwq4GRtZlYBJ2szswo4WZuZVcDJ2sysAk7WZmYVcLI2M6uAk7WZWQWcrM3MKuBkbWZWASdrM7MKOFmbmVXAydrMrAJO1mZmFXCyNjOrgJO1mVkFnKzNzCrgZG1mVgEnazOzCjhZm5lVwMnazKwCTtZmZhVwsjYzq4CTtZlZBZyszcwq4GRtZlYBJ2szswo4WZuZVcDJ2sysAk7WZmYVcLI2M6uAk7WZWQWcrM3MKuBkbWZWASdrM7MKOFmbmVXAydrMrAJO1mZmFXCyNjOrgJO1mVkFnKzNzCrgZG1mVgEnazOzCjhZm5lVwMnazKwCTtZmZhVwsjYzq4CTtZlZBZyszcwq4GRtZlYBJ2szswo4WZuZVcDJ2sysAk7WZmYVcLI2M6uAk7WZWQWcrM3MKuBkbWZWASdrM7MKOFmbmVXAydrMrAJO1mZmFXCyNjOrgJO1mVkFnKzNzCrgZG1mVgEnazOzCjhZm5lVwMnazKwCTtZmZhVwsjYzq4CTtZlZBZyszcwq4GRtZlYBJ2szswo4WZuZVcDJ2sysAk7WZmYVcLI2M6uAk7WZWQWcrM3MKuBkbWZWASdrM7MKOFmbmVXAydrMrAJO1mZmFXCyNjOrgJO1mVkFnKzNzCqwbukKzAV33/MgLr7sCSPH3/jCj4xe+AtHDwV49KmvyYp/zN98La8CU2zDMy4fOXbfM3bqriIjeAx1/t7vjHtKV2FkPrM2M6uAk7WZWQWcrM3MKuBkbWZWASdrM7MKOFmbmVXAydrMrAJO1mZmFXCyNjOrgJO1mVkFnKzNzCrgZG1mVgEnazOzCjhZm5lVwMnazKwCTtZmZh2Q9AlJt0v6zsC+LSSdK+n65nHzgZ+9VdINkv5X0r7DPt/J2sysG58E9pux72hgUUTsACxqXiNpR+AQ4AlNzIclzVvbhztZm5l1ICIuBu6asfsAYGHzfCFw4MD+z0TEbyPiB8ANwK5r+3wnazOz8dk6IpYANI8PbfY/HPjRwPtubfatke/BaGZTa9+9Noo771o29H1Xf/u33wXuG9h1QkSckFG0VrMv1hbgZG1mU+uOu5Zx+dnbDH3fegtuvC8idhmhiJ9KWhARSyQtAG5v9t8KPGLgfdsAt63tg9wNYmZTLFgWy4duGc4CDmueHwacObD/EEkbSNoe2AG4Ym0f5DNrM5taASxfe+9Da5I+DewJbCXpVuAdwDHAaZIOB24BDgaIiO9KOg34HrAUeG1ErLU/xsnazKZWEDyw9hzZ/rMiXrSGH+2zhvf/E/BPbT/fydrMplpXZ9bj5mRtZlMrgGVO1mZm/RbAA3kDiBPjZG1mU62OVO1kbWZTLAh3g5iZ9V0EPFBHrnayNrNpJpat9srv/nGyNrOpFcByn1mbmfVbAPdXsuqGk7WZTbXl4W4QM7NeSxfFOFmbmfVaIJa5G8TMrN/SFYxO1mZmPSeWVZKsW9VS0oaS3ibpY83rHSQ9e7xVMzMbr7Se9TpDtz5oW4sTgd8CT2te3wr841hqZGY2IRHi/pg3dOuDtsn60RHxXuABgIj4Dau/4aOZWVWWo6FbH7Tts75f0oNo7r4r6dGkM20zs2qlqXv96OYYpm2yfgfwFeARkk4BdgdePq5KmZlNQiAeiDrmWbSqZUScK+nrwFNJ3R+vj4g7xlozM7MJWDYXrmCUtPOMXUuax20lbRsRXx9PtczMxm8uXRRz7Fp+FsDeHdbFzGyi0kUxc6AbJCL2mlRFzMwmLdDc6AZZQdJ84Ejg6aT/jC4BPhIR942xbmZmY9eXi16GaXv+fxJwN/Ch5vWLgP8CDh5HpczMJiGCai43b5us/zAinjjw+gJJ3xpHhczMJiVN3evHFYrDtP0v5RuSnrrihaQ/ARYPC5K0u6QLJF0o6SJJB0n6pKQLB95zlKSfNc93k/QdST8Z+PmTJV3SxJ8mab1m/19JuqLZntPs+3BT1oWSfiNpC0lbN3W4SNIZktZv+Z3NbAosY52hWx+0rcWfAJdKulnSzcBlwDMkXSPp26sLkLQl8GHgxRGxJ/BM4Lbmx/Mlbd083x34UfP8u8CupLVHVvgxsG9EPAO4ATiw2X8ksFvzuX8HEBFHNmW9FLgsIu4CXgKc2sR/D/izlt/ZzOa4QCyP4VsftO0G2W+Ez94fOCMilgBExAPAZZJeDZwBHCjpy6TE/PDmPb8EkFY2TkT8ZOAzHwCWNs9vAB4EbALcOaPs5wOnN8+vBf64eb4Z8LMRvouZzUE1Td1rdWYdET8EfgU8GNhyxRYRP2x+tjoLaC6ikbR30zVxVvOzc0lnxM8HPt+mDpK2bWK+0Oz6CulM+Urg32a8/XkDn3sV8BJJ3wEeB1zepjwzmwZiWYutD9quZ/1u4NukpHhss71/SNhtrDxjPr/pnnhY87P7gXuA5wAXtSh/U9Lsk1dExAPN61cBO5AS8HvUnI5L2gZYOnBG/ibg2Ij4I+CLwBFtvrOZzX0BLI91hm5tSPobSd9txt0+LWl+M252rqTrm8fNR61r2/P/F5CWSb1/Fp/9JeBCSR+JiNskzRxy/QSwc0QsHez2mKmJOwV4V0R8v9m9HLiPtPLfUmAD0polwapdICvc1Tz+AnjoLL6Dmc1hEd3MBpH0cOCvgR0j4jeSTgMOAXYEFkXEMZKOBo4G3jJKGW2T9XdI/b23t/3giLhT0muAT0kKUlI9jtSVQURcDFw8GCPp8aS53I+VdB7wRtKX3Q3YRNLbgOMj4lRJp5MGOucB/xERy5uPOYhV53//O7BQ0puAZaQGNDMDOp1nvS7wIEkPABuSehfeCuzZ/HwhcCFjTtb/TJq+9x0G1rGOiOeuLSgiFrOyoiucvJr37dI8XkuTzAd8C/j0amLez2q6YiLi/854/SO8homZrUa6rVd+n3RE/FjS+4FbgN8A50TEOZK2HphksUTSyH/Zt03WC4F/Aa4hdUGYmVUvEA8sb9UNspWkqwZenxARJ6x40fRFHwBsT+pu/aykQ7usa9tkfUdEzJxxYWZWvZYXvdyxogdgDZ4J/CAiVlzg93lS9+1PJS1ozqoXMIuu5JnaJuurJf0zcBardoN4PWszq9aKi2I6cAvwVEkbkrpB9iFNG74HOAw4pnk8c9QC2ibrJzWPTx3Y5/Wszax6Xay6FxGXN5Mevk6aTPEN4ARgY+A0SYeTEvrIi9+1va2X17U2szknAh5Y3s1skIh4B+l+tYN+SzrLztb6OktJ+wNPAOYPVO5dXVTCzKyE1A3Sj4Wahml784GPkOYN7gX8J+nCkyvGWC8zs4noy+Xkw7T9L2W3iHgZ8POIeCfwNOAR46uWmdn4BWLp8nlDtz5o2w2y4vZd90p6GOny7e3HUyUzs8np4qKYSWibrP9H0mbA+0ijnQF8bFyVMjObhHRbr7mVrK8DlkXE5yTtCOwM/PfYamVmNgErukFq0LbP+m0RcbekpwPPAj4JHD+2WpmZTchyNHTrg7bJelnzuD/wkYg4E/C9DM2samk967l1W68fS/oo6fr3f5G0Ae0TvZlZP8Xc6wZ5AXA2sF9E/ALYAnjzuCplZjYJK5ZIraEbpO3l5vcycK/EZn3WJeOqlJnZpPSlm2OYOm7ra2Y2BgEs7WhtkHFzsjazqdXhEqlj52RtZlOtL33SwzhZm9n0CvdZm5n1nvuszcwq4D5rM7NKhJO1mVm/RcDSuXSnGDOzucpn1mZmvec+azOz3gtgmWeDmJn1XKR+6xo4WZvZVPMVjGZmPRfI3SBmZjVwN4iZWQU8dc/MrOcinKzNzKqwbLmTtZlZ7/nM2sys5wJVk6zrmLNiZjYOzc0Hhm1tSNpM0umSrpN0raSnSdpC0rmSrm8eNx+1qk7WZjbdosXWzgeBr0TE44AnAtcCRwOLImIHYFHzeiRO1mY21SI0dBtG0qbAHsDH02fG/RHxC+AAYGHztoXAgaPW033WZja1AljebjbIVpKuGnh9QkScMPD6UcDPgBMlPRG4Gng9sHVELAGIiCWSHjpqXZ2szWx6BdCuT/qOiNhlLT9fF9gZeF1EXC7pg2R0eayOu0HMbKpFDN9auBW4NSIub16fTkreP5W0AKB5vH3UejpZm9kUE7F8+DZMRPwE+JGkP2x27QN8DzgLOKzZdxhw5qg1dTeImU237hZyeh1wiqT1gZuAV5BOiE+TdDhwC3DwqB/uZG1m06vDtUEi4pvA6vq19+ni852szWy6eYlUM7MKVHK5uZO1mU03n1mbmfVc+3nWxTlZz/CvB65t3vuqtjv6iycAbPf0g/iLLe8aucydXvGXI8fm2iQz/sknPTYr/qzFj8+sQb3WvXnkKbc2ovWu//1/p7G8QEVG4GSd4eZj9n8VwC5PPOiI93zgESN/ztUv26mrKk3ce078Zlb8JWe8upuKVGjDsy8f/ibr1M3xjd/f6TNrM7P+k/uszcx6LgS+rZeZWQV8Zm1mVgEnazOzngvcDWJmVgMPMJqZ1aCSZO31rM3MKuAzazObanKftZlZzwXVdIM4WZvZdHOyNjPrP3khJzOzCvjM2sys3xSeZ21mVgfPBjEz6z+fWZuZ1cDJ2sys58KzQczM6uAzazOz/nOftZlZDZyszcx6zvOszcwq4WRtZlaBSpK1bz5gZlNLpKl7w7bWnyfNk/QNSV9oXm8h6VxJ1zePm49aVydrM5tesXJ9kLVts/B64NqB10cDiyJiB2BR83okTtZmNt2ixdaCpG2A/YH/HNh9ALCweb4QOHDUarrP2symWstujq0kXTXw+oSIOGHGe44DjgI2Gdi3dUQsAYiIJZIeOmo9nazNbLq1O3O+IyJ2WdMPJT0buD0irpa0ZzcVW5WTtZlNr+7uwbg78FxJfw7MBzaVdDLwU0kLmrPqBcDtoxbgPmszm2pdzAaJiLdGxDYRsR1wCHB+RBwKnAUc1rztMODMUevpM2szm2pjvoLxGOA0SYcDtwAHj/pBTtZmNt06TtYRcSFwYfP8TmCfLj7XydrMppbvwWhmVgsnazOz/vOZtZlZDZyszcx6zvdgNDOrhM+szcz6z33WZmYVcDeImVnfdbc2yNg5WZvZdHOyNjPrtxW39aqBk7WZTTVFHafWTtZmNr3cZ21mVgd3g5iZVcDzrM3MauBkbWbWc17P2sys/zx1z8ysFp66Z2bWf+4GMTPruwAtK12JdpyszWy6+czazKz/3A1iZtZ3AVpeR7Z2sjaz6VZHrnayNrPpJdwNYmbWfxHuBjEzq0IdudrJ2symm7tBzMz6LoBKukHWKV0BM7OStHz4NvQzpEdIukDStZK+K+n1zf4tJJ0r6frmcfNR6+lkbWbTLWL4NtxS4I0R8XjgqcBrJe0IHA0siogdgEXN65E4WZvZVFMM34aJiCUR8fXm+d3AtcDDgQOAhc3bFgIHjlpP91mb2dRS+ysYt5J01cDrEyLihNV+prQd8CTgcmDriFgCKaFLeuiodXWyNrPp1u7mA3dExC7D3iRpY+BzwBsi4leSMiu3krtBzGyqKWLo1upzpPVIifqUiPh8s/unkhY0P18A3D5qPZ2szWx6RaSpe8O2IZROoT8OXBsRHxj40VnAYc3zw4AzR62qu0HMbKp1dFHM7sBLgWskfbPZ93fAMcBpkg4HbgEOHrUAJ2szm24d3IMxIr5KWhdqdfbJLgAnazObZgFaVscVjE7WZjbd6sjVTtZmNt3azvYozcnazKZXAO4GMTPrN9F+HnVpTtZmNt2crM3MKuBkbWbWc566Z2ZWCZ9Zm5n1XeubCxTnZG1m08tT98zM6uCpe2ZmNXCyNjPruQhY1u5WMaU5WZvZdPOZtZlZBZyszcx6LgKWLStdi1acrM1suvnM2sys54JWN8TtAydrM5tuPrM2M+u7gOWeumdm1m+Bk7WZWRXcDWJm1ne+gtHMrP8CIpyszcz6z1P3zMx6zlcwmplVwgOMZmb9F566Z2bWcxWtZ71O6Qq0IWk7Sac3z/9Q0lWSzpN0paQLJb154L1vkLRI0sWStpX0ZEmXSLpI0mmS1iv3Tcysd2L58K0FSftJ+l9JN0g6uutqVnVmLWkb4GTgxcDfAW+IiO8M/PzJwLYRsc/AvvuBfSPiXknvAQ4EPjvRiptZLwUQHcwGkTQP+A/gWcCtwJWSzoqI72V/eKOKM+vGFsDpwGsi4vukdv6YpHMlPbF5zwHAfEnnS/qQpHkR8ZOIuLf5+QPA0slX3cx6KaKrM+tdgRsi4qaIuB/4DCkfdaamZL0zcHtEXN28flNEPA14HfDRZt/WwNKI2Bu4Fzh4RbCkbYFnAl+YXJXNrO9i2bKhWwsPB3408PrWZl9nauoGOQ9YIukfIuIfIuJOgIi4TtKKP0N+AVzavP98YHcASZsC/wW8IiIeWFshf/HUrTXbil397d/eMW/BDT9cy1u2Au5Y849vWNvHD4kdaqzx8xbklv3mtf2w5Hfvdbv3uOzc+HGX/cjBF3fz87PPi9O3avG58yVdNfD6hIg4YeD16vJGp3MCa0rWAG8A/lvSS4EzI+JXkh4KbBARyyQtBp4EnNk83tQk8VOAdzXdJ52LiIes7eeSroqIXUb57JzY0vGuu+ve97IjYr9Ry5rhVuARA6+3AW7r6LOBypJ1k5BfRDrLfmUzs2Me8MbmLV8EniPpQuBnwEuAFwC7AZtIehtwfEScOvHKm9lcdiWwg6TtgR8Dh5AmQnSmimQdETcDz2+e/xp46hretww4YsbuTzebmdlYRMRSSX8FnE06gfxERHy3yzKqSNZzwAnD3zKW2NLxrnuZeNe9gIj4EvClcX2+opLr4s3MpllNU/fMzKaWk7WZWQWcrM3MKuBkPSGSDm35vg0lvaRZgGozSUdJeq2k+S1i15V0iKR/k3Ry8/hCSa0GkiXtMlCHt0j6pKS3S9qsZfxzmseHSjq+WWTrE5K2axF7rKT/26acNcRvL+kDko6W9AeSPi3pS5JWO3NoNfHzJB3ctNlJkt4rabeWsZtJel3T9vMl/bOkD0t6zCzKPkjSAc11ASv2H7y2uCGfeWzL92058PxASf8g6VBJWblhFsf7yN9d0iaS3irpfyR9VdJZzXG7SU7d+8oDjB2TtOPqdpOueNq9Rfz/kOZsbgjsCXwO+CWwd0QcMiT2v4Brga80MZsB+wKPj4iXtij7wojYU9JC4DLSVaBPAl4WEfu3iL8gIvaS9DngY038zsB7miUA1hZ7HXAJaVrmhcBnI+LiYWUOxF8EvBvYHDgGOAj4FXByRAxNupI+AXwP+AZpWYIHAT8HfhMR7x0Sew5wGqm9Xwq8oyn7HyJijxZlfwa4kbR2zT7AyyPiRknnD2u3Jv524BZgOSuvpNsB+H5E7Dok9vyI2FvSPwEPAT4P7AE8LCJe3qLs3ON95O8u6SzStNyzWfV4PyQinjus7OpEhLcON9I/0k8AJ87Ybm0Zf8HA868PPF/UIvbi2exfzfsWNY/nrqlOQ+LPI00H/Qqw7sD+i9p+b9Ic1WeRpmBdA/x7y7IvHnh+7cDzC1vGnz/zuzSP57SIvXDg+TUjtNvg73xb0n9We82s01riX0Ra5OyFA/u+PJvvPfN31OZ31ryvy+N9Vt8d+Ops9te+eZ51974HvDmatUtWkPTFlvGD620fNfB83sw3rsbXJJ0EnEP6R7Qp8KfA11qWfYKkTwG3Svo8cAGwEyvXWxnm3aQzs18CiyV9FdiRtAJZK5EubDoXOLf5s3ivlqE3NXUH+Lqkj5HOjH/cMv42Se8EvgnsDXyr2d9m/fO7lZbf3RC4XdLfA3cB97Qse56k+RFxX0Tc0nQnnQL8cZvgiPh0c4b6Ykn/TTrLb7vGzR9JOo109d2DIuI3zf6NWsbnHu8zv/tzScsgt/nup0s6j5TgfwU8GHgG6T+uOcfdIB2TtDFwb4x4f3tJTwCua5LWin3rA8+KiKH/ACTtBDyN9CfhXcDlEfHNWZS/ANiPtILhL4DFEXHNLOI3IF3evyL+ioi4q0XcUyLiyrblrCZepP9YbiUt5LNv86Nz2vwumj7a5wGPInUlfSEiQtLWEfHTIbHzgT8Drif9SX9Y86NTIuLuFmXvTDoTvX1GfV4QEa3/oxuIOxTYMSKGLoAvaXBhoyURcX9zDD89Ir7SIj73eM/67pK2Ap5COt5/DlwVETkLQfWWk/UYSdoaeDxwU0TcMu54SdtExK3NwX4w6az2RuBTETHRdbxzvvuk263r+Bwl6z5K7MAxJ1YeczfR8pjLjV/DZz4zIs4bJbbPPBukY80gH5JeCJxFOktdKOmvM+Nf1yL85ObxX0lnmYuBBaQ/qUcm6fiW7xus+5nM4ruPsd1GiZ9V3dfymaO022x/51nfPbdsVh5zx5EGky9ldsfcyPFKs5ZmbhsBb21Zdl1Kd5rPtY2BARtg4+b5POCyccezclBs0Yz9F/T9u5dsty7iKz5mcsvOOuZy4kk3GDmfNLZywcDzO8f9OyuxeYCxe9tIOhLYKtIKgURa2nUS8f8j6f3AdZKOAxaRpt5dP8vv8DuSjo2INw5/J5BX95LtlhWvNKf5vyPikraFdVV2B/G5Zececznx1wEHxIxxAUnntiy7Ku4G6d4/kWYBvFfpDjUoTdL/8rjjI+KDwKdIA2zzSXfK+QbwmjYFS7pd6c7xVyjdOf5K4HBJV4y77pmxpeP3B14m6Rqle38OnVvdo7pnlZ17zGXGP580P3umkS8m6jMPME6ApCdFxDdKxM8mVunGDgeRLkg5tdn35Yj4s1HKnm35XcZOMl4rLwaaR5r2dzBpRs5FEfFX4yx7HPG1tPu0cTdIx5o/KVfZBfylpA9HxIfHGb+aWIAj25YdefN1u677xNqti3gYfY54z46Z3LJhFsdcbvwaPvMjEdHqzL4m7gbp3muBZ5MGP+4Bfk26DPjXE4ifGXvPLMsmklNIc47XJV0k0laXdZ9ku+XGHzVzR0Qsi/bTx/p0zOSWPdtjLvuYnWkuJmrAs0G63khnJi8GziBdBrwO8KVJxOeWvYbPPKyGuvcg/uHA/Ob5rsAf1VD3HrTbOI7ZQ3Pi+7q5z3pMlC5MeQnwQuChMWRBnS7jR41Vuux4lV2kAZ+vRsQLxl1+bmypeEn/AuxCumLzTtLVm/eQrkR9V5/r3oeyR41X5iJStXGyHrPmINw6IpZMOn62sZI+CvwB8H7gh6SznE+Szq5/OO7yu4qddLykSyNiN6WlaP8XeExEhKSLo8Wqezlldx1fWbv/irQOyMxxlWdFxDajlN9rpU/t59pGmsZ1KelqsGeSljv9FnDEuONzy24+45HAh4F/b57P5k/aYnUv3O6DK/69aeD5BRXUveZ2/xqw5Wr2f7HtMVvTVrwCc21rDqBNSH2YtwIbkwbqLh93fG7ZMz5re+CjwNmT+O4l262Dur8MmDdj3/rA31dQ95rbfWNgndke27VunrrXPZFGsoM0qv3biFgqqe2qZDnxuWX/TkT8AHj1LMNK1r1YfESc9LsPkR4GPBa4OSL+se91L1x2Vnw0V1yu8mHSYRGxsGXZVXGy7t5/Ad8hXS77z8AVkn4DfHYC8VllS3oBaRGc75LmC/8daZ3gd0fEWX2ue8l4SZ+PiOdJehVpZsOlwBMlLY6I9/S57oXLzopf04C4pP1jFgPitfAA45g1l+4uj4i2C9F3Fj/bWEmXky7k2IQ0v/pxwH2kxXZmfX/ESda9ZLyk8yLimZIuAPaJZm1nSZdMut1y4ytr904HxPvOZ9Yda2YEPJ+0AP8WpBsALJb0uWi3vu/I8bllk+43eC9wr6RFEfHL5nNbrStcsu6F42+Q9GzS4NhzJK24d+XQGw+UrnvN7R4Rr1a6ecJbSF0o7yPdCGHOJWrwmXXnlH/T2pHjOyj7OOCN8ft3qflURDy/53Uv2e7rA68jrQX9CNJ8668C742BO6D0tO7VtvuMz9keOBp4ZETs1zauKqVHOOfaRv5Na0eOzy275Hcv2W6l267mY6Zw3Tcb9++mT5u7QbqXe9PanPissiXtExGLJG0GvBN4InAD8I6IaHPj2WJ1Lxkv6eHA24E/Ii3cvwz4NvCPE2i33Phq2510k+NLSRfGfD5a/BVTM3eDjIFW3rT2waSbeM72prUjx2fGnh8Re0s6mfRn6emku0X/bUTsu/bosnUvGa90h+23R8SlA/t2B94VEfv0ue6ly86JbwZ0X0rq834eaW3r04HPzcXE7VX3xuP2ZrsT+Bmw1rtjdxyfWzbAwyPi5Ii4LyLOBjaYUPkl2y0n/kHA5TP2XdHsH3fZXcTX2u4REbdGxHGRLut/GelipLbTBqviZN0xSW8BPg5sSxqh3h74hKTfW0az6/jcskm3eLoS+ANJmzefuT7pSrFe171w/AdJMxj+U9IHJH2cdE/D4/pe98rb/drBFxHx44j4YEQ8o03Z1SndaT7XNtY8YHLJuONzy15D7LqkVdB6XfcexK8LPJ705/zjgHWn4ZgpHT9NmwcYu3eHpFfy+wMmd0wgPqvsNc15BT7X97qXjFe6K8wBwFLgC9FMfZR0cES0+ZO82mOmZPzAwO4fk3oJZjuwWxUPMHZM0kbAEaQzrM1ICe9y4GPR7qqskeM7KDt3zmzJupds988AN5IGuPYBXh4RN64YsO153Wtu9+yB3Zo4WY+JpC1JZ6d3RsRdk4wfNVZrWH95Tfu7Lj83tlS8mhvmNs+3BU4iTX18W5tkXbLufSh71HhJi4E9YtWLuNYj3ah4t9nWoe/cDdIxSfsA7yL9GfdLYPNmsO4dEbFonPG5ZdPBPO1SdS8cP0/S/EizZ26R9FzgZNKf50PVfMwUjl8xsPsd0vH6YNK4wXHDyq1S6U7zubaR+ng3mbFvU2DxuONzy27evxPwl6TV914D7DSJ716y3Tqo+87MGIQl9aEeUkHdq2335r0jD+zWtvnMunvLgIew6iI+D2n2jzs+t2wiXYzwzcF9kv44Iq4Zc/kl2y0rPiK+Lulhkp7HyoHZyyLiM+Muu4P4atu9g4HdqrjPumOSHg+8B1gAiDR3dAnw/yLi2rXF5sbnlr2Wzz0nIv60z3Uv3O5HAXuR1gBfMTD7TOD8iHhfz+tec7tnDezWxsnafkfSbaRbK4l05w6a5ztExGal6tV3axmYHWk9a2unq4HdWrgbpGPKXNQnJz63bNLlvntExH0zPvfcFrFF6144Pnd+e7XHTOH4rIHd6pTuNJ9rG3AesNuMfbsDi8Yd30HZT2PGYE+z/3EV1L1ku28EvAE4FTgb+HTzeqMK6l5zu2cN7Na2+cy6e7mL+uTEZ5UdEZcBSNqCZqAsIu6KiOvaxGeWX7LdsuIjXbxxXHNR0Shzjas9ZkrGR/7AblWcrLu3prmf/zqB+KyyJe0NvJv05/uvgM1mOU+7WN1Lxncwv73aY6Zk/GoGdrcHjmgGGIcO7NbGA4xjoLTGxg6sXJ/3xmh3D8Ts+MzYxcB+EXH3wL5NgS9HxO59rnvJ+NLtlhtfcbtP1cCuz6w7popvfkrmnNmSdS8cX6zdcuMrb/fcRaSq4jPrjqnum5/mzpmdyhu3lmy3Dupec7tnLSJVndIjnHNto+Kbn5b87iXbrXTb1XzMlI5v3rslqRtli3H/rkpu7gbpXrU3P+1gnvZU3ri1cLvlxtfc7rkDu1VxN0jHJG0IPJaVf5r9HPhatL956cjxHZSdtT5w4bqXbPdi7dZB3Wtu9+yB3aqUPrWfaxtpPYgi8R2UvRiYN2PfesClFdS9ZLsXa7cpb/eLgUfN2PdoJtTtN+nN3SA2KHfO7LRyu5XxauBYSTMHdl9dtFZj4m6Qjkn6FXAdv78YUkTEruOMzy27+Yyc+brF6t6D+CLtlhvfg3bLPmanhc+su3d1NCuBFYjPKruDedrF6l4yvnC75cbX3O65A7tVcbK2QSeS5rx+klXnvJ4IDJ0zO8XcbmUsZPUDuyeR1reeW0p3ms+1DViveRRwEHAk6X/9nccd30HZuXNmS9a9ZLsXa7cpb/esgd3atnVWl8BtdBHxQPN0IfBI4KWRbjf03nHH55ZNM+dV0qGSnts8nkTLObMl6144vli75cZX3u4rBnb/U9IHJH0cuIg5OrDrZD0+D4uIDwC/aV5rgvEjxUbEUcAHgE2AHUnrNH+g2T8bE697yfietFtufI3tfhrwdOBY4DRSgt8j5uD9F8F91uN0t6Q9SXez2B34xQTjc2Jvb7alpDvH/HQWsV2UX7LdcuJLt1tufHXt3sHAblU8dW9MJG0JHE0607oWeE/MYkH6nPhRYyW9BdiTlZf+bka68esFEdG2K6VI3UvG96HdcuMrbfesRaSqU7rTfK5uwPrN4zqkP9U2nFT8qLGseaDskr7XvWR8H9rN7T58f+2b+6zH55zm8Z2k6Vuz7UfLiR819g5Jr5S0jaRNm8dXMvv1gUvUvWR8H9otN77Gds8a2K2Nk/X4rOhf2iYiXk26sm1S8aPGvpS06tmxpH8s72teHzqLsnPKz40tFd+HdsuNr67do7uB3Sp4gHF8fi3pNOAiSSvWLZhU/EixkX/j16zyO4gtEt+TdsuNr67dG10M7FbBA4xjImk94BERcVPzfLuIuH4S8aPGajXrAzfbO2IW6wOXqHvJ+D60W258pe3eycBuLZysx0TSiaz80w6AiHjlJOJHjVVH6wOXqHvJ+D60W258pe3uG+ZaJ45pHgU8GZjtYug58aPGZt34tYPyc2NLxfeh3XLja2x33zDXuifpnIj40xLxbWO18savf8DKJStb3/g1t/yuYycV38d2y42vpN2n6oa5PrMeE0nvY+WfddsC904qPiN2r4j4i9aV7L78ou2WEV+83XLja2z3Dgd2q+Az6zGR9IzmaQA/j4hrJhU/aqyk8yNi79nUs8vyc2NLxfeh3XLjK233TgZ2a+FkPSZK6xa8mHRPuOuBz8Qs1ivIiR81VtL3geNW97OI+HC7mpepe8n4PrRbbnyl7T5VN8z1RTHjczJpycdLgUc1rycVP2rsMuDXwD2r2WajRN1Lxveh3XLja233h8zYN8rAbh2iB9e8z8WNNNdz8PWFk4ofNXZmXInvXrLdRo3vQ7tNabs/HjgDuIx0ifllwOdJCzll/z76tnmAcXzulvQq4EpgV1ad1jXu+FFjPzW7KnZefm5sqfg+tFtufI3t3snAbi3cDTI+h5Lmfb6KtGbBSyYYP2rswyQ9GEDS/pK+JelKSQfNouyc8nNjS8X3od1y42ts9+fPsoyqeYBxjCQ9hjSlCICIuGJS8aPESloczcCMpGuBPUgXG5wfsxywmXTdS8b3pd1y4yts904GdmvhbpAxkXQ66QKJH7PyQonZ/MMZOT4jNpr4xwJLIuJnzetZ3XWjUN1Lxhdvt9z4Stt9xcDubG8hViUn6/HZOCL2KxQ/auxlSiufPYrm8l9JW6w9pNPyc2NLxfeh3XLja2z3n0TESRllVsXdIGMi6UhgHvBtmjOviLh4EvGZsTsC90XETc3rzYEHR8TNfa97yfjS7ZYbX2O7SzoiIj7Wtoza+cx6fPYlXVX1sOZ1AK0P3sz4kWIlvX3gOaQ7TV8TEV9pWW5W+R3EFonvSbvlxlfX7jQDuxHxS0n7k9ZnuR84JiI+N4uyq+BkPT4bRMTLCsWPGnv5jNfzgWdJenZE/NUEys+NLRXfh3bLja+x3f80It7ZPH8/AwO7gJO1tXanpLey6p91X5pQ/EixEXH2anafKemCluVmld9BbJH4nrRbbnx17b7ifbkDu7Vwsh6f7wPrA7s0rwOYzcGbE59b9u9Imt981myUrHvpeKBIu+XGl263UeK7GtitggcYx0zSnhFxYYn42cZK+iwrl6mE9Of8NsDbI+IL4y6/q9hJx/ep3XLja2r35v3ZA7u1cLIeM2Uun5kTP9tYSY+cseu+iBj5BqSTrHvJ+D61W258Ze3+9hm7Rh3YrYK7QcbvRwXjZxv7hJk7mtkNs+1/HLX8rmInHd+ndsuNr6nduxrYrYLPrDsm6VEDf5JtGBH3Ns//T0R8e5zxHZT9AHA1cB5pCtSKK8MiIt7V87qXbPdi7dZB3att97V85gURsdcosb0WPVj6by5tpPUg1vp8XPEdlL018JekaU8nAs8F1p/Edy/Zbh3UvVi7TXO7r+Hz5gOLR4nt++ZukO6pxfNxxWeVHamf9XjgeEk7AZ8A/gJ4RZv4zPJLtltWfOF2y42vtt3XNrDbsuyqOFl3byOlu12vA2w4+HwC8VllNwNlBwFPJ92d+83AhS3rXbTuJeMLt1tufLXtDrxpxuusgd2+c591xySduKafRcTQM62c+A7KXgZcBXwVWM7AWUtEHNUivmTdS7Z7sXbLja+83f98LbEjXVfQZ07WYyRpnYhYXiJ+lFhJe7Lqn5W/ExEXjbv8LmJLxPel3XLjK2z3rIHd2jhZd0zSRsC/AE8BlgLrkdblPToifj3O+A7KXi8iHhh4/SzSn6TnRIsDpXDdS7Z7sXbroO41t/vWwPOAZ5LWBDkD+EpE3D+s3CqVHuGcaxvwQeDwGfteDnxo3PEdlH0JsEnz/EOkgbL3Ah+voO4l271Yu01zu8+I2Qn4OnDibOJq2opXYK5twLlr2H/euOM7KPui5nF94MaB/RdUUPeS7V6s3aa83R8J/C3pjub/AewDzGtTbo2bZ4N0L/cWQznxuWWvJ2lTYG9g8cD+DSZQfsl2y40v2W658TW3+02sOrC7L7CvJKLFwG5tnKy7t6ukKxgY7GieP3YC8bll/z3wZdJgzSsBlG5i+s2W8SXrXjK+ZLvlxtfc7vuwhoHducgDjGZWpdyB3do4WY+BpB2AF5BuUXQbcGpE3DCJ+MzYC1h5pqKB5xER+/S57iXjS7dbbnzF7X4J8OcRcbekDwEbAXcAW0bE4W3Lr8U6pSsw10h6PvAp4CfAmaQr2k5p9o81PrdsYD/gz5rtz0mzGjYG/rfvdS8cX6zdcuMrb/flTaJen5S0X9n0VT+qTdnVKT3COdc20mDHRjP2bQR8ddzxuWU3718POIK0/OS7gYdM4ruXbLeO4ou02zS3O3ApsClwIHDS4P62bV/T5gHG7i2LiHsGd0TEPc0lyeOOzypb0ptJa1ycCuw987PGXH7JdsuKL9xuufHVtjv5A7tVcTdI974n6ShJ60EaBGn+MV87gfjcso8kTTd7MXCBpCskXdmM1ve97iXjS7Zbbny17R4R50fE7hGxV0T8oNl3Q0Qc2bLsqjhZd+8NpD/Nvi3pBuBbwCbA6ycQn1v2yRHxpIh4CrAwInZtni8eFtiDupeML9luufEly86Kl3SBpPObbfD5opZlV8WzQex3NHD/uzU9t9/nditD0uBFR+uQuqL+GrhqLp5d+8y6Y5I+JukZM/btIemEccfnlp2rZN1Lx+eo+ZgpGR8RvyVdufgy0vrhfwjsPxcTNeDZIF1vwCVr2H/RuOM7KPt24DTgszOe/7SCupds92LtNuXt/mbga8DfMGNGyVzcPBtkcuYVjG8b+5SMMroov+vYScX3sd1y42to9yOBX5AGdl+kdEd5kS5G2jWz/N5xN0j3zpV0vKRHNiPb20k6HjhnAvFZZUfED9e09b3uJeMLt1tufLXtTv7Abl1Kn9rPtY30P/sRwCLge8C5wOE0g7njjM8tu+R3L9lupduu5mOmcN07vTN63zfPBjGzKk3bLBwn645p1UV9VtHmAMqJzy07V8m6l47PUfMxU7jut5NmgQh4xsDzPSJi62Fl18YDjN3bb+D5KnM/JxCfW3auknUvHZ+j5mOmZPy4Bnb7qXQ/zFzcyFjUJzc+t+yS371ku5Vuu5qPmdLx07L5zLpjylzUJyc+t+xcJeteOj5HzcdM6fhp4j7rjkn6AWnu51JmLEgfLeZ+5sTnlp2rZN1Lx+eo+ZgpHT9VSp/az7UNePfA89cOPP/Xccfnll3yu5dst9JtV/MxUzp+mjZfFNO93QeeHzTw/IkTiM8tO1fJupeOz1HzMVM6fmo4WZuZVcB91h3LnfuZE1963mnJupeOz1HzMVM6fpp4Nkj3cud+5sSXnndasu6l40uW7XafAj6zNjOrgPuszcwq4GRtZlYBJ2szswo4WZuZVcDJ2sysAv8fRpWVyTRdZHYAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 338.4x432 with 3 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "GridSpec(2, 4, height_ratios=[0, 6], width_ratios=[0.2, 4.5, 0, 0.2])"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# plot heatmap? \n",
    "var_names = sg.adata.var.index.tolist()\n",
    "sg.adata.X = sg.adata.layers['pi']\n",
    "sc.pl.heatmap(sg.adata, var_names, groupby='sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "printable-blanket",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6/16/21 \n",
    "# calculating tss abundance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "alien-wilderness",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding transcriptome to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Transforming to str index.\n",
      "Transforming to str index.\n",
      "AnnData expects .obs.index to contain strings, but your first indices are: Int64Index([0, 1], dtype='int64'), …\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding abundance for datasets PB65_B017, PB65_B018 to SwanGraph.\n",
      "hewwo?\n",
      "                        PB65_B017  PB65_B018\n",
      "ENCODE-mouseT000008451        3.0        1.0\n",
      "ENCODE-mouseT000008452        1.0        0.0\n",
      "ENCODE-mouseT000008453        1.0        1.0\n",
      "ENCODE-mouseT000008454        1.0        0.0\n",
      "ENCODE-mouseT000008455        1.0        0.0\n",
      "\n",
      "Adding abundance for datasets D12 to SwanGraph.\n",
      "hewwo?\n",
      "                        D12  PB65_B017  PB65_B018\n",
      "ENCODE-mouseT000008451  2.0        3.0        1.0\n",
      "ENCODE-mouseT000008452  0.0        1.0        0.0\n",
      "ENCODE-mouseT000008453  0.0        1.0        1.0\n",
      "ENCODE-mouseT000008454  0.0        1.0        0.0\n",
      "ENCODE-mouseT000008455  0.0        1.0        0.0\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "db = 'data/chr11_and_Tcf3_no_gname.db'\n",
    "sg.add_transcriptome(db, include_isms=True)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance_PB65.tsv'\n",
    "sg.add_abundance(ab)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance_D12.tsv'\n",
    "sg.add_abundance(ab)\n",
    "meta = 'data/chr11_and_Tcf3_metadata.tsv'\n",
    "sg.add_metadata(meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "deluxe-valley",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_end_adata(sg,kind):\n",
    "    \"\"\"\n",
    "    Create a tss / tes-level adata object. Enables calculating tss / tes\n",
    "    usage across samples.\n",
    "\n",
    "    Parameters:\n",
    "        kind (str): Choose from 'tss' or 'tes'\n",
    "    \"\"\"\n",
    "\n",
    "    end_exp_df = swan.get_ends(sg.t_df, kind)\n",
    "\n",
    "    # get a mergeable transcript expression df\n",
    "    tid = sg.adata.var.index.tolist()\n",
    "    obs = sg.adata.obs.index.tolist()\n",
    "    data = sg.adata.layers['counts'].transpose()\n",
    "    t_exp_df = pd.DataFrame(columns=obs, data=data, index=tid)\n",
    "    cols = ['gid', 'gname']\n",
    "    t_exp_df = t_exp_df.merge(sg.t_df[cols], left_index=True, right_index=True)\n",
    "#     print(t_exp_df.head()) \n",
    "\n",
    "    # merge counts per transcript with end expression\n",
    "    end_exp_df = end_exp_df.merge(t_exp_df, how='left',\n",
    "        left_index=True, right_index=True)\n",
    "    \n",
    "    # sum across the same gene\n",
    "    end_exp_df.reset_index(drop=True, inplace=True)\n",
    "    end_exp_df = end_exp_df.groupby(['vertex_id', 'gid', 'gname']).sum()\n",
    "    print(end_exp_df.head())\n",
    "    \n",
    "\n",
    "    # sort based on vertex id\n",
    "    end_exp_df.sort_values(by='vertex_id', inplace=True, ascending=True)\n",
    "    end_exp_df.reset_index(inplace=True)\n",
    "    end_exp_df.index = end_exp_df.gid + '_' + end_exp_df.vertex_id.astype('str')\n",
    "    end_exp_df.index.name = '{}_id'.format(kind)\n",
    "#     end_exp_df['{}_id'.format(kind)] = end_exp_df.tid\n",
    "#     end_exp_df.set_index(['gid', 'gname', 'vertex_id'], inplace=True)\n",
    "    print(end_exp_df)\n",
    "\n",
    "    # obs, var, and X tables for new data\n",
    "    var_cols = ['vertex_id', 'gid', 'gname']\n",
    "    var = end_exp_df[var_cols]\n",
    "    end_exp_df.drop(var_cols, axis=1, inplace=True)\n",
    "    X = end_exp_df.transpose().values\n",
    "    obs = sg.adata.obs\n",
    "    print(obs)\n",
    "    print(obs.shape)\n",
    "    print(X)\n",
    "    print(X.shape)\n",
    "    print(var.shape)\n",
    "    print(var)\n",
    "\n",
    "    if kind == 'tss':\n",
    "\n",
    "        # create end-level adata object\n",
    "        sg.tss_adata = anndata.AnnData(var=var, obs=obs, X=X)\n",
    "\n",
    "        # add counts and tpm as layers\n",
    "        sg.tss_adata.layers['counts'] = sg.tss_adata.X\n",
    "        sg.tss_adata.layers['tpm'] = swan.calc_tpm(sg.tss_adata, sg.t_df).to_numpy()\n",
    "        sg.tss_adata.layers['pi'] = swan.calc_pi(sg.tss_adata, sg.t_df)[0].to_numpy()\n",
    "\n",
    "        if sg.has_abundance():\n",
    "\n",
    "            # some cleanup for unstructured data\n",
    "            sg.edge_adata.uns = sg.tss_adata.uns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "velvet-spirituality",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                         PB65_B017  PB65_B018  \\\n",
      "vertex_id gid                    gname                                          \n",
      "1         ENSMUSG00000020167.13  ENSMUSG00000020167.13         3.0        1.0   \n",
      "6         ENSMUSG00000020167.13  ENSMUSG00000020167.13         1.0        0.0   \n",
      "38        ENSMUSG00000020167.13  ENSMUSG00000020167.13         0.0        1.0   \n",
      "57        ENSMUSG00000020457.12  ENSMUSG00000020457.12         2.0        1.0   \n",
      "61        ENCODE-mouseG000002988 ENCODE-mouseG000002988        1.0        0.0   \n",
      "\n",
      "                                                         D12  \n",
      "vertex_id gid                    gname                        \n",
      "1         ENSMUSG00000020167.13  ENSMUSG00000020167.13   2.0  \n",
      "6         ENSMUSG00000020167.13  ENSMUSG00000020167.13   0.0  \n",
      "38        ENSMUSG00000020167.13  ENSMUSG00000020167.13   0.0  \n",
      "57        ENSMUSG00000020457.12  ENSMUSG00000020457.12   0.0  \n",
      "61        ENCODE-mouseG000002988 ENCODE-mouseG000002988  0.0  \n",
      "                           vertex_id                     gid  \\\n",
      "tss_id                                                         \n",
      "ENSMUSG00000020167.13_1            1   ENSMUSG00000020167.13   \n",
      "ENSMUSG00000020167.13_6            6   ENSMUSG00000020167.13   \n",
      "ENSMUSG00000020167.13_38          38   ENSMUSG00000020167.13   \n",
      "ENSMUSG00000020457.12_57          57   ENSMUSG00000020457.12   \n",
      "ENCODE-mouseG000002988_61         61  ENCODE-mouseG000002988   \n",
      "ENCODE-mouseG000002989_63         63  ENCODE-mouseG000002989   \n",
      "ENSMUSG00000020368.14_66          66   ENSMUSG00000020368.14   \n",
      "ENCODE-mouseG000002990_67         67  ENCODE-mouseG000002990   \n",
      "ENSMUSG00000020368.14_83          83   ENSMUSG00000020368.14   \n",
      "ENSMUSG00000020368.14_97          97   ENSMUSG00000020368.14   \n",
      "ENSMUSG00000020368.14_101        101   ENSMUSG00000020368.14   \n",
      "\n",
      "                                            gname  PB65_B017  PB65_B018  D12  \n",
      "tss_id                                                                        \n",
      "ENSMUSG00000020167.13_1     ENSMUSG00000020167.13        3.0        1.0  2.0  \n",
      "ENSMUSG00000020167.13_6     ENSMUSG00000020167.13        1.0        0.0  0.0  \n",
      "ENSMUSG00000020167.13_38    ENSMUSG00000020167.13        0.0        1.0  0.0  \n",
      "ENSMUSG00000020457.12_57    ENSMUSG00000020457.12        2.0        1.0  0.0  \n",
      "ENCODE-mouseG000002988_61  ENCODE-mouseG000002988        1.0        0.0  0.0  \n",
      "ENCODE-mouseG000002989_63  ENCODE-mouseG000002989        1.0        0.0  0.0  \n",
      "ENSMUSG00000020368.14_66    ENSMUSG00000020368.14        0.0        0.0  1.0  \n",
      "ENCODE-mouseG000002990_67  ENCODE-mouseG000002990        0.0        0.0  1.0  \n",
      "ENSMUSG00000020368.14_83    ENSMUSG00000020368.14        0.0        0.0  1.0  \n",
      "ENSMUSG00000020368.14_97    ENSMUSG00000020368.14        1.0        0.0  0.0  \n",
      "ENSMUSG00000020368.14_101   ENSMUSG00000020368.14        1.0        1.0  2.0  \n",
      "             dataset cluster   sample\n",
      "index                                \n",
      "PB65_B017  PB65_B017       2  GM12878\n",
      "PB65_B018  PB65_B018       2  GM12878\n",
      "D12              D12       1     K562\n",
      "(3, 3)\n",
      "[[3. 1. 0. 2. 1. 1. 0. 0. 0. 1. 1.]\n",
      " [1. 0. 1. 1. 0. 0. 0. 0. 0. 0. 1.]\n",
      " [2. 0. 0. 0. 0. 0. 1. 1. 1. 0. 2.]]\n",
      "(3, 11)\n",
      "(11, 3)\n",
      "                           vertex_id                     gid  \\\n",
      "tss_id                                                         \n",
      "ENSMUSG00000020167.13_1            1   ENSMUSG00000020167.13   \n",
      "ENSMUSG00000020167.13_6            6   ENSMUSG00000020167.13   \n",
      "ENSMUSG00000020167.13_38          38   ENSMUSG00000020167.13   \n",
      "ENSMUSG00000020457.12_57          57   ENSMUSG00000020457.12   \n",
      "ENCODE-mouseG000002988_61         61  ENCODE-mouseG000002988   \n",
      "ENCODE-mouseG000002989_63         63  ENCODE-mouseG000002989   \n",
      "ENSMUSG00000020368.14_66          66   ENSMUSG00000020368.14   \n",
      "ENCODE-mouseG000002990_67         67  ENCODE-mouseG000002990   \n",
      "ENSMUSG00000020368.14_83          83   ENSMUSG00000020368.14   \n",
      "ENSMUSG00000020368.14_97          97   ENSMUSG00000020368.14   \n",
      "ENSMUSG00000020368.14_101        101   ENSMUSG00000020368.14   \n",
      "\n",
      "                                            gname  \n",
      "tss_id                                             \n",
      "ENSMUSG00000020167.13_1     ENSMUSG00000020167.13  \n",
      "ENSMUSG00000020167.13_6     ENSMUSG00000020167.13  \n",
      "ENSMUSG00000020167.13_38    ENSMUSG00000020167.13  \n",
      "ENSMUSG00000020457.12_57    ENSMUSG00000020457.12  \n",
      "ENCODE-mouseG000002988_61  ENCODE-mouseG000002988  \n",
      "ENCODE-mouseG000002989_63  ENCODE-mouseG000002989  \n",
      "ENSMUSG00000020368.14_66    ENSMUSG00000020368.14  \n",
      "ENCODE-mouseG000002990_67  ENCODE-mouseG000002990  \n",
      "ENSMUSG00000020368.14_83    ENSMUSG00000020368.14  \n",
      "ENSMUSG00000020368.14_97    ENSMUSG00000020368.14  \n",
      "ENSMUSG00000020368.14_101   ENSMUSG00000020368.14  \n",
      "hewwo?\n",
      "                           D12  PB65_B017  PB65_B018\n",
      "ENSMUSG00000020167.13_1    2.0        3.0        1.0\n",
      "ENSMUSG00000020167.13_6    0.0        1.0        0.0\n",
      "ENSMUSG00000020167.13_38   0.0        0.0        1.0\n",
      "ENSMUSG00000020457.12_57   0.0        2.0        1.0\n",
      "ENCODE-mouseG000002988_61  0.0        1.0        0.0\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Value passed for key 'pi' is of incorrect shape. Values of layers must match dimensions (0, 1) of parent. Value had shape (3, 0) while it should have had (3, 11).",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-7f036532a3ac>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mcreate_end_adata\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msg\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'tss'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-4-fa297552f8e5>\u001b[0m in \u001b[0;36mcreate_end_adata\u001b[0;34m(sg, kind)\u001b[0m\n\u001b[1;32m     59\u001b[0m         \u001b[0msg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtss_adata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlayers\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'counts'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtss_adata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     60\u001b[0m         \u001b[0msg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtss_adata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlayers\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'tpm'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mswan\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcalc_tpm\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtss_adata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mt_df\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_numpy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 61\u001b[0;31m         \u001b[0msg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtss_adata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlayers\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'pi'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mswan\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcalc_pi\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtss_adata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mt_df\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_numpy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     62\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     63\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0msg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhas_abundance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/anndata/_core/aligned_mapping.py\u001b[0m in \u001b[0;36m__setitem__\u001b[0;34m(self, key, value)\u001b[0m\n\u001b[1;32m    148\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    149\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setitem__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mV\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 150\u001b[0;31m         \u001b[0mvalue\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_validate_value\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    151\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    152\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/anndata/_core/aligned_mapping.py\u001b[0m in \u001b[0;36m_validate_value\u001b[0;34m(self, val, key)\u001b[0m\n\u001b[1;32m     50\u001b[0m                 \u001b[0mright_shape\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparent\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0ma\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0maxes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     51\u001b[0m                 raise ValueError(\n\u001b[0;32m---> 52\u001b[0;31m                     \u001b[0;34mf\"Value passed for key {key!r} is of incorrect shape. \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     53\u001b[0m                     \u001b[0;34mf\"Values of {self.attrname} must match dimensions \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     54\u001b[0m                     \u001b[0;34mf\"{self.axes} of parent. Value had shape {val.shape} while \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Value passed for key 'pi' is of incorrect shape. Values of layers must match dimensions (0, 1) of parent. Value had shape (3, 0) while it should have had (3, 11)."
     ]
    }
   ],
   "source": [
    "create_end_adata(sg, 'tss')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "thermal-survey",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6/15/21\n",
    "# calculating edge abundance\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "thermal-ticket",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding transcriptome to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Transforming to str index.\n",
      "Transforming to str index.\n",
      "AnnData expects .obs.index to contain strings, but your first indices are: Int64Index([0, 1], dtype='int64'), …\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding abundance for datasets PB65_B017, PB65_B018 to SwanGraph.\n",
      "\n",
      "Adding abundance for datasets D12 to SwanGraph.\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "db = 'data/chr11_and_Tcf3_no_gname.db'\n",
    "sg.add_transcriptome(db, include_isms=True)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance_PB65.tsv'\n",
    "sg.add_abundance(ab)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance_D12.tsv'\n",
    "sg.add_abundance(ab)\n",
    "meta = 'data/chr11_and_Tcf3_metadata.tsv'\n",
    "sg.add_metadata(meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "tested-following",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[    0.        , 11904.76190476, 11904.76190476, 11904.76190476,\n",
       "        35714.28571429, 11904.76190476, 11904.76190476,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        , 11904.76190476,\n",
       "        11904.76190476, 23809.52380952, 23809.52380952, 11904.76190476,\n",
       "        11904.76190476, 11904.76190476, 11904.76190476, 23809.52380952,\n",
       "        23809.52380952, 23809.52380952, 23809.52380952, 23809.52380952,\n",
       "        23809.52380952, 23809.52380952, 23809.52380952, 23809.52380952,\n",
       "        23809.52380952, 11904.76190476, 11904.76190476, 11904.76190476,\n",
       "        11904.76190476,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        , 11904.76190476, 11904.76190476,\n",
       "            0.        , 11904.76190476,     0.        , 11904.76190476,\n",
       "        11904.76190476,     0.        , 11904.76190476, 23809.52380952,\n",
       "        23809.52380952,     0.        , 23809.52380952, 23809.52380952,\n",
       "            0.        , 23809.52380952, 23809.52380952, 23809.52380952,\n",
       "        23809.52380952, 23809.52380952, 23809.52380952, 23809.52380952,\n",
       "        23809.52380952, 23809.52380952, 23809.52380952, 23809.52380952,\n",
       "        23809.52380952, 11904.76190476, 11904.76190476, 11904.76190476,\n",
       "        11904.76190476,     0.        ],\n",
       "       [12820.51282051,     0.        , 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051,     0.        ,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051,     0.        ,\n",
       "            0.        , 12820.51282051,     0.        , 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        , 12820.51282051, 12820.51282051,\n",
       "            0.        , 12820.51282051,     0.        , 12820.51282051,\n",
       "            0.        ,     0.        , 12820.51282051, 12820.51282051,\n",
       "        12820.51282051,     0.        , 12820.51282051, 12820.51282051,\n",
       "            0.        , 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051,     0.        ],\n",
       "       [    0.        ,     0.        ,     0.        ,     0.        ,\n",
       "        26666.66666667,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        , 13333.33333333, 13333.33333333, 13333.33333333,\n",
       "        13333.33333333, 13333.33333333, 40000.        , 40000.        ,\n",
       "        40000.        , 40000.        ,     0.        , 40000.        ,\n",
       "        13333.33333333, 26666.66666667, 13333.33333333, 26666.66666667,\n",
       "            0.        , 13333.33333333, 26666.66666667, 40000.        ,\n",
       "        40000.        , 13333.33333333, 26666.66666667, 26666.66666667,\n",
       "        13333.33333333, 26666.66666667, 26666.66666667, 26666.66666667,\n",
       "        26666.66666667, 26666.66666667, 26666.66666667, 26666.66666667,\n",
       "        26666.66666667, 26666.66666667, 26666.66666667, 26666.66666667,\n",
       "        26666.66666667, 26666.66666667, 26666.66666667, 26666.66666667,\n",
       "        13333.33333333, 13333.33333333]])"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.edge_adata.layers['tpm']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dependent-charm",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sg.create_edge_adata(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "distributed-listening",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         edge_id\n",
      "edge_id         \n",
      "96            96\n",
      "98            98\n",
      "95            95\n",
      "94            94\n",
      "97            97\n",
      "['96', '98', '95', '94', '97']\n"
     ]
    }
   ],
   "source": [
    "print(sg.edge_adata.var.head())\n",
    "print(sg.edge_adata.var.index.tolist()[:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "frozen-pathology",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "      <th>cluster</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PB65_B017</th>\n",
       "      <td>PB65_B017</td>\n",
       "      <td>2</td>\n",
       "      <td>GM12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PB65_B018</th>\n",
       "      <td>PB65_B018</td>\n",
       "      <td>2</td>\n",
       "      <td>GM12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D12</th>\n",
       "      <td>D12</td>\n",
       "      <td>1</td>\n",
       "      <td>K562</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             dataset cluster   sample\n",
       "index                                \n",
       "PB65_B017  PB65_B017       2  GM12878\n",
       "PB65_B018  PB65_B018       2  GM12878\n",
       "D12              D12       1     K562"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.edge_adata.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "excessive-singing",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[    0.        , 11904.76190476, 11904.76190476, 11904.76190476,\n",
       "        35714.28571429, 11904.76190476, 11904.76190476,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        , 11904.76190476,\n",
       "        11904.76190476, 23809.52380952, 23809.52380952, 11904.76190476,\n",
       "        11904.76190476, 11904.76190476, 11904.76190476, 23809.52380952,\n",
       "        23809.52380952, 23809.52380952, 23809.52380952, 23809.52380952,\n",
       "        23809.52380952, 23809.52380952, 23809.52380952, 23809.52380952,\n",
       "        23809.52380952, 11904.76190476, 11904.76190476, 11904.76190476,\n",
       "        11904.76190476,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        , 11904.76190476, 11904.76190476,\n",
       "            0.        , 11904.76190476,     0.        , 11904.76190476,\n",
       "        11904.76190476,     0.        , 11904.76190476, 23809.52380952,\n",
       "        23809.52380952,     0.        , 23809.52380952, 23809.52380952,\n",
       "            0.        , 23809.52380952, 23809.52380952, 23809.52380952,\n",
       "        23809.52380952, 23809.52380952, 23809.52380952, 23809.52380952,\n",
       "        23809.52380952, 23809.52380952, 23809.52380952, 23809.52380952,\n",
       "        23809.52380952, 11904.76190476, 11904.76190476, 11904.76190476,\n",
       "        11904.76190476,     0.        ],\n",
       "       [12820.51282051,     0.        , 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051,     0.        ,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051,     0.        ,\n",
       "            0.        , 12820.51282051,     0.        , 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        , 12820.51282051, 12820.51282051,\n",
       "            0.        , 12820.51282051,     0.        , 12820.51282051,\n",
       "            0.        ,     0.        , 12820.51282051, 12820.51282051,\n",
       "        12820.51282051,     0.        , 12820.51282051, 12820.51282051,\n",
       "            0.        , 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051, 12820.51282051, 12820.51282051, 12820.51282051,\n",
       "        12820.51282051,     0.        ],\n",
       "       [    0.        ,     0.        ,     0.        ,     0.        ,\n",
       "        26666.66666667,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ,     0.        ,     0.        ,     0.        ,\n",
       "            0.        , 13333.33333333, 13333.33333333, 13333.33333333,\n",
       "        13333.33333333, 13333.33333333, 40000.        , 40000.        ,\n",
       "        40000.        , 40000.        ,     0.        , 40000.        ,\n",
       "        13333.33333333, 26666.66666667, 13333.33333333, 26666.66666667,\n",
       "            0.        , 13333.33333333, 26666.66666667, 40000.        ,\n",
       "        40000.        , 13333.33333333, 26666.66666667, 26666.66666667,\n",
       "        13333.33333333, 26666.66666667, 26666.66666667, 26666.66666667,\n",
       "        26666.66666667, 26666.66666667, 26666.66666667, 26666.66666667,\n",
       "        26666.66666667, 26666.66666667, 26666.66666667, 26666.66666667,\n",
       "        26666.66666667, 26666.66666667, 26666.66666667, 26666.66666667,\n",
       "        13333.33333333, 13333.33333333]])"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.edge_adata.X\n",
    "sg.edge_adata.layers['tpm']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "sacred-attitude",
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://towardsdatascience.com/dealing-with-list-values-in-pandas-dataframes-a177e534f173\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "maritime-think",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def pivot_path_list(t_df, path_col):\n",
    "    df = pd.DataFrame([[tid, x] for tid, path in zip(t_df.index, t_df[path_col]) for x in path])\n",
    "    if path_col == 'path':\n",
    "        df.columns = ['tid', 'edge_id']\n",
    "    elif path_col == 'loc':\n",
    "        df.columns = ['tid', 'loc_id']\n",
    "#     df.set_index('tid', inplace=True)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "exclusive-passing",
   "metadata": {},
   "outputs": [],
   "source": [
    "# edge_exp_df = swan.pivot_path_list(sg.t_df, 'path')\n",
    "# edge_exp_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "traditional-adobe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                        PB65_B017  PB65_B018  D12\n",
      "ENCODE-mouseT000008451        3.0        1.0  2.0\n",
      "ENCODE-mouseT000008452        1.0        0.0  0.0\n",
      "ENCODE-mouseT000008453        1.0        1.0  0.0\n",
      "ENCODE-mouseT000008454        1.0        0.0  0.0\n",
      "ENCODE-mouseT000008455        1.0        0.0  0.0\n"
     ]
    }
   ],
   "source": [
    "tid = sg.adata.var.index.tolist()\n",
    "obs = sg.adata.obs.index.tolist()\n",
    "data = sg.adata.layers['counts'].transpose()\n",
    "t_exp_df = pd.DataFrame(columns=obs, data=data, index=tid)\n",
    "print(t_exp_df.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "united-polls",
   "metadata": {},
   "outputs": [],
   "source": [
    "# edge_exp_df.loc[edge_exp_df.edge_id == 20]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "single-hammer",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                        edge_id  PB65_B017  PB65_B018  D12\n",
      "ENCODE-mouseT000008451       97        3.0        1.0  2.0\n",
      "ENCODE-mouseT000008452       92        1.0        0.0  0.0\n",
      "ENCODE-mouseT000008452       93        1.0        0.0  0.0\n",
      "ENCODE-mouseT000008452       94        1.0        0.0  0.0\n",
      "ENCODE-mouseT000008452       95        1.0        0.0  0.0\n"
     ]
    }
   ],
   "source": [
    "# merge with transcript expression data\n",
    "edge_exp_df = swan.pivot_path_list(sg.t_df, 'path')\n",
    "edge_exp_df = edge_exp_df.merge(t_exp_df, how='left', left_index=True, right_index=True)\n",
    "\n",
    "print(edge_exp_df.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "noble-ordinary",
   "metadata": {},
   "outputs": [],
   "source": [
    "# groupby edge id and sum counts\n",
    "# edge_t_df.reset_index(inplace=True)\n",
    "edge_exp_df = edge_exp_df.groupby('edge_id').sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "medium-distinction",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # edge_t_df.set_index('edge_id', inplace=True)\n",
    "# edge_exp_df.head()\n",
    "# edge_exp_df.loc[20]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "realistic-elder",
   "metadata": {},
   "outputs": [],
   "source": [
    "# edge_t_df.transpose().head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "threaded-stockholm",
   "metadata": {},
   "outputs": [],
   "source": [
    "# edge_t_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "european-certification",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sg.adata.obs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "architectural-fight",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         PB65_B017  PB65_B018  D12\n",
      "edge_id                           \n",
      "96             0.0        1.0  0.0\n",
      "98             1.0        0.0  0.0\n",
      "95             1.0        1.0  0.0\n",
      "94             1.0        1.0  0.0\n",
      "97             3.0        1.0  2.0\n",
      "[ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True]\n"
     ]
    }
   ],
   "source": [
    "edge_exp_df = edge_exp_df.merge(sg.edge_df[['v1', 'v2']], \n",
    "    how='left', left_index=True, right_index=True)\n",
    "edge_exp_df.sort_values(by=['v1', 'v2'], inplace=True)\n",
    "edge_exp_df.drop(['v1', 'v2'], axis=1, inplace=True)\n",
    "print(edge_exp_df.head())\n",
    "print(edge_exp_df.index == sg.edge_df.index)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "statewide-payday",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PB65_B017</th>\n",
       "      <th>PB65_B018</th>\n",
       "      <th>D12</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>edge_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         PB65_B017  PB65_B018  D12\n",
       "edge_id                           \n",
       "96             0.0        1.0  0.0\n",
       "98             1.0        0.0  0.0\n",
       "95             1.0        1.0  0.0\n",
       "94             1.0        1.0  0.0\n",
       "97             3.0        1.0  2.0"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edge_exp_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "aging-montana",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Transforming to str index.\n"
     ]
    }
   ],
   "source": [
    "var = edge_exp_df.index.to_frame()\n",
    "X = edge_exp_df.transpose().values\n",
    "obs = sg.adata.obs\n",
    "adata = anndata.AnnData(obs=obs, X=X, var=var)\n",
    "adata.layers['counts'] = adata.X\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "prompt-handle",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         edge_id\n",
      "edge_id         \n",
      "96            96\n",
      "98            98\n",
      "95            95\n",
      "94            94\n",
      "97            97\n",
      "...          ...\n",
      "23            23\n",
      "22            22\n",
      "21            21\n",
      "20            20\n",
      "49            49\n",
      "\n",
      "[102 rows x 1 columns]\n",
      "edge_id    int64\n",
      "dtype: object\n",
      "[96, 98, 95, 94, 97]\n"
     ]
    }
   ],
   "source": [
    "print(var)\n",
    "print(var.dtypes)\n",
    "print(var.index.tolist()[:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "wooden-square",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['96', '98', '95', '94', '97']\n",
      "[96, 98, 95, 94, 97]\n",
      "[ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True]\n"
     ]
    }
   ],
   "source": [
    "print(adata.var.index.tolist()[:5])\n",
    "print(sg.edge_df.index.tolist()[:5])\n",
    "print(adata.var.index.astype('int') == sg.edge_df.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "regulation-maine",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # def calc_total_counts(adata, obs_col='dataset'):\n",
    "# # \t\"\"\"\n",
    "# # \tCalculate cumulative expression per adata entry based on condition given\n",
    "# # \tby `obs_col`. Default column to use is `adata.obs` index column, `dataset`.\n",
    "\n",
    "# # \tParameters:\n",
    "# # \t\tadata (anndata AnnData): Annotated data object from the SwanGraph\n",
    "# # \t\tobs_col (str): Column name from adata.obs table to group on.\n",
    "# # \t\t\tDefault: 'dataset'\n",
    "\n",
    "# # \tReturns:\n",
    "# # \t\tdf (pandas DataFrame): Pandas DataFrame where rows are the different\n",
    "# # \t\t\tconditions from `obs_col` and the columns are transcript ids in the\n",
    "# # \t\t\tSwanGraph, and values represent the cumulative counts per isoform\n",
    "# # \t\t\tper condition.\n",
    "\n",
    "# # \t\"\"\"\n",
    "# # \tadata.X = adata.layers['counts']\n",
    "# # \tdf = pd.DataFrame(data=adata.X, index=adata.obs[obs_col].tolist(), \\\n",
    "# # \t\tcolumns=adata.var.index.tolist())\n",
    "\n",
    "# # \t# add up values on condition (row)\n",
    "# # \tdf = df.groupby(level=0).sum()\n",
    "# # \t# df = df.transpose()\n",
    "\n",
    "# # \treturn df\n",
    "\n",
    "# def calc_tpm(adata, sg_df, obs_col='dataset'):\n",
    "# \t\"\"\"\n",
    "# \tCalculate the TPM per condition given by `obs_col`.\n",
    "# \tDefault column to use is `adata.obs` index column, `dataset`.\n",
    "\n",
    "# \tParameters:\n",
    "# \t\tadata (anndata AnnData): Annotated data object from the SwanGraph\n",
    "# \t\tsg_df (pandas DataFrame): Pandas DataFrame from SwanGraph that will\n",
    "# \t\t\tbe used to order the rows of resultant TPM DataFrame\n",
    "# \t\tobs_col (str): Column name from adata.obs table to group on.\n",
    "# \t\t\tDefault: 'dataset'\n",
    "\n",
    "# \tReturns:\n",
    "# \t\tdf (pandas DataFrame): Pandas datafrom where rows are the different\n",
    "# \t\t\tconditions from `obs_col` and the columns are transcript ids in the\n",
    "# \t\t\tSwanGraph, and values represent the TPM value per isoform per\n",
    "# \t\t\tcondition.\n",
    "# \t\"\"\"\n",
    "\n",
    "# \t# calculate cumulative counts across obs_col\n",
    "# \tid_col = adata.var.index.name\n",
    "# \tconditions = adata.obs[obs_col].unique().tolist()\n",
    "# \tdf = swan.calc_total_counts(adata, obs_col=obs_col)\n",
    "# \tprint(df.head())\n",
    "# \tprint(df.dtypes)\n",
    "# \tdf = df.transpose()\n",
    "# \tprint(df.head())\n",
    "# \tprint(df.dtypes)\n",
    "#     # we use ints to index edges and locs\n",
    "# \tif id_col == 'vertex_id' or id_col == 'edge_id':\n",
    "# \t\tdf.index = df.index.astype('int')\n",
    "# \tprint(df.index.tolist()[:5])\n",
    "\n",
    "# \t# calculate tpm per isoform per condition\n",
    "# \ttpm_cols = []\n",
    "# \tfor c in conditions:\n",
    "# \t\tcond_col = '{}_tpm'.format(c)\n",
    "# \t\ttotal_col = '{}_total'.format(c)\n",
    "# \t\tdf[total_col] = df[c].sum()\n",
    "# \t\tdf[cond_col] = (df[c]*1000000)/df[total_col]\n",
    "# \t\ttpm_cols.append(cond_col)\n",
    "\n",
    "# \t# formatting\n",
    "# \tdf.index.name = id_col\n",
    "# \tdf = df[tpm_cols]\n",
    "# \tfor col in tpm_cols:\n",
    "# \t\tnew_col = col[:-4]\n",
    "# \t\tdf.rename({col: new_col}, axis=1, inplace=True)\n",
    "\n",
    "# \t# reorder columns like adata.obs\n",
    "# \tdf = df[adata.obs[obs_col].unique().tolist()]\n",
    "# \tdf = df.transpose()\n",
    "\n",
    "# \t# reorder in adata.var / t_df order\n",
    "# # \tprint(df.head())\n",
    "# # \tprint(id_col)\n",
    "# # \tprint(sg_df.head())\n",
    "# # \tprint(len(df.columns))\n",
    "# # \tprint(len(sg_df[id_col].unique()))\n",
    "# \tprint(set(df.columns.tolist()) == set(sg_df[id_col].tolist()))\n",
    "# \tprint(df.columns.tolist()[:5])\n",
    "# \tprint(sg_df[id_col].tolist()[:5])\n",
    "# \tdf = df[sg_df[id_col].tolist()]\n",
    "\n",
    "# \treturn df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "median-bleeding",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>edge_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>edge_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         edge_id\n",
       "edge_id         \n",
       "96            96\n",
       "98            98\n",
       "95            95\n",
       "94            94\n",
       "97            97"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "plain-departure",
   "metadata": {},
   "outputs": [],
   "source": [
    "tpm_df = swan.calc_tpm(adata, sg.edge_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "gothic-poster",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>edge_id</th>\n",
       "      <th>96</th>\n",
       "      <th>98</th>\n",
       "      <th>95</th>\n",
       "      <th>94</th>\n",
       "      <th>97</th>\n",
       "      <th>93</th>\n",
       "      <th>92</th>\n",
       "      <th>91</th>\n",
       "      <th>90</th>\n",
       "      <th>89</th>\n",
       "      <th>...</th>\n",
       "      <th>28</th>\n",
       "      <th>27</th>\n",
       "      <th>26</th>\n",
       "      <th>25</th>\n",
       "      <th>24</th>\n",
       "      <th>23</th>\n",
       "      <th>22</th>\n",
       "      <th>21</th>\n",
       "      <th>20</th>\n",
       "      <th>49</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PB65_B017</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>11904.761905</td>\n",
       "      <td>11904.761905</td>\n",
       "      <td>11904.761905</td>\n",
       "      <td>35714.285714</td>\n",
       "      <td>11904.761905</td>\n",
       "      <td>11904.761905</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>23809.523810</td>\n",
       "      <td>23809.523810</td>\n",
       "      <td>23809.523810</td>\n",
       "      <td>23809.523810</td>\n",
       "      <td>23809.523810</td>\n",
       "      <td>11904.761905</td>\n",
       "      <td>11904.761905</td>\n",
       "      <td>11904.761905</td>\n",
       "      <td>11904.761905</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PB65_B018</th>\n",
       "      <td>12820.512821</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12820.512821</td>\n",
       "      <td>12820.512821</td>\n",
       "      <td>12820.512821</td>\n",
       "      <td>12820.512821</td>\n",
       "      <td>12820.512821</td>\n",
       "      <td>12820.512821</td>\n",
       "      <td>12820.512821</td>\n",
       "      <td>12820.512821</td>\n",
       "      <td>...</td>\n",
       "      <td>12820.512821</td>\n",
       "      <td>12820.512821</td>\n",
       "      <td>12820.512821</td>\n",
       "      <td>12820.512821</td>\n",
       "      <td>12820.512821</td>\n",
       "      <td>12820.512821</td>\n",
       "      <td>12820.512821</td>\n",
       "      <td>12820.512821</td>\n",
       "      <td>12820.512821</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D12</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>26666.666667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>26666.666667</td>\n",
       "      <td>26666.666667</td>\n",
       "      <td>26666.666667</td>\n",
       "      <td>26666.666667</td>\n",
       "      <td>26666.666667</td>\n",
       "      <td>26666.666667</td>\n",
       "      <td>26666.666667</td>\n",
       "      <td>26666.666667</td>\n",
       "      <td>13333.333333</td>\n",
       "      <td>13333.333333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 102 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "edge_id              96            98            95            94  \\\n",
       "PB65_B017      0.000000  11904.761905  11904.761905  11904.761905   \n",
       "PB65_B018  12820.512821      0.000000  12820.512821  12820.512821   \n",
       "D12            0.000000      0.000000      0.000000      0.000000   \n",
       "\n",
       "edge_id              97            93            92            91  \\\n",
       "PB65_B017  35714.285714  11904.761905  11904.761905      0.000000   \n",
       "PB65_B018  12820.512821  12820.512821  12820.512821  12820.512821   \n",
       "D12        26666.666667      0.000000      0.000000      0.000000   \n",
       "\n",
       "edge_id              90            89  ...            28            27  \\\n",
       "PB65_B017      0.000000      0.000000  ...  23809.523810  23809.523810   \n",
       "PB65_B018  12820.512821  12820.512821  ...  12820.512821  12820.512821   \n",
       "D12            0.000000      0.000000  ...  26666.666667  26666.666667   \n",
       "\n",
       "edge_id              26            25            24            23  \\\n",
       "PB65_B017  23809.523810  23809.523810  23809.523810  11904.761905   \n",
       "PB65_B018  12820.512821  12820.512821  12820.512821  12820.512821   \n",
       "D12        26666.666667  26666.666667  26666.666667  26666.666667   \n",
       "\n",
       "edge_id              22            21            20            49  \n",
       "PB65_B017  11904.761905  11904.761905  11904.761905      0.000000  \n",
       "PB65_B018  12820.512821  12820.512821  12820.512821      0.000000  \n",
       "D12        26666.666667  26666.666667  13333.333333  13333.333333  \n",
       "\n",
       "[3 rows x 102 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tpm_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "controlled-deviation",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>edge_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>edge_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         edge_id\n",
       "edge_id         \n",
       "0              0\n",
       "1              1\n",
       "2              2\n",
       "3              3\n",
       "4              4"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "sapphire-crisis",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "      <th>cluster</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PB65_B017</th>\n",
       "      <td>PB65_B017</td>\n",
       "      <td>2</td>\n",
       "      <td>GM12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PB65_B018</th>\n",
       "      <td>PB65_B018</td>\n",
       "      <td>2</td>\n",
       "      <td>GM12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D12</th>\n",
       "      <td>D12</td>\n",
       "      <td>1</td>\n",
       "      <td>K562</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             dataset cluster   sample\n",
       "index                                \n",
       "PB65_B017  PB65_B017       2  GM12878\n",
       "PB65_B018  PB65_B018       2  GM12878\n",
       "D12              D12       1     K562"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "improved-dining",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 1., 1., 1., 2., 2.,\n",
       "        1., 1., 1., 1., 1., 1., 1., 1., 2., 2., 2., 2., 2., 2., 2., 2.,\n",
       "        2., 2., 2., 2., 2., 2., 2., 2., 0., 0., 0., 1., 0., 0., 0., 0.,\n",
       "        0., 0., 1., 1., 1., 0., 0., 0., 0., 0., 1., 1., 0., 0., 0., 0.,\n",
       "        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1.,\n",
       "        0., 3., 1., 1., 1., 0.],\n",
       "       [0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 0., 0., 1., 1.,\n",
       "        0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "        1., 1., 1., 1., 1., 1., 1., 1., 0., 0., 0., 1., 0., 0., 0., 0.,\n",
       "        0., 0., 1., 1., 1., 0., 0., 0., 0., 0., 1., 0., 1., 1., 1., 1.,\n",
       "        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "        1., 1., 0., 0., 0., 0.],\n",
       "       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "        0., 0., 0., 0., 1., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2.,\n",
       "        2., 2., 2., 2., 2., 2., 3., 3., 1., 1., 1., 3., 3., 3., 3., 3.,\n",
       "        1., 1., 2., 2., 2., 1., 1., 1., 1., 1., 0., 0., 0., 0., 0., 0.,\n",
       "        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "        0., 2., 0., 0., 0., 1.]], dtype=float32)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "abroad-climate",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENCODE-mouseT000008451\n",
      "[97]\n"
     ]
    }
   ],
   "source": [
    "for tid, _list in zip(sg.t_df.tid, sg.t_df.path):\n",
    "    print(tid)\n",
    "    print(_list)\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "crazy-proof",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tname       object\n",
      "gid         object\n",
      "gname       object\n",
      "path        object\n",
      "tid         object\n",
      "novelty     object\n",
      "loc_path    object\n",
      "dtype: object\n"
     ]
    }
   ],
   "source": [
    "print(sg.t_df.dtypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "forced-seattle",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"None of [Float64Index([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,\\n              nan, nan],\\n             dtype='float64', name='tid')] are in the [index]\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-7-513381bf98ad>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0msg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mt_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0msg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mt_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontains\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'97'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    877\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    878\u001b[0m             \u001b[0mmaybe_callable\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_if_callable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 879\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmaybe_callable\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    880\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    881\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_is_scalar_access\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mTuple\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_getitem_axis\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1097\u001b[0m                     \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Cannot index with multidimensional key\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1098\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1099\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_iterable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1100\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1101\u001b[0m             \u001b[0;31m# nested tuple slicing\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_getitem_iterable\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1035\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1036\u001b[0m         \u001b[0;31m# A collection of keys\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1037\u001b[0;31m         \u001b[0mkeyarr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_listlike_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mraise_missing\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1038\u001b[0m         return self.obj._reindex_with_indexers(\n\u001b[1;32m   1039\u001b[0m             \u001b[0;34m{\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mkeyarr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mallow_dups\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_get_listlike_indexer\u001b[0;34m(self, key, axis, raise_missing)\u001b[0m\n\u001b[1;32m   1252\u001b[0m             \u001b[0mkeyarr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnew_indexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0max\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reindex_non_unique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkeyarr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1253\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1254\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_validate_read_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkeyarr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mraise_missing\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mraise_missing\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1255\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mkeyarr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1256\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_validate_read_indexer\u001b[0;34m(self, key, indexer, axis, raise_missing)\u001b[0m\n\u001b[1;32m   1296\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mmissing\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1297\u001b[0m                 \u001b[0maxis_name\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_axis_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1298\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"None of [{key}] are in the [{axis_name}]\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1299\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1300\u001b[0m             \u001b[0;31m# We (temporarily) allow for some missing keys with .loc, except in\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: \"None of [Float64Index([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,\\n              nan, nan],\\n             dtype='float64', name='tid')] are in the [index]\""
     ]
    }
   ],
   "source": [
    "sg.t_df.loc[sg.t_df.path.str.contains('97')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "numeric-avatar",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "102\n"
     ]
    }
   ],
   "source": [
    "for ind, entry in sg.edge_df.iterrows():\n",
    "    \n",
    "    # transcripts that use this edge\n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "inappropriate-oregon",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6/14/21"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "prime-costs",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding transcriptome to the SwanGraph\n"
     ]
    },
    {
     "ename": "UnboundLocalError",
     "evalue": "local variable 't_df' referenced before assignment",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mUnboundLocalError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-7-13658fc52e15>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0msg\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mswan\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSwanGraph\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0msg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd_transcriptome\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'../new_testing/files/test_full.db'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/swan_vis/swangraph.py\u001b[0m in \u001b[0;36madd_transcriptome\u001b[0;34m(self, fname, pass_list, include_isms, verbose)\u001b[0m\n\u001b[1;32m    191\u001b[0m \t\ttids = self.add_dataset(fname, pass_list=pass_list,\n\u001b[1;32m    192\u001b[0m                         \u001b[0minclude_isms\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0minclude_isms\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 193\u001b[0;31m \t\t\tverbose=verbose)\n\u001b[0m\u001b[1;32m    194\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    195\u001b[0m                 \u001b[0;31m# fill NaN annotation transcripts with false\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/swan_vis/swangraph.py\u001b[0m in \u001b[0;36madd_dataset\u001b[0;34m(self, fname, pass_list, include_isms, annotation, verbose)\u001b[0m\n\u001b[1;32m    242\u001b[0m                 \u001b[0;31m# create the dfs with new and preexisting data and assign them to the\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    243\u001b[0m                 \u001b[0;31m# df fields of the SwanGraph\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 244\u001b[0;31m                 \u001b[0mloc_df\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0medge_df\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mt_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcreate_dfs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mt_df\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexon_df\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfrom_talon\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    245\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mloc_df\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    246\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0medge_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0medge_df\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mUnboundLocalError\u001b[0m: local variable 't_df' referenced before assignment"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "sg.add_transcriptome('../new_testing/files/test_full.db')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "minor-appearance",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6/11/21\n",
    "# idk\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "scenic-arctic",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding transcriptome to the SwanGraph\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "sg.add_transcriptome('../new_testing/files/test_full.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "alternate-poland",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tname</th>\n",
       "      <th>gid</th>\n",
       "      <th>gname</th>\n",
       "      <th>path</th>\n",
       "      <th>tid</th>\n",
       "      <th>loc_path</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>test1</th>\n",
       "      <td>test1_tname</td>\n",
       "      <td>test1_gid</td>\n",
       "      <td>test1_gname</td>\n",
       "      <td>[0, 1, 2, 3, 4]</td>\n",
       "      <td>test1</td>\n",
       "      <td>[0, 1, 2, 3, 4, 5]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test2</th>\n",
       "      <td>test2_tname</td>\n",
       "      <td>test2_gid</td>\n",
       "      <td>test2_gname</td>\n",
       "      <td>[5, 6, 7, 8, 9]</td>\n",
       "      <td>test2</td>\n",
       "      <td>[12, 11, 10, 8, 7, 6]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test3</th>\n",
       "      <td>test3_tname</td>\n",
       "      <td>test2_gid</td>\n",
       "      <td>test2_gname</td>\n",
       "      <td>[5, 6, 13, 14, 9]</td>\n",
       "      <td>test3</td>\n",
       "      <td>[12, 11, 10, 9, 7, 6]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test4</th>\n",
       "      <td>test4_tname</td>\n",
       "      <td>test4_gid</td>\n",
       "      <td>test4_gname</td>\n",
       "      <td>[10]</td>\n",
       "      <td>test4</td>\n",
       "      <td>[6, 7]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test5</th>\n",
       "      <td>test5_tname</td>\n",
       "      <td>test2_gid</td>\n",
       "      <td>test2_gname</td>\n",
       "      <td>[5, 11, 12]</td>\n",
       "      <td>test5</td>\n",
       "      <td>[12, 11, 8, 7]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             tname        gid        gname               path    tid  \\\n",
       "tid                                                                    \n",
       "test1  test1_tname  test1_gid  test1_gname    [0, 1, 2, 3, 4]  test1   \n",
       "test2  test2_tname  test2_gid  test2_gname    [5, 6, 7, 8, 9]  test2   \n",
       "test3  test3_tname  test2_gid  test2_gname  [5, 6, 13, 14, 9]  test3   \n",
       "test4  test4_tname  test4_gid  test4_gname               [10]  test4   \n",
       "test5  test5_tname  test2_gid  test2_gname        [5, 11, 12]  test5   \n",
       "\n",
       "                    loc_path  \n",
       "tid                           \n",
       "test1     [0, 1, 2, 3, 4, 5]  \n",
       "test2  [12, 11, 10, 8, 7, 6]  \n",
       "test3  [12, 11, 10, 9, 7, 6]  \n",
       "test4                 [6, 7]  \n",
       "test5         [12, 11, 8, 7]  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.t_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "specialized-subject",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6/9/21\n",
    "# browser path tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "distinguished-cruise",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding transcriptome to the SwanGraph\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAxsAAACmCAYAAACyYpL/AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAEfUlEQVR4nO3dQW4SYRiA4a9EozEmLvUexr26YuGSeBXvZDgAO+Nt7A3UxOKmCzdt45SXtvR5NoRM+PgXMMwbZuBsv98PAADAoa3uegEAAMBpEhsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEDiyXUbV+vNj5l5faS1AI/b+cVu++auFwGcPsc3LORzaoGbvtnwRgSOxf4GOBb7G5bwulnAaVQAAEBCbAAAAAmxAQAAJMQGAACQEBvASVitN19W683BfiXEPPPMMw+4vbP9fn/lxtV6c/VGgMP7fovHvp2ZpzPz80BreX55a5555p3evFcHeh4emYvd9uyu1/DQXPs/GwAPyO+ZeTkzzy7v/5rbHbj8mZkX5plnnnnAcmIDuDcudtv3Sx63Wm8+zMzXmfl4sdt+u+06/pn3yTzzzDu9ec7cgONxzQZwCt7NzOdDHKiYZ5555gGH45oN4N5wLixwDI5vWMrn1P/zzQYAAJAQGwAAQEJsAAAAiZti4/woqwCwvwGOx/6GJbxuFrj2AnEAAIClnEYFAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJD4C5mp9l8u812yAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 1008x201.6 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "sg.add_transcriptome('../new_testing/files/test_full.gtf')\n",
    "sg.plot_transcript_path('test5', browser=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "seeing-process",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6/8/21 \n",
    "# plotting tests\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "imposed-basics",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tname</th>\n",
       "      <th>gid</th>\n",
       "      <th>gname</th>\n",
       "      <th>path</th>\n",
       "      <th>tid</th>\n",
       "      <th>loc_path</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>test1</th>\n",
       "      <td>test1_tname</td>\n",
       "      <td>test1_gid</td>\n",
       "      <td>test1_gname</td>\n",
       "      <td>[0, 1, 2, 3, 4]</td>\n",
       "      <td>test1</td>\n",
       "      <td>[0, 1, 2, 3, 4, 5]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test2</th>\n",
       "      <td>test2_tname</td>\n",
       "      <td>test2_gid</td>\n",
       "      <td>test2_gname</td>\n",
       "      <td>[5, 6, 7, 8, 9]</td>\n",
       "      <td>test2</td>\n",
       "      <td>[12, 11, 10, 8, 7, 6]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test3</th>\n",
       "      <td>test3_tname</td>\n",
       "      <td>test2_gid</td>\n",
       "      <td>test2_gname</td>\n",
       "      <td>[5, 6, 13, 14, 9]</td>\n",
       "      <td>test3</td>\n",
       "      <td>[12, 11, 10, 9, 7, 6]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test4</th>\n",
       "      <td>test4_tname</td>\n",
       "      <td>test4_gid</td>\n",
       "      <td>test4_gname</td>\n",
       "      <td>[10]</td>\n",
       "      <td>test4</td>\n",
       "      <td>[6, 7]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test5</th>\n",
       "      <td>test5_tname</td>\n",
       "      <td>test2_gid</td>\n",
       "      <td>test2_gname</td>\n",
       "      <td>[5, 11, 12]</td>\n",
       "      <td>test5</td>\n",
       "      <td>[12, 11, 8, 7]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             tname        gid        gname               path    tid  \\\n",
       "tid                                                                    \n",
       "test1  test1_tname  test1_gid  test1_gname    [0, 1, 2, 3, 4]  test1   \n",
       "test2  test2_tname  test2_gid  test2_gname    [5, 6, 7, 8, 9]  test2   \n",
       "test3  test3_tname  test2_gid  test2_gname  [5, 6, 13, 14, 9]  test3   \n",
       "test4  test4_tname  test4_gid  test4_gname               [10]  test4   \n",
       "test5  test5_tname  test2_gid  test2_gname        [5, 11, 12]  test5   \n",
       "\n",
       "                    loc_path  \n",
       "tid                           \n",
       "test1     [0, 1, 2, 3, 4, 5]  \n",
       "test2  [12, 11, 10, 8, 7, 6]  \n",
       "test3  [12, 11, 10, 9, 7, 6]  \n",
       "test4                 [6, 7]  \n",
       "test5         [12, 11, 8, 7]  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.t_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "wired-sociology",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>v1</th>\n",
       "      <th>v2</th>\n",
       "      <th>strand</th>\n",
       "      <th>edge_type</th>\n",
       "      <th>edge_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>edge_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>-</td>\n",
       "      <td>intron</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>-</td>\n",
       "      <td>intron</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "      <td>-</td>\n",
       "      <td>intron</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>intron</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         v1  v2 strand edge_type  edge_id\n",
       "edge_id                                  \n",
       "0         0   1      +      exon        0\n",
       "1         1   2      +    intron        1\n",
       "2         2   3      +      exon        2\n",
       "3         3   4      +    intron        3\n",
       "4         4   5      +      exon        4\n",
       "10        6   7      +      exon       10\n",
       "9         7   6      -      exon        9\n",
       "8         8   7      -    intron        8\n",
       "12        8   7      -      exon       12\n",
       "14        9   7      -    intron       14\n",
       "7        10   8      -      exon        7\n",
       "13       10   9      -      exon       13\n",
       "11       11   8      -    intron       11\n",
       "6        11  10      -    intron        6\n",
       "5        12  11      -      exon        5"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.edge_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "soviet-valuation",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "experimental-forward",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "perceived-sheep",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding transcriptome to the SwanGraph\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAxsAAACmCAYAAACyYpL/AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAu70lEQVR4nO3deXwcdf3H8dckbdI7PdOGFATKISBXFwRUBK0cEhCq1C4IkiDiDwV+/mT5gaN4gC4CKSqIIohbUWECgqJUAQ3nj7MM96Ec5SghpbRJ09IkzTW/P2bb9EibPb6zO7v7fj4eeaSPZGfmm7wz0/ns9zvfr+V5HiIiIiIiIqaV5bsBIiIiIiJSnFRsiIiIiIhIIFRsiIiIiIhIIFRsiIiIiIhIIFRsiIiIiIhIIFRsiIiIiIhIIFRsiIiIiIhIIFRsiIiIiIhIIFRsiIiIiIhIIFRsiIiIiIhIIFRsiIiIiIhIIFRsiIiIiIhIIFRsiIiIiIhIIFRsiIiIiIhIIFRsiIiIiIhIIFRsiIiIiIhIIFRsiIiIiIhIIFRsiIiIiIhIIEbkuwEiIrng2s4YoAbYLvkxFajI8mPkVr7eA6wGOpKfN/5I6WuReLQ3oF+FiIhIzlie5+W7DSIiGXNtp5JNi4j1H5t/bWKempipLrYsStqBN4ElwOvJz29G4tGePLVRRERkm1RsiEgoubZj4RcMM9mykNj4Y0q+2hgSHrCUTQuQ9Z+XAG2ReFQXehERyQsVGyKSd67tjAU+AuyT/Ng7+XlSPttVJDoYLDw2L0bejsSjfXlsm4iIFDkVGyKSM67tlAE7MVhUrP+YBVh5bFqp6gfewi88XgCeAB4H3lBviIiImKBiQ0QC4drOJAZ7KNb3VuwNjM1nuyQlKxgsPJ4AnojEo235bZKIiBQiFRsikhXXdkYAu7Flb8X2+WxXCtqBTvyZo3qA3o3+PdzHtl7bC4wCJmz2UTXE1yZQOD06rzJYgDwOPBuJR9flt0kiIhJ2KjZEJC2u7ZQD+wKfTn4cCozLa6M21QG8O8zHskg82p23FiYlh5WNZegiZPMCpQr/gfid8YeiVeahyRvrAZ5hsPfjceA1Db8SEZGNqdgQkW1Kzgq1J35h8SngcPLz4HYn0MK2i4jWSDy6Ng9ty6lkkbK+8JiV/Lzxv6flqWntbNr78UQkHl2Rp7aIiEgIqNgQkU0ki4tZDBYXnwaqc9iEJcBzG328DLwDrNG75qlxbWc8WxYg6/+9I7ld0PU14G5gEXBfGHqUREQkd1RsiAiu7WzPpsVFLp636GCwoHg++fmFSDy6JgfHLlnJYXAzGSxCZuE/uH8Q/qrqQeoEmoG/A4si8ejSgI8nIiJ5pmJDpAS5tjMdfzjU+ucudgnwcP3AK2zaW/EcsFQ9FeGR7NHaEb/oOAj4KDAb/2H3oDyH3+OxCHgsEo/2B3gsERHJAxUbIiUgOaxmDoPFxV4BHep94Fk2GwaloTOFybWdkfgzi32UwSLkwwEdrg24C7/wuEtT7YqIFAcVGyJFKllgHAd8ETiaYGYvehG4D7gXeDQSjy4L4BgSIq7tVAEHMtj7cRAw3fBhBoBHGez1eF69YCIihUnFhkgRSRYYx+IXGJ/FfIHxGoPFxf0qLiQ5/GoHNu39iACjDR7mHQYLj3tLYcYxEZFioWJDpMBtVGDMwy8wTI6xfwe/sLgXfyahtw3uW4pUcqHHj+BPOFAHfBIYaWj36/AL3kXA7ZF49F1D+xURkQCo2BApQK7tjGPTHgxTBcb7DBYX9wKva/iKZMu1nQnAZ/ALj2OAGYZ2PQDcA/wOuCMSj3YZ2q+IiBiiYkOkQCQLjDr8AuMYzBQYq4D7SfZcAC+quJAgJRck3B//b7kO//kPy8CuO4AmYCH+zFb6OxYRCQEVGyIhtlGBMQ+/wMh2HHwn8ACDz108o+lGJZ9c26nG752rA44CJhjY7av4RcfvtZaHiEh+qdgQCRnXdsYyWGDUYabAuBO4BfhHJB7tzHJ/IoFITrX7cfwhgnVkP82uh7+I4O/wn+/Q376ISI6p2BAJgeTQkiOB0/FvtEwVGLcCf9dNlhQi13Z2ZnC41eFkN7vaGvzz4XfAQxpmJSKSGyo2RPLItZ2p+AXG14Cds9xdF5sWGJoeVIpGssdvDn7hMReYlsXuluAXHTdG4tE3s2+diIhsjYoNkRxLrktwCHAW/lCpbN6t7cKfAvQWVGBIiUgOtzoaqMdfuDKbaXXvxy88/hSJRz/IunEiIrIJFRsiOZJ82PtL+EXGvlnsqgv4O36BsUgFhpQy13amACcBpwEHZLGrtcBt+A+WPxCJRweyb52IiKjYEAmYazsfwS8wTgXGZ7ibbvwejFvxCwy9AyuyGdd29sIvOk4lu7U8lgA/BRIq5kVEsqNiQyQAru1UAp/HLzIOzXA33Wzag6ECQyQFyRXMj8AfZnU8mQ9VbAN+CfwiEo++Z6Z1IiKlRcWGiEGu7ewInAmcQeYPsN4PXA/8VQWGSHZc25mEvxBmPXBwhrvpAW4ErozEoy8bapqISElQsSGSJdd2yvEXI/s6/sJ7mayGvBp/rPi1upkRCYZrO7vjD7P6MlCb4W7uBBqBBzV9rojI8FRsiGQoufLx+mlrd8xwN08BvwJu1thwkdxIvkHwafzejs8DozLYzZP4RcdtkXi0z1zrRESKi4oNkTQkp639GH4vxjwym3KzG2jCHwu+WO+OiuSPaztV+OfyacAnMtjFW/gPk/82Eo+uMdk2EZFioGJDJAXJIuMI4CIyuyEBeBW4FlgYiUfbTLVNRMxwbWcf4DzgZGBEmpuvwj+/r47Eo+8abpqISMFSsSGyDckiow6/yPhoBrvoB/6KP1SqWXP3i4SfazszgXOA/wImpLl5L/BHYEEkHn3BdNtERAqNig2RIbi2UwbMBb4L7JfBLt7Fn1Hq+kg82mKwaSKSI67tTAC+AnwT2CGDXdyF/1zHvRouKSKlSsWGyEaSD45+EfgOsFcGu2jGfxbjb5F4tNdk20QkP1zbGQmcCMSA2Rns4hn8ouMWXRdEpNSo2BBhw83ElwAb2DXNzdsZnLb2FcNNE5GQSA6rPBy/6Dgmg10sBS7Gf25LM1iJSElQsSElLbnS92nAhcBOaW7+PP4sNE2ReLTTdNtEJLxc29kL+BZwClCR5ub/xu89/bOGV4lIsVOxISXJtZ3R+GOxLwBmprs5cAn+UCk98C1SwlzbqQHOBs4CJqW5+RPAhZF49D7jDRMRCQkVG1JSXNsZiz/DTAyYkebmj+IXGXfp3UgR2ZhrO+OABuB/SL+X9G7g25F49GnjDRMRyTMVG1ISkrPKfAN/2MPUNDd/AL/I0IwyIrJNru2MwJ/J7nzgwDQ3d4CLIvHoa8YbJiKSJyo2pKi5tjMJOBf4b9If4nAP8KNIPPqQ8YaJSFFLPkz+Cfyi47g0Nu0DrgMuicSjy4Jom4hILqnYkKLk2s5U/F6Ms4HxaW5+J36R8bjxholIyXFtJwJcChyRxmad+BNQXBGJRzsCaZiISA4YLzbmL1o+GYjgdx8fBuwGVAIj8VdWXQe8gj80ZTHgNtVVtxlthORMR8uctPOuqm0OLO/kg9/fxJ/Cdlyam9+OX2Ro3PRWhC1vCZbyNsu1nTn4RUc6w6vagDhwTSQe7Q6kYUnKu7S0LrTSzrum3lPeBSqfeRspNuYvWj4BOBU4D6gFuoAx+D/A1vTiv3MzGmgBFgC/b6qrXp11gyRQHS1zjOVdVdtsJO/kkIX5wGWkt9LvANAE/DgSj75ooi3FJox5S3CUd7CS16ovAD/G/88+Ve8A3wduNLlGh/IuLa0LLWN519R7yjvkwpJ3VsXG/EXL98T/AU7Cv2kbm/HOYC1QBtwMLGiqq34pi31JADpa5gSWd1Vtc8Z5u7ZzMP5wg4PT2Kwf+AMQ10J8Qwtr3hIM5Z1byQfJG4AfANulsenL+D23d2QzYYXyLi2tC63A8q6p95R3yIQt74yKjfmLllfir4J6Dn51NCLtnWxdH35VdRXw/aa66nUG9y0Z6GiZk7O8q2qbU87btZ0PAT8Bomkcrxd/te+fROLRJWlsVzLCmrcEQ3nnl2s7Y/B/9xcCE9PY9DH8NToeSOd4yru0tC60cpZ3Tb2nvPMsrHmnXWzMX7Q8AtwKTMfviglKJ/AeMK+prtoN8DiyDR0tc3Ked1Vt8zbzTk5jeyH+A+CVKe5/HXA9cHkkHl2aTUOLWRjzluAo7/BIzpx3Af7MeaPS2PQfgB2JR58Z7oXKu7S0LrRynndNvae88yTMeadVbMxftPxC4Hv4F0IrkxamyQO6gYub6qp/koPjyUY6WubkLe+q2uYt8nZtpxx/1e9LgOoU99kLXINfZLSaamgxClveEizlHU6u7dTi5/IVoDyNTX+L39Px/lDfVN6lpXWhlbe8a+o95Z1jYc87pWJj/qLlFv4DIl8j2GppazqBXwPnNdVVa67egHW0zAlN3lW1zR6AaztHJNu0dxr7uQ24IBKPvm6+icUjjHlLcJR3YXBtZ3f8N1bmpbHZKuAi4Nr1D5Er79LSutAKTd419VpbIWiFkneqY7kWAGeSnx+E5HHPxK+kzstTG0pJaPJ2bec3QCNwTBrbPwl8S4vxpSw0eaPzOxeUdwGIxKP/Ab7o2s6B+NPlzklhs4nA1cAZru2cDby73fE1N4yuee9jVvlARXCt3SblnVs6v0tLQeQ9bM9GcujUReTvB9lYJ3CJhlQFJ9nVHoq8vf6yvpVPzC5b9dT+ZSlu0gJ8G/hjJB4dCLBpRSNMeZM8vzXkIjjKu3C5tvMZ/AkxIqluM2n20/2TDnimvGyksZlys6G8A5YcShOq81tDqoJTSHlvs9hIPgz+EP5cu2HRBXyiqa76qXw3pNgkHx4MVd4DveW03P451q2Yuq2XdeKvr7EgEo+uzU3LCl8Y8yZ5flfVNuv8Nkx5Fz7XdsqAE4EfAbtu67WV096ndu7fKBvZn5O2pUh5ByT5cHAoz++aek95GxbWvGvqvSELn+GKjSXAjuTmYZNUecAbwB5NddU9+W5MsUhOh/gyIcvb86Bv9XjeumkeDGzxrKQH/A74TiQefTf3rStcYc2bjc7vqtpmnd+GKO/i4trOSOB0/Gc6pm3xgrJ+PnTyrYyYsAYrTGkr70AkpzsN9fldU+8pb0PCnHdNvTfkSJThhqdMJ1w/CPjtmYE/j7CYczEhzNuyoHxMF5M/+uTm33oAOCASjzao0MhIKPNG53dQlHcRicSjvZF49NfA7vjPaGwybHTKQU9SPqYrbIUGKO+g6PwuLWHOe+hvDNOzEeaZBDqBA5rqql/Od0MKXXIl2ScJV3fcJgZ6y1l661x62ye9BpxPlqvnlrJCyJvk+V1V26zzO0vKu/i5trMP8Avg0JGT2tl+3p/DNnxqc8rbkORK0QVxftfUe8o7S2HPu6beG7LgSPXB2zCqQDMdmHIe/kqToWWVDzD90w88BuwViUf/okIjK6HPG53fJinvIheJR58DDgPOmrjfc1AW+vkxlLc5Or9LSyHkvYVC7tkA/+Gj6U111Wvy3ZBC1dEyZwKwjJBWyZvpAqZX1TYr7wwp79KivEvLktsu2H5S5Om3Q96rsZ7yzlLrQqvgzu+aek95Z6gQ8t5az0aq62ykbaC/j5dvuYo3/3UTXW3LGTt9B3Y97ivsUtdg9DDAKcCvTO60xJzKZuN9M/HB2j6u+tXbPP3sGp56djUr23q54H925Nvn7WygiRso7+wZyfvpZ1fTdPsyHny4nbeXdjNmTDl77D6Wb53zIQ77+GQDzQSUtwlG8n7m+TUsuOpNnntxDcuX91BZWcasncbwldNqiX5hBmVlRoYOK+8sTTn4yc8N9JX3k97K48N6+LF26k58GoBnHz2ED21v5F5HeWfPyPm9dAUcdP7Q3/vW8RA7IdsjAMrbBCN5r7esHRbcAfc+ByvXwJTxMHsWXHk6jDdczgRWbLjX/C9v/PMmdj7qVCbvui/Lnn6Ap351IT1rO9jzi980dZix+OP39cebufPwf49ZWdnWy+U/e5Pamkr2+cg47nuw3UDTtqC8s2ck759e8xYPP7aKzx0zjTPrZ/JBZz833dLK8fOf4eeX7c5pX6o10FTlbYCRvN96q4vu7gFOOrGGmukVdK8boPn+Nr7+Py/z/IsfcOkPtjkLa6qUd/bOKxvRb7TQ6Osb4PzvvMLYMeWs7TTaY6K8s2fk/F7vs7PhmM1Wcdlje1N7V94GGMv71Vb4wk9g/Cg49XCYPhFWrIHFr0JXj/liI5BhVO1LXuCf587hwyeezT71F234+qOXfZV3H7+HuhsWM2pSdSa7HkoPUNNUV91maoeloqNlzhTP413LIuuVZdetG6CtvZeaGZW8tbSLfQ95NIieDUjmXVXbrLzTZDLvxxavYv99JlBZOfjYV3d3P4cetZi29l5efeYTpt7tVt4ZMpn31sw/7Vnue6iNN1/8JGNGG7nHVd4Z6miZMwV4F8zm/Yvr3uZn17zFiSdM59ob3jHZswHKO2OtC60pnkerZWU/fn99z4bBXoyt6QFqauo95Z2mZN5GrueeB5+92J866rYLYUylgQYm5fQB8aUP/gWAXY49Y5Ov73rcV+nv6abl8btMHq4LmG1yh6XizuVVc9f0lxl50KiysoyaGQb/YrdOeWfogyUfmjvQM9JI3gcfOHGTQgNg1KhyjpwzhZVtvSx/39iU6so7QybP763ZYeYoeno8uruN9ewr78zNxv/9GbPsvXVcduUbfO+CWVRNCGQghPLO0ANdn/z6GiqMn99dPdDda3qvg7tHeWfkLm/WXFN5/9/L8NybcN4JfqHR1QO9fSb2vHWBXD3aX3+O0VO3Y8zUmk2+PmnX/bDKymh/7XljxxrwBsa9s+4/MStx+QHGdloivj1r1IVHTu0I2zzN2+QNMO6D13eOvXaNo7zTNPnAyReO/dDSQPNe9t46Ro60jN2Y9HmMu2PZxNhX7okp7zQFcX6v7eynu7ufNWv6eejRdv54Syv77j2eyZOM3fOMBg4E/mVqhyXkQGDI1Xsz9Z2LX2XWzmM4JVrDZT99w+SuAegbYMJdr21/zY23xZcY33mRmz2WIw/B7B3ir++GK+/w/73bdnDusfD5Q8ztv8+zxt3JbrGvJ3Q9T1eMaRd+mjeMXM/vf8H/PLoCjr0EnloCZRYcvDtccrLRoXMbBPNWxcpljJ48fYuvl4+soGL8JLpWtho7VplVVj7SqjwKOMrYTkvEAVWdVBTY5MdWGeVlFT3KOwOVM5ZjlQc3JeYrr63lb/94n2OPnsZoM0NqGGFRXjWyX3lnIIjz+7sXv0riD/4ampYFhx86iauu2MPkISrwp3C91OROS8RhGJwS86FH2rn9r8u5+y8RU0MitzCiDGvU5FW73TO+bbdADlDE6nmLCsvM9dyy4JN7wdGzYbvJ8G4bJJrh7Oug7QM44wgjh2GE5ZVXed26nmdgNq3G8l6yzP/8tV/CQbvBr78O77XDT/8KX7gMmi+BmklGDrVBIMVGf083FeOHbml5xSj6e7qNHm/8CGOz35SUXcauy3cTMlIxsSPfTShIQf7eOlb3cepXn2fsmHJ+dNEuRve9y5jC/DvNtyDO77PO2J7j66p5f0UP/7x3Je8uW8eaNcb733XjmRljv7fe3gHO/+4rfPHzM/hopMrUboek8zszO2NuEpaZU8CJbfq1kw6FI38Al90O8z9h7oHhWQbbXUpM5r02ecrtuQPccM7g1/feEeZeCtfdDd+PGjscENAzG+UVoxjoHfoC0t/TTXnFKLPHMzvLX8moDP/CT0OyygtiDvnQCer31tXVT7ThWZa+083NiX2o3c7s+V2of6f5FsTvbbddxnL4oZOZN3cG1129F/vuPZ66E59iZZuxZ3QAcvLwVxEy9nv71W+W0vJuNz+0Z5na5Vbp/M5MJcH+P1g5Ek6fA2u7wX3N3H4rAm53sTKZ96hk/+eJmw2RO2g3v/B84lVjh9ogkGJj9JQZdLW9t8XX+3t76FnTzugpM4wez7JUbGRihBX2NRuHFuRQoGJmlZnPu6dngFPOeJ4nn1rNjdftzcEHTjR+jJEBtLsU5OL8nnfCdNpX9XHnP943udvAZs8qckaGUHWs7uPyn73JSfNq6OzsZ8kbnSx5o5P2VX4P1tJ3unlrqbnn0HV+Z2aEueUWtmrmFP9z2wfm9jkyB+0uRibznjHR/zxtiE7L6iroWGvsUBsEMoxq0qx9eO/pB/xnNzYqLNpffQZvYIBJu+xj9Hiep0o5E31Dz1AWel5/gT1oEhLegNm8+/oGqD/rBe57qI0brtmLz3xqitH9r9druN2lIhfnd/c6/z/AVR1Gh1IZ7SYpIUbmEFrV0csHa/u5LvEO1yXe2eL7x857muppFbzy9CdMHE7nd4b6gnmveBNvJt9DmDLB3D57c9DuYmQy7313gj884D+bs7nWdqgN4L/yQIqN7Q89nn//6Wpe/dv1m6yz8erfrqe8YhS1Bx1t9HgeXh/wttGdloB1A2XGF8HIBc+zlHcGvP5yY3kPDHicee5L/P3uFVx1xYeZe9yWE0KY0q+8M2Ly/H5/RQ/Tpm7Z4fDb37cAMHs/g3cjoEH8mTHye5s2tYLfXfuRLb7+5zuX85c7l7MgvpvRoZL9A+XM6K0IeOLN4uLh0TOy3Nj924rVMHWzU3hNF1x/D1SNgQMNPobXT5mu5xnowdz/30fvDxfdBDc/CNFDoTxZx/zzGb/Y+NJhpo40KKCejb3Z6YiT+Pdt19C7dg2Tdt2X956+n6UP/ZWPnHKByQX9ABhTPqHZa2g0W8GUgKde+ey/gd1N7e+6xDt0rO6lY7X//8Yjj6/iip/70yXO/8IMdphp5gmzkePXNkfiUeWdpvdfbDKW93cveY3b/7qcjx88kcqKMppuW7bJ94/97DTGjjEzvHHm6F6d3xkweX43nPUC5eUWBx9YRW3NKFa29fD3e1aw+KnVnHBsNYd+zOjUJa+Y3FkJeQXYMdudjBldzvHHbvl/9Ev/8cfSfOZTU0wu6sfMMevubj0zrvM7TY/85vfvMWK1kZupH93qz1D0yb38WYha2/0b0WWr4MrTzS76VmutafbqdT1P1yOJG4xdz6dMgP+dCxc3wYmXwXEHwrJ2uOFfsFO1udnHNhZIsQEQ+cYVjJk2kzf+5fDGP29m7PTt2f9rcXY59nTTh+oBHjC901Iwa2zPjf0ePyy3zPwdXP3rt1n6zuBMYw89soqHHlkF+IvAGSo2lHeGKiauvtEb4IdWWfZ5P/v8GgAefmwVDz+2asvvf/QQxo5R3vlk8vyOnjiDW//8Hok/vEtbey9jRpexx+7juPLS3Tnt5O1MNHc95Z25B4BPYXD62xxQ3hnaaUTLz/s965Jyy8t6fM1he8HS9+HG+2DVWhg3CvbbCX52Bhy6p4nWbqC8M7STtcro/dp/HQ0Tx/q9Vxc3+Zkf91H4zokwwehqPb7Aio2yESPZ6+QYe50cG/7F2ekCFgd9kCK1uNxiLWBkbsPnH/uYid0MR3lnbrFVZibvRX/K2SKwyjtzxs7vU+ZvxynzjRYVW6O8M7cY6MTQ9Xxz3z5vZ759nvGRt8o7c4vLLW8NBvKee7D/kQPKO3NG79fAH0IVPdTU3ratGJ7UGQ08le9GFKin8H9/hUR5Z055lxblXVqUd2lR3qWlEPPeoBiKjZamuuohnqmX4VTVNq8EWvLdjjS1VNU2K+8MKO/SorxLi/IuLTX1XkHmXVPvKe8MFGjeGxR6sbEWuCLfjShwC/B/j4VAeWdPeZcW5V1alHdpUd6lpZDy3kShFxtlwB/y3YgC93sK5+9AeWdPeZcW5V1alHdpUd6lpZDy3kRBNjqpD7ipqa56Tb4bUsiqaptXAzfj/z7DrA+4qaq2WXlnQXmXFuVdWtbn7XnKuxTU1HsFlXdNvae8s7A+b8J/Pd9CIRcbvfhdSpK9BRhafTZAytsc5V1alHdpWeCF/2ZEeZuzAOVdSgrher6F4YqNzpy0In2dwFVNddUv57shxaCqtvkl4GpCnndVbbPyNiDseQ94dKO8jQl73p7yNqqqtvmlt7oqbu7ss/LdlK3R9dygmnrvpf7e6psHBiry3ZSt6QSuqqn3lLcBNfVeqK/nWzNcsfEe4OWiIWnwgGXA9/LdkCLzPZR3KQll3gMevN1VUbnzfXs/nO+2FJlQ5u150LdmXOWSG05V3oZYidjsgx/Z47PLe0YyEKq0AV3PjXNtZ/byt6747EBfFZ7yLhWhvJ6zjfYMV2zMA7qHeU2udQPzmuqqe/LdkGJSVdu8jhDnXVXbrLwNCmve6wYsTnt2J6utd8RfrETsnHy3p1iENW+vv5xl/zjCGuge9RfXdpR3lqxE7ATgwXUDZTNOe3YnugdC17uh67lBru2cADyIVzGjvfWb4IWud6MbmFdT7ylvg2rqvVBez9lGe7ZZbDTVVbvAxYSnu6YTuLiprlqLwgSgqrY5lHlX1TYr7wCELe+1fWVc9voMnl0zBvxr01VWIvYLKxEbkeemFYWw5T3QO4K2xfuzbsVUSObt2s4vXNtR3mmyErFyKxH7MfBnYCzAs2vGcPnrM1jbF5pHM3U9N8S1nXLXdjbJu3fdzqxeOZcQDafqBC6uqfeUdwBq6r1QXc9J5r21b1peCv1u8xctvxI4k+QfdZ50Atc21VWfl8c2lISOljmhybuqtll5BywMea/tt0gsncp3X5k51LfvAuZ7DY2rc9ysohSGvAd6R9DxwodZ+cghQ337LmB+JB5V3imwErEpwE3AkUN9/8e7vUP9zJWMHTGQ24ZtStdzQ1zb2WbeE6beyJiJ91JWti63DdtUJ3BtTb2nvAPWutDK+/WcFPJO9S2P84DryF8F1QlcC8TydPxSo7xLS17z7u63BvxCo3ZrLzkaWGwlYvvlrlVFLa95D/SVD/iFxsFbe8nRwGLXdvbLXasKk5WI7Q88yVZuPAG+80ott7ZOWtkzYOVrBhtdzw1xbWfYvFevOJWu1R9f6Xnlyrs0FMT9Wko9G+vNX7T8QvwHU0YBuRgM6uGPAbu4qa76Jzk4nmyko2VO3vKuqm1W3jmWr7y7+634jOb99gBOHub1PfgX1mu8hsawPRhXcPKV90BfeXzJr09PK+9IPKq8N2MlYqfi32SMGual9wNzVx359H+h63nBcm0nrby32+2kvOVdU+8p7xxrXWjl7X4tlbzTKjYA5i9aHgFuBaYDYzJpYYo68WcxmKdnNPKno2VOzvPWmN78yVfeViJmARcBP0xh2zuA072GxrYA21cS8pW3aztp5x2JR5U3YCViFfhz7Z+dwsuvB87xGhrXga7nhci1nbTzjsSj6wBaF1o5z1vPaORPmPNOu9gAmL9oeSX+fxLnAiMBkw/09eG/o3U18D3NOpV/HS1zcpa3ZinJv3zmbSViUWAhUDnMfpYCJ3sNjf9nsG0lKZ95u7aTVt6ReLSk87YSsRr8m4mPD/PSHuBsr6Hx+s2/oet54XBtJ628I/HoFnm3LrRylrdmncq/sOadUbGx3vxFy/cEvoXfHT5Adg+orMV/huQmYIEW7AufjpY5geWtBZ7CJ195W4nYIfjvZk8bZp8DwPeBS72Gxv4s2ibkL2/XdtLOOxKPllzeViL2ceBPwIxhXvoO8AWvofGJbb1I1/Nwc20nrbwj8eg2825daAWWtxbsC5+w5Z1VsbHe/EXLxwOnAOcDtUAXMBrY1hxsPRu9rgW4AvhDU131mqwbJIHqaJljLO+q2mblHXL5yNtKxGYCfwQ+mcLL7wNO8Roa301l37Jt+cjbtZ20847EoyWRd3KI4deBnzH8u5T348/ctjzV/et6Hi7JIYZp5R2JR1POu3WhZSzvmnpPeYdcWPI2UmxsbP6i5ZOB2cCBwGHAbvhd5BX4P8A64BXgAWAx8FRTXbXG4haojpY5aeddVdusvAtULvO2ErFy/HH9FzH8zHkrgC97DY3/yORYMrRc5u3aTtp5R+LRos7bSsRG48/08uUUXr4AuNBraOzL9Hi6nueXaztp5x2JRzPOu3WhlXbeNfWe8i5Q+czbeLEhImKSlYgdjv+u93YpvHwBYHsNjRo7XKBc2zmcNPOOxKNFl7eViO0I3A7sP8xLO/EnTGgKvFESGNd2diSNvCPxqPKWgqFiQ0RCz0rEpuI/SFyXwssXA1GvoXFJoI2SwLi2k3bekXi0aPK2ErFjgN8Dk4d56WvAXK+h8YXgWyVBcW0nrbwj8ajyloKiYkNECkJy7Pp/A5fjz7KxLauBM/Vub+FKjl1PK+9Cf7fXSsSqgCuB01N4+Z3AqV5D46pAGyWBcW0n7bwj8eiqQBslEgAVGyJSUKxE7ADAAWal8PLfAN/0GhrXBtsqCYprO2nnHYlHCy5vKxE7Cr/9M4d5qQf8APiR19A4EHS7JBiu7aSddyQeVd5SkFRsiEjBsRKxCcCvGH4VaoA3gK97DY13BdsqCYprO2nnHYlHCyLvZG9GI3BGCi/vAL7kNTQuCrZVEpRkb0ZaeUfiUeUtBU3FhogUpOSwqnrgF6S2Wuot+L0crUG2S4KRHFZVT5p5R+LR0OZtJWJHAjcw/LvbAC/gP5/xWrCtkqC4tpN23pF4VHlLwVOxISIFzUrE9gCagL1TePlqwAau1UKAhcm1nYzyDtNCgMmeuQWk9u42+ItpnanhgIUp2TOXdt6FOBxQZCgqNkSk4CXXI2jEXwwrFYuBr3kNjU8H1yoJSnI9grTzjsSjec872ZvxG2D7FF6+AviG19B4S7CtkqAkezPSyjsSjypvKSoqNkSkaFiJ2OfxhylMTOHlA8DPge95DY0fBNkuCYZrOxnlHYlHc553sjejEfhqipvchv+sUcqrQ0t4JHsz0s47ndXARQqFig0RKSpWIrYD/rj+41Lc5B3gbK+h8Y7gWiVBcW0no7wj8WjO8rYSsSPw393eIYWXrwS+AdziNTTqP+gC5NpORnlH4lHlLUVJxYaIFJ3kw+MnAFcDtSludgdwjtfQuDSodkkwkg+Pn0AGeUfi0cDyTvZmXAGcmeIm6s0oYMnejLTzVm+GFDsVGyJStKxEbDxwMXAuUJbCJmuB7wFXeQ2NfUG2TcxzbSfjvCPxqLG8rUSsDH+a3jipjdVfCZwNNKk3o/C4tpNx3urNkFKgYkNEip6ViM0Gfg0ckOImz+A/QP5EYI2SwLi2k1HekXg067ytROwz+O9u75fiJn8GzvIaGt/L9tiSe67tZJR3JB5V3lIyVGyISEmwErFy4Cz8dx/Hp7CJhz8F5cVeQ+MrQbZNzHNtJ+O8I/Fo2nlbidi+wGXAUSlu0ob/7raj3ozC49pOxnmrN0NKjYoNESkpViK2HfAzYF6KmwwANwKXeA2NS4JqlwTDtZ2M847Eo8PmbSVi2wOXAF8GrBSP8Rf83oxlKb5eQsK1nYzzjsSjyltKkooNESlJViJ2DHANsGOKm/QDC4EfeQ2NbwbTKgmKazsZ5x2JR9/c/JtWIjYRuBD4b2BUivtsA84BblZvRmFxbWciWeSt3gwpZSo2RKRkWYnYGPwHhM8DRqS4WR/+2g4/1sxVhcW1nazyjsSjS61ErBJ/eNZFwOQ0Dn8rcK56MwqLaztZ5a3eDBEVGyIiWInY3sC1wMfS2KwHuB641GtobAmkYRII13bSztvD6/l3Zed959e8vseykT2prJ+w3sPA+V5D46PptlPyJzmd8heBS4Gd0tj0YeD8SDyqvEWSVGyIiLBhutKv4I/Hnp7Gpuvwb1x/onetC0dyutK0815nDXBb1fv8blIrK0dsc7bcV4ALgDs0ZKqwuLZzGP4MUwemsdmGvDVkSmRTKjZERDaSHFp1Fv6Nw7Q0Nu0CfglcrkXZCkdyaFXaeXdb/dxa9T43TlpG+6ZFx3Lg+8ANXkNjr9HGSmBc2xkJzMV/xuITaWy6Ie9IPKq8RYagYkNEZAhWIjYO+Abwv6Q3VrsT/0HkX3kNjW8E0TYxz7WdccA3BvAuKMOalOp2XVY/t058n9uqlne1jOy5HFjgNTSuCa6lYpJrO9PwV/w+i9RXnwf/PL8CWBCJR5W3yDao2BAR2QYrEZuA/25nDJiYxqYe8C/gN/hDadaZb52YYiViEeCcsf1l0fkd0ytPaZ/OhIFUnyEHD8+zsDbkHYlHlXeIubYTwT+vo0BlGpv242f8w0g82hpE20SKjYoNEZEUWIlYFfBN4FvAhDQ3XwH8DviN19D4b8NNkwxZidhI4PPAuWz2sPjY/nJOXlXNl1ZNZ1waRUfShrwj8ajyDonkUKkh807RX4ELI/Hoy0YbJlLkVGyIiKTBSsQm4Rcc3wTGZbCL/8N/Z/RWr6Gx02DTJEVWIlbN4NCZ7bb12vH95Xxp1XROap/OWK88k8NtyDsSjyrvPHBtJ+W8t+IJ/BmmHjTaMJESoWJDRCQDViI2FX+9hnOBMRnsYjXwR+B6r6HxaZNtk6FZidgBDA6dqUhn26r+EXx1ZU3bvI7qseVY6Qy7WW9D3pF4VHnngGs7Geed9B/89TX+pBmmRDKnYkNEJAvJd8nPx3+YfHSGu3kKf82Om7yGxtWm2lbqrETMAvYBjgdOAPbPYDce8HfgKuBfT756wFQM5h2JR5W3Icm1MYzmHYlHB4w1UKREqdgQETHASsRm4A/TOB2YmeFuOvHHhd8F3OM1NOoB1DQln8M4FP+G83PAjhnuajXwW+Aar6Hxtc2/6dqO8bz1wHH6ks9hGM07Eo9ukbeIZE7FhoiIQVYiVg4cDZwBHAdkNNA/6Tngbvyb0Yc1o9XQrERsPP7v/HigjvRmDdvcy8DVwO+9hsYPhnuxazuB5a0ZrYbm2k4geUfi0WHzFpH0qdgQEQmIlYjVAKfh34jOynJ3ncB9+DejdwOvlvLK1FYiVot/c3888GkyG5O/ngfciT90pjnT36trO4HmXcrPDbi2E1jepfx7FckFFRsiIgGzErEy4HD8m9AvkN2N0npvMngjeq/X0NhhYJ+hZSVi04H9gIPwbzoPMLDbDvyZon7pNTQuMbA/AFzbCTzvSDxa1Hm7thNo3pF41FjeIrJtKjZERHLISsSmAKcCXwX2NLTbfuBR/AePX8IfGvKS19C4wtD+cyZZmO2Mf6O5/0afawwe5iX8d7X/4DU0rjW43y24tpOzvCPxaMHlnSzMcpZ3JB4NNG8R2ZKKDRGRPEjOlHQw/rvfUTKbPnc4K0jeiG72uSUMQ7CsRKwC/wZ8fwZvNPcDxgdwuBeBO5Ifi3P98ydnSspb3mEYKuTaTt7yDsPPL1KqVGyIiOSZlYhNwJ+q82jgSGBKwIdcg38Tuv6G9I3k19YCH2z0+QNgrdfQ2J/qjq1EbDQwGf9n2Pjz5l/bEdgLGGniBxrCAP6CencAd3gNja8HdJy0ubYT6rwj8WjKebu2E8q8I/FoaPIWKXUqNkREQiQ5m9Vs4KjkxyFkN8ORCd1sdkOa/NyJv4r6xjeWma49YUIn/jMNdwCLCmEYWXI2K+WdmU3yLsRhZCKlQMWGiEiIWYlYFf7sO+tvRnfMa4PC5z3gb/g3nM1eQ2NXntuTFdd2lPe2bZJ3JB4t6LxFSoGKDRGRApF8zmNXBm9EP0UwY//DrBt4HrgX/4bzca+hsShXeU4+56G8N8tbq3qLFBYVGyIiBcpKxCqBjwNzgI8Ae+Cv71CWz3YZ1A48DTyz0ed/ew2NfXlsU964tlOSeUfi0ZLMW6RYqNgQESkiViI2Cv/d8D3xb0b3SP57N8ys9xCUtxm8yVx/o/l2GGbNCjPXdooqb80aJVJ8VGyIiJQAKxEbgb+ewcY3pHsAH8Z/6DdIncBKoC35sf7fr5K80fQaGlcG3IaS4tpOqPOOxKPKW6REqNgQESlxyUJkLP5N6LiN/r35582/NhZ/pqLNbyo3vtFs8xoau3P448gwkoVIYHlH4lHlLSIbqNgQEREREZFAFMtDZSIiIiIiEjIqNkREREREJBAqNkREREREJBAqNkREREREJBAqNkREREREJBAqNkREREREJBAqNkREREREJBAqNkREREREJBAqNkREREREJBAqNkREREREJBAqNkREREREJBAqNkREREREJBAqNkREREREJBAqNkREREREJBAqNkREREREJBAqNkREREREJBAqNkREREREJBAqNkREREREJBD/D9898GR6OGmZAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 1008x201.6 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "sg.add_transcriptome('../new_testing/files/test_full.gtf')\n",
    "\n",
    "sg.pg.new_gene(sg, 'test2_gid')\n",
    "sg.pg.graph_type = 'swan'\n",
    "\n",
    "sg.pg.new_swangraph()\n",
    "sg.pg.calc_node_edge_styles()\n",
    "sg.pg.plot_graph()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "offshore-shadow",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6/7/21 \n",
    "# plotting tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "angry-floor",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "married-recruitment",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding transcriptome to the SwanGraph\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "sg.add_transcriptome('../new_testing/files/test_full.gtf')\n",
    "beep = sg.subset_on_gene('test2_gid')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "heated-reservoir",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>coord</th>\n",
       "      <th>vertex_id</th>\n",
       "      <th>internal</th>\n",
       "      <th>TSS</th>\n",
       "      <th>TES</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vertex_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>80</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>75</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>65</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2</td>\n",
       "      <td>60</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr2</td>\n",
       "      <td>50</td>\n",
       "      <td>5</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr2</td>\n",
       "      <td>45</td>\n",
       "      <td>6</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          chrom  coord  vertex_id  internal    TSS    TES\n",
       "vertex_id                                                \n",
       "0          chr2    100          0     False   True  False\n",
       "1          chr2     80          1      True  False  False\n",
       "2          chr2     75          2      True  False  False\n",
       "3          chr2     65          3      True  False  False\n",
       "4          chr2     60          4      True  False  False\n",
       "5          chr2     50          5      True  False   True\n",
       "6          chr2     45          6     False  False   True"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "beep.loc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "proprietary-management",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "confused-copper",
   "metadata": {},
   "outputs": [],
   "source": [
    "ctrl_edge_df = pd.read_csv('../new_testing/files/plot_minus_gene_summary_edge_df.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "technical-assist",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "otherwise-tolerance",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strand</th>\n",
       "      <th>edge_type</th>\n",
       "      <th>v1</th>\n",
       "      <th>v2</th>\n",
       "      <th>color</th>\n",
       "      <th>line</th>\n",
       "      <th>edge_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>edge_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>exon</td>\n",
       "      <td>None</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>-</td>\n",
       "      <td>intron</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>intron</td>\n",
       "      <td>None</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>exon</td>\n",
       "      <td>None</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>-</td>\n",
       "      <td>intron</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>intron</td>\n",
       "      <td>None</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>exon</td>\n",
       "      <td>None</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>exon</td>\n",
       "      <td>None</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>-</td>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>intron</td>\n",
       "      <td>None</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>-</td>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>intron</td>\n",
       "      <td>None</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>exon</td>\n",
       "      <td>None</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        strand edge_type  v1  v2   color  line  edge_id\n",
       "edge_id                                                \n",
       "9            -      exon   5   6    exon  None        9\n",
       "8            -    intron   4   5  intron  None        8\n",
       "12           -      exon   4   5    exon  None       12\n",
       "14           -    intron   3   5  intron  None       14\n",
       "7            -      exon   2   4    exon  None        7\n",
       "13           -      exon   2   3    exon  None       13\n",
       "11           -    intron   1   4  intron  None       11\n",
       "6            -    intron   1   2  intron  None        6\n",
       "5            -      exon   0   1    exon  None        5"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctrl_edge_df\n",
    "ctrl_edge_df.drop(['v1_old', 'v2_old'], axis=1, inplace=True)\n",
    "ctrl_edge_df = swan.create_dupe_index(ctrl_edge_df, 'edge_id')\n",
    "ctrl_edge_df = swan.set_dupe_index(ctrl_edge_df, 'edge_id')\n",
    "ctrl_edge_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "stupid-investigation",
   "metadata": {},
   "outputs": [],
   "source": [
    "ctrl_loc_df = pd.read_csv('../new_testing/files/plot_minus_gene_summary_loc_df.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "smoking-vertical",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vertex_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>coord</th>\n",
       "      <th>internal</th>\n",
       "      <th>TSS</th>\n",
       "      <th>TES</th>\n",
       "      <th>old_vertex_id</th>\n",
       "      <th>color</th>\n",
       "      <th>edgecolor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>chr2</td>\n",
       "      <td>100</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>12</td>\n",
       "      <td>TSS</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>chr2</td>\n",
       "      <td>80</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>11</td>\n",
       "      <td>internal</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>chr2</td>\n",
       "      <td>75</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>10</td>\n",
       "      <td>internal</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>chr2</td>\n",
       "      <td>65</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>9</td>\n",
       "      <td>internal</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>chr2</td>\n",
       "      <td>60</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>8</td>\n",
       "      <td>internal</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>chr2</td>\n",
       "      <td>50</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>7</td>\n",
       "      <td>TES</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>chr2</td>\n",
       "      <td>45</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "      <td>TES</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   vertex_id chrom  coord  internal    TSS    TES  old_vertex_id     color  \\\n",
       "0          0  chr2    100     False   True  False             12       TSS   \n",
       "1          1  chr2     80      True  False  False             11  internal   \n",
       "2          2  chr2     75      True  False  False             10  internal   \n",
       "3          3  chr2     65      True  False  False              9  internal   \n",
       "4          4  chr2     60      True  False  False              8  internal   \n",
       "5          5  chr2     50      True  False   True              7       TES   \n",
       "6          6  chr2     45     False  False   True              6       TES   \n",
       "\n",
       "  edgecolor  \n",
       "0      None  \n",
       "1      None  \n",
       "2      None  \n",
       "3      None  \n",
       "4      None  \n",
       "5      None  \n",
       "6      None  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctrl_loc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "gothic-memorial",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>coord</th>\n",
       "      <th>vertex_id</th>\n",
       "      <th>internal</th>\n",
       "      <th>TSS</th>\n",
       "      <th>TES</th>\n",
       "      <th>color</th>\n",
       "      <th>edgecolor</th>\n",
       "      <th>linewidth</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vertex_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>TSS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>80</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>internal</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>75</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>internal</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>65</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>internal</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2</td>\n",
       "      <td>60</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>internal</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr2</td>\n",
       "      <td>50</td>\n",
       "      <td>5</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>TES</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr2</td>\n",
       "      <td>45</td>\n",
       "      <td>6</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>TES</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          chrom  coord  vertex_id  internal    TSS    TES     color edgecolor  \\\n",
       "vertex_id                                                                       \n",
       "0          chr2    100          0     False   True  False       TSS      None   \n",
       "1          chr2     80          1      True  False  False  internal      None   \n",
       "2          chr2     75          2      True  False  False  internal      None   \n",
       "3          chr2     65          3      True  False  False  internal      None   \n",
       "4          chr2     60          4      True  False  False  internal      None   \n",
       "5          chr2     50          5      True  False   True       TES      None   \n",
       "6          chr2     45          6     False  False   True       TES      None   \n",
       "\n",
       "          linewidth  \n",
       "vertex_id            \n",
       "0              None  \n",
       "1              None  \n",
       "2              None  \n",
       "3              None  \n",
       "4              None  \n",
       "5              None  \n",
       "6              None  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.pg.loc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "immune-census",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding transcriptome to the SwanGraph\n",
      "             tname        gid        gname               path    tid  \\\n",
      "tid                                                                    \n",
      "test1  test1_tname  test1_gid  test1_gname    [0, 1, 2, 3, 4]  test1   \n",
      "test2  test2_tname  test2_gid  test2_gname    [5, 6, 7, 8, 9]  test2   \n",
      "test3  test3_tname  test2_gid  test2_gname  [5, 6, 13, 14, 9]  test3   \n",
      "test4  test4_tname  test4_gid  test4_gname               [10]  test4   \n",
      "test5  test5_tname  test2_gid  test2_gname        [5, 11, 12]  test5   \n",
      "\n",
      "                    loc_path  \n",
      "tid                           \n",
      "test1     [0, 1, 2, 3, 4, 5]  \n",
      "test2  [12, 11, 10, 8, 7, 6]  \n",
      "test3  [12, 11, 10, 9, 7, 6]  \n",
      "test4                 [6, 7]  \n",
      "test5         [12, 11, 8, 7]  \n",
      "         v1  v2 strand edge_type  edge_id\n",
      "edge_id                                  \n",
      "0         0   1      +      exon        0\n",
      "1         1   2      +    intron        1\n",
      "2         2   3      +      exon        2\n",
      "3         3   4      +    intron        3\n",
      "4         4   5      +      exon        4\n",
      "10        6   7      +      exon       10\n",
      "9         7   6      -      exon        9\n",
      "8         8   7      -    intron        8\n",
      "12        8   7      -      exon       12\n",
      "14        9   7      -    intron       14\n",
      "7        10   8      -      exon        7\n",
      "13       10   9      -      exon       13\n",
      "11       11   8      -    intron       11\n",
      "6        11  10      -    intron        6\n",
      "5        12  11      -      exon        5\n",
      "          chrom  coord  vertex_id  internal    TSS    TES\n",
      "vertex_id                                                \n",
      "0          chr1      1          0     False   True  False\n",
      "1          chr1     20          1      True  False  False\n",
      "2          chr1     25          2      True  False  False\n",
      "3          chr1     30          3      True  False  False\n",
      "4          chr1     35          4      True  False  False\n",
      "5          chr1     40          5     False  False   True\n",
      "6          chr2     45          6     False   True   True\n",
      "7          chr2     50          7      True  False   True\n",
      "8          chr2     60          8      True  False  False\n",
      "9          chr2     65          9      True  False  False\n",
      "10         chr2     75         10      True  False  False\n",
      "11         chr2     80         11      True  False  False\n",
      "12         chr2    100         12     False   True  False\n",
      "\n",
      "         v1  v2 strand edge_type  edge_id                         curve  \\\n",
      "edge_id                                                                   \n",
      "5         0   1      -      exon        5   arc3,rad=0.9600000000000002   \n",
      "6         1   2      -    intron        6                arc3,rad=-0.96   \n",
      "7         2   4      -      exon        7  arc3,rad=0.48000000000000004   \n",
      "8         4   5      -    intron        8  arc3,rad=-0.9600000000000002   \n",
      "9         5   6      -      exon        9   arc3,rad=0.9599999999999996   \n",
      "11        1   4      -    intron       11                arc3,rad=-0.32   \n",
      "12        4   5      -      exon       12   arc3,rad=0.9600000000000002   \n",
      "13        2   3      -      exon       13                arc3,rad=-0.96   \n",
      "14        3   5      -    intron       14   arc3,rad=0.4800000000000001   \n",
      "\n",
      "          color  line  \n",
      "edge_id                \n",
      "5          exon  None  \n",
      "6        intron  None  \n",
      "7          exon  None  \n",
      "8        intron  None  \n",
      "9          exon  None  \n",
      "11       intron  None  \n",
      "12         exon  None  \n",
      "13         exon  None  \n",
      "14       intron  None  \n",
      "          chrom  coord  vertex_id  internal    TSS    TES     color edgecolor  \\\n",
      "vertex_id                                                                       \n",
      "0          chr2    100          0     False   True  False       TSS      None   \n",
      "1          chr2     80          1      True  False  False  internal      None   \n",
      "2          chr2     75          2      True  False  False  internal      None   \n",
      "3          chr2     65          3      True  False  False  internal      None   \n",
      "4          chr2     60          4      True  False  False  internal      None   \n",
      "5          chr2     50          5      True  False   True       TES      None   \n",
      "6          chr2     45          6     False  False   True       TES      None   \n",
      "\n",
      "          linewidth  \n",
      "vertex_id            \n",
      "0              None  \n",
      "1              None  \n",
      "2              None  \n",
      "3              None  \n",
      "4              None  \n",
      "5              None  \n",
      "6              None  \n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAxsAAACmCAYAAACyYpL/AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAABDwklEQVR4nO3deXwcdf3H8dckadqmR1p6hnAjICDncCqXVgEtCAilA4J0EREU0J8MioMIog4CW0UQOQSnXjAFQRGqgJZT5BzuQ2kpRyktPdKmR5Lm2Pn98Z20aZrNXjO7s7uf5+ORR/poZ2e+zTvznfnOfA/N932EEEIIIYQQImw1pS6AEEIIIYQQojJJY0MIIYQQQggRCWlsCCGEEEIIISIhjQ0hhBBCCCFEJKSxIYQQQgghhIiENDaEEEIIIYQQkZDGhhBCCCGEECIS0tgQQgghhBBCREIaG0IIIYQQQohISGNDCCGEEEIIEQlpbAghhBBCCCEiIY0NIYQQQgghRCSksSGEEEIIIYSIhDQ2hBBCCCGEEJGQxoYQQgghhBAiEtLYEEIIIYQQQkRCGhtCCCGEEEKISNSVugBClCPPcuuAZmCb4Gvb4Pt4YCXwPrCw75duG+2lKa0olOaYOeftJ5KStxBlQOpzIaKl+b5f6jIIEXue5U4ApgFfBHZDXZhyfTO4HHgTuBeYrdvGB6EWUgzKs9wGoAnYMvgaD9QP9PVB3frR9zQu3eXZhjXbLxnSOaa1prvB13I73uie2p7mrqHtB7aNXnLM6vHzt+sathRYDbQG3/t+bfZ3um10FfyfFkJsRupzIYpLGhtCpOFZ7ijgeOBU4HNAbYi794EngNuBP+u2sSLEfVcVz3KHsmkjover/9+NGWw/67QeHh25igdHreCZhtX05Ni4GIzmwz7tIzlqzTimrB3LmFRWL5Xb2bxRshJ4F1gAvB18f1e3jc7wSitE5ZH6XIjSkcaGEH14llsPfAF1QToWGFaEw3YDDwF3APfottFWhGPGnme5GqrBsBWbNyT6fo3L9xhdpHhyRCsPjGrhiRGrWF8TfX1Y62sc1Daao9ZswWfWjmGYX/A9j4/q2tG3AdL7fQHQotuGVPSi6kh9LkQ8SGNDCMCz3BrUBelK1M1tqSwFLMDRbSNVwnIUlWe5I4BPAHsGX3sE38dGcbwUPg+MauGGcR/w0ZDS9VbaoruOb6xo5ourx1NDiK9SNtXKxoZH/8bI+7ptdEd1YCFKQepzIeJFGhui6nmWewDwS+CgUpelj+eBC3TbeKrUBQlTcBOwPRsbFb1fO0J0d9t9vTZ0LTMnLOTV4euKcbis7NbRgLlsG/bsGFnsQ/cA76EaHq8BzwLPAO/I2xBRjqQ+FyJ+pLEhqpZnuU2ADcwIYXdLUDOWfIQaeLwNqgtQodNL/x64WLeNxQXup+g8yx3LxjcUvW8r9gBGlKI8y2s7uWH8Iu4bXXh36nHdQ5jcXc8W3XW01nazpK6T5XVdpApsLk1dPY7zlzczvqe+4DIWaDkbGx7PAs/qttFS2iIJkZ7U50LElzQ2RNUJBhR/G/gBkOuj5JXAn4H/oC5G76OmQVw/wHHqUGMKtg6+DgKmA5NzPOZa4MfALwc6TqkF/8+d2fxtxdalLFevTi3F7WM+4rdbLKatJreeDA2pmq7920Yv2r1jxNKm7vrWbTqHrdq+c9jKBr92PdDZ92u9lup+Y+i64W8NbR+5ZMj6SfPq23d6ddjaPdbVpnJqXDWkavhqSxOnrJpEvR+rpZDmsbEB8gzwchx/H0V1kfpciPiTxoaoKp7lHg7chuq2k6021PSGdwAPFjLzj2e5tcDhqP7EJ5JhhqR+3gJO023juXyPH4bg/7AX8Jng61Byv8hHqRX4EPjwoZEtPVdPfP+AVbXdY3L4/CZ5+4lk3nlrjpl33qN6apd8Z9nWtx+7ZnwrMBpoRN3s7IDqijY033KFpBN4iY1vP54B5kv3K1EsUp8LUR6ksSGqQjCz0QXATLKf8vDvwB+Bv+m2EXoH/+CJ3FHAV1AXqmysB76q28afwi5POsHPbjdUw+LTwBFENHA7gzZgEUFDIs3XYt021mmOWVDefiIZet6aY+adt59IbpJ3MPalt+GxY/C9758nhFTsXK1k07cfz+q2sbxEZREVSupzIcqLNDZExfMsdxhwE3BGth8BvqXbxpPRlarfAdWgxuuAA7P8yFXAJbpt9ERQFg1109rbuPgMMDHs4wxiAfBKn683gQ+ANdk8NdccM6+8/USyaHlrjplX3n4imVXewZoC/RsgvX/eDshqoY+QzAceBOYAj+i20VHEY4sKI/W5EOVHGhuionmWuxVwD7B/Fpt/BHwf+F0ppikMnlafjrrwTMriI/cDX9ZtY3UIx96aTRsXxRhv0crGBsWrwffXdNtYk+8ONcfMK28/kSx63ppj5pW3n0gWlHfQ9WMrNjZCdkQN3D8QNRg2Sm3AXNRT5jm6bSyM+Hiigkh9LkR5ksaGqFie5X4KuJvMFX0X8Avgp3Go6D3LHY0a7PhtYEiGzd8Avqjbxts5HmMSqjtU77iLj+Vc0Oz1oPonv9Lva2GY/fs1x8w570Jv3MOgOWbOefuJZE55ZyN4o7UdqtFxIHAAsC/RLoT2CuqNxxzgaXmyK9KR+lyI8iWNDVGRPMv9OnA9mSv3B1Dzn8+LvlS58Sx3Z1R3gU9n2LQFmKbbxsOD7GsUMIWNjYvdwypnP8uAl+nXDSrqrjOaY+aUt59Ixi5vzTFzyttPJNPmHRbPcoegZhY7gI2NkI9HdLgWVD5zgAdkql3RS+pzIcqbNDZERQmmJ7weOCfDpj5wKWDHefac4P+TBL6VYdMe4BzdNm7t89lRwLHAycDRRDN70evAI8DDwFO6bSyJ4BhpaY6Zc95+IhnbvIP/T9Z5+4nkrRm2C51nuY2obiy9bz8OJLtuIrlIAU+x8a3Hq3E+T0U0pD43in5+CxEFaWyIihH0RZ8FnJZh09WovrH3R16okHiW+1XgRjI/2TsXNRbiZODzhN/AmM/GxsWjxW5c9BVMKzuLLPP2E8myyVtzzGzzTviJ5KzoS5Re0P1qGzZ9+6EDw0M8zAdsbHg8HMVsQiJepD4HIKHbxqzoSyREtKSxISpCMBjvZuCsDJv+Dzhet43/Rl+qcHmWewhqcGQxpzX9ANWweBg1k9D7RTx2WsHg6qzz9hPJsstbc8xs8k4BJ/uJ5N3FKVV2gie4n0B1GZkKHEbmG6tsrUc1eOcA9+i28WFI+xUxIfX5BingZN02YnV+C5EraWyIshc8Wb0e+GaGTe9HLaLUGn2pouFZ7raoBan2iugQy9jYuHgYeDtu3RKCNTSyzttPJMs2b80xs8m7CzjWTyQfLE6pchcMkv0squHxBXJfdTmdFPAQ8DvgXt022kParygRqc830wUcq9tGbM9vITKRxoYoa8GFKQl8J8OmPwV+WIopEMPiWe5I1M3aKaixGDUh7HYV8CjBmwvg9bg1LvoKGhpZ512KKW3DpjnmSNTN9JcG2awd+Fwx1wrJV/DUeh/U7/JU1PgPLYRdtwKzUV1vno7z77EYWDXV572Cej2r87uYa4UIESZpbIiy5lnuT4BLBtmkB/X0yy1SkULVp4ExDfVEuNB+8G3AY2wcd/FSOU03qjlmVnn7iWRZ5p1O0G3sl8B5g2y2GjjCTyRfLE6pwuFZ7kTU+KKpqBWYR4ew23moRscfZC2P8lHp9Xk6QQM8q/Nbt42yOr+FAGlsiDLmWe4PgB8PskkKNXCwrC5MnuWOYGMDYyqFNzDagfuAO4F/6LbRVuD+SkJzzKzyrrSGRq+gwfFbBl85eTlwaDmOUYENU+1+CjgG9btf6DS7PmoRwd+hxneU5e9+NajU+jxbQYMjq/O7HMeoiOomjQ1RljzLvQi4OsNmM3Tb+F0xylOo4EJzJHAm6kYrzJl8PgQ+pdvGuyHus6g0x8wqbz+RLIu88xVMjTubwbtcLAIO8RPJd4tSqAh5lrsDG7tbHUFhs6utAe5CNTyekG5W8VFp9Xm+gokVsjq/y7k+F9VHGhui7HiWey7w6wybfV23jVuKUZ5CeJY7HtXA+DqwQ4SHmg8cpNvGigiPEQnNMbPK208kY593GDTHHAr8DdU4TWc+cJCfSJZd3ukEb/ymoBoeJ1DYrGwLUI2O38tNW2lVUn0eBs9ysz6/y7E+F9VJGhuirHiW+xnU7DO1g2z2Ld02ritSkXIWDII8GLUmxjQKe1rbjpoC9F7gG8F+0/k7alaTshlUqTlmVnn7iWRs846C5pgjgAdRXY7S+TtqlqqyyTtbQXero4EZqMkSCplW91FUw+PPum2sLbhwImuVUJ9HIWhYZ3V+l1N9LqqXNDZE2fAsdxvAA8YPstl3ddu4pkhFykkw2PvLqEZGIVPXtqMuNHcCc3oXOAtWdn4ENdNPOj/QbeOnBRy7aDTHzCpvP5GMZd5R0xwzq7z9RLIs8s6XZ7njUDO0nQHsV8Cu1gF3owaWPyY3cdEq9/o8apVWn4vqJo0NURY8yx0OPIFamTidH+q2MdgAw5LwLPcTqAbG6cCoPHfTgXqDcReqgTHgE1jPcicAj5N+YG0KOFK3jbl5lqMoNMfMKm8/kYxd3sWkOWZWefuJZKzzDotnubujGh2nU9haHguAXwCOrFYevnKuz4upUupzIaSxIWIv6HY0C/jKIJv9DLDiMugz6Hf7JVQj49A8d9PBpm8wsuri4VnuVsBTwFZpNlkK7KvbxqI8yxWpYC2NWWSRt59IxiLvUtIcM6u8/UQylnlHIRho+zlUN6vjyL+rYgtqPMGvdNv4KJzSVbdyrM9LqdzrcyFAGhuiDHiWex5qRdl07gamxeHC5FnudsDZwFnkP4D1UeA3wN/y7UPuWe7BqCdidWk2eRL4tG4bXXmVMEKaY2aVtzQ0NtIcM6u8/UQydnlHzbPcscDJqIbHQXnuphP4PfBz3TbeDKloVamc6vO4KOf6XAiQxoaIOc9yD0UtPpeukn0DNSvHmuKValOe5daiFiP7BmrhvXxWQ16Netp3U1g3M57lfgu4dpBNZuq2YYZxrLBojplV3n4iWbK840pzzIx5+4lkrPIuNs9yd0F1s/oK0Jznbu5HrXL9uNwQ56Yc6vO4Ksf6XIhe0tgQseVZbjPwAjAxzSargf1123ireKXaKFj5uHfa2u3y3M0LwI3AHWH3DQ+6K9wJnDTIZifqtnFPmMfNl+aYWeXtJ5IlyTvugu5nGfP2E8lY5F1KwQOCz6DednwJGJbHbp5HNTru1m2jO7zSVaa41+dxV271uRB9SWNDxFIw5uEx4MBBNjtWt437i1QkYEOF/0nUW4xp5DflZgdq4aZfA89F+XTUs9zRwHPAzmk2WQ3sp9vGvKjKkI1g7YiMefuJZFHzLjeaY2aVt59IljTvOAlm/ZmGeuNxSB67eA81mPy38kR+YHGtz8tNudTnQvQnjQ0RS57l3gJ8bZBNLtNt44oilkdDDTi9lPxuSADmATcBs3TbaAmrbJl4lrsH8AzpVyV/BThYt422YpWpP80xM+btJ5JFy7ucaY6ZVd5+IlmyvOPKs9w9gQuBU0nf1SedVajz+3rdNj4MuWhlLW71eTkrh/pciP5qSl0AIfrzLPdUBr8w/Q34SZHKonmWewzwNGqRpVwbGj3AX1CrwX5ct42fF7OhAaDbxqvAOYNssidQsikmNceMTd6VwE8kY513nOm28YpuG2cA2wNXo54UZ2sMcDHwrme5TjDlddWLU31eCeJenwsxEHmzIWLFs9wtgddRF+6BvAUcoNtGa8TlqAFOAH4A7J3HLj5EzSj1m7hMSZjh6WIK9TTs2SIWCc0xs8rbTyQjzbsSZXhblEK93Shq3uUm6LbyVeDbwDZ57OIB1LiOh6txMHlc6vNKFMf6XIh0pLEhYiPoqnQ/akangaxFXZgim3oyGDh6MnAJsHseu5iLGotxX9ymIfQsdxjwH9KvSPsaoOu20VmM8gQDmjPm7SeSMtVoHjTHzCpvP5EsSt7lzLPcIaiBuSawbx67eAnV6LgzbvVCVOJQn1eyuNXnQgxGulGJOEmQ/sIE8JWoLkye5Q7xLHcG8CZwO7k1NFaiBojuotvGZ3XbuCeONxS6bXSgbpjSDWL9BPC94pUoc97S0Mifn0jGLe+ypdtGl24bdwD7oWax+nuOu9gb+CPwtme5ZwWLDla6ktXn1SCG9bkQacmbDRELnuVug3oSMyrNJjfptnFuBMcdipqF5mJUP+1cvIpqZMwup8F4nuWeg5pudyBdwN66bbwRZRk0x8yYt59Ihp53NdIcM2PefiIZad6VyLPc3YHvAKcB9Tl+/L+ot6d/qcTuVaWqz6tRHOpzITKRNxui5ILX7beR/sL0DnBRyMccHqxkOx+4mdwaGh5wPKoSd8qpoRG4BbUa7UCGALcG3ckiEXSfKmreVS5j3ppjRpZ3pdJt43XdNr6KWmPHRr3hzNbHUStlP+1Z7qcjKF7JlKI+r3Ilrc+FyIY0NkQcnAN8dpB/T+i2sTaMA3mWO8Kz3AuBBcD1wFY5fPwpVLeA/XXbuFe3jVQYZSq2oNxfA9an2eRg4JsRFiFj3n4iGUreAvxEstR5VzTdNhbrtnEJagD5Baib6WwdADzsWe4DnuWm63tfbopWn4tY1OdCZCTdqERJeZa7A2pe8BFpNrlWt43/C+E4o1EV7neA8Tl+/DHUVIIVNaOMZ7kXA1em+ed1wO66bbwX5jE1x8yYt59IFpy32JzmmBnz9hPJUPOuRsF4jBNQT+/3z/HjLnCpbhvzQy9YERSrPhebK0V9LkS2pLEhSiaYXvYR4LA0m7yF6qrUXsAxxqKeNn4LGJvjxx8CfqLbxhP5Hj/Oghl2niH9bCYPAp8Pq4GlOWZWefuJZN55i/Q0x8wqbz+RlItCCILuRIegGh3H5vDRblTXmB/rtrEkirJFoRj1uUiv2PW5ELmQblSilC4g/YUpBZyR74XJs9zxnuXawHvA5eTW0LgfOEi3jaMqtaEBaoYd1BoCPWk2OQo4PcRDZsxbGhrR8RPJYudd1XTb8HXbeEK3jS+iZrH6Z5YfrQO+gZq56iee5TZGVshwRVafi8xKUJ8LkbXQ32xMn7N0C0BHvT4+HNgZGIoaqNSF6lf4FqprynOAN3vqxKKuqCzC07poSs55NzbPbfEsdxfU3PPD0uz6Kt02Ls61PJ7lDkctwGUBI3P8+D2oNxkv5nrccuZZ7pWo2bgG0gLsptvGR5B/3ppjZszbTyRzzlvkTnPMjHn7iWRBeUf6HyhjnuVOQXV1yaV7VQtqAPoNwXSnkYlbfS5yl0t9vniWlnPeTTN8Ob/LVCnzDqWxMX3O0tGoFvOFQDPQDjSg/gPpdAFtwHBgETAT+MPsqRNXF1wgEanWRVMKytv3WdTyjJ5a9condvS7Bpwx8nXUYkTpBrxtJuiyMB24itxW+k0Bs4Gf6rbxeg6fqxhBA+1lYKeB/r2mfv2CsfpLyxr3eGP7miHdY8gj75/On5y6+f2JO67pGXBSlNdRi8tlnbfIn+aYg+bdWNe94Nvbf7Ts7K2Xbz+iLjWGAurzxua5Up/3E9RVJwI/RV3ss/UBcBnwe902usMqTxzrc5G/TPW5VrNuwcgt7l02YsxD29fUrB9DAed30wxfzu+YWzxLC+3+vJC8C2psTJ+zdDfUf+AU1E1bukFh2ViH6tZ1BzBz9tSJMi90zLQumhJa3qmuOsBnzbwdWfXSnnSt3NDLqQc4ULcNL9t9eZZ7EGq9i4NyKEIPapEtW7eNt3L4XEXyLPcw1NOMDYaMXcmYvV9h1E5vg69RU5///c26bg1Ng7uXjOVX707kf+uG9/5TD3Cgn0hmnbconOaYm+W9y4h2zttuKSdOXknK1xhZV9Bka5vU543Nc6U+7ycYSJ5AdfPcMoePvol6c3tvIf3v41qfi8INVJ/X1X/AiLH30zDqKXxfo6a2oLbfJud30wxfzu+YWTxLi+z+PJ+882psTJ+zdChwBXA+qnUU5mqo3ahW1XXAZbOnTpSnISXWumhKZHmnejRI1bDqld1peXY/SNVeodvGZdl81rPcbYGfAUYOh+wCZgE/021jQe4lrlye5d4InENND+MOfJ7GPV6HmhQ1teF1texKQZevcfP7E7hyfhOdfs0VfiKZVd4iXJpj3gicU6+lsD62mK9vs4w6zWdIuCP5NqnPG5vnSn3ej2e5Dai69WJgTA4ffRq4WLeNxzJu2Udc63MRrg31udbFqHF3MmLMQ2h0o9WEOmP7Jud30wxfzu8SWzxLK9r9eS5559zYmD5nqQ7cBUxCvYqJShvwETBt9tSJ8lSkRFoXTSlK3qmuOnrW13dqNanDxu0055nBtg2msb0YNY3t0CwPsR74DXC1bhsLCyttZfIsd9uhE5bNn3zU3LrahnZqhoTWU2Mz67o1VnXXddZr/mEf2+HBQfMW0dAcc9u9R7fNn7XnO3UT67toqIt0kpoN9Xlj81ypzwcQzJz3PdTMeenGPgzkH4Cl28ZLmTaMY30uouFZ7rZDhi6YP7bp2rqaulZqajqjPNyG87tphi/nd4ksnqUV/f4827xzamxMn7P0YuCHqIpQy6eEOfKBDuCK2VMn/qwIxxN9tC6aUtS8fR9f01Tejc1zN8s7WAX1q6g1LyZmudsu4AZUI2NxaIWtQIv+nbi7YatFX9LqetCKcHZnyltE644XT7j7iHFrvjSsxqemGLV5n/pc8k7Ps9xmVL37VSCXlZ9/i3rTsWygf4xbfS6iteBX+t3DGl77ElpnUepz+pzfTTN8ybvIFs/SSnZ/nk3eWTU2ps9ZqqEGiHydaFtL6bQBNwMXzp46UeaIjljroimxybuxea4P4Fnu54Iy7ZHDfu4GvqfbxtvhF7Fy9Obt99Scr9Wmwnzlmq3N8hbR6c27M6WdX1/jS94xFczw9GNgWg4fWwVcCtzUO4g8jvW5iM7iWZqqz/3a8zWtp6Tnd9MMWcgtar15E4Pze7C8s/1FnAmcTWn+IwTHPRvVkrqwRGWoJrHJ27PcW4Ek8IUcPv888J1KXiMjZDOBs0vU0AA5v4ttJnB2iRoaIHlnRbeN/wEne5a7P2q63ClZfGwMcD1wlme55wEfbnlc023Dmz76pFabGnCqqCKQvItL1eelaWiA5F1ssblfY5C8M77ZCLpOXUrp/iN9tQE/li5V0Qletccib7+npnvFs/vWrHphn2yHrC4Cvg/8SbeNUEfBVao45U1wfkuXi+hI3uXLs9zPoibE0LP9zNh9X+wZu99LtVGOv8qB5B2xoCtNrM5v6VIVnXLKe9DGRjAY/AnUXLtx0Q4cMnvqxBdKXZBKEwwejFXeqa5aFt3zRdYvHz/YZm2o9TVm6raxrjglK39xzJvg/G5snivnd8gk7/LnWW4NcBLwE9Kso9Br6IRlNJ9wHzVD0i0oXRKSd0SCwcGxPL+bZviSd8jimnfTDH/Ahk+mxsYCYDuKM9gkWz7wDrDr7KkTI51eoZoE0yG+Sczy9n3oXj2K926fBqnNxkr6wO+AS3Tb+LD4pStfcc2bPud3Y/NcOb9DInlXFs9yhwBnosZ0TNhsg5oetj31LupGrynW4OBsSd4RCKY7jfX53TTDl7xDEue8m2b4A/ZEydQ9ZRLx+o+AKs9k1DzCIjxXEMO8NQ1qG9rZ4oDn+//TY8B+um0kpKGRl1jmjZzfUZG8K4huG126bdwM7IIao7FJt9FxBz5PbUN73BoaIHlHRc7v6hLnvAf+hwxvNuI8k0AbsN/sqRPfLHVByl2wkuzzxOt13CZSXbUsvOsEulaOnQ9cRIGr51azcsib4PxubJ4r53eBJO/K51nunsCvgEOHjF3J1tP+ErfuU/1J3iEJVooui/O7aYYveRco7nk3zfAHbHCEu1ZscdUjMx2E5ULUSpOxpdWmmPSZx54Gdtdt46/S0ChI7PNGzu8wSd4VTreNV4DDgXPH7P0KhLtKdBQk7/DI+V1dyiHvzZTzmw1Qg48mzZ46cU2pC1KuWhdNGQ0sIaat5H7agUmNzXMl7zxJ3tVF8q4uC+7+3tZj9Rffj/lbjV6Sd4EWz9LK7vxumuFL3nkqh7zTvdmIbB7mVE83b955He/+63baW5YyYtI27HTsV/nY1ESohwFOA24Mc6dV5nT69ffNx9p13Vx34/u8+PIaXnh5NStauvje/23H9y/cIYQibiB5Fy6UvF98eTWz71nC40+u5P2FHTQ01LLrLiP4zvnbcvintgihmIDkHYZQ8n7p1TXMvO5dXnl9DUuXdjJ0aA07bt/AV89oxjhxMjXhLEEueRdo3EHPfzHVXdtDbiuPZ/Tk0yuZetKLALz81MFsu3Uo9zqSd+FCOb8XLocDLxr4375zHJjHF3oEQPIOQyh591qyEmbeCw+/AivWwLhRsO+O8PMzYVTIzZnIGhveDd/lnX/ezg5Hnc4WO+3Fkhcf44UbL6ZzXSu7nfztsA4zAtV/X35583ch6udYkBUtXVx97bs0Nw1lz0+M5JHHV4ZQtM1I3oULJe9f3PAeTz69ii9+YQJnz9iKtW093H7nYo6b/hK/vGoXzvhycwhFlbxDEEre773XTkdHilNOaqJpUj0d61PMfbSFb/zfm7z6+lquvHzQWVizJXkX7sKaup5QGxrd3SkuuuQtRjTUsq4t1DcmknfhQjm/e31+X/hCv1Vcdt06rL1L3iEILe95i+HEn8GoYXD6ETBpDCxfA8/Ng/bO8BsbkXSjWrngNf55wRQ+ftJ57Dnj0g1//9RVX+PDZx5i6m3PMWzsxHx2PZBOoGn21IktYe2wWrQumjIO+BDVn7Ig69enaFnZRdPkoby3sJ29Dn4qijcbEOTd2DxX8s5RmHk//dwq9tlzNEOHbhz21dHRw6FHPUfLyi7mvXRIWE+7Je88hZl3OtPPeJlHnmjh3dcPo2F4KPe4kneeosr7V7e8z7U3vMdJx0/ipts+CPPNBkjeeVs8Swst7943GyG+xUinE2hqmuFL3jkKM2/fh89foaaOuvtiaBhacPE2KOoA8YWP/xWAjx1z1iZ/v9OxX6Ons4NFzzwQ5uHagX3D3GEV2Rf18yvY0KE1NE0O8Tc2Pck7f6HlfdD+YzZpaAAMG1bLkVPGsaKli6XLQptSXfLOX2h5p7PNVsPo7PTp6Ajtzb7knb/Q817y0Xqu+vk7/PB7O9I4OpKOEJJ3/iI5v9s7oaMr7L1u3D2Sd75Cy/vfb8Ir78KFx6uGRnsndHWHsef0ImlsrHz7FYaP35KG8U2b/P3YnfZGq6lh5fxXwzzccGD/MHdYRfYnHsvc50Lyzl/keS/5aD1Dhmhh3phI3vkLPe91bT2saOnk3ffa+YP7IX+6czF77TGKLcaGNjmK5J2/0PO+5Ip57LhDA6cZTZk3zo/knb/Q8775Qdjx67DD2XDEJXDPU2HuHZC8CxFa3o++pr4Pr4djfqwy3/5sOOkqeHNhGEfYXCSNjfYVSxi+xaTN/r52SD31o8bSvmJxmIerR035J3J3OOU3hZrknb9I835r/jru+8cyjjl6AsPD6VIDknchQs/7B1fMY8c9/83en3qKCy76Lwfu38gfb90jzENI3vkLNe8n/rOSe/62lKt/vHNYXSIHInnnL7S8NQ0O2x0umQazvgX26Wrp7/NugVv/GcYRNpC88xda3guWqO9f/7Uaq3HzN+ByQzU0TrwKFkcw5DaS96I9nR3Ujxo74L/V1g+jp7Mj7EPuHPYOq0S5/tzKtdylFtnPrXV1N6d/7VVGNNTyk0s/FvbuJe/8hP5zO/esrTlu6kSWLe/knw+v4MMl61mzJvT375J3fkL7uXV1pbjoB29x8pcmc4DeGNZu05G88xPaz22rceCam/7dKYfCkZfDVffA9ENCHTAseecntJ/buvXq+27bwG3nb/z7PbaDE66EWx6Ey4ywjqZE8majtn4Yqa71A/5bT2cHtfXDwj5kUQYLVKBy/bmVa7lLLZKfW3t7D0biZRZ+0MEdzp40bynnd0yE/nPb+WMjOOLQLZh2wmRuuX539tpjFFNPeoEVLaGN0QHJO1+h/dxuvHUhiz7s4EfWjmHtcjCSd34i/bkNHQJnToF1HeDND3fXoe6teoT2cxsWvB856eBN//7AnVXD89l5YR1po0gaG8PHTaa95aPN/r6nq5PONSsZPm5y2IeMbLaVClduXah6Sd75CT3vzs4Up531Ks+/sJrf37IHB+0/JuxDgOSdr8jP72nHT2Llqm7u/8eyMHcreecnlLxbV3dz9bXvcsq0JtraeljwThsL3mlj5Sr1BmvhBx28tzDUccmSd34iP7+3Gqe+t6wNdbeSd35Cy3vyGPV9wgAvLSc2Quu6sI60USTdqMbuuCcfvfiYGrvRp2Gxct5L+KkUYz+2Z9iHDPWxWhWJbs6JaEne+Qk17+7uFDPOfY1Hnmjhtht257OfHhfm7vuSvPMT+fndsV7NQrWqNdSuVJJ3fkLJe1VrF2vX9XCL8wG3OB9s9u/HTHuRiRPqeevFQ8I4HEje+Yr8/H43eIYwbnSou5W88xNa3nttD398DD4cYALixSuhOYJLeSSNja0PPY7//vl65t33m03W2Zh332+orR9G84FHh33IgftsiUzK9edWruUutdB+bqmUz9kXvMHfH1zOddd8nBOO3XxCiBBJ3vkJ7ee2bHknE8Zv/kDyt39YBMC+e4d6NyJ55yeUn9uE8fX87qZPbPb3f7l/KX+9fykz7Z3D7iopeecntJ/b8tUwvt8pvKYdfvMQNDbA/uEOw5O88xPaz+3ofeDS2+GOx8E4FGqDPk7/fEk1Nr4cwRD+iN5s7MH2nzuF/959A13r1jB2p7346MVHWfjE3/jEad8Lc0G/Xm+FvcMq8RawXVg7u8X5gNbVXbSuVk85//PMKq755TsATD9xMttsFdoIM8k7P6Hl/YMfz+eevy3lUweNYWh9DbPvXrLJvx/z+QmMaAhtRirJOz+h5Z049zVqazUO2r+R5qZhrGjp5O8PLee5F1Zz/DETOfSTA08IkifJOz+h5N0wvJbjjtn8Gv3G/1Rfms9+elyYi/qB5J2v0M7vn9ylZig6bHdoGqtuOO94HJasgp+fGe6ib0je+Qot73Gj4bsnwBWz1XS3x+4PS1bCbf+C7SfCWZ8L4yibiqSxAaB/8xoaJmzFO/9yeeefdzBi0tbs83Wbjx1zZtiH6gQeC3unVeIx4NOE1Bfw+pvfZ+EHG2cae+I/q3jiP6sAtQhcSI0NyTt/oeX98qtrAHjy6VU8+fSqzf/9gIMZ0SB5l1hoeRsnTeauv3yE88cPaVnZRcPwGnbdZSQ/v3IXzjh1y8JLupHknb9Q6/MikbzzF1reh+8OC5fB7x+BVetg5DDYe3u49iw4dLfCC9qH5J2/UM/vc46GMSPU26srZqvMjz0ALjkJRkewGldkjY2auiHsfqrJ7qeamTcuTDvwXNQHqVDPAW1AKHMbvvr0J8PYTSaSd/5Cy3vOn4u2CKzknb/Q8j5t+pacNj3URkU6knf+Qq3P+/v+hTvw/Qt3CHu3knf+Qsv7hIPUVxFI3vkL/fw2DlVfxRDJbFRFNhx4odSFKFMvoH5+5UTyzp/kXV0k7+oieVcXybu6lGPeG1RCY2PR7KkTBxhTLzJpbJ67AlhU6nLkaFFj81zJOw+Sd3WRvKuL5F1dmmb4ZZl30wxf8s5Dmea9Qbk3NtYB15S6EGVuJurnWA4k78JJ3tVF8q4uknd1kbyrSznlvYlyb2zUAH8sdSHK3B8on98Dybtwknd1kbyri+RdXSTv6lJOeW+iLAsd6AZunz114ppSF6ScNTbPXQ3cgfp5xlk3cHtj81zJuwCSd3WRvKtLb96+L3lXg6YZflnl3TTDl7wL0Js38a/PN1POjY0u1CslUbiZxH81cck7PJJ3dZG8q8tMP/43I5J3eGYieVeTcqjPN5OpsdFWlFLkrg24bvbUiW+WuiCVoLF57hvA9cQ878bmuZJ3COKed8qnA8k7NHHP25e8Q9XYPPeN99rr72jr1kpdlHSkPg9R0wz/jZ6uiXekUvWlLko6bcB1TTN8yTsETTP8WNfn6WRqbHwE+MUoSA58YAnww1IXpML8EMm7msQy75QP77fXD93hkT2eLHVZKkws8/Z96F4zcuiC206XvEOiOea+B/1n188v7RxCKlZpA1Kfh86z3H2XvnfN51PdjfiSd7WIZX3OIOXJ1NiYBnRk2KbYOoBps6dO7Cx1QSpJY/Pc9cQ478bmuZJ3iOKa9/qUxhkvb6+1dNX9VXPM80tdnkoR17z9nlqW/ONzWqpj2F89y5W8C6Q55vHA4+tTNZPPeHl7OlKxe7sh9XmIPMs9Hngcv37yysXfBj92bzc6gGlNM3zJO0RNM/xY1ucMUp5BGxuzp070gCuIz+uaNuCK2VMnyqIwEWhsnhvLvBub50reEYhb3uu6a7jq7cm8vKYBVN10neaYv9Ics67ERasIccs71VVHy3P7sH75eAjy9iz3V57lSt450hyzVnPMnwJ/AUYAvLymgavfnsy67tgMzZT6PCSe5dZ6lrtJ3l3rd2D1ihOIUXeqNuCKphm+5B2Bphl+rOpzgrzT/aPmZ/HebfqcpT8Hzib4pS6RNuCm2VMnXljCMlSF1kVTYpN3Y/NcyTticch7XY+Gs3A8P3hrq4H++QFgup9Iri5ysSpSHPJOddXR+trHWfGfgwf65weA6bptSN5Z0BxzHHA7cORA//7TnT9gxlYrGFGXKm7BNiX1eUg8yx0079Hjf0/DmIepqVlf3IJtqg24qWmGL3lHbPEsreT1OVnkne0jjwuBWyhdC6oNuAkwS3T8anMhcEswULcUJO/iKun53dGjpVRDozndJkcDz2mOuXfxSlXRSpp3qrs2pRoaB6Xb5GjgOc9y9y5eqcqT5pj7AM+T5sYT4JK3mrlr8dgVnSmtVDPYSH0eEs9yM+a9evnptK/+1IqUX1uqGaok7+Iqi/u1rN5s9Jo+Z+nFqIEpw4DIO4OqsvnrNa3m8tlTJ/4s6uOJTf3l5eP/dPDYtacOrfGpKU7XXx/V5++Kxua5kneRtS6aUtTzmyDvjh7Nnjx3712BUzNs34mqWG/wE8m4DYwrO6XKO9Vday+4+cyc8tZtQ/LuR3PM01GNxmEZNn0UOGHVkS+eQxHzTvngw/pajculPi+cZ7k55f3WThffcCCLTh1Kd1Gu3ypvbX2t5l/eNMOXvIvsDme7PxUzb/rcr2WTd06NDYDpc5bqwF3AJKAhnxJmo9vvpKOnjadX37eopXvxx/1Ecm1UxxKbC17Nv7nXqLYJv9vrHSbWd9FQF+n1vg01a8U06dNbOq2LphTl/KZf3ppjasClwI+y+Oy9wJl+ItkSYfmqQqny9iw357x125C8Ac0x61Fz7Z+Xxea/Ac73E8n1ULy813VrLO0cwtde3W7R860j5PpdAM9yc857v52eHwm8uQcfTbiF+5jAOhq06F50tPl1LGME3+QLi15gS8m7yHrv14qVN33q82zH5OQ8ciwYNL4rap7fdkJeTCbl99DtdzGv7QUeaLmVlu7FzcA1YR5DZOXnwISX1zRwwJO7cvPCCbT3aHSF3+23G/WLez2wqzQ0SisYRBzZ+U2avP1E0vcTySuAU4BMnY2PA17SHPOQkMtWdUqVt24bvm4bOeXtWW7V5605ZhPwMJlvPDuBs/1E8uzehgZEn3dXCtp6NG5ZOIEDntyV51tHyPW7AJ7l5pS3bhtn67axnuD6/SqTOJwZ/JZ9aPfr6PLDfeTd5Wu0+XX8ln04jAQvsKXkXRpFyZt+9Xkug/9zfrPR1/Q5S3cDvoN6HZ6igAEqKb9nfYrU0Pc73uR/bc+ypmezh1hH+YnkQ3kXVmRNc8yjgX/0//tdRrRz3rZLObW5pbMGujStoAFJ61CN3duBmbLAU/y0LpoS2vlNDnlrjnkw6mn2hAz7TAGXAVf6iWRPAWUTlC5vz3Jzzlu3jarLW3PMTwF/BiZn2PQD4EQ/kXx2sI3CzLsrxfouXxt69+KxXP/eJN5at1lPH7l+58iz3Jzy1m3jWUh//d6JFZzD85zM6501+AVdv7t8bX03tUP/yse5kf2Yz7j+m0jeRVKMvOlXn+ezQGNBjY1e0+csHQWcBlwENKOemAwHBpuDrbPPdot837/m3uXXH9vpd3w+zfYfAHv4ieSqggss0tIccxTwGrBNmk1mrTryxQsoMG/U048/NjbPXRNS0UVEWhdNKfj8Jse8NcfcCvgTcFgWmz8CnOYnkh9ms28xuFLk7VluznnrtlEVeQddDL8BXAtkmhb4UdTMbUuz3X8Yefs+1+zwyB7Hruyuk+t3gYIuhjnlrdvGUsju+v2hNrPg67fvc83ufOPYVQyXvEuoWHkT1OdNM/y879dCaWz0NX3O0i2AfYH9gcOBnYGhqP9YJ+qV+VvAY8BzwAuzp05sgQ2viF8HxqbZ/Sw/kUyEWmCxCc0xrwPSLa71EbBb377yrYum5Jx3Y/Nc6XtdpoqZt+aYtah+/ZeSucvncuArfiK52RMekb9i5u1Zbs5567ZR0XlrjjkcNdPLV7LYfCZwsZ9I5t01qpC85fpdOM9yc85bt40Need6/V48S8s576YZvuQdE8XMu+Cyht3YKJTmmKeinnClc6yfSN5frPJUE80xPwn8m/QzlUzzE8k/F7FIQqA55hGoOmHLLDafCVh+Iikr1pYpz3KPIMe8dduouLw1x9wOuAfYJ8OmbagJE2ZHXqgM5PqdP89ytyOHvHXb2CTvUly/Je/SKbf7tdgsLdrHHagTLp3fBCPvRYg0xxwK3Er6X9y/AncXrUBCBPxE8lFgL2BOFptfCPxbc8wdIi2UiIxuG4+SY96e5VZU3ppjfgHwyHzjOR84MA4NjYBcv/PgWW5OeQ/Q0CjV9VvyLoFyvF+LXWMjmD//XNRr8oFMRo2EF+G6BDVLyUBagW/K2gaiVPxEcjlwLPB/QKbFyfYHXtQcc3rkBROR0G0j57w9yy37vDXHbNQc8zZUQ2uLDJvfD+zvJ5KvRV+y7Mj1Ozee5TZ6lptT3rptDJR3Sa7fknfJlN39Wuy6UfXSHPNE1EwM6ZzkJ5KxarmVK80x90Q9VUk3GO1rfiJ5axGLJERammPuB7jAjllsfivwbT+RXBdtqURUPMvNOW/dNsoub80xj0KVf6sMm/rA5cBP/EQy/MnIQyDX78w8y805b902Nss7Dtdvybt44pB3PmLb2ADQHPN21DzsA1kO7CUz0BRGc8w64D+op4MDeQSYErdWsqhummOOBm4k8yrUAO8A3/ATyQeiLZWIime5Oeet20ZZ5K05ZiOQBM7KYvNW4Mt+IplNF7OSkuv3wDzLzTlv3TYGzDtO12/JO3pxyjtXsetG1c95qFUKBzIe+HOwmqrI309J/4vbjmolx+4XV1Q3P5FcjZrO70zUgMnBbA/8Q3PM2cEMKqLM6LaRc96e5c4OFkWLLc0xj0RNXZnNjedrwH7l0NAIyPW7H89yc847XUMjEKfrt+QdvTjlnZNYv9kA0BzzGOC+QTa5yU8kzy1WeSqJ5pjHoQYSpXORn0gmi1QcIfKiOeauwGxgjyw2Xw1YqHqj6haGqwSe5eaVd5wWAgzezM0ku5tOUItpnV1u3QHl+q0Eb+Zyznuw7oBxvH5L3tGJY965iH1jA0BzTAeYMcgmZ/mJ5G1FKk5FCGbreQFoTLPJ88DBhczZLkSxBOsRJFGLYWXjOeDrfiL5YnSlElEJ1iPIOW/dNkqed/A241Zg6yw2X44a7HlntKWKTrVfv4O3GTnlrdvGoHnH+fpd7XlHIc55Z6tcGhuNqIvFTmk26QQO9RPJZ4tXqvKlOeYwVL+/dNPsdQAH+Inkq8UrlRCF0xzzS8BtwJgsNk8BvwR+6CeSa6Msl4iGZ7l55a3bRtHzDt5mJIGvZfmRu1FjjbJeDTyOqvX6HbzNyDnv3tXA04n79bta845K3PPOVtzHbADgJ5KtwAlAuleK9cA9mmNOKl6pytp1DD6f9zfj/osrxED8RPIe1BoNg73K71WDmlr1zeAVtSgzum3klbdnuUXNW3PMzwGvkt2N5wrAQC3KVdYNDajO67dnuXnlnamhEYj19bsa845YrPPOVlm82eilOeZJwF2DbPI48Fk/kcw0L3vV0hzzDGDWIJs4fiJ5ZpGKI0QkNMfUgONRc7w3Z/mxe4Hz/URyYVTlEtHwLDfvvHXbiCzv4G3GNcDZWX6kIt5mDKQart/B24yc886ykVFW1+9qyDtq5ZR3JmXV2ADQHPNK4OJBNrnOTyS/VazylBPNMfcAngGGp9nkFeAgP5FsL16phIiO5pijgCuAC8juTe464IeoeiS2/V/FwDzLzTtv3TZCy1tzzBrUNL022fXVX4GazWd2XGeTCUOlXr89y807b902ssq7HK/flZp3MZRj3oMpx8ZGLfB34MhBNvuKn0j+oUhFKgvBE7bnSd+PcjWg+4nk/OKVSoji0BxzX+BmYL8sP/ISagC59CsuQ57l5pW3bhsF56055mdRT7f3zvIjfwHO9RPJjwo9dtxV4vXbs9y88tZtI+u8y/X6XYl5F0O55j2YsmtsAGiOuQUqiO3TbNIBHC43CkrQpeRO4KRBNvuSn0j+pUhFEqLoggvfuainj6Oy+IiPmoLyCj+RfCvKsonweZabd966beSct+aYewFXAUdl+ZEW1NNtt5LfZvRXKddvz3LzzjvbtxlQ/tfvSsm7WMo973TKsrEBGyr2p0j/imkl8Gk/kXy5eKWKn+AXdyZqYGQ6M/1E0ixSkYQoKc0xtwSuBaZl+ZEU8Hvgx34iuSCqcoloeJabd966bWTMW3PMrYEfA18BtCyP8VfU24x0i6BVtHK+fnuWm3feum3klHelXL/LOe9iqpS8B1K2jQ0AzTFPQT2JSmcZcISfSL5RpCLFjuaYlwGXD7LJv4HPyCAtUW00x/wCcAOwXZYf6UEN1vuJn0i+G02pRFQ8y807b9023u3/j5pjjkH1R/8WMCzLfbYA5wN3VNPbjIGU2/Xbs9wxFJB3Lm8zelXS9bvc8i6FSsq7v7JubABojjkT+M4gmywBDvMTyXlFKlJsaI75beAXg2yyFNjHTyQ/LE6JhIgXzTEbUAOELwTqsvxYN2pth5/KzFXlxbPcgvLWbWOh5phDUd2zLgW2yOHwdwEXVOvbjIGUw/Xbs9yC8s71bUavSrx+l0PepVKJefdVCY2NOuABYMogm32A+gV+pzilKj3NMc9EXSDTSQFH+onk3CIVSYjYCmb+uAn4ZA4f6wR+A1zpJ5KLIimYiIRnuTnn7eN3/ndo2yMXNb2965IhndvkcLgngYv8RPKpXMtZ6eJ8/Q6mUz4ZuJL04w0G8iRwkW4beeddqdfvOOddSpWad19l39iADSP3HwIOHGSzd1G/wBX/JFJzzGmAy+BTP57lJ5KD/XILUVWC6Uq/iuqPncuCU+tRN64/k6fW5SOYrjTnvNdrKe5uXMbvxi5mRd2gs+W+BXwPuLfau0wNJo7Xb89yD0fNMLV/Dh/bkHc+XaZ6Vfr1O455l1Kl592rIhobsKH/7Fxg30E2m4/6BV5clEKVgOaYn0ctVjVkkM2+4yeSg72uE6JqBV2rzkXdOEzI4aPtwK+BqytxUbZKFXStyjnvDq2HuxqX8fuxS1i5aaNjKXAZcFs59q0uhThcvz3LHYJa+fp84JAcProhb902Csq7Wq7fccg7Dqolb6igxgaA5pjjgEeAPQbZ7E3UIKSKuxnQHPMw4EEGH7z2Iz+RvLw4JRKifGmOORL4JvBdcuur3YYaiHxjNXUFKHee5Y4EvpnC/14N2thsP9eu9XDXmGXc3bi0fdGQzqtRs8Wsia6klalU12/PciegVvw+l+xXnwd1nl8DzNRto+C8q+36LfdrVZZ3JTU2ADTHnAQ8Cnx8kM0WAMf5ieRrRSlUEWiOeQxwBzBykM2uRbWSKyt0ISIUvPY/HzCBMTl81Af+BdyK6kqzPvzSibBojqkD54/oqTGmt04aetrKSYxOZTuGHHx8X0PbkLduG5J3jop5/fYsV0ed1wYwNIeP9qAy/pFuG6E8da/W67fcr1VP3hXX2IAN8+g/Duw4yGbrUCtX3lOcUkUjmJf5u6hBbIPN+f1bVL+/ygtciCLQHLMR+DZqNpXROX58OfA74FY/kfxvyEUTedIccwjwJeAC+g0WH9FTy6mrJvLlVZMYmUOjI7Ahb902JO8cRHn9DrpKDZh3lv4GXKzbxpt5fHYzcv2W+7U0Ki7vimxsAGiOuQ3qF3jbDJv+BLjMTyRT0ZcqXJpjDkfNhvPlDJveBZziJ5I90ZdKiMqmOeZYVIPj2wz+ZCqdf6OejN7lJ5JtIRZNZElzzIls7Dqz5WDbjuqp5curJnHKykmM8GvzOdyGvHXbkLyzEPb127PcrPNO41nUDFOP5/HZAcn1eyO5X9tEReZdsY0NAM0xd0D9Amfqh3k/cJqfSLZGX6pwBE8D/krm2TL+ARzvJ5KdkRdKiCqiOeZ41HoNFwANeexiNfAn4Dd+IvlimGUTA9Mccz82dp2pz+WzjT11fG1FU8u01okjatFy6XbTa0Peum1I3hmEcf32LDfvvAP/Q62v8edCZpjqT67fm5P7NaCC867oxgaA5pg7o2Y92CrDpv9D9Qv8X/SlKozmmPujfnEzPaF5DPiCPD0VIjrBU/KLUIPJh+e5mxdQT71u9xPJ1WGVrdoF3Rb2BI4Djgf2yWM3PvB34DrgX8/P2288Ieat24bknUau1+/n5+33FiHnrdtGqE/R5fqdntyvVW7eFd/YANAcczLwZ+BTGTZdjZp3/e449pULLpynA7eQeUDb74Bz/ESyI/KCCSF665lzgTPJfLFMpw3VL/wB4KFKnvYxKsE4jENRN5xfBLbLc1erUX2nb/ATyfn9/9Gz3NDzDmvAcSXJdP2u9TX2aR/JZ9aO7Txm9bhVDX7txDwPtSFv3TY2y7tQcv3OjtyvVaaqaGwAaI5ZD1yP6reZyT+B//MTydejLVX2NMfcHfglg6+8CWqlyYuAX8TxBBSi0mmOWQscDZwFHAvk1dE/8ApqesQHgCdlRquBaY45CvUzPw6YSm6zhvX3Jupa8Qc/kVybaWPPciPLW2a0UvpfvxtSNRy8rpHD143hkHWNOc0aNoANeeu2kTHvfMj1Ozdyv1Z5qqax0UtzzK+jfokHW0QF1PR2vwYu9xPJlsgLlkaw+M3lwHlkvoitBgw/kfxHxMUSQmRBc8wm4AzUjehgs61kow01L/2Dwde8Sr9ADUZzzGbUzf1xwGfIr09+Lx/VF/w6YG6+P1fPciPNO8xxA+XGs9zm+0ctv3qLniGn7Nc+Sqv3B1twOaNN8o7q5yrX78LI/VrlqLrGBoDmmIegXtNNymLzFtQAsVv8RLI708ZhCZ6OngnYwPgsPjIP+KJMqylE/GiOWQMcgboJPZHCbox7vcvGG9GHy2nAZD6COfn3Bg5ENTL2C2G3raiZon7tJ5ILQtgfAJ7lRp63bhsVnbdnuZHmrdtGaHn3J9fv8Mj9WmWoysYGgOaYWwF/IfsK7DXgMmBOlF0Zgl/az6J+affN8mP/BKb7ieTKqMolhAhHsHLu6cDXgN1C2m0P8BRq4PEbqK4hb/iJ5PKQ9l80QcNsB9SN5j59vjeFeJg3UE+1/+gnkutC3O9mPMstWt66bZRd3kHDLNK83x/SsX5+ffustbU93//RxedFdp2U63c05H6t/FVtYwM2zHv8a2BGDh9rBe4GbgceDWMu5GAg0YHAKcB0smvB9/olYBazFS+EKFxw3h+EevptkN/0uZksJ7gR7fd9URy6YAV9s3dD3Vz23mjuDYyK4HCvA/cGX88V+//vWW5J845DFyzPcouW99v17Tw2YhWPjVzF60PX9S6hJtfvMiX3a+WtqhsbvTTHPAr4BbBrjh9dAsxGLTv/fC6/yMEv7O6oX9hTgO1zPPYbwLf8RPJfOX5OCBEzmmOORk3VeTRwJDAu4kOuQd2E9t6QvhP83TpgbZ/va4F1OdZtw4EtUP+Hvt/7/912qDowU3/sfKVQC+rdC9zrJ5JvR3ScnHmWG+u8ddvIOm/PcmORdw8+Lw1fy+MjVvHoiFUsqs/4QFuu32VI7tfKkzQ2AsF0id8AfgQ05rGLbuAD4D3g/T7flwATgG2Arft8bUN+T7ZaUa8Hf+0nkl15fF4IEWPBq/l9gaOCr4MpbIajMHTQ74Y0+N6GWkW9741lvmtPhKENNabhXlQXith3Kwpms5K889MGPLheS91/2tZvTH5naMd3ket3xZP7tfIjjY1+NMecAFyBmnKtoOkuQuYDtwGX+Ink0lIXRghRHJpjNqJmW+q9Gd2upAWKn4+A+1ANjLl+Itle4vIUxLNcyXtwm+St28aGvOX6XV0k7/IhjY00NMfcG7gWOLy0JQHUQMDz/UTSK3VBhBClE7zO34mNN6KfJpq+/3HWAbwKPIy64XzGTyRDXeU5LoJxHpJ3v7wzreot1+/qInnHnzQ2BhFc2I9FzSJyNFDQykE58oHHgRuBO+MwmFMIES+aYw5FrbQ7BfgEqh/zjsTrKV8hVgIvAi/1+f7fahxgCeBZblXmrdtGznnL9bu6SN7xJo2NLAXTVZ6EGhx0GL1zW4TvedQAptl+IrkoomMIISqU5pjDUE/Dd0PdjO4a/HlnwlnvISrvs/Ems/dG8325cA/Os9yKyjuKWbPk+l1dJO/4kcZGHoI5n6cDp5L93MqD+S9qajbXTyTnhbA/IYTYhOaYdaj1DPrekO4KfBw16DdKbcAK1KJbLX3+PI/gRtNPJFdEXIaq4llurPPWbaMkecv1u7pI3vEgjY0CaY65NaoC3wbYNvje++etUa/yelAzHyzs9/U+6hd3njy9E0KUStAQGYG6CR3Z58/9v/f/uxGomYr631T2vdFs8RPJjiL+d0QGQUMksrx12yiLvOX6XV0k79KRxkaEgiksRwFrwlhMRgghhBDRk+t3dZG8oyWNDSGEEEIIIUQkKmUGCyGEEEIIIUTMSGNDCCGEEEIIEQlpbAghhBBCCCEiIY0NIYQQQgghRCSksSGEEEIIIYSIhDQ2hBBCCCGEEJGQxoYQQgghhBAiEtLYEEIIIYQQQkRCGhtCCCGEEEKISEhjQwghhBBCCBEJaWwIIYQQQgghIiGNDSGEEEIIIUQkpLEhhBBCCCGEiIQ0NoQQQgghhBCRkMaGEEIIIYQQIhLS2BBCCCGEEEJEQhobQgghhBBCiEj8P28zRzIv9a7nAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 1008x201.6 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "sg.add_transcriptome('../new_testing/files/test_full.gtf')\n",
    "print(sg.t_df.head())\n",
    "print(sg.edge_df)\n",
    "print(sg.loc_df)\n",
    "sg.plot_graph('test2_gid')\n",
    "print()\n",
    "print(sg.pg.edge_df)\n",
    "print(sg.pg.loc_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "answering-supervisor",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>v1</th>\n",
       "      <th>v2</th>\n",
       "      <th>strand</th>\n",
       "      <th>edge_type</th>\n",
       "      <th>edge_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    v1  v2 strand edge_type  edge_id\n",
       "0    0   1      +      exon        0\n",
       "1    1   2      +    intron        1\n",
       "2    2   3      +      exon        2\n",
       "3    3   4      +    intron        3\n",
       "4    4   5      +      exon        4\n",
       "10   6   7      +      exon       10"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.edge_df.loc[sg.edge_df.strand == '+']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "naked-insulin",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding transcriptome to the SwanGraph\n",
      "                         tname                    gid                  gname  \\\n",
      "tid                                                                            \n",
      "ENSMUST00000020637.8  Canx-001  ENSMUSG00000020368.14  ENSMUSG00000020368.14   \n",
      "ENSMUST00000146979.1  Canx-003  ENSMUSG00000020368.14  ENSMUSG00000020368.14   \n",
      "ENSMUST00000153068.1  Canx-002  ENSMUSG00000020368.14  ENSMUSG00000020368.14   \n",
      "ENSMUST00000155801.1  Canx-004  ENSMUSG00000020368.14  ENSMUSG00000020368.14   \n",
      "ENSMUST00000179865.6  Canx-201  ENSMUSG00000020368.14  ENSMUSG00000020368.14   \n",
      "\n",
      "                                                                   path  \\\n",
      "tid                                                                       \n",
      "ENSMUST00000020637.8  [0, 29, 30, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 1...   \n",
      "ENSMUST00000146979.1               [32, 21, 22, 23, 24, 25, 26, 27, 33]   \n",
      "ENSMUST00000153068.1                                       [34, 23, 35]   \n",
      "ENSMUST00000155801.1                                       [36, 19, 37]   \n",
      "ENSMUST00000179865.6  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,...   \n",
      "\n",
      "                                       tid  \\\n",
      "tid                                          \n",
      "ENSMUST00000020637.8  ENSMUST00000020637.8   \n",
      "ENSMUST00000146979.1  ENSMUST00000146979.1   \n",
      "ENSMUST00000153068.1  ENSMUST00000153068.1   \n",
      "ENSMUST00000155801.1  ENSMUST00000155801.1   \n",
      "ENSMUST00000179865.6  ENSMUST00000179865.6   \n",
      "\n",
      "                                                               loc_path  \n",
      "tid                                                                      \n",
      "ENSMUST00000020637.8  [37, 36, 35, 33, 32, 31, 30, 29, 28, 27, 26, 2...  \n",
      "ENSMUST00000146979.1                  [15, 13, 10, 9, 8, 6, 5, 4, 3, 2]  \n",
      "ENSMUST00000153068.1                                      [11, 9, 8, 7]  \n",
      "ENSMUST00000155801.1                                   [17, 16, 14, 12]  \n",
      "ENSMUST00000179865.6  [37, 36, 34, 33, 32, 31, 30, 29, 28, 27, 26, 2...  \n",
      "    v1  v2 strand edge_type  edge_id\n",
      "28   3   0      -      exon       28\n",
      "31   3   1      -      exon       31\n",
      "33   3   2      -      exon       33\n",
      "27   4   3      -    intron       27\n",
      "26   5   4      -      exon       26\n",
      "hewwo? wevewse stwand?\n",
      "[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37]\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAxsAAACmCAYAAACyYpL/AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAB9lklEQVR4nO2dd7gcVfn4P3NveggJSQgJodyE3suF0BXpHRSRVSwsItgbi/pbC9ar4mIBBEF0UUQu6JeONOnSWYrS61IDKYSQkH7v/P6Ys8ncc87MTjlTbpjP8+SBPbP33XdmTj9vsWzbpqCgoKCgoKCgoKCgwDQdWStQUFBQUFBQUFBQULB6Uiw2CgoKCgoKCgoKCgoSoVhsFBQUFBQUFBQUFBQkQrHYKCgoKCgoKCgoKChIhGKxUVBQUFBQUFBQUFCQCMVio6CgoKCgoKCgoKAgEYrFRkFBQUFBQUFBQUFBIhSLjYKCgoKCgoKCgoKCRCgWGwUFBQUFBQUFBQUFiVAsNgoKCgoKCgoKCgoKEqFYbBQUFBQUFBQUFBQUJEKx2CgoKCgoKCgoKCgoSIRisVFQUFBQUFBQUFBQkAjFYqOgoKCgoKCgoKCgIBGKxUZBQUFBQUFBQUFBQSIUi42CgoKCgoKCgoKCgkQoFhsFBQUFBQUFBQUFBYlQLDYKCgoKCgoKCgoKChJhSNYKJEWj2jsW+DAwAri8u6c0y+N7awNHApOAW4H7u3tKdmqKGqJR7e0A9gG6gRu6e0qPZaySQqPauz1wEPAgcFt3T6k/W40G0qj2bgAcDbwGXNvdU1qcsUoDaFR7xwEfwWm3V3T3lGZnq9FAGtXeYcAhwEbAVd09peczVmkAjWqvBewB7Anc2t1TeiBjlRQa1d4tgMOBx4GbuntKKzJWaQCNau9k4KPAPODq7p7SgoxVygWNau8aOO9tY+B+nPqVm3dn1StDgP2AnYFngWvtcu29pH+3Ue0dIX5ze2ApcC/wRN76fpM0qr0TcMaRFTj99Lw48qx6ZQxwFDAGuNwu196MraRBGtXeIcCBwJbANd09paczVkmhUe3dFdgb+A9wd9w5nlWvbIMz1j0C/Nsu15T6bNUrFjAD2BeYDVxpl2u5GrPTxLLtQTevbkuj2rsHcCkwVRTNB07o7ild7vrO9sBPcCa/7kXXC8DvgbPyNFj4ITq3vwMHuIrPBr7e3VPqy0arVYjO6NfAV1zFNwEf7+4pvZ2NVgNpVHtPAs4Ehouip4GPdveUnshOq1U0qr0fBHqByaJoHvCZ7p7SNdlptYpGtXca8H/ADqJoBfDd7p7S6dlptQoxGazjTJRbXAR8rruntDQbrVYhFkI/Ar4HWKL4XuCY7p7S65kp5qJR7f0Y8CdgDVH0CvCx7p7S/dlplS1ig+KnOJPLUa5LbwEXAj/p7iklPqn3QkxUvwuUcTbUWizCaa9Vu1x7zeRvNqq9o4Ev4Wzi7QQMk74yD2fSdxHwz8G4uedFo9p7EPA3YIIomgV8orundEsUeVa9sgtwGbCBKFoAnGiXa5fF1dUEjWrvVOAfwG6iqB/4GXBaHt6rWOyeD3zKVfx/OGNn6HZp1SudwOnAN13FtwKl1kJCfOfL4t/Gru/14cx7vmeXaw+H/e3BzmpnRtWo9m4E/ItVCw2AsUBvo9q7s/jOgcADwGGopzsb4UyMLxWT5FwjJin/ZOBCA5yK/oP0NdLyMwYuNMDR9xKhf6Y0qr0fAc5j1UIDYHPgOnGakCmNau9mwHWsWmgArAX8o1Ht3UH/V+nRqPaOAm5g1UIDnHb1y0a19zPZaKVwIQMXGuAMQL9JXxUtpwDfZ9VCA5wB/Io89EONau/ewCWsWmiAMwG6Vkw43nc0qr1b4exsfoqBCw2AdYBvA3c2qr1rpq0bgFWvjAPuEnpMki6PwtH7Eate2dzUbzaqvd3Aw8Avgd1RFxrg9F2H40yir2hUeyea+v0sEX3xlaxaaIDz3K8SfXgorHplQ+B6Vi00wDnd+LtVr+weQ1UjiJPs61i10ABnTvl9nPlHHjiHgQsNcDYGzoso7zQGLjTAsSi5zKpXLHGCeCnwWwYuNAA6gYOB+6x65aCIvz9oWe0WGzi7OLrOfShQFaceV4jPfnwEOD8Pk+E2HIJzPKjjlKwGuhaNau+6wNc9Lh/AwI4qdRrV3k7gFx6XN8TZEcyaHwCjNeXDge+krIuOzwKbelw7TZj4ZUaj2jsDZ4DRcVLWk2Vh8vl9j8s7A4emqI6C6AN/jn68mIizi/2+Qpzk3QSMb/PVHYGrG9XekclrtQqrXhmNMxHcrs1XJwI3WfXKBm2+50uj2tvZqPZ+B7gP775Ax5HAY41q7z5xfj8n/JiBG1YtRgOVCPK+g7Mwk+nEOQHNmk/gXb++JxYjmdGo9m6J9/h9XKPaG6aeYtUrk4BTPS7vDXwI5xTFa6xpMRS43KpX9gjz+4Od1Wqx0aj2rgd80ucrh+OcegTt+Ms4g2ye+bbPtdE4HUKWfB397laLk1LSw4ujgE18rp+U5YKzUe3tAo71+cqHG9VeedcyNRrV3qE4u/JeTMOxFc8SvzbSCZyQliIenIx+g6RF1m1kL2BXn+sniHrwvkD4+f0bWDfgn3wQ52Q9lfFWmHH8E+dkIQjrAzdb9YpuYtsWcbJ5Pc5YGeUUbl3g341qr187zTWNau/WOJYSXnw8zMafVa+sg/9G10Hi5CMTRF3+ls9XJgFHpKSOF14LgxafCynvqzg+wF6cTfDNyZHAdVa9smVIHQYtq9ViA6fy+A16naiD+nKco7ESzpG4zLcb1d5dzKhnlka1d1uciYAfJ2c1WRY7G+0mSsc2qr2RBjlDtDvu3Zz2zzhJPo9Tb70YChyfjipaDsU5AfLj5DQU0SFOLT7c5mufEydcqSPaZruTgYMb1d7MJha0byPrkP3EIk1qwHRN+e9xTsRv0Fw7Av+NMJOchOOLKPMvHP3O1VzbFMfcNRRi0nkRsL/HV+7C8UU6AMeM8XeALnCEBfyiUe39dFgdcsIX2lwPu/H3WfSnJC0s4MQQ8kyzN7BFm+9k2e+vBRzX5mvHN6q9fs94JcI86vNtvqZ7Ho8CH8c5uV4uXRsLXCAcyVd7VrfFhuy30A4b+GR3T+nH3T2lS3F2YB/XfK+WU3OqIPe7PY6TXhbMwGlQfowgvUF4AMKRcc8AX81yZznIO/5chqZKQfQ7olHtnZK4Jnr2Y6AfhI71caKpZMGmDLTJ1mHhTD5SRyzCgpxMZX36kgqNau9OgG5C/P3untKXu3tKV+AsLP6l+U6POAVIDOGn8WPNpauAI+1y7Qq7XPsijlO7zElWvdJuAinTg7OAkXkXOK67p/SB7p7SD7t7Sjd395T+r7un9HVga7x9pf7YqPZm7o8QgSD9YJiNvyDyPismwVkQRL/9hA9tFuxNe1P5ibTfiGqxAwN9cYLwOLCvXa712uXaT3HmObLT/G60N7taLVhtFhuNau8YHPtmNzfjOGx58Y3untLKqA4iMtIBOBEk3OxJ8EqZJrKdaz+qIzZkZ0qls8PV7eJmpd+eqMf+P0Z9/0cH3QExSaPaOx5nsejmWhyTBTcbk92CUn7Hs3F2Mt0MQXXOTgtdHfyipixPbeQbqLtgWem3Hard+FnAM1LZAcK8aLVFTBTP0Fw6D9epQHdPaTlwDM6uppupqM6lpqniTKLc3I8TLccdXfEHwF+l77Ui7QSiUe0tozdR/A+wbXdP6e+6v+vuKS3t7il9E8dZVu5rh+E4jWd5khcKEZFMdga+BLhbKtseaOsobtUrI1F9Ge/EcTx2MwXHRC8L5H7rPfR1oZSCLjpM9/s6eV9BXTy0eBM4wC7XVkbbFBHEdO3/l1a9kvr8Im1Wm8UGjqmLbApxG/Bnj+8/gjNoDqC7pzQTfRSnX2bt8ORG2EjL5j2P4ISmfFcqn5GKUipyA12OExXoLql8h4xsvj+kKbsRdRAeAWyTvDoKH0Tdlfeq06m/Y+H8Lw+etwEXaL6ehX4W6jt+GWdyKIf7zEsbAWdSca1UtpEIcZ02ujZyC/o6mNWCNy2OAj4glc0Bvi2H+ezuKS0CvqaR8R2Rq8Q4Vr0y3eM3v2SXa0vcBXa5ZuPYtMt5Ug6z6pW2ztoiOtn5mksPAAd095Rebieju6d0A87mnhyCdBKOU/0a6l/lEl0buRVnLJYJ0s/oonjdSn76/bE4+bzc3IXT78uh9rPqV+V38jZOv/+UVD4j4GmT3Cb6cOYJt3l8v2qXazM15WehWs9MR78QWq1YnRYbug7yVuC/Ht8/xSex0J+AJ6WyjcnWNl5mJwaGoQQnkdRi4CGpfPu0bdJF9BV5d+Y+MQjLi43hwFapKDYQ3e7Mg6j6gdq5poFXndbpt2PCuujQDbK3dfeUXgOaUnkWz28jHBMpN61kkvKu4yZiEE0NYfq2t1T8tNjw0L3jLMIcy3XQxtllzUsbSQXxrnTBQn7Q3VOar/ub7p7Snagn66NJLoLcj1AnqX+1y7WG7st2uTYLxwxKxtd3Q7STv6OeCr8CHBkmGapIPqszo90WxzdmMKDtB4neRnT9/m3APag76Vm0uQ+gzh1vE5Yhcl6q1PVrVHvXQZ1P3C76/f9I5evQJtCDVa8MQzW3btjl2rvAMs2fPIu6YQmAXa71oY9M9n2rXknUxDJrVqfFhtzgFwANnNB6Mgu7e0peK1JEMj9dJIN2DkJp4tXBgXPfbkYR4PjWMF67M6DqBylPlr12Z4QJROb6CXS7M//t7im9BciJ3vK0GAL1GW4ufGTSxGvHEfTvePvkVNGyNarJi59+qb5jcdoo7+Q/LDIiP4ZjtukmizaSFvui9qFPAX9s83ffxklw6eYzpkPhWvXK2qhR6xbjhIL343c4iwQ3u1r1il/I3B/hmPC4WQgc3t1TCp3durundCWO+ZfMSSJvR24Ru+JyP/gK8CJOgmB5IRqkjcj91mLgfrtcW4iTbDasPNOE6fenisl/muytKYsz95iBmkfnVqte6UAf+vdfYlGhxS7XbsQJm+1mLRzTy9WW1WKxIWzb5V2/u3COunQREUYLHw8/rsfZSXCzQ446P92xXmvVnvlEhXCLIUhfP+3ujPjvGzgZgN2kPdHT7c7c4TqNk5/hVmnH8kd9x28Az4n/l/WzSH8y77VDCPmog35t5BGy38XsRj09vQ1WmgnJp7956RuTQDeOfFdsTHnS3VN6FviLVDwO8z5Mx6M6xJ7VLju4Xa4tBn6iuaQNC9qo9m6P3i+w1N1T8rIiCMIvcE5L3FjAWVnn6WnDdPSnp7YwrZMzRe/gZ2UgMr7Lpkd32+XaUvH/cr81zapX2uV6MY3cb81nVSTPPGzUme73vfrpfVAX3eAkXmyHbhMgbCjeQUWeG3EY9kS1bb9VlOviGLed+IiOQpdlMvOoK2LHUU4I80B3T6llf5uHidTe0uclOAmfwNn5mStdT1s/nWPdrbDy3cvPcJuUfXbkHWVYtTsDqn6dOKYHqSCcIqdJxbe6bNfzUAfld/ycMPECdRIA2bcRgNsBRFt+VrqWB/386uAGq0s2aDfCx0I+IX8NuCagiD9oyoyNI2KHVZZne/yujktQfTc+KZt1iEn/Oajzhnp3T+m6gL+lRfQbX0EdF3ZDzQCdJ/bWlPm1kdH4Jz3cHdX31E8epDiZFxu78m7+Hd09pdZOfh76/b2lz2+xylfjf6jBN9rpJ8tbjmOG6xXat+392uXaQ6jPag+rXsnCnDwVVpfFhm6SdQf+cZ6DNIB/AO9IZZ/IgePaxqjJZe5w/b/u+Da1Bi+OluV3cm93T2kpeE7mt2tUe9MM4yfrt5CBeVZk/YbhmL2khVedbpH1DlIU/dKsgxNRd51W6tfdU3oHp524SXtQlJ/h4909pTmuz/IznJ5yThpZP5uBNs95mFikwQmo/gkXtDvVcNFAjUy1Z6PaGzbMrBcfQo2GdKNdrr0U5I/tcu094GKpeCyqWcfxqH548/BPmhkYYfP//zSXTk/bnyoE7frBsJsaue5X0Y+Bbv0eQ3UST7PfH4WapPeO1iaYmIPIDtrt9JPfyYM4pxdHeXx/a6te8Uv+10Jngpll7pREWV0WG5tLn22c1azfUXXbBiAc3f4mFa9BduHcWsj3C86KHVg5mQ91fGuYKahHif+TPsud5gj0p1BJIT/DJ1y7M5B9py7rt4KB9rp50w8G1sG5OJGf3ORGP4H8DDcNYF5phEa1dwTqyVA7/SBdJ3H5Gb7kOj2F7Otg4og+Uz416EcfaUiL6I91kZtMmU3ofAmDnmq00E18Vt63iISmC4tb7e4pzQ75W378GbVeTQJ+aPA3TCK3kQUM9IEJ20Z0/ZZ7cvwo2ZpXtuv3F5OteeUmqFYu7frVKV55oKx6ZSKqX93/0G9AtBhCsOiVl6BGYvt0wIXKoGN1WWzIO0Qv4yQa8otdHLQB6AaJEwL+bVLodsRkxzF5sdHu+NYkUfSDlDolcTIl29nmRj+B/AxfEM7rjiKOI6YcWi9L/UDNvSA/wy1SdBKPUgct0pvMb4o6KOamDgqTGXliIeuncxJfrRYbOHbZcs6Ha13meEH5O7BIKvt03JDfYjJ0lFT8OhDKrMku1x5GnYTtbtUrrTpQRU1q9iDtHeRDITZ8dBnrv9Ko9uqytmeN3M88LYVBfh7VRM2vjcjy3hBRjwCwy7WszSuj9Kvrp5iDR7cYitOveslrdwIRxJTqXZwFh5vx6IMaDXoG/WLDY1B8ivZJ+DYPYg7V3VP6H05SJDe7ZZgRGfQNQJ7oZbnrGKTBZ6mfLjKXrN9rOAnq3KQ10RuCehQsxwcH9RluLXbM00B+x28K0yQ3sn5e0TuSIO91UKef/I4f0XwnLf3WB+SAAwOeX3dP6T25jNVvsaEbR8KeGiDC48pJ2SaghtQMy6FoTLykBH5B0S0cjhKnGrJJsg18UToNNkJ3T+k+nHxMbjpJPiFiKIJsWomAHjorA2XuZdUrFu0X+KAxr7TqlbTMK2X93kPNWZRlvxpkMRRGP10/vRT1VDqoPBldm8tjAunYDPrFBk5WVjks2Uuonbhs0mERfOKjy4R6eMC/TQK5AbwiBn43Wdr0B5lINXHsfd3kRj8Pv5JtU0o+2IUaNjjIoDOEFPxKhE+O3KkHWQxBdu9YNyjqdriyrIPyRGU+q6J7tchNGxHI73jDjJIPGkfU88Ok4rmoYSuDohtHjogoq4Wsn9fvBOEfqPb2R+I4bssnkvXunpKcz8kk38EJ+ermhJxlqddZCujaiNzPrIG6mQSwNk4I1HbysjSvVBZDckJL8tXv96P2of9FDUftpZ+uH9SdsL0TUJ7Mg6i+gweL3B6rFavDYkO3kh2GGtHhFs33gq4+r9aUxR0kIiEGwCC7H2GPb00iv5P5SKFkPSbz26fkJB5k9wP0TuJpRIuIqh+k8451g2Ke9AP1GT4jJ/EUDqmyE21W+ukGRVCfYVrJB+PUwdUl38a2qDvX/4qxm38n8K5UdmTADMYKYkJyoFT8rF2uyWY2gbDLtbeRE9HZ7DK7c5mclbwffYJDY4hcQnLG7JHAl5L83ZCY7qdz3e8L52vZpFCn36NkZ14pz41ebAWmWalIT2kJwZMPyu9kMbCXVLYC1fplG6te8TPjB8Au12zU+eWa6KNlDmpWh8WGbuUpNwjQO/QFOtno7ik1UTOR75dRVKp1UZ2vld0Pj+Pb7aIObCEJsvsBaqc5EjWqShLonK9f1HxP16mnYQYUdVcZstNPGXSE4+irUnHi+ol8I11SsW5QBH3ywTRM0XTO10s039O94yDOh3EJ9I7Jrg6mge7UIGi4W4XuntIy4F9S8TSib2B8AHUsiKyf4KoBnyysu0fPHyd95x/dPaXnY/5OEH6NOmn9spj05gHTbSSoPJ15ZRptLoifWSsHjzxepdHvd6CaSAft96d6nMjK7+QF1Dwod6AuNoZq/taLqzRlq53fxuq62NhZ+twE7kVN1NbO7s6NXCGGAweE+HtTBO2QQF29j8VJKJUYIprPVKlYN1EGVT8I906iIj/D59zO1y7yoh+oPjl095TeQD2+zUq/oO84Df10EUmC6mcBGxjXyEXIQTEvdXCOFJa3RVb6pYFsKruC6CZULUxOLHSmvNdGlNVCOcW/Y/Q7ctEvYv5GILp7Si8C/5SKJwDlNH4/ALpNK90i7DlUsx1dGwnUrwrHYjnre977/Q1TiIa5IWpKgMhzDxEVSn6u76GOLdcGkefD3cDbUtkRwodntWF1WGzIx1zvok6orxE7602pvCvE7+gGiSxMqYIetYJ6vxDunqMQZjHU1JR1GdNEgzDTkm1tvfSTO3RI/vmB+o5nCvt9HU3pc5dxbVTi1MFxjWrvOKPaqOS9jWxAG+drF01NWZdJZTxQoux4fO8dVNOgLtPKpE2j2rsOmh1Mn3YYlBtQJ56hFxtiIiIvNt7BmbhExi7XXkTKQ/DAqHdZbK20HLu+u6f0aJzfCMmvNGWnpJyTyQvfiIEthNmdPJZ0BZC3EHjD47ebAeSZJk6/OgTHKiNJTM89dCc5utDo1wSUp0UEc5Cjx61Pm8TTg43VYbGhi3Mt09rtaUrl64fotB7GCSno5rAMOr0wuwtNTVmXMU305HqxgbPbIDt5a/UTZi1yeNmuBHRaiYdPjtf7Bc2gk4KpnM75Wm4bLZqaMp2Zo0nyXgfDtGE5sAUkXwfH4+Q2cOPVRuJu4uSVQ9DvYMZCRGy7XSreuVHtDTsR2wJ15/QGu1zTndCGZcDG2tIOm/tHrVxP9hiQHxjhhH6bVDwNJ7R9ZnhEDPTqYyBYG1HMj4VNfxB56wbxEYiJzvnay5yuqSnrMqmMBtP9fhDn8Kftcu2FgPL8WO1NqQb1YkPskE6WiuWoGQtZleGyKV0LvNoWg6p8xDwB1WQraeQG8A4wy+O7TU1Zl0FddIRp8G+g7vJ1GdVGJYx+kP5EKqjzdYum9HkE6kTRNPIzVJyvXTQ1ZV1GtVEJEpGkRVNT1mVSGQ2B66Cwf5bbd5dphSSChIZ205Q+p7HgTRqdiVJcf4gWuoAjB4WUYdSfRELR7+7R8wH+091T+o/69cTRJRP8SupaDKSLYBEDWzSlz2u7cw5Z9UpQ52sveZCw+ScBnK9dNDVlXUa1UUljsSGbabU2IGajRk+T5flxE7BMKjskxN/nnkG92EA/KI6TPt/oahBy5BkIVyF0O1tyZIKkCep8DfHvNwqyfsvRO1+HOV42SZhdZVA7pfUSPs0KuxhK9R17RCQJ8/wg/XfsNyi+hhrus8u4RgOJ+467zKmiJW4bGYWadXfQ0Kj26vzxnuruKckhKqNiYhyRF0N9OCZaJnhojb7OAe3lkZELISVfDQ03omaB3rNR7U0jmIgXYduIrp9296NBw+j6yevy+X4shL+FrGPe+n3ZzGuWiDioYx6qFUyX9DmIg/c1sDKqVLONPE9EssbbpeIdrXoliyBEiTDYFxs6G0IZd8fe1FzvCvF7d+MkM3KT2mKjUe1dk+DO1+DEhJfzb3SZ1EmD/E6e93C+btGUPncZ1UYlSOZrN03pcyewnjFtVHT65alTDxSRxEVTU9ZlShmZkM7XdPeUVqDm3+gyrJaM/I5nd/eU5vp8vyl93iBhZ8swttmQnV9JUuyNekIe24TKRRO1zgUeR6x6ZQKwu1R8twhdG5uHnttp7RmLxwzYtW8OW8J+0x990IT8sIjNtPM0lz6dti4uTLeRvLc5nfO1n36pm38SLCUAENj8U34nshXGPOAe1+d28tpxp/S5E9g1pIzcMtgXG+1WnjYDQw02Nd8JHMVBOAc+KhXvocsGmhChzBs8GlRiUStEwjt5t8mvQwJ1h2ZSwqEN5TrzRndPSXZwdaPbQUoy8kdcMy/Il36ziHe8HJYwztct5HecdGSXwIOioCl9HoK66WASWb+l6CcPLbI4QU2SJE2UWv2yPLHYqFHtnRJQxMGoY7cx/YBP77hojGIG907nij0M/kZYelHNTD6T4tgrY6Kf7jIsLzf9vjD/lKN/dplUyI0IWysnfAzb7680/7TqFd2mlWzRcL1w7vaSN9aqV8a10cGN3CdA+pYzibG6Lzbu7+4pue2dTUQXukv6vBbpJHqD8B0SpGtPrXO+9tuVhxQdiCM4X0P6O0iyfgvxdr6G9HeQQpkPZOBAHNa8AVT9JotcHcYJ43ztoqkp6zKhjwc6nxy/RHZNTVmXMW3SR15szMMJnW4SeRyB4BOLQzVlRk5eRB954vZLdEF3spv4iJM/2ZdkA5xTqCyQ24hfxEAIv9jow9v5GtI3/zTRr3YZ0USPbiM2rH6jcfxwwYkG1W4MkNucLA/C3fODqAvqYrGREzZqc31AZejuKS0G3pS+0xXyN3WDxAdCyoiK7n79TIBAbQBjUB2QTSFHagBol822qSnriq2JnnGo954n/UB9x8/6+OS0TtvmScVdppVyIb9jG/9BEdIddHRtJMo7TsrZUtdGwrZhSOgZip1ieYc0b20kMRrV3vVQdb9emNuZJNI4IkLeyt97gfZ1KCh7AJtusnQko/sUS720xjkvLtSUHZ+yDi3kdtyujczE8V900+X6f7nfesku1+SJ50rEjnqa5p8mxvYkzT9N9ftd4r+6+3Wj85Hyk9cWu1xbAjwgFe9m1StyIIJByWBfbLSznZdD5kH8iU+cHam46O5Xd1rjpqkp64qtiZ71NWV+5heQrn6659dOv9RybYhdRVnHdvpBupN5+R2/6eN83aIpfU4y10aUd9zUlHXF1kRP3vVbB9VcwFc/Ec5V3tXtMqdSquymKdONI3F5Csenzk2QcWQ91AiKt/mESA3LiQCdWGy3RHZbYQerXtEeeaTEjaimOR8Vvoyp0aj2DsNpJ27atZF2wVAGW7//HmoiOpmm9DnJXBum+1XdXMbNo3a5Jm/y+ckLijy/HAF0h5SRSwbtYqNR7V0D/2zYy4FHNOVN6XOYXBt095TeQl0x75VSqEe5Qc0VpzV+NDVlXUa0UdE1eHn3RaapKeuKrYme0PqJibScWKnLlEISa6Ee3bZ7fqA+ww0TrI/yM4yiHySXa0PWrx81V4pMU1PWZUIZDVHaSJqmclH0g9Un14bOIfM+0z8iQkXLYWS3aVR72506J6Zfo9o7Fjim9XmHxcq6ogP9YiwVxOnSRVLxSOCjKauyLmqQjMhtROxcy4uXKPKmiKzXSaD0+34n7oKmpqzLiDYqpuce7TaydW3OT15QVlu/jUG72KC9g+SjHhNx2YmnM4AsGXn1uS7tj91MIDeAVwP8TZrOm1EavO54OSlHN51+QZ5hU/rcFVsTPVH1k9/xCNTBKzYeJy9Rnh+k945nBjCBaWrKuoxooxL6HRsy/wyKsTYySHNtyJP5BbS3/Y6KPI5YOGZMfiS5GCrhhC0GYIfF2qibWZtS/UVTdnzKOpjqp9cWm6a63f6oY3tS5p+Drd9f0MaHBvz1i7LYmAMs8pAXlHtwNsjcZN3mjDCYFxtRKgMkc9QF6aw+Te0qd8XWRI+s3xyRhduTlHNtmNq1XU9E3jKNKf0gmWc4HjX8YZ70g2ht5HXSc7aU9etDXUjoaEqfu0woo8FUHRyFGh0m1wjzGNlk4YE2zvFxiLKLKS823sXcYuh494ctl47GsvPlsNrdU3oceEgq3qtR7U1js6+FyX56Q8PyugL8XShE3hk5qEVu9BNE6fffwWk/bro85Mko88u4uTaEjHeBx6TiPUR0rEHNYL6B99Vio1HtHQOMlYqDNKi3cSIauekyoZOGKA0esp1IySZSOprS5w4PWXHJ9aBDzvXzOHlpq584+ZB36roMqSWjO3kJMpltSp9DmX+GINfvOGG2A4ZLZcZNqFw8gpoHyXMcEeY2ymLILtfkndDQNKq9U5EWMsPsjntsS7n/Xax6RX5GaXOhpuwjKf6+6TaS9zanO3kJol+W5p9B+n2/xYHf+D4HJyiDDkWeCOoQBnl+OQ7YOqSM3FEsNhzCHnW9hBqONOkdH52pV9QGldZRZtTFxtqNaq/inWgAWb/Z7U5eBE1NWRLP0OSgkyf9ZpNOro2oJy8w+NpIUrk2TJ28wOBbbKTir9FCLHLlkLo7+eQZ2p7kFkNHacp6USc+w4GdDP1mVP6BampyVIq/n+fFRm76fWH+mXiujRgnL+Bt/uk3v7zPJyCDLG9N/H2LdayWfhur62JjFnp7RjAQXUhM4OVOeJNGtXdyGDkhidohQQr21CIiiBwVJKp+kIwDsamJHqR3chDk5CWtHaSog05auTZMtpF1TOfaiHryImhqyrri6OOBqZMXGHyLDZ3zc5InG6BOLIYCu3h8V7cYMpX/48OasivRT3wytSEXubPulop3b1R7jfupeSC3kaWokcV0NDVlXRp5kC/zT6Nzj1ia6Il68gIa8897Rs2fir8JqF+fIMuD8PcsB46A1cBvY3VdbNznFSlB7GTL0Wm6Ivx+2qZUJhv8Gji7wCaJdPIiaGrKuiJr4s1gm+jNChBWFpEBXQ5D2GVKKRd5H3RM6gfmF7wTUHem814H86ZfksiT+ee7e0pzEv7NMOOIbrFxf1wFRKLJvaXiB7t7Sq/iLGbkU4Q87LJeIX22gCNT+u0okZnAO9eGLG8J7cPKtnJtpGH+abJfTcL802i//+rQJe1CzSa62LDLNW3E0wjmWLlitV1stPnbpvQ5ytHjYF5sgPnj1lzrF/Pk5RWc5HVu0jiuDqofpGMGFPXkBVT9xgYI8xmWXNdBcq5fnJMXj1wbSZmiGUfsisv6Jn2qAc5iQZ6ABl1sPGeXa0F21NtxGE5URjdXANjl2gLUEPJ7WPVKUsnZgiIvNkB/OpMEUdtIP+op9DSdvBB5U5oaeaYx2W8lYf5ptF9d0Nm3vc/3bZxM34HlYWZ+OYV0Ip4mxmBebPglXWk3SMgNfl2ROTcMT6Bmbp4RUkYYdPcr+414oTOzyVODz7V+4nRBtls3qp+Y6MnvOMxiQ36GUxMIPRrp5EWQxjvWtZE81cG86zcRkLPVxqqD8dRJFZ3pUuKLDWHXLkdXmiGPR1a9kuRiSOdc7Z7MyxOfNYFNDf12JLp7Sk3gUal4X5ErJDHErvwUqThuGzHZ769j1SumIyVGOnkRDLp+dUFH38Y+339cRIwKLI9o96vbzE5yfpk4g3KxIWypJ3hc7kftvGXk3dghOANtYMQuhfw72yUUIQbUBv9Od09JjjLlhW732XQmz6ixx8GZyMs7OWnoF6ZTl5+hPODEZU0c8zY3cfQbhXqSE5cosdZbZFUHg568ZKVf0AXvEtTNjTy1YVCfYVLZgpMgVedwiQekz2uiLiwS0U8E4jhQKn6qu6f0tOuzrB/AjnF/2wDy6cZQ4JCEf3My6rwpchtZTv862LEWL7p+y7TvqNLvBzQbg5z3q2j0W9jRF2cjew4g53WKcr95bXORGZSLDfxXio9395QWtPl7XUbhKJPHhvR5BLB5BDlBiGNiY+p+/dA1+EAnL909peU4jdRNGvrFeYZ56jAh4Xds4OQlizr4VndPSc4T4EVWbSTOM8y7fhNE7orBgDyZXwz8N6XflscRUCcWSS2GDkSN4CZP4oPolwVZmFIZbSNzh6ywsCJlI9fKEyTdL+Rdv8WoGzNezEbySVrQscJvsebb5kQYatkKIsr9PosaFjsPbS4yg3Wx4eevESQ6h6kG8LCmLKkKMdga/LzunpLcWPwYbBOpNX1CVEYh14sNnN1WORxxnvSDGG1EbFDI9TVp/Wz0z8WLwdZGwPwuq3HEabRsovCQ2ARJgyDjSFKLId3kXJ7EP4+aqykPE5/HUfMdHNyo9sqLJ5MYbSOzhmj3QnLTr4rktXFOXjLp94OevIhIewPC877b2ecXPCfIAj92P22Xa32oZoI7DmYn8dVxsRGlMkCOFxtiUis3gDATqSU42TLd5GaiJ8hiIhXU5wWS7zTzvtjItX4xw8q2SLsOhjl5AVW/SY1qr0lH3Vy/4wTZCnUhnZYJFcAzqHloVo4jVr0yBNhZuv6giEYUGXHqdLhU/CrSSYbYrZWdxHfMOquxmFDKC6M1gP0S/Nn31WIDZ7MgzsmLnGcDcj73mN+5YozH9+YDT3tc85RH9PuV55fjGHwR/lbyfl1s6BpAqVHtDZus6EXUCCyHN6q9x4tEM6aIE1a2hWymtGOj2vsxg07EcRu87HA2sVHt/YLB5H5xT150ITC/2qj2msokHncQ09Xp4xrV3h0i6iMTV795gDyy7tOo9sr24lEZS7yTF1Bj5W/WqPZ+yqAfluk20gGcajCqV9yTl9mass83qr1+Dpd5IEt/jVZyv8ek4t0b1d7Pi3e7NckshvbGaTdurvDYFZYnPmsCp1n1StbvVmdK9flGtXePBAJkQPx+cJb7w1vxFxu6hJvHWPWKKWfiWPcrNlPkfmvPRrX3CBPvR5y8yKensfrVeZ0rvE7G7hcL71DygDFWvfJVq16RfTLbodvM/oFVr3jl4ck1q9ti4x3U+MQDaFR7twJu1Fw6ALi3Ue0NsyuyPWrYwulAHbg5wYk8BHRKa1R7hzeqvdcC8qAwBbgU+F1M3WKfvDSqvT8HSppL5wC3GHqOkZ2bG9XeEvBLzaWvAg8bSiYVuVNvVHu3B67TXDoEuL9R7f1gDL1aRI74IeL4348a6Whr4IZGtfdrMXWDeG2ks1Ht/RtqRKKxwF+Bi2Lq5nXyEqYOngp8SXPp5zj9lgnfCPkdvxnUlKhR7T0I+KPm0vFAI+cLjkwXG41q74aoWYbXAM4FHh3T16lrvyb0C2JChZgk6d7fD4D/WfXKtgZ0icp9qIvcQ3ESo12dwIJDbiPLNb+vpVHt3Q34p7ts1hBt8wrUL1j1ymbArZpL+wD3WvWKCWf5OP3+mEa1907UucHGwFXA92LqBs48Rn7HQft9q1HtPQ/XSdhy+pnb6dnltW1zVr3yY+DTmku/A+4MehoozKW20Fw6HufdfiyInDyxui027hdRovz4Fd4O5kOAchAFGtXecTiLFq8oVnvhHM+bIM7uwidxOl8vPmegQ44zUd4a+A5qh9FiF8DEYBZpV1mYqZwPeGWTXhs4KrpaK5H1myvCYgbhDLxt44cCn4ms1Sri1MFTUM1A3JwUXh2FOPodChznc/1YAyE1x+FECHMTtA5OBk5HzYXQYjPMZJiNc/JyLuBlfrAmcGwkjdJBXmy82t1TChrFLBai770K78AiG0xcMfQoTXmsZH4itK4sdy767MVn4D2GjMBM/xKVU/DO9nwY+oVkHOQ28nqAOUeLM5Em3hozqmXoT9F1/ALYwONaB3BCQDl+xOlXv4R/7rGs+/0PyjrMGbIc23s25LvYsOqVTYDv4z2v3gEIaj1zMvAtr58S1wcVq9tiI8huz7g212e1ud5iS/xT2oPqVBeVOA1qXJvrs0OEsfMiSf0g4M6RF8IUSzY1CarfCNRoLTJB64wfcSZ649pcT0I/CO7z0s7MJyn9gj7DdvotJn5bjqNfkIVOrGcYx+dF/O24Nl8z8Y6NI8yU5Il+mv4aY4Ht/L7w1tBl8qnCK3a5FncxtCvqBsVVwqRLZu82skyNc6EQIfB/1uZrpp3FjfbTGjOq1wOa6mjlSWTdr45rcz1r/ZR+38OsrUW7Bf64AL8Z9J73bnM9jAl4Lhisiw2vOMhBIlH9rc313oA6/A/wC7H7vEg8ZII4SWv+D/BLvBb0fv2IE5//PhzfFy/uNLDLGCe/wXvo7YJbzEdvlheWOGFlTdVpP+RnOCfEycvFba6b0C9OG7kW1ffKzWUiakkc4uj3LP65g57C6Y/isBbq6V2gNiw2K/ze8XLgymhqJY7O/jnIOGKK+TgJYvUXO1YsX9TRL7c9E/oFMqES6E473NwcU5eodKDmZ5KJ2y5WIk655ZDnYXJsKP30W6oZVd77/WUE3/y7BP/3k3W/fzPS5N/DrA3gGbtca5fIsIET7MGL++1yrRlQt7y2ucgMusWGcLye5HFZlwhF5nzUyBotHg8ooxUq8+s+X/lXEDkBkRv8gu6ekl8Wy5WIBU+Pz1f+FFUpF3Em8yuAL/p85YJIGg0krlNfBe/du4u7e0qLwqu0ika1dwzq7nUY/c7Gqbta8d09pUej6CURJ6zs3Ti+DzoW4wxKcYmT52UuUPX5SqZ1UEzmv4j3wH1BxqeT4Njve01CruzuKcU6nUyQrJ3DbeDzgHYxe8/o+bp2HUs/cRIlLzYWAv/2+JPvoQ9AAc5iKc2ToJWIjaDf+nzlle6eUlCTpCBMwjG1dhOmjZyOa2OtD5s56k56GHl1vHfbn0OfhTosccLKPgb83uNyH95jQhji9KsLcczwVuIRHQwC1HFxIvUFn6+EGUfOBx70uW5yfpkKg26xgXcytae7e0ptE7mIya2XvdtFIQftOt4NRuewG5W4UWx+iX4H5tnunpLfSjwocSNW3IjkOCdYhnMyE5e4+r2KM5nSYWIiGivamHDi9arTF0ZRSEPcOlhBDe8JcHvQhXMbdCcvS0L8/Xnod0FnAfdE1moVcevgg+gHbhsDDuzE1+9t4Jsel020kaSQFxvL8d6MSoTuntJ/8HCWvXWNee9oiuNO7rcCNpLKrvdqL3a5NhNvn6bH4obgjUkVx+dFR9g+qh1x28hiXBtrb3cup0/1DwgsT0xuvfr9S+xyLe4GBMTv97+HGnYfHP9arwVsGHQnL2EWmBfjOkXwMaMK1Obscu029P3xCpyAPIGwy7VlOH5uOuuZOXa5JueXyT2r02IjsMOcGLhv0lwKtcvv2nGUQ2bawJ1hZLVBvudQZkXdPaWlwJc1ly6LrNFAZP0WBsjiLvNVpEyewINxTw0EujoT1jTrLNT3PK+7p2RiYhJbv+6e0j2odc7IRFREG5NPXsLqNxsncpLMH6LqJRG3jfQBJ2ouXWfg1ADM1MHvoYYPfsbQqYEJ/S4GmlLZErx3zPPA9tLnR0MuUk1xOpokfY2RSjdqYjGkCzftZyqKXa7dgn6xe0dMXWIh2u3HUUMHA7xi+OdM9NM3It7fbL3JTih5drn2GHCN5lLsflUEEZBzRIS93/nAdzWX6lH1klD6/TD9tfju51qf53ibUYUJyHAK6knlo3a5FmpOZJdrL6F38jeRzDN1BuNiwyvMaNgOuMTAeMjnCXOKUIij3CMYOFG+xNSAJexEZUd0XWxtX7p7Slfj2Ka3eBn4SQzV3MjvJIp+M4HTXEWLMRflRFdnQukoTsRKODsU4EzkPx9Trxax9RMcw8BdpLNEZx8XU/r9jIGnB/eJemkCE3XwAQaeBM0BTITlBVW/+WH7CPEuv8Iqc6rlwKcM6AZm2ogNHM1AH7FKiGg9qSJCVsv3neqpRgvxjA5k4Onfo+929k2TvvqkXa75+eAFYX/ps00wv7OvAS+5Pr+LPiR4qogTg31QTV1rhn/KVD/4EWDh2/oQq1HkHcdAE8YLxWlUXNZCNRuLot+5DPQz+h9mzLfBTL//NE6kMK+wtyvw8auSscu12cD/cxUtQR8ON4isfzLQSqYPZ2N20DEYFxte/hqhVnvC5GptYF9gendPKfLEUewqb4gzmTqqu6fkF0YzLONR31Ok48funtLhOLkN9gCmhcxe7Ifc4KPq91OcY9H9gXHdPSVTR4WyfitwksyForun9G+ciBOHAuO7e0qmToZ0dTr0M+zuKc0CJuDEDe/q7iklNVGGaPr140Te2RnYsbuntFtcxQBE0j05BHXUOlgGNsEJi7hOhBM6L+R3HFW/84Wsg4Cx3T0lP8fxMJh6xw/jhLo9BJjU3VPystnOA9toyjLbNezuKb2J825PA07ce/oje6NmDI6lX6PaOwI1TPJDwgzOF7tc68Mxv/oK8G1gXbtcy0VUHKH/Fji27A8Ch3f3lBr+fxUaU22kCYy7YuycM0zIEzvmk4EPARvb5Vqg8P0BMHW/dndPaXdgR2AGsJ2h02IwN/f4GjDtqRGLdKdhTwmzpsDY5dqvcE5dDgDG2uXaU1H0ErIOAz6Kszm8oV2uBV745AlTmXHTxOtkI3TUCTH50SXFCU13T+k1zCSpkTHS4Ft095SSqKhGGjxAd0/pdYKHVA2KrN+sqLut4iTLtHOWsXcs7uuWeOoomNTPxj+yUhQmouZpiVMHnweej6WRisk2MgczEdDcyPotwz9ClydiE+P62Boljy5/T6YmCsJp9ccAC+sP7aH5Slz9dkONOhbYzE34AZwdU4dEEGOwX06puJjsB/vuqFd0vgWR5An/jduj/K0PpuceRk8NhXmvnJU7jn7NxfWHdLmCIkU0E6dLJk6YsMu1/8OM/2pmDMaTDV0DmGk46kSeMNrgTdOo9q6BmqwsN/oJjE30EkL3jvOUlyDXdZCc6yei/wy2OviWwd3HvKJbbBgLlWqAJBZDsgkVDMIwmhkht5El+Ie/DysP8tUvvK/0s+qVYehzLg1KH4m8kejJxrHXzZqOk6FxWxwH0/k4L+78Sw+d5JdbQcv81/edPmnfTfbtHLmYjmHL6V82lGVzx7PgmU2ejaLf/Nf39dRv7NRbIulnWt7UD6/ztf7lQwfc79I5EyJl/E5Cv67PjPrW0rnjB+j3XnPDyOZZSeg4YY+tpw1b652VOq54b9TY+a//cXpe3vHa+2z6wSGjFq16hm+PW7LWDo+vB6Vc6Dfl0A0+idU/4B3b/Z3DHBeW7PVb72MTqn2LRg7Qb/EbUyKdXJnWD2DdI67bZunsiSOGTXh7pY79y4dOz1MdnLjXllsPHfvuSv2WLxgzNC/6JTGOACdNPWryh+mwV9aZd5/c/I1tv33yO3nR745dR358cZ/FghWdPLFwJPXXJvLy4uGhJz7u97Hex8bv1bdoFMvmjmf+E1uw4t01FxEh2trMCy3Pe55yvJ2bd6yTF7mf3nuzXYas8d6qNvLO2KXjtntiWtR++ryt1zpg/NAVjBnSx4IVnTy5cGT/Z9efMw41EElbTL+P+a/vO33dw6eeaNvWgH51xXujI80Zk3gfG3x87A+XLxgzQL9Fr06NFBlt/uv7Tn/ug0NOefTdkSvfR5w2B+bfSd7ltcOybfObV8deN2t/nLTtO+OcngxzXV6G40z9IPCTSw+d1HZXZf7r+66UZ/dZw63OVQnl7b4OsOw+q8O+B/jJ2Km3hJLXTr+s5dn9DLU66Gxds/s6oKN/mWVxfy70sxliWasWreJ9rLA67HuDyktSR9tmZ/o7Rlidq+aedr/Vb3XYy6LIM60f2jpt2VanvTQ3+vVbw6wOe+UpqKiDSy2LB/KhH0OsDrkO9i+zOrJrI26Zts0M+juGD6yD9FkdLA8jM9k6KLWRPqvf6sy2jSQ6jth0WNYqeXZfB0C/1dl/dx70k+Ut7QfbthjeYd9lWfHfh93XgW1bLH9n7LzhE98+NmidnnmhFfiepxxv5+oZyvKy7qeX9zN8aMcq889l/TCsgyVh5Jl+HwHmHvnp9/Vzj+VWh31fFHn9Np0dFkNb11ptDrh/RKf9/azaSN7lBcXoYuPY62ZZOM5tp6Ka1uhYhBPy78eXHjpJUWT+6/tGljd26i2FvJzLGww6FvIKeWHkDQYd8y4v7+NI3vVLok7PvNCKLHPK8eok4/32DE3LM/0+8n6/eZcH5t9J3uWFxbTPxmk4ybuC3Ajie6finTCtkLd6yxsMOhbyCnlh5A0GHQt5hbww8gaDjoW8Ql6W8gaDjkncc2CMnWyI49ArCX4jbhYBR7mPScXxVix57mOvQl6+5A0GHQt5hbww8gaDjnmXl/dxJO/6JVGnhdlFLJluc4z32zM0Lc/0+8j7/eZdHph/J3mXFwWTJxvfJ9qNIP5ODhtbyFu95SUhs5BXyMtSXhIyC3mFvCzlJSGzkFfIW53kJSEz7/JCY+RkQ0SSeAIYIV/rW7aExjnfZsiIUQwbsxZbH/ctLzFLgC0vPXTSSyJygVbey68u5vTfNnny6YXcdt3OfmotAbYcO/UWX3m33jGXy6+exaLFfWy1xRqc8pWuWPKaLy/mt+e8TH+/TX8/nH3GFrHktfj8159k5IgOfvOLzWPLe/nVxRz7mf+y24yxTJk8nG99XU5SG06ebdv88jdN3p63nHFjh1CtTPeVByujU2hl3nP/O/zjCicR6L9umsODd+zKmmO0QTAC6fjWrKV8/TvPsPaEoazos/n9GVtgWdqAXoHkPf/iIn7Y8wKT1h5G1wYj+OoXNtTJaitPrssvvLiIn5/xEkOHWRx64NocdtDaseT9+7a5XHzZTMaNHeJVb0LJO/eCV3n62fd4a/YyPvupqey/z4RY8m6+dS7X3zyHBQtW0L3Dmnz+s+vHkgfwzjvLOexjj/CNL23I0UcqkRFDybv4spn844o3mbbhSPbafS0+coS3PNDXaVnm3LeX8avfNelbYbPTjmM59ujJsXT86yVv8Mhj77JocT8vv7KYG67ojiXvrnvm8Yc/vcqaaw5hy83W4Cuf3yCWvGuun80V17zFmmOGcND+EzloPzn34ip5unHkzYdv59W7rmTFkkWMnbYl6+9xBE/8/XQ6hgxj6q4HMXW3Q7TyvMYRub/fbusxgdtIEP2GDB/Fu688w5J5s9jokOOZstO+sfTbdqsxsdqIbnwL2kZgZcSalTJvfxyufgAWLYUt14d1xsHl90LXJNh9czhyF90jdGROOd5WnqH8/DY68FM8eelvsPv6mLB5Nxt+6KOe8oI+w7UnDovVRmR5M7rHxmojsrxNNx4duI3I7wPg5Vnw+39Bv+38+/IhULsShg2BA3eEg3f0fh86/eQ5zFGHTYo1jsjyttpijVjjiCzvyEMnxWojujlbnDYiv49dNg3XRoK84+8eA7+7Blb0QfdGcPTu8eTttDE89hIsXgYvz4arqt7ytNoHwNTJxklesl675zrW2W4vdvz8z1k8dyaL5njmOOkATm4nb8P1R/L7M7Zg/Lih7XQKJG+fD07g7DO24M/nbM1d98yLLa9rw5H89pebc+avtmDhe30sXtwXSx7A+fXX2PeD4/10CyUPYPSoTpYu62fahnJ+p/Dybvj3XJ574T06OmDS2sO8vuaW5ytz913G8ZtfbM4XTlyfA/ed4LXQCKzj08+9x567jePMX23B0qX9zJ/vGR0vkLwbb5nDpz4+hV//fDOefOY9Zr65NJI8uS6f+YdX+PH3Nubc32zJhX97Pba8/T40gR9WN/LSLbS8L5y4Pr87fXPO/fUW/P0f2nYcSt7++0zg1z/fjD/8bkvu1Le9UPIAflZ7idJHlQl8JHmWBaNGdbJ4ST/Tu7TtpG2dlmX+5vcvM2xoByv6bKauOzy2jp/++Lr85hebs+vOYyl/cmpseXfePY9vfWMa5/x6C+74z9ux5V3yz5mc+avN+c0vNuO8P78aWt7kHfdm56/9lt2+fT6z/3s3z1z+e7Ytn8aMb5zJCzdcFFqe3N/HbSOyfpseeRI7feUMdv7GmTRv6Y2tX9w2ohvfQrQRRebeW8OvT4A/fAHufsrJpDlqOCxZBtO80u366Cg/v6f+eRYdQ4bR37eCkRPXDS1Pd89x24gsL24bkeXFbSMbToLTj4daGd5bAudcD98/Fn57Ivztdp04f3nyHGaPXcfFaiOyvOOPWzfWOCLL23O3cbHaiG7OFqeNyO8jbhvRyTz7Ohg6BPr6YV391DCUvA/vCr/8DOy8MXx670D3HBpTeTa2ZWD4rJUsmvUaE7d0lnKj1l6fxXNeZ9TEKbqvDgO2aScvBKHk/fWSNzh4f2VHIZK8ex94h/pFrzNxwlBGjPCcoweS9+h/32Xx4j4O3G8C99z/jhH9NlhvBLdcuxO2bfOpzz3OXruvxeR1lIlPYHnPPPseW27u7Jp9+ZSnaL68mC51EeOW11YmwB/rr3Hi8ev5fSWQjtttPYaz//Aq9z0wn0lrD2Oc90I1kLzS0ZP5xRkvcefd85g7dzmvv7GEKZO1E8dQdXDmm0tZd4ojx+PkJdM2AtDfb/PTX73IF09Udo8iyWudHpTUHf7Q8i74y2sceejavPzqEq+vhJJXOnoynzhmCgsWruCTJ/6Pq3p38JMXSObTz7zHqV+fxk47rMlHjnuUPXdTckhFesdXXjuLf/x1O92lUPKOOmwSp1SfYfjwDo77WPx++ltf6+K7P3qecWOH8PY87SI/kLwXb7qYdXc5kDcbt7nGj+htJEB/H0k/ALu/n8cv+gWbHqkdm0PrZ6KNtOSFbCOeMv9+JxywA3x0d/jYnrBwMZxwNlx2aluZWnmt5zfzoVvYsvRNxm+6I3f+4FgmbaNu2QaRB+o7jttGWvL23G0tI22kJW+XncbGbiP3PwsX3Qbjx8Brc2CK6FY8EnK1lRdwDhNZXtxxRJYXt4245f3t0pmx24j7fRxtqI24ZT77Bnz9cNhxI/j4GbCbeuAUSt4IMSW65iH42zcC3XNoTJ1sjPW6MGrtqSya4+zSLpr9mt9uhVuOp7yIevnKO+u8V1i+vJ+TT9BW/NDydpsxjvPP2or+fnjmuUWx5N14y1xeeGkRP/3li9z7wHweaMyPrV9rImtZFhPGD+W99zxPXwLJm7rucMav5dTWtcYNZWF7eW1lvrtgBS+/uoRttxrj97VAOv7t0pl87vipXPTHbRg5soP/Pemb9LWtvAnjh/Grn23Gz36wCUOHWUzT73oHltdi8jrDVp6S+Jg3ZtJGAJYu7ecb33mGY4+ezM7dnl8Ppd9xH5vCFX/fnr9e4nniGVjevQ+8w/9d9RaX/GMmF/W+wdvzlseS19HhtJM1RnfS39/2fQSSOXXdEYxfawidnRYd3r1vqGd4593zmNE9lmHDPAUGlvfTX73IP/+2HVf17uC16xhK3vbbrsnvTt+c73xzGmuN89zb8pX3zOXn0r9iOZscfiIjxq/D4rlviivR2kiI/j60fn3Ll9L4/als+KGPMmHznYzoF7eNuOWFbCNamX+4AZavgM/ux8o6PHoE2P3yN4Pp6H5+oyauy7Axa9HR2Ynl00D85IH6DOO2Ebc8E23ELc9EG9llUzj7ZLBtWLYC3hQb/D5G8r7yQsxhQst7/KmFsccRWb+4bcQt7/a73o7dRtzv4wWhUpw2Isvs64e11oDODujwTvEcWN5zbzgnlTtt7JjftZEXCVMnG54z4Km7H8rD536Ht599hBFrTWKU/2JjvvRfhXcXrOC0nz3PM8+9xze+8zQ9p23CyJGdXl9vK+/iy2Zy4d9e5wN7rMU3vvO0n09EIHkPNObzj8vfdGzh+m02nu47EW0r79vfcPwpXn51Mb/9/cvM8G6cgeQB3HXPPC79vzfp6IDxaw1lo+mefkOB5B1xyCS+WX2G515YxPJl/Wy95Rrt5LWV+ffLZvLxYzyPMUPpuP+HJvDT01/kxlvmMvft5Uzv8vWTaivvjZlL+fkZL7Kiz+bg/ScyYbzv5q6nPLkuf668Hqf97HmGDevg05/wbCeB5R3z4cnUL3qd/z25kNqZTSpf7Yol7535K5g1exmX/GMmDzTm8+WTFFvlUPJ27h7LI4++y5Kl/Rx8gOcuc2B5Z9e2YOTITi6+bCYjhnesXABHlbfl5mvwxFMLWbBwBccfpzW/kOW0lXlSeT1+9bsmo0d1cvjBk2Lfc89pm1D/2+v0nLaJl6xQ8g4/eG2+/q1nGDWqg2229FzoB5Z3wL4TuemWOby7YAXf/5anKYanvJf+3csLN/yVSdvuyUNnn8pmH/kij9V/TOfQYUw/8JOh5cn9/ac/vi7n/PHVyG1E1q9/+VLee+tVmrdcxtynG2z24c/H0i9uG5Hl/en3W68sD9BGFJmX/scxzdljC/j2X2CL9eHJV2DBEjhuby/1vHVU3++XeLL3DDqHj2LqboeGlte6N3lMj9NGZHkf+/A6sdqILO+IQyZx9b9mRW4jDz3v+ATYNvT3w88+CT+9DIYNheM+EP5+5TnM4sV9nHFmM3IbkeWdde4rvD5zaeRxRJb38GPv8qe/vBa5jcjy/nLe1gwZ0hG5jcjv484n4YKbo7cRncyeT8JvrnbMsw5RXJDCy5s+Gc64Cn708UD6RcLUYuO/wL5ojmqGDB/JjK//LoiMZcD/2slbc4ynk1Ikecd9bIrXUWgkeTO6x7ZbEISS12LD9Ue2u+/A8vbafS322l0x34gsb/jwDn7v7Qivk9dWpsbJK7KOm20ymov+GOgEMJC8dacM56xa2/ttK09Xl88/ayuj8nbfZZxReW0ILe8Tx/i2vUj6+bTnJO83sMzzzzT7juvnbm1Unul30sZcyVfetP1KTNuvNOAPdq2cE1merr//49nR34dOvzaE1i/O+/Aa3wK2EUXmsXs6/0LiqaPu+e1yyu8jywP9PcdpIzp5Hz7c2/g+irwPfcDXJ9NX3k4bO//cnO1vXe8rTzeHidNGQsyJIsuL00a89IvaRnTvIwCh3/FZJ5mVd94XA8uLhCkzqvNxUpzHoR84r5D3vpCXhMxCXiEvS3lJyCzkFfKylJeEzEJeIW91kpeEzLzLi4SRxcalh056EXgwppgHLj100ksAY6feYkReKzRZIS9f8pKQWcgr5GUpLwmZ7zd5eR9H8q5fEnV6yvG2EZmtkJnvt2doWp7p95H3+827PDD/TvIuLyqmTjYAfoKTaTAKi4CfFvLeV/KSkFnIK+RlKS8JmYW8Ql6W8pKQWcgr5K1O8pKQmXd5oTG22Lj00Ek3A78C3gv5p4uA08Xfr0Skg48sT04nX8jLl7zBoGMhr5AXRt5g0DHv8vI+juRdvyTq9JTj7Vgyxd+v5P32DE3LM/0+8n6/eZcH5t9J3uVFweTJBsCPcG4o6ApqEXA68ONC3vtS3mDQsZBXyAsjbzDoWMgr5IWRNxh0LOQV8hKTZ9tFG4mL5RPPPzLHXjdrf+B7wIx+u29Yh9W5clHTZ6+g0xqyBHgA+Km8S6Fj/uv7rpSHs0Aa5rq8DMd55QHgp7pVZxh5S/vAxmJ4h32nZYWX12czpNNaFeVraR8M7WBZh8V9JvSLe7/9Np0dFitjuS3tg06LFUM6uCcr/dwybZsZy2xGDHctg5f30z+0g2VhZCb5DJf3M3xox6p8Scv6YVgHK+t01vqt6GfokA5WxoNe2gfDOlhqWdyfB/10bWRIB8s7Le7Ng362TYdlDewTOixWDA3RRpLWcVk/I4YNbCP20A6WRpVnWj9Z3tI+wLLs4R32XUHlSePI0A6rc2Wd7rNX0EHnUsuy7iejccStnyyvz16ZnO2+TmvID8Lqp+2nO1g+JMM2AjDzQiuwzCC7oe5naNt2p2VZK++5z16BhbWiw+q8hwjveFk/w4cN7KftYatZGzH9PhKeewzpyHG/L8trzQUb80fx4uLh3/jyjH/8tp08MP9O8i4vKIksNloce92saS8veeKaodaIrYZaw1luL+XdFXPsjUbusPHlh6/3Ylh581/fdxpOyvRtcBKMzMcJx3We22EnrLz/vjvy2Pf6OroWrOjkyYUjqL82kZcXD9/ALtdeDSvvoXdGnfH28iEfHjOkj5a8h+aP/vw1h14Q2pM/ifuds6zzlIfnj/6SW79rZo0776GPnKsNCJ+mfi2Z57488dGNRi1bs6XjzKVDX/nMenP3jvOOTek4//V9p13yxlrXTxjat1lLv6cXjlhx8oazN520/r9zod+/54z5U79tfcj9jmcvG3rI2fv89fo86PfUwhE/eXXxsOPc+t0+d0z1tiPP/3ke9Fvezxdvm7tmxa3f5W+uddljHz3n2LDyktLxT69OuHf9EcvXaen48uJhcz63wZwZeWkjwMn3zRt9og0TWs/wwtcmvtdcPHxNu1wLFRnl2OtmTXtj6Qt/AfZqjSPzV8xmhb18vyePPfqWqPph6H6PvW7WNODkBSvePmxJ/6KtWvq9uPgx3uuff5Bdrt0YVr9XFw+tPrVw5InuOnjD7LFn3PvhP1TC6pdEPz3zQstTZhRH0mOvmzWt3+7/wlvLmqcMsYZ1tJ7hy0uevG3+p7+zT1h581/fd9pFr42/ZZ3hK6a1nuFz7w1f9KWu2VvnqY005o/61LJ+a13X3MN+efHwMXa5FsrcxfT7SOJ+X1o07PvPvTei7K7TN89Z8+f/Oeq8ah70a8mbu6xzw8b80Vu55oIAP7LLtR+GkWn6neRdXjsSXWwAWPXKGcA3peJJdrk2O9EfDoFVr5wI/FEqPswu166LIGt34G6p+At2ufaHqPqZxKpXhuKsXt38xS7Xjs9AHS1WvfIg4E6/+7RdDpbYIg2seuVs4EtS8Vi7XHs3C31krHrlU8BfpeID7XLtpiz0kbHqlU2AZ6XiH9jl2k+y0EeHVa+8DbiT0Vxnl2uHZaWPjFWvXA8c5Cqaa5drvgkt0saqV34NfEMqnm6Xa6EHMo8+em+7XLsjqn6mseqVPYD/SMWn2uVaLYKstYC3peKz7HLtq1H1GwxY9cpDQLer6DW7XAuc6l2S9VfgU64iGxhml2srPP4kdax65f8BPVLxLna59kAW+iSJVa+sA7wpFdfscu3ULPTxwqpXRgHzGLjjf49dru2RkUpGEePv5jhjWtyQuIExldTPj7c0ZesAoRYbVr3SAVwMHALMBA6xy7XQpyMe/FdTti0QerGB9/0GRtzrjTiJWCxgAfANu1z7UwR9BmCXa8utemUuMCGqfi49dwL+ibMq/rtdrskT8KjIzzCqfhNwnuMmQAM4wNBA4/WOQy02rHqlE7gU2B94HTjILtdeia9e/DrYwqpXfgh8FVgMnBB2l9YDk/odCfwBGAqcbpdrp8dRzMVbDFxsRNVvM+AaYBJO536cAd1AfYYTrHplqF2uLQ+p32jgemB74CngQ3a5FtSmtx1e/WqUXTMT/eoY4H6cgRacMehYu1y7PYI+Oh7XlG0bUdY7OJtC7glP6Dpo1St/AE4EOoElwJl2ufbtiDqlwX8ZuNhYz6pXxtvlmrzwCoJcZyxgIuqE1xOrXpkB/AtnvOwHngD2NbhZ6tVGMllsiM3Iu3E2+xYBn7DLtasNiZ+Ds+CzXGVR+9X9gQuBEcDv7XLtB7G1E9jl2iKrXrkb+JCreBerXlkzLxuKUbHqlTpwvPi4wKpXNrDLtXfS+G3TDuI6TE0szgFKwJrAZsA9cZSSeAKnEbjZLqIsE/d7PrAfqxrlGOACMXExQezJvFj93wlsCIwDvmjVK6Z23WZJn9ey6pVh2m/6czfOwLUmTsdxZUy9Wpiq038GjsbRbwvM1Wkj+ln1Sgk4DWfSvS5wrXjvcVkALJXKoui3HnA5MBlnMvBLq17ZO7Z2DibayBDgXpzF7ljgE1a98ksDuoH+Ha8dQc7NwF44fcwM4LY4SknoJlKZ9KtWvWIBt+C0M0v8mwTcZNUrwyPqNAC7XJsPNKXiSPdrl2s2aj8YdnF1Mo6ZRMvXZQTwLateMbXgTQKvyXcU4taZNYB/s2pjrgPH5ES2XIiDyTZigiuAnXHax2igV7Sd2NjlWh/OgsNNlH51PM4CcF1gPPB9q175SHwNByD7KnQCoc358oRVr5zEqoUGOH3+OWn9fhqLDd0ugm9ueRmrXlkf+KxUvI5Vr2wTWSsXwj7yeal454iyFqKGFwt8v1a98gHUe21hakdKfieh3oegBxgplX0xmjoKujozOYwAq145AmdR6ma/iIsWGRN1ehogD/pTrXpl08harcKEfsOBM6XiITi7pLEQEykTdbCO2ocl1UYmixPHMHyNgacjAJ+JrtIATLzjPYDdpOKdrHplbGStBvIk0CeVRepXiX+/n/X47aEMNLWJizx53EqcHkUhchux6pXpwG89LsumbXlCN/lOvc6ICfYfcCZkMpuIjQ4TvIZziuUm6v3GwqpXjgIOlYpHYnbxY6Lf/wOqVc4p0dTxROcYfaTh30gNq145BDhXc2mrtHRIY7HxmqYsbEP9HXqTL69JeRTkY8vpVr0yKaIs+Z4D3a/4vV6frxwsTG/iIuu3TphJuFWvbInqswCwmVWvbBBLM4dYdUYM7mdpLg0HTNjdx9XPAs5m1Y6jGxN1ejYgm9OEbXPfRr9TXoqkkUqkNtLCqlcOwDn9k9nLqlfkRXAUZP2GEuLkQExGfqS5tI5Vr+wQRzFB3Do4FGfQlunA0Du2y7UlqJPHXSPulM5EPX0O2q9ui74/aPGJCPp4IY8jnQw0CwqD0kaCPDuxUXAZzkmGjh2teiVX/j0uHkJ9z7tGlBWnjZyAuhnk5vjQ2mgQGy9yndnB1GlbUMTm14Uelw8x+FNx+/3dgWM0l2ZY9cq4qEppeBinz3FzuDitHlRY9crOwD/Qz/ejznFDk/vFhlWvHA582OPyp6x6xatDDcu9mrJdIsqK2qDOx3+lPxn9BCsssn5Wm99d9UVnd1e3s9CiHEOvFnEXqD8EvBY9sXfmia/f0Xh34MeLiWBkhNPX61JxmDa3CeAVIWQ3Q6cv8jOcGvTkQCwmdLs04Bz9Hx1HMUHcd3ym0EWHiQVlXP2+CWztcc1EG2kh96sTgelhhQhfFHlXtO39uvyi/MaJD4nJlgl044ipyfIoHJPVdvwQ/wWOhdnTHGMIm/gnpOLdIi5QI7URURfObvO1z0Y46fRCrjPDABMbEoEQz/ZvOKaeOuTTjjjI72StoCd/Ylz0iuo5BIObBmIMvVIqnoBjcjrYOA+n79AxOa2NhzQWG+8CC6WyMNEl/MKijcfc0dZ9mjJTg0TbkwOrXplMsHsxMVHRhfQNOlHZA/8G91kDNp6R9RNOoH6+IwcZOAKfgxrRy1SdnoSZzl1+hmEnon47ayeEV0dB1m8IwXdZPob/hDWpyXygd2zVK5vjvUECcJyBTZI4bWQI8C2fr+wkTgNMoOtXZdOtoESp03uyyiHcj+NDa6PnQdSdeVP3C8Hu+XMBvmPSKsA0cp2ZQviTWVA3XAgo51P4L04BuoC9Q+rjhcm5RxS2BHb3ub6rCLZiAl2dnhrwbw/De4MEzNfpyzVlpn1D0sDvmUH0k9dQJL7YEMeEcY7Outpc36nN9aD8Fydah5uog4Tu5GDdNn/zHhAkUpKJ+42zK7phm+vrE81R1U0c/SYzMIKLjEXMXaM4JwdiIdbuGSbxjtcNYYKXhX4Q/B13tbmetX7tnt84IuzuS7yJEx3HTVD91sDZqPHDVL+a5CZOELOi+QFlG7lfu1xbgBqVKqrpWNQ6GOSetzJkbpgERhaodrm2GJgrFZt6fmCujegiT0Wde0RhQZvrHcTvr1ok2e9vb9jM6Q6cELhujjLlMJ8i17S5LvsVJkIaJxsQb5f1wjbXHwunih5xTP+QVDwjoo9E6B0pMUh9BdWhUsbE/UbetcVxnJJtGWXZUcIUuolzcvAC/tFCbJzENXGJVKfF4vvCNl9L4h0PIXjkj7+0uZ51HfwnahAGNyb0i3P6dw/wnM/1+cDLoTVyIUI4y+0w0PMToQ6vavM1I/0qTuANuT8wtdgYSZtFk12uPQqcEUC2qfsFdbI8GW+zTj+itpEv4YQt9eMZ1M21vGBygSq34yDP73zg1gDfMzX3eBvnfbhJ7WRDhFuvom5etFgImEozEKffvwrVmd7NE7SfPwVGzAmvlYrXw9wiMy0+g/d4tAizkV09SWuxIVewSSEcoL4LfF9TPgv4CXBJHMUk5E5uNNG89SOt3kXiv/s9Li/DCUv3hQj6yETeXbDLtbdwOsKn5Es4Ca2OjJvLIs5pmDh1OBTQxQZ/GijZ5Vozjn6COCcH38KpuzJvAj8A/i+OYoI47/hS4GDNpQU4EaB+GF2tlcTR7wmcMK1yrPvlwA2YsUd/E3XgCqrfAhxTQ92O5YPAEWEzBHsQ58T4E6iJH8HJgXGSXa41ImvlQrRluV/dLmII5ah15lS8N0Dew0kW+LMI+nhharIcdRy5Ae9x0cYJb3y0eDd55CnUnEXGTsPa/YE4Efmyz1dm4tQpk0lS5TqzgVWvtLOGMIZdrv0cNSLnMpyJ6JF2uSafEEUlTr//Ik6/Ly8g+3DCWh+TQJ2+QlM22EypdNYUb+P46XzIUG6vtmS12ID2ZkXAysnj05pLJ9nl2g9E7GZTmHLui+qY1orjreNSu1z7iF2uBU5I5MNc1DwHgScqonLKfjjP2eXaXna59nBc5QSRJ1Ii3r1u53h3u1y7LJZWq5D16yTgyYGos/JOFkDZLtd+YiirZ1wHYnngATjDLtdOEJPpuMTSzy7XnkTtv26xy7WDo2So1sjvQz05CKPfW6gx5efZ5doMu1y7M65+gkjRisBJXIVzCihzkF2uyZm64yL3q0OIZicctc6sh/cJyM/scu0kQ4u/FrpxJIpZzBuasqB10Mvn5mG7XNtHLNhziej/5E23HSNGaJLrzJpWvbJmgL/b0efaR+1yrWZ4YmsyQE1ohJP2JlLx1Xa5toddrgU55QlK3H7/OVRz8/vtcm0/u1zTjalxuREnoa0bP3+8PHIwqmn52Xa59qk0M9VnudgIM/HRRUryM+WJSqY7UsDG6ON6g8H7NeBHA+o7Mf0+ZP2mhLTHlPVbiv8RbFjyXqfzrl/kkwMAEXBBdlpMug7mvY0MR30mfrxf+lW/iWMS9/sMqt1/6PsVoYPl07sgm1ZD8V5sJHG/SSDXmeFEy/egqzNBHJL9FsN5biNR2Z6Bmb3BCf9qFLFRJZ9ahQ0bn3S/uhKxKXODVLyZVa9skdRvJoBucaQ7sUmUwbzY0O36xMIu195APaKL0uDfRrWHjTsomr7fyBMpcQIjJ9lLeiIV+ORAINeZNwzvROW9Tud6sRH35AB9ksfE20hI50ClDsbUR8b0O15o6NRKRhehKc3Fht/EMYlxRLczHzV3QpR+eku8o8kZv9+EeL8tUJ9A9UNLc7Ghu18jppQa4mzijEWNFJZ0ndZNzAfF6Yboc+Tolk3M+qgFYrAuNmzgrejq+CJ3cltY9Uoob/0YJwdpdnBxTg4moObZSFo/iFdn8q7fCtTIKXF4ixgnB6Sz6x1n0MlCv8AnB6ItyaF8814HE9khFGaNT0rFUXInRDUrSnviCOo4EjV3Qt7HkaTQ+S5GMUUL3UbEZprXu5onTpyMIvwcH5SKd46bcykEujrzSEK/lfd+X+ZaVNOtweK3sQ+qtcwVWfhrZbnYCJOXQPbvmCMiBSSBbkdlRgQ58j0Hud80j25l/XSnFV5kMdGDcOFls1hsxKnTbxry1QBWnhzIk7O8depxTg50Pl95eseTUPvX3LQRQWrmCBjInWCXa0txAoO4aTdxtEjfJAaS25kfE8DnIIv7NYpwSJYj6KR1srER4PWM02wjI/H24TSNXGdetcs12YTPFPI7mRgi71Aa/f4A7HJtHk5QBTfdVr0SJcJc2uTChArSW2zMQ3WyCTPQyN9N8tgsKSfxKX67FOKan0NYnkwwdBOuPOk3HqejdmNav1moux15q9NxdpDkd9yPOtGLi6zfMJwM00HQ3Uue6mCu9RMnL/LAnXa/amKnut39TsN7E2UZ8cN0e5HZzjxO4lUvBosZFah1pkskvw1DlMR+WT0/k9nnA2PVK+NQE78Z99dwEdWPBtLpV3XoJuhHpfC7kRHRMeVE0bNJKdStTCqLjTgOyWJnqksqbsbXypNHcEJoujGx2LDwPznYASfUrg4bfdz/OMTJI9ClKWtG1kRPrvWLc3IgOgF5V6RpQC0Z+RlOFSYCQeiSPr+WwGlirt8x5hNQNSNroucNVF+IoPpNxfGDctOMq5APpnb6lbwJbU7DPuBz7WWTp4luDOZOCNVGxMTRz5G6GUGHrNDVmVARmuxybSFqYJB2bcSvzjTD/H5IdAvUNPw29kB1Dv9Pgr+X935fhy4vUd79NnZDNeW9ynAE18CkdbIB0XdZJwJyTPZmbG08EPaYsq3iLiEmaS3CTlT28rn2ul2uyUnu4pL3iVSck4MuTVkzjjIeRK3TUwD5lKsZWxuVOCcHXdLnZlxlNJisg0sw78eV6zYi+gT5nnOjn4Sp3AnyOxmNt8kL+PerzQi/HwYTuRPC1sHdUSeObmIlk0yZpEzR2rWRTOqMCJcth+1OY7Ghu9+7Evw90/1q4nkiRPAguT5+wKpX1k76t2Og8yvJxIQKsl1sTA7o/NSlKWvG1sYfuVKtBWwaUobJxUYz5G8HwWSDX45hu0mx4ygfgedtIiU/w6AnB12asmZsbVSi5nsZjmpi0zShkITJOvhyAk5vM4l+ctClKUtiohd1wdulKWvG0sQH0Z7lmO5RcifkvV91E3tnHrP3+7ZdrskLvjzzP1Tz60RN76x6ZQpOCHovmhF+PwxyndnEqleCbhBFRa4zi0jfjCpqvzUzCYd9D+SJegdweEq/HQpx2iufvCzASX6YCVkuNnROvDq6NGXNuMq0wcSOSuAGJSaoaQ+Ks1HNxeJM9JIwR4g6kZqmKWvGU0WLrN9QIMhOR5emrBlXGQ1RO3Wd41sznipa4gw68jtuxlNFxfDJwTt2ufZOXJ00RHWy79KUNWNr448ud8L2IWWE6VcnoyYqc9MM+dth0Y0jYSfLYX0OslxcGcUnQlOYfEug1pnxPhns/Z4fDI4FamCsemUksLNUfG+CAXgg5/2+D7pTgY+m+Pth2AG1j/+XCLKRCVkuNiBYBevSlDVjadIeE45aYe53C7wz3EIyEymTJwfNuPp4YOrkYAFOkALT5L1O512/SCcHVr0yFue00U3TkE4ypk4OmrE10aMzKxob4O+6NGVJmyOk3a9mPXF8nJi5E0RSMdmJ3WtxNRL/yInNML+dE+Q6MwrVmbkdYRySs64zaTuJ74Jq0pukCRU4CS/ldhGk3x+i+V7TkE5tEdnLH5eKD7TqlaDO7WnyGU1ZZiZUUCw2vHgZdUczbIMPc3KQVQcXOjyvVa+sgWr33zSlkITu5EB2eNLRJX1uJhRX2mSdTsPEBoKFbu3SlDVjaaJB7J69KRUH0W9DTVkztkJ6Qp8cpBzUwtQ7ftMu12STFdOYcIANc7+ZThw9duZ3ipA7IWg/PQN14uimGfJ380BSVgZR6oxxc2ENjwHy7nOSiw3d/d6Z4O95BQwK0metR7pBLXT8XfrcgX5inxlWvTIMOE4qXoCTLyQzsl5sBKlg8rHZ/ITMEVYiGoPcyW0jJtpBZehODrzu1y/6BaQ3UVlXREryo0tT1jSijUqUhEyDYaIn1+lZYgfTNG/ihKx1E2QxlJYZGkQ7OejSlDVja6JH1m8U6qmKzCTU0MtNUwpJRF3wpm6OYCh3Qhizoqx3qcFM7oSgbSQP92uapBaoyjMUkby29ZGTWPSyFsJ0U87cvUuAcTkq8txjOfpnbpq89/te/AV1TD0hQgChJDkUNfnsZXa5Jp8mpUrWi40oFawZW5NgyINEB6ptYzvaNigxOc7LyYYu67FMl6asaUIZDVHqzATUEMJNI9qo5LpOe5wcRNGvH/29miCKz0GXpqxpRBuVKO+4S1PWjK2JnigL8izNEeR+NVTuBHH6Mlcq1vWrY/EPAbsUtW0kgQm/Dfkdj/PY+FrtFht2uTYT9dQ37vMDfRvRhYB10wz5u1GR68wYHFNro4h+QH6WjYQ2vmTkd7KO2JH3o0tT1jSiTUBEVKrrpOKNgA+mqUcbdCctF6athEyai405OEmU3PjuAmeQY8ONznbSL9mPjiAnBxviPzlIIsdGiyg7812aMjlcnynyrt9M1F2OdnU6rRwbLaIcV3dJn19N0GFQ1m8k/v5LMDjrYDO2Jnqi6KczR0jq+ckk0a/q7rfdxDHxXWqBbrER935BGjPExHH3NnLSesemkevMpiFDjgZtI3lZrJloI0HQ5fZK2l+jRZTEfl2asizq9J80ZZ9NXQsNVr0yCedkw80LwN0ZqDOA1BYbomOXHRD9QsyBE9knLXMEmYdQ8zwcHFKGvCPTidpg2plQJZFjo4XOT6DdO+nSlDVja6In1/oJm2zZrKOdfuvinCC5aZrSSYP8DDcMYDPeJX1uGtNGxcQ7XoL57OYtcl0HcQZt2R8pT/rJ6CZScfvV6ZpNnFxMHEXuhBel4gNDmsUEqYPbA+3MfAdTjg03ujpzUNA/tsu1+ThOyW50bSQrc2YZE20kCLr7TdRfw4WpfjXxHBsa/oXq03u0Va+0M69Ng0+gzi8uTMhnNRRp25nJGVU3a2Pr1qUpaxrTxgdh3yav8ncLuaMi3y/A5tLnLAfFIPrJdEmfl5GcOcKbOI5Nbtrpl6a/AajPcPM2ZkBdmrKmMW1UZP2G4Bz7arHqlRGoIambhnVyE6UOKv4GCXamJtoIJDTRE6EM5d29KPo1TegTgP+i9heHhrR5lt/JCNTTwlwsNgQ3SJ/HE87vwMQ4Mtcu1+S+dLAgPz8In99A6afdH0Qkr53ayGiG/M1I2OXa66hRj/YXfbNJ5Dpjk94OuIl+/40sQrmKU/4LpeIROBP9rDle+mwDf81AD4W0FxtPS591g4SbLk1Z05QyAZC99y3C7TDI9wv5Wmy8AMip68NOVBIzRxATSPkZtrNd7dKUNU3o44Gs35qAnw16l6asaUoZDUHqoJu0cmy00OkX9h03jWiiZw5q6NGwbSSpHBst5GeY502cflSb58lAdwgxvnXaI3eATDPE78VFFwUmzGQ57+NIotjl2rOogQUODGDj70Z+hutZ9coY12ddCFiZZojfi4tcZ0YBHzIlXPQPe0rFj9vlWhIh4nXkvd9vx581ZZmaUln1yvaofmq32uVaFqc/CmkvNp7SlPlVsC5NWdOIJsG4RlMWd5BYeb/Cvm6zNjKaIX4vFMI863mpOG8NXq4z67eJCtYlfX4XeMekQhJ5r9N51+8lVF8uz8m8iBgzTipuGtXIhVjwys8w721kFP5+G12asjQHJF2/eliIv29Xp2cA7SaizRC/F5fbcLIyuwl8v8IMSA656h5HdEFG5JO+ZtDfyylynVmT9gssN7o64x5725lQweCae7RjC9SIRWmZUIGTGiDwJo4w/c0sx4aMWADLli87WPVKmE0T0+gWOxemrYQXWZ9sgP8uYZemrGlEkwCIJC7PSsWBd1TELoFs2+e+3wMCiEnaAUp+J5t62ROLnSC5g2omoZQLXZ3Z1Of7XdLnJE1sIP91Wq6/kCP97HKtD1VHP/3SzLHRQn7H63jZ53oEtUi7DUO4dzzTLteWmFOnLbeg5hIIM5FqZ4IRpF9thvi9WIhne7NUvJVVr+hMPr1QTnhd5po7oOY+kk05myF+K4/oTofCLFDbtRG5zshjxjKSz7Hh5n7UqGuHBYjUFxRdG0nLOdzLasGvz1oPdb7aNKlTBC7QlJ2auhaAVa+MB8pS8QIyTuTnZrAtNpI2R9Ahd3JjCLYL0sKvQR0R4O+bIX4rCrJ+w9BPOCEfEz3w2FnOKHpZ3Dr9VpLJ1OxybSFqNLPcLDYE8jPc2GdB36UpaxrVRkX3jr1OJCfhmIe6aRrVRiVuHWwa0yQAok7eJhXvEDQTr12uvY0aEMCvX5UDfUD6ExXTk+W1WLXAyMM4kjT/QXXyPjzE5NuzjQg/TDmSl5yTIK3oZcDKTZh/ScXr458HJAxynbGB2w3JDor8TqaIkNU6ujRlTaPahOefgGx2doxVr7RzdE+CL6BGFvt71rk13KS62BBJnWZLxbkdFAWmj/wnWPXKRKteGY7q/yEfK0L6Jhjg/U66NGVNY5roCaPfRBwTEjdNo9qozMQx1XKTtzotP0M/J/Yu6XMfyeXYaCHr14m3E3uXpqxpUhkNq00byTjHhhvT/Wpr4jgd2Fq6Jo85S1FPnJNG9lOBcKc5fu/4SKlcF6a6GeK3codwypUdxTeivRlyixdQF52t53cY6klQHk6G4rYRLWIXXDZBu09ETkuTvPervoh8JGdKxR3At9LUQwQO+KpUbANnpKlHO7LIehjI4TfjHBtu7sb8jsoWOM5esu+BHKUlyWRqLcI4anVpypI2EfEbJGS6NGWJ6hfGiT2DHBstdE7scsSpFl3S51dFiN8kCbMz36Upy8JMKUwbaRrTRIPYxJkjFXvpl2WODTe6yXecnf61rXplAurEG1SfoGaau9SwMjndQ1Lx3pKTsh/aOmjVKxvihL11I4fahcGbY8NNZEd7sVjx8k/U1Zm0EsP6cRPq2GfCb+NQ1D7gKgNyw5L3fj8IZ6Gegn0m6CmtIT6Nmoz5CuEGkBvysNhoDRIy65Jdjo2ViE7qeql4Ou0j0rTwalBBjr6TzLHRIkyD30RT1jSniop4/i9IxV4TqdT1E7SLdNJiQ9LNsdEi0DsWC2j5GTaTUEgizGRe1m8x6s61aZqoPga5aSOCoPbPWek3ALtcexknDK6b/ax6RT6Z9CJov2qjJolsBvwN08g71UOB/QP+rdf96iafuhPywZpjw831qElU4yxQNxHBRmT/Bd1irRnid4wgAgPcIRXPsOqVdWKK1s09ro4pMwpx+v0kkx0HRmz0nC8VDwO+mcbviw3MiubSr9L4/TBksdhoFxWixY6aMl3lTIM4oQu9jgrlBj8XNY504vfrEenEa6Iiv5P5pGOOID/DTYQ5iExWdUb3jnVO7HnST/eO10N1NE1DvzAx1+Vn+HTSCYtCOrHL+vWj7qgmgfyOvZzY89yvjgD2Cfi3ujrdjWoe8giOn52bvNwvBB9HXkPdQd0c/a68bPf+mvCTGdSIid09UvEewiwoCHKdGYqTG0He1NTVj7zUGQs4JKowYb59kFT8HNncX5hIhHK/9WIWOTY8+DWq6eLJHpvopjkCdSF2l12u3ZfCb4ciDycboF/N6kKINQzrEpQ4OyqvoYY93BWQj9nuRu300rpfv0gnwMoV9A7S9x5OyRwhqBO7XGcWoo/GZJq81+lc6yec2OTQq7qTlymo5l9ZtZGNPJzY5Wf4pLDtTZqgJ5SyfjbOhDwL4tik6+73IFTzEN33shpHHgHekMoCJTT0MNfcCthbKnsIdfMuq/tNArnOdBI8m7iuLnxUU6YL2JHVMzTtt7EPqvn21VlkmA66iSOsBOTNu9zUabtcew34i1Q8Gvhykr8r5mg6/5DcnWpAfhYbQQbFFcD/zKvTHhH9RM6suYeY/LT7237UndstNV/VxbnPaiI1HnWHe1NUO9as9ANpsiwGbHn349GMFkMQrE4vQb9Da5q3UP2OgugH2b1jnRO7Tr+HE9JHRtavExgQdcSqVyai+uRk2UaCvONnMtz1fhDVBO5Ij1NLmVdRN3HkzRDI0cRRTOjkneq1UZOreSG/4/VRzTIfQl1w5WZiZgDd6dDRAf9W10bkKFRvoQZQmEu6eWhWYpdrL6DqfWCbXFN+fERTloW/RosgkQh3QHXYz1udPh11Q/qbVr0yOcHfPAJn49rNU+j94TIni8XGyziTLDdBBsXHMz42k3cYOlDjGnshTyjHSZ+Xos9emtZEKoiZTZYTvSD6bYTj+OwmrQ6prRO7mDjLz/CxFJyvvRLTBWlzy4AnElFKRdZvDI7flpssF0ODvo0Is6rp0nfS0k9B7Gzqsom33bn12MSRB/ZnUU1TF2n+Lk10O9WfC/i38jvWBSmRAwVAhu84AZ5C9ak4IuCkTrfYkDfQrkXNwvxwFjv/LuQ6Mxr4eFghYoFSkop1pmlpEiQSYZb9aiCEM/Y/peI1gVoSv2fVKyOB32gu/Srt4BdBSX2x4TFIyINiluYSXlyCunL9nFcCPIl29pBXAttIZfNIL9pCkF1Rna13Wu8kiE1/Zvp5RDqR9dsA1VE1zTrt68TusRj6XwoBClpEqYMrUJ2MkyLvbeQV2m/i6Hb+s+5XL9KUfT7g37brVy9Gf9rZF1B+EtyEeppzTED77nb3+1/0UeayfsfGEJN+uc4MAU4I8LfvopqxydyNmicn6+f3N01Z0DbipoRqQnVJxu0har+aq8WG4Eeom47HWfXK3gn81qmoGynP4vR5uSSLkw1QK9h0ESu4RZaDthZhlycf4XYRLFttu0HiAtTwhWnupuj0k0295InoAtJxfPVyYm+nH2Q7md/Uqlfcp1V50w8GdurroobPy1q/du/4iRQzX+t8f9rp1w88mog2Eh72z3lrIzpuw3FQdXOAyJfRDr9+tR/4N+opctbjyDLgz1LxcOAzAf683ThyPuo7fsMu194MqN5g4QKS2fh7DjVzOGRfZ/4L3CsV72jVKzuFFKU7QZMjKaVNlH7/JWHanivscu1J4LeaS+f4JKkNjVWvdAH/T3PpqyluDoYmq8WGfHTWAcxwfc7joAhwnqbs5AB/52eX/wJOfg055GOa9/sajjO1m91a/yP8IbJyDm8hP8MdpAWqXGfSNpfQRTpxL5qzrtO6Orib6/9zrZ8I9ygHVUhtd0s4scvhQ3eTPsvP8OmUM7jKz3BjkR25he4dZ+UcDqzcqZb7VYtgpkV+/eq/UOsL5GMc0U3wTg6Qu+l59NnQwfFN+SdqQsM83K9RxMafbH7XRbCNvyd9rv2R7PtBL3Rzj8CnG1a9sh0D51gA99vlWiZ+sC6eQV3gufv9NVBPOvLwPrz4EfC6VLYF8HWDv3EG6unblXa5dqPB3zBOVosN2dkanCR3LXTO4WmZS/hxI+qE47AACVyeAt7xuPZHMj7JEQO+bLe5k1WvtHwgNkENH5l2g5frzHBEpyQG6azNJcLW6aX4D3ymkXfGwF8/SLcOvoWaT+VDrglYHiYB8jueLpKqIcxguqTrWesHA6MVyc/wWWFakjV/QQ2BeUKA3UA/W3PdLj/kYKJil2sv4phTudkUNbKU/HdLURMDtrgMx1RTdhjP/H4T4g+asiAbf7o2Ak7o0r+g1pl55CP7+mWoc4iPW/WKHObYizyearQ2cR6Vij/gChKxPfl3Dl+JCLbxdc2l06x6RTZ7Co1VrxyM6uS/hJTyesQhq8XGPahxid3x1bM0l/BETF7/KBV3At8J8He6Tm4FcCH5GBRvkz53sipmfR71g1WT5elkby5xFyAvbvYBX+dwuQ0khl2uzUaN5ra3y/RA1m858Hjiig1EfseTWbWrlfc6mAfTT51+rTo4Fil6FjkZtO1ybQ7wD6l4EnBim797Hb1f2xs44crlOrOY7PIlyOgmy98LcLqhe8eQ7135JLgRNULUEWIH34/bPcqvxEmGuL1U3sjYORwAu1xbjBpedRQBdsyF8/zxUvEC4FITuhlArtNjWNWfDsY6/X849dPNKOBq1wZuaKx6ZRP0Phm/tMu1PGRT9yWTxYaIOy/vtO5q1SujPMwl8lS5/ox6lP0Fq17xSkbTQhdtqiF2dOUGNR99FtMkuVVT1ppI5aHB34uaxbm1QM1cP7FDLO867il2Z9dHDSWcRZ2W3/E4VkVe0TmHpx39LUwd7AMeS1Ydhby3kSdQnY/ztBjyQzf5/lGInVs3V+PUD/mdPJpG9LeAXIvqrLwPcGibv9MtRmbjbODloQ6mgtjAk3fmLeCMNgu2d9Gbol2As7GRVa6rIOhMqb5l1Sty1D6ZH6NG3fp7yiaefuj6Vb+xPY/O4SsRi9OvoJ7Wbg1cGjC09wCsemUcTlQyOVFrE/hleC3TJ6uTDVAr2DCcmNc6p6fcVC67XJsJ/Ekq7sQnkYpVr6wPfFBz6XVR8baXyrMItfcwzm6Hm1aDl9/JQlSnzkQRJ1uy2cQuVr0yGn2dycNkfiSwC/mp09qdb2EGKIeOzEK/2zVl+4jJg/wMnxS7fWnyEqoZpZd+Nik5h6/8QafPkN/xZmIykpc66MXdwJ1S2UTgu15/YNUrH0CNyAJODo5pqANzbu5XnGrqJgk1KbDESkS4y09qLs0V715+x2+K8Wp15VwcMyc3++K/YPsmqqkZwCxy3kbscu0p4HKpeBTwM6+/seqVrYHPSsUrcOz+84LOKqC1SSK/k5dFJvlcI0Lhfk9z6SD0IWs9EXPES1GTdfYDJ2UwDkYiy8WGl0nCsZryPO0uAJyGOjE/zKpX9vf4fg+Oj4HMVjgdY1bJ8lYidvzkwX57q17ZFjXp1CMZxXKW68wQ4APAx6TyrMwl8l6n70R1xsuNfmJiJL+3vXHqX2bO4S08JvPr4ZwaHCyVZ5UsL0wdzNQ53I14tjq7469Z9Yocd78VtMIrhv1O5KROt+Fc1AhimwFf9Pj+11ETzoHjO7Q9GhOgGLrlHhGR6MeaSzWdv48wJ/Iyed6bwVFnvo1qgv4Zq15RTgDEJkgNdZ53jpgM5wJhFSA/5z3FPW0hleftffhRQzV9A/iyVa98T/Rhvlj1ynCcEy1d8INT7HLt5pg6pkaWi437UbO7fhJ14jibnFUwYfrUo7l0sVWvDFh9WvXKl9HvRoEzsOh2t66Pp2Fk5J15C8cxTa4nedEP4NfAhlLZTRmZS9yNOhAcD3xYKnuDDAIe2OXaPNRJ+gfRT/Kyimwhv+MJqH5SkK86eDFqNLk86fd9VHOEe+xy7Z3k1QmOXa41gL9KxcNwbJ1X5qEQk6gzgZ09RO0LfFUq60d1ys4UcbpxqubS6Va9sq+7wKpXDsKJdKNjGNCrKc+qDqbJOain7JsBf3WHwhVRja5E3dhr8UXgQKnsOdI3Z/bFLteeB86Sii3gcmFB4eY01Huaj36BljVyvzUKfX6RQVOnxQbKyTgnNzI/AW7wM4ETc8l70eeQuQD4nQk90yK07Zgp7HJtqVWv/AdwnwZsoPlqPU1H2hD8Fif0nHuiuzZws1WvHIgT0q2M2jHIyEdjz+PtxJY0uomKrN8KoJ6CLjoeBN5j4ICh85XJJMqGXa4tsuqVe3FOW1rozDz+lGEipVsZOPEcjToA35Shw9mtqDu7ch2cgzNxyALdyYGsH+gXSGnwHE7oRfdJkE6/zCPRePBd4BgG2s5vCVxv1StH4Djx/gj4ko+MNcU/N9cJh/K8cQ1OnXJHhhsGXCXu9zZgPxzzGa15lUB+x4vJcYIvU9jl2jKrXjkVtT84FnjXqle+jmNVcBmOSasXm2jKzs+Dc7iGn+DkZXEngtwAZ+5xEE4o+2/gLDZkfppTM6RbUU+d5LF9AfpFdW4R89yPAPehZkbfH/ivVa98F7iZVcEuNsexePkR6iYWwB3Al3JaNz3J8mQDgnWGWQ3avggfAnn3DBxn4CdxTmQu0P1pG9HnZ5hu/jEcJ1M/rsoqSZRIWNMugsYrZLcrD/rdGDc2+nqRFkHanM4RMS1uwJlQ+lHPwHkdWBnj//Y2X7tT2FenjhiA2tXBd3AmX7lDPF/dDv7OOJOoWbSJ/udBlnXaE/G+voYa/GI0cAvOOHITquNyu3GkN28nVwlyNU5eFZnPAW/h5LHaL6TMZTiRInOHeK/f0lzaDOckZhZwuub6U7Tf/MyKO1DzU8hcnJFpaixEtL1DAd0ibwJOcIwXcPq3t3Dmj79Cv9B4DvhonpP3eZH1YuMSnAfsxS3i2DCX2OXa1cApHpfHa8rexn9l3or1nQli4NN1Um6y3hH1stNucUGGpwYAF+F0GF5cb5drcsjG1LDLtcfwX4y9ibPbmgkiQsrZbb6W9QZE3tvI71Ajobi5KOdOhaejX5B3ArroVM/inT8BHIfxGwzolQgisdoxqE6yMHD3usVyHH8PP7Kug6khxq2Poze3XgPnpEjmahyTIi8uF5PEXGKXa38Gfq65ZKEGRgBnIn9IVps07RCT53aO04O2Ttvl2jM4J2v3+3xtXRzrGC+uBnbPc730I9PFhqhgnlGcaD/pyBy7XPs18NMAX30POATHTMDLLOxSu1ybZUq3iFyCdxKjZ4B/p6eKikdEjhZLcEITZ4Y48fKL9JGHOq3zN2pxXg7MFs9CzWjf4sYcODfegLdz9Zs4cdYzQzjae5k69qEPM5sbxOTx8wQ7fXkVxxzBy58B4NyMNyDaYpdr1+CYxrQ7sbBxfABPxXtTo4H/pGa1QzgZH0SwwCD/xvEN9dvl/70JvRLmu7RfdIKzo76/Xa41k1UnNuej3/0Hx8csNwEtomCXay/g5C77CY4PWVCW4JiNHjVYFxqQ/ckGOI1aN6j83i7XrkxZl6j8AKjivZv4MnCQXa7dL2zhP4sa6/spnOP0TBETzWNQQwq+AxyboYmXm5NRs2/3AZ/NiV32b9D7FPzaLtcyd3Czy7U7cZyGZW7DfyGSCqJD/QROJ+vmFdQwjqkjJsMlQA4ruggo5SEBKc5k9AFN+Tfsci3NzPWREIuDT+G/MHoY2Ncu114RUVl0mz430f40NBfY5drFwHGomaJbzAM+bpdrl4lcVcegRkWcDRw32Oy5TSD6jf3wN3O8FPiw2OH/KY6tvMwP7XLtP+Y1NIt4x1/GCTLjtZh+GtgvK7POMNjl2gKcRaCc/+Mt4NPpa2Qeu1xbbpdrP8Dx6wwSJOY/wM52uXbOYG/Tlm1nr7+ILnKS+LcIJ8JE72B7uCJiyjE4E5EunKP9i4Gb5d1iq17ZFfh/OHaWlwM/y1GSnVZukB/jNIrbgdOEPXUuEPk1vgccjbPw6LHLNd3kKhNEWLsv4GRBfhdn8ZwrO3mrXjkExwxwIo75129zlPQMq17ZBmdRtANOErQfi4haucCqV9bG2VE/AGcn+Yc5OHVZiQib+C0cE5Mm8Cu7XPPKQJ1brHplA5x7+CiOichNwN9xdjv7pe8egeMcuxaOSeqZeT/VkBF92xE4C49tcCYlfweulscIERb4R8CuODv2p4loie9bxHxie5zF6qE4zvJX4iSye1b6bifOJt/xOL4OvxOnTIMKEda3hDNZn4wzZl8M3D4I6/+mwA+BGTgmvz+0yzU5WemgR9TT6TinHXviBG7pxxlL7gTuysnmqRFysdgoKCgoKCgoKCgoKFj9yIMZVUFBQUFBQUFBQUHBakix2CgoKCgoKCgoKCgoSIRisVFQUFBQUFBQUFBQkAjFYqOgoKCgoKCgoKCgIBGKxUZBQUFBQUFBQUFBQSIUi42CgoKCgoKCgoKCgkQoFhsFBQUFBQUFBQUFBYlQLDYKCgoKCgoKCgoKChKhWGwUFBQUFBQUFBQUFCRCsdgoKCgoKCgoKCgoKEiEYrFRUFBQUFBQUFBQUJAIxWKjoKCgoKCgoKCgoCARisVGQUFBQUFBQUFBQUEiFIuNgoKCgoKCgoKCgoJEKBYbBQUFBQUFBQUFBQWJUCw2CgoKCgoKCgoKCgoSoVhsFBQUFBQUFBQUFBQkQrHYKCgoKCgoKCgoKChIhGKxUVBQUFBQUFBQUFCQCP8fRilyB2OupDoAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1008x201.6 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "sg.add_transcriptome('../new_testing/files/Canx.gtf')\n",
    "print(sg.t_df.head())\n",
    "print(sg.edge_df.head())\n",
    "\n",
    "sg.plot_graph('ENSMUSG00000020368.14')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "korean-uganda",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6/4/21\n",
    "# graph.py tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "embedded-assurance",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         0\n",
      "0  sample1\n",
      "       0\n",
      "0  gene1\n",
      "[[0.]]\n"
     ]
    }
   ],
   "source": [
    "# has_abundnace\n",
    "sg = swan.SwanGraph()\n",
    "obs = ['sample1']\n",
    "var = ['gene1']\n",
    "X = np.array([[0]])\n",
    "sg.adata = anndata.AnnData(var=var, obs=obs, X=X)\n",
    "print(sg.adata.obs.head())\n",
    "print(sg.adata.var.head())\n",
    "print(sg.adata.X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "comparable-cholesterol",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6/2/21\n",
    "# graph.py tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "intelligent-penetration",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       path   loc_path\n",
      "tid                   \n",
      "0    [0, 1]  [0, 1, 2]\n",
      "1       [2]     [0, 2]\n",
      "         v1  v2\n",
      "edge_id        \n",
      "0         0   1\n",
      "1         1   2\n",
      "2         0   2\n",
      "          chrom  coord\n",
      "vertex_id             \n",
      "0          chr3     20\n",
      "1          chr1    500\n",
      "2          chr1     20\n"
     ]
    }
   ],
   "source": [
    "# dfs_to_dicts\n",
    "sg = swan.SwanGraph()\n",
    "\n",
    "# loc\n",
    "data = [[0, 'chr3', 20],\n",
    "     [1, 'chr1', 500],\n",
    "     [2, 'chr1', 20]]\n",
    "cols = ['vertex_id', 'chrom', 'coord']\n",
    "sg.loc_df = pd.DataFrame(data=data, columns=cols)\n",
    "sg.loc_df = swan.set_dupe_index(sg.loc_df, 'vertex_id')\n",
    "\n",
    "# edge\n",
    "data = [[0, 0, 1],\n",
    "        [1, 1, 2],\n",
    "        [2, 0, 2]]\n",
    "cols = ['edge_id', 'v1', 'v2']\n",
    "sg.edge_df = pd.DataFrame(data=data, columns=cols)\n",
    "sg.edge_df = swan.set_dupe_index(sg.edge_df, 'edge_id')\n",
    "\n",
    "# t\n",
    "data = [[0, [0,1], [0,1,2]],\n",
    "        [1, [2], [0,2]]]\n",
    "cols = ['tid', 'path', 'loc_path']\n",
    "sg.t_df = pd.DataFrame(data=data, columns=cols)\n",
    "sg.t_df = swan.set_dupe_index(sg.t_df, 'tid')\n",
    "\n",
    "print(sg.t_df)\n",
    "print(sg.edge_df)\n",
    "print(sg.loc_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "plastic-planet",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'numpy.ndarray' object is not callable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-9ccf6dbde2b9>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0msg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdicts_to_dfs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/swan_vis/graph.py\u001b[0m in \u001b[0;36mdicts_to_dfs\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    254\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    255\u001b[0m                 \u001b[0;31m# loc_df\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 256\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfrom_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc_df\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0morient\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'index'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    257\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnames\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'vertex_id'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    258\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36mfrom_dict\u001b[0;34m(cls, data, orient, dtype, columns)\u001b[0m\n\u001b[1;32m   1297\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1298\u001b[0m                 \u001b[0;31m# TODO speed up Series case\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1299\u001b[0;31m                 \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mSeries\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdict\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1300\u001b[0m                     \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_from_nested_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1301\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: 'numpy.ndarray' object is not callable"
     ]
    }
   ],
   "source": [
    "sg.dicts_to_dfs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "retired-sharing",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6/1/20\n",
    "# todo - add get_loc_path checks etc - done\n",
    "# todo - add find_ir_genes, find_es_genes - done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "awful-requirement",
   "metadata": {},
   "outputs": [],
   "source": [
    "sg = swan.SwanGraph()\n",
    "sg.annotation = True\n",
    "\n",
    "# t_df\n",
    "data = [[0, [0,1,2,3,4], True, 'g1', 't1'],\n",
    "        [1, [0,5,4], False, 'g1', 't2']]\n",
    "cols = ['vertex_id', 'path', 'annotation', 'gid', 'tid']\n",
    "sg.t_df = pd.DataFrame(data=data, columns=cols)\n",
    "\n",
    "# edge\n",
    "data = [[0, 'exon', True, 0, 1],\n",
    "        [1, 'intron', True, 1, 2],\n",
    "        [2, 'exon', True, 2, 3],\n",
    "        [3, 'intron', True, 3, 4],\n",
    "        [4, 'exon', True, 4, 5],\n",
    "        [5, 'intron', False, 1, 4]]\n",
    "cols = ['edge_id', 'edge_type', 'annotation', 'v1', 'v2']\n",
    "sg.edge_df = pd.DataFrame(data=data, columns=cols)\n",
    "\n",
    "# loc\n",
    "data = [0,1,2,3,4,5]\n",
    "cols = ['vertex_id']\n",
    "sg.loc_df = pd.DataFrame(data=data, columns=cols)\n",
    "\n",
    "ctrl_gids = ['g1']\n",
    "ctrl_tids = ['t2']\n",
    "ctrl_edges = [5]\n",
    "\n",
    "sg.get_loc_path()\n",
    "sg.create_graph_from_dfs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "selective-carpet",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Finds all unique genes containing novel exon skipping events.\n",
    "Requires that an annotation has been added to the SwanGraph.\n",
    "\n",
    "Returns:\n",
    "    es_genes (list of str): A list of gene ids from the SwanGraph with\n",
    "        at least one novel exon skipping event\n",
    "    es_transcripts (list of str): A list of transcript ids from the\n",
    "        SwanGraph with at least one novel exon skipping event\n",
    "    es_edges (list of tuples): A list of intronic edges in the\n",
    "        SwanGraph that skip at least one exonic edge\n",
    "\"\"\"\n",
    "\n",
    "# get only novel edges\n",
    "if 'annotation' not in sg.edge_df.columns:\n",
    "    raise Exception('Cannot find novel IR events without '\n",
    "        'annotation in SwanGraph.')\n",
    "\n",
    "edge_ids = sg.edge_df.loc[ \\\n",
    "    (sg.edge_df.annotation == False)& \\\n",
    "    (sg.edge_df.edge_type == 'intron'), 'edge_id']\n",
    "print('Analyzing {} intronic edges for ES'.format(len(edge_ids)))\n",
    "\n",
    "# get subset of transcripts that are novel to look for ir edges in\n",
    "nt_df = sg.t_df.loc[sg.t_df.annotation == False]\n",
    "\n",
    "# for each edge, see if the subgraph between the edge vertices\n",
    "# contains an exonic edge\n",
    "es_edges = []\n",
    "es_genes = []\n",
    "es_transcripts = []\n",
    "for eid in edge_ids:\n",
    "    # subgraph consisting of all nodes between the candidate\n",
    "    # exon-skipping edge coords in order and its edges\n",
    "    entry = sg.edge_df.loc[eid]\n",
    "    v1 = entry.v1\n",
    "    v2 = entry.v2\n",
    "    sub_nodes = [i for i in range(v1+1,v2)]\n",
    "    sub_G = sg.G.subgraph(sub_nodes)\n",
    "    sub_edges = list(sub_G.edges())\n",
    "    sg.edge_df['tuple_edge_id'] = sg.edge_df[['v1', 'v2']].apply(tuple, axis=1)\n",
    "    sub_edges = sg.edge_df.loc[sg.edge_df.tuple_edge_id.isin(sub_edges)]\n",
    "    sg.edge_df.drop('tuple_edge_id', axis=1, inplace=True)\n",
    "\n",
    "    # find edges that are exonic; if there are none, this is not\n",
    "    # an exon-skipping edge\n",
    "    sub_edges = sub_edges.loc[sub_edges.edge_type == 'exon']\n",
    "\n",
    "    if len(sub_edges.index) > 0:\n",
    "\n",
    "        # transcripts that contain the candidate exon-skipping edge\n",
    "        skip_t_df = nt_df[[eid in path for path in nt_df.path.values.tolist()]]\n",
    "\n",
    "        # circumvent the ISM bug\n",
    "        if len(skip_t_df) == 0:\n",
    "            continue\n",
    "\n",
    "        # does at least one of the skipped exons belong\n",
    "        # to the same gene as the skipping edge?\n",
    "        else:\n",
    "            # genes that contain the candidate exon-skipping edge\n",
    "            skip_genes = skip_t_df.gid.values.tolist()\n",
    "            skip_g_df = sg.t_df.loc[sg.t_df.gid.isin(skip_genes)]\n",
    "\n",
    "            # check if the skipped edges are in one of the\n",
    "            # exon-skipping genes (wow this is confusing)\n",
    "            for gid in skip_genes:\n",
    "                if gid in es_genes: continue\n",
    "                for skip_eid in sub_edges.index:\n",
    "                    # transcripts with the exons that are skipped\n",
    "                    temp_df = cand_g_df[[skip_eid in path for path in cand_g_df.path.values.tolist()]]\n",
    "                    tids = skip_t_df.tid.tolist()\n",
    "                    if len(temp_df.index) > 0:\n",
    "                        es_edges.append(eid)\n",
    "                        es_genes.append(gid)\n",
    "                        es_transcripts.extend(tids)\n",
    "\n",
    "# drop tuple from edge_df\n",
    "print(sg.edge_df.head())\n",
    "sg.edge_df.drop('tuple_edge_id', axis=1, inplace=True)\n",
    "\n",
    "es_genes = list(set(es_genes))\n",
    "es_transcripts = list(set(es_transcripts))\n",
    "es_edges = list(set(es_edges))\n",
    "\n",
    "print('Found {} novel es events in {} transcripts.'.format(len(es_edges),\n",
    "    len(es_transcripts)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "known-netherlands",
   "metadata": {},
   "outputs": [],
   "source": [
    "# fix find_ir_genes\n",
    "sg = swan.SwanGraph()\n",
    "sg.annotation = True\n",
    "\n",
    "# t_df\n",
    "data = [[0, [0,1,2], True, 'g1', 't1'],\n",
    "        [1, [3], False, 'g1', 't2']]\n",
    "cols = ['vertex_id', 'path', 'annotation', 'gid', 'tid']\n",
    "sg.t_df = pd.DataFrame(data=data, columns=cols)\n",
    "\n",
    "# edge\n",
    "data = [[0, 'exon', True, 0, 1],\n",
    "        [1, 'intron', True, 1, 2],\n",
    "        [2, 'exon', True, 2, 3],\n",
    "        [3, 'exon', False, 0, 3]]\n",
    "cols = ['edge_id', 'edge_type', 'annotation', 'v1', 'v2']\n",
    "sg.edge_df = pd.DataFrame(data=data, columns=cols)\n",
    "\n",
    "# loc\n",
    "data = [0,1,2,3]\n",
    "cols = ['vertex_id']\n",
    "sg.loc_df = pd.DataFrame(data=data, columns=cols)\n",
    "\n",
    "ctrl_edges = [3]\n",
    "ctrl_tids = ['t2']\n",
    "ctrl_gids = ['g1']\n",
    "\n",
    "sg.get_loc_path()\n",
    "sg.create_graph_from_dfs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "fewer-lancaster",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analyzing 1 exonic edges for IR\n",
      "Found 1 novel ir events from 1 transcripts.\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Finds all unique genes containing novel intron retention events.\n",
    "Requires that an annotation has been added to the SwanGraph.\n",
    "\n",
    "Returns:\n",
    "    ir_genes (list of str): A list of gene ids from the SwanGraph with\n",
    "        at least one novel intron retention event\n",
    "    ir_transcripts (list of str): A list of transcript ids from the\n",
    "        SwanGraph with at least one novel intron retention event\n",
    "    ir_edges (list of tuples): A list of exonic edges in the\n",
    "        SwanGraph that retain at least one intronic edge\n",
    "\"\"\"\n",
    "\n",
    "# get only novel edges\n",
    "if 'annotation' not in sg.edge_df.columns:\n",
    "    raise Exception('Cannot find novel IR events without '\n",
    "        'annotation in SwanGraph.')\n",
    "\n",
    "edge_ids = sg.edge_df.loc[ \\\n",
    "    (sg.edge_df.annotation == False)& \\\n",
    "    (sg.edge_df.edge_type == 'exon'), 'edge_id']\n",
    "print('Analyzing {} exonic edges for IR'.format(len(edge_ids)))\n",
    "\n",
    "# get subset of transcripts that are novel to look for ir edges in\n",
    "nt_df = sg.t_df.loc[sg.t_df.annotation == False]\n",
    "\n",
    "# for each edge, see if the subgraph between the edge vertices\n",
    "# contains an exonic edge\n",
    "ir_edges = []\n",
    "ir_genes = []\n",
    "ir_transcripts = []\n",
    "for i, eid in enumerate(edge_ids):\n",
    "    entry = sg.edge_df.loc[eid]\n",
    "    v1 = entry.v1\n",
    "    v2 = entry.v2\n",
    "    sub_nodes = [i for i in range(v1+1,v2)]\n",
    "    sub_G = sg.G.subgraph(sub_nodes)\n",
    "\n",
    "    # need way to get this working, bc (v1,v2) ind does not work anymore\n",
    "    sub_edges = list(sub_G.edges())\n",
    "    sg.edge_df['tuple_edge_id'] = sg.edge_df[['v1', 'v2']].apply(tuple, axis=1)\n",
    "    sub_edges = sg.edge_df.loc[sg.edge_df.tuple_edge_id.isin(sub_edges)]\n",
    "    \n",
    "    # find edges that are intronic; if there are none, this is not\n",
    "    # an intron-retaining edge\n",
    "    sub_edges = sub_edges.loc[sub_edges.edge_type == 'intron']\n",
    "\n",
    "    if len(sub_edges.index) > 0:\n",
    "\n",
    "        # transcripts that contain the exon-skipping edge\n",
    "        cand_t_df = nt_df[[eid in path for path in nt_df.path.values.tolist()]]\n",
    "\n",
    "        # circumvent the ISM bug\n",
    "        if len(cand_t_df) == 0:\n",
    "            continue\n",
    "\n",
    "        # does at least one of the retained introns belong\n",
    "        # to the same gene as the retaining edge?\n",
    "        else:\n",
    "            # genes that contain the intron-retaining edge edge\n",
    "            cand_genes = cand_t_df.gid.values.tolist()\n",
    "            cand_g_df = sg.t_df.loc[sg.t_df.gid.isin(cand_genes)]\n",
    "\n",
    "            # check if the retained edges are in one of the\n",
    "            # intron-retaining genes (wow this is confusing)\n",
    "            for gid in cand_genes:\n",
    "                if gid in ir_genes: continue\n",
    "                for cand_eid in sub_edges.index:\n",
    "                    temp_df = cand_g_df[[cand_eid in path for path in cand_g_df.path.values.tolist()]]\n",
    "                    tids = cand_t_df.tid.tolist()\n",
    "                    if len(temp_df.index) > 0:\n",
    "                        ir_edges.append(eid)\n",
    "                        ir_genes.append(gid)\n",
    "                        ir_transcripts.extend(tids)\n",
    "\n",
    "ir_genes = list(set(ir_genes))\n",
    "ir_transcripts = list(set(ir_transcripts))\n",
    "ir_edges = list(set(ir_edges))\n",
    "\n",
    "print('Found {} novel ir events from {} transcripts.'.format(len(ir_genes),\n",
    "    len(ir_transcripts)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "little-impression",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['g1']\n",
      "['t2']\n",
      "[3]\n"
     ]
    }
   ],
   "source": [
    "print(ir_genes)\n",
    "print(ir_transcripts)\n",
    "print(ir_edges)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "hidden-buffer",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[3]]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nt_df.path.values.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "sound-judgment",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding transcriptome to the SwanGraph\n",
      "             tname        gid        gname               path    tid\n",
      "test1  test1_tname  test1_gid  test1_gname    [0, 1, 2, 3, 4]  test1\n",
      "test2  test2_tname  test2_gid  test2_gname    [5, 6, 7, 8, 9]  test2\n",
      "test4  test4_tname  test4_gid  test4_gname               [10]  test4\n",
      "test5  test5_tname  test2_gid  test2_gname        [5, 11, 12]  test5\n",
      "test3  test3_tname  test2_gid  test2_gname  [5, 6, 13, 14, 9]  test3\n",
      "  v1 v2 strand edge_type edge_id\n",
      "0  0  1      +      exon       0\n",
      "1  1  4      +    intron       1\n",
      "2  4  5      +      exon       2\n",
      "3  5  2      +    intron       3\n",
      "4  2  3      +      exon       4\n",
      "   chrom coord vertex_id\n",
      "0   chr1     1         0\n",
      "1   chr1    20         1\n",
      "2   chr1    35         2\n",
      "3   chr1    40         3\n",
      "4   chr1    25         4\n",
      "5   chr1    30         5\n",
      "6   chr2   100         6\n",
      "7   chr2    80         7\n",
      "8   chr2    50         8\n",
      "9   chr2    45         9\n",
      "10  chr2    75        10\n",
      "11  chr2    60        11\n",
      "12  chr2    65        12\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "sg.add_transcriptome('../new_testing/files/test_full.gtf')\n",
    "\n",
    "print(sg.t_df.head())\n",
    "print(sg.edge_df.head())\n",
    "print(sg.loc_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "cardiovascular-jackson",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_transcript_loc_path(path, edge_df):\n",
    "    start_v1 = [edge_df.loc[path[0], 'v1']]\n",
    "    v2s = edge_df.loc[path, 'v2'].tolist()\n",
    "    loc_path = start_v1+v2s\n",
    "    return loc_path\n",
    "\n",
    "sg.t_df['loc_path'] = sg.t_df.apply(lambda x: \\\n",
    "    get_transcript_loc_path(x.path, sg.edge_df), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "stable-retrieval",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tname</th>\n",
       "      <th>gid</th>\n",
       "      <th>gname</th>\n",
       "      <th>path</th>\n",
       "      <th>tid</th>\n",
       "      <th>loc_path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>test1</th>\n",
       "      <td>test1_tname</td>\n",
       "      <td>test1_gid</td>\n",
       "      <td>test1_gname</td>\n",
       "      <td>[0, 1, 2, 3, 4]</td>\n",
       "      <td>test1</td>\n",
       "      <td>[0, 1, 4, 5, 2, 3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test2</th>\n",
       "      <td>test2_tname</td>\n",
       "      <td>test2_gid</td>\n",
       "      <td>test2_gname</td>\n",
       "      <td>[5, 6, 7, 8, 9]</td>\n",
       "      <td>test2</td>\n",
       "      <td>[6, 7, 10, 11, 8, 9]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test4</th>\n",
       "      <td>test4_tname</td>\n",
       "      <td>test4_gid</td>\n",
       "      <td>test4_gname</td>\n",
       "      <td>[10]</td>\n",
       "      <td>test4</td>\n",
       "      <td>[9, 8]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test5</th>\n",
       "      <td>test5_tname</td>\n",
       "      <td>test2_gid</td>\n",
       "      <td>test2_gname</td>\n",
       "      <td>[5, 11, 12]</td>\n",
       "      <td>test5</td>\n",
       "      <td>[6, 7, 11, 8]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test3</th>\n",
       "      <td>test3_tname</td>\n",
       "      <td>test2_gid</td>\n",
       "      <td>test2_gname</td>\n",
       "      <td>[5, 6, 13, 14, 9]</td>\n",
       "      <td>test3</td>\n",
       "      <td>[6, 7, 10, 12, 8, 9]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             tname        gid        gname               path    tid  \\\n",
       "test1  test1_tname  test1_gid  test1_gname    [0, 1, 2, 3, 4]  test1   \n",
       "test2  test2_tname  test2_gid  test2_gname    [5, 6, 7, 8, 9]  test2   \n",
       "test4  test4_tname  test4_gid  test4_gname               [10]  test4   \n",
       "test5  test5_tname  test2_gid  test2_gname        [5, 11, 12]  test5   \n",
       "test3  test3_tname  test2_gid  test2_gname  [5, 6, 13, 14, 9]  test3   \n",
       "\n",
       "                   loc_path  \n",
       "test1    [0, 1, 4, 5, 2, 3]  \n",
       "test2  [6, 7, 10, 11, 8, 9]  \n",
       "test4                [9, 8]  \n",
       "test5         [6, 7, 11, 8]  \n",
       "test3  [6, 7, 10, 12, 8, 9]  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.t_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "accessory-reality",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 1, 4, 5, 2, 3]\n"
     ]
    }
   ],
   "source": [
    "path = [0, 1, 2, 3, 4]\n",
    "start_v1 = [sg.edge_df.loc[path[0], 'v1']]\n",
    "v2s = sg.edge_df.loc[path, 'v2'].tolist()\n",
    "loc_path = start_v1+v2s\n",
    "\n",
    "print(loc_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "executed-iraqi",
   "metadata": {},
   "outputs": [],
   "source": [
    "sg['loc_path']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "veterinary-fellow",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding transcriptome to the SwanGraph\n",
      "\n",
      "Adding annotation to the SwanGraph\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "sg.add_transcriptome('../new_testing/files/test_novel.gtf')\n",
    "sg.add_annotation('../new_testing/files/test_known.gtf')\n",
    "# sg.add_transcriptome('../new_testing/files/test_novel.gtf')\n",
    "\n",
    "edge_df = sg.add_edge_coords()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "english-browse",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tname</th>\n",
       "      <th>gid</th>\n",
       "      <th>gname</th>\n",
       "      <th>path</th>\n",
       "      <th>tid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>test4</th>\n",
       "      <td>test4_tname</td>\n",
       "      <td>test2_gid</td>\n",
       "      <td>test2_gname</td>\n",
       "      <td>[0, 1, 2, 3, 4]</td>\n",
       "      <td>test4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test5</th>\n",
       "      <td>test5_tname</td>\n",
       "      <td>test2_gid</td>\n",
       "      <td>test2_gname</td>\n",
       "      <td>[0, 5, 6]</td>\n",
       "      <td>test5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test3</th>\n",
       "      <td>test3_tname</td>\n",
       "      <td>test2_gid</td>\n",
       "      <td>test2_gname</td>\n",
       "      <td>[0, 1, 7, 8, 4]</td>\n",
       "      <td>test3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test1</th>\n",
       "      <td>test1_tname</td>\n",
       "      <td>test1_gid</td>\n",
       "      <td>test1_gname</td>\n",
       "      <td>[9, 10, 11, 12, 13]</td>\n",
       "      <td>test1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test2</th>\n",
       "      <td>test2_tname</td>\n",
       "      <td>test2_gid</td>\n",
       "      <td>test2_gname</td>\n",
       "      <td>[0, 1, 7, 8, 4]</td>\n",
       "      <td>test2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             tname        gid        gname                 path    tid\n",
       "test4  test4_tname  test2_gid  test2_gname      [0, 1, 2, 3, 4]  test4\n",
       "test5  test5_tname  test2_gid  test2_gname            [0, 5, 6]  test5\n",
       "test3  test3_tname  test2_gid  test2_gname      [0, 1, 7, 8, 4]  test3\n",
       "test1  test1_tname  test1_gid  test1_gname  [9, 10, 11, 12, 13]  test1\n",
       "test2  test2_tname  test2_gid  test2_gname      [0, 1, 7, 8, 4]  test2"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.t_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "surprised-resort",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>v1</th>\n",
       "      <th>v2</th>\n",
       "      <th>strand</th>\n",
       "      <th>edge_type</th>\n",
       "      <th>edge_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>v1_coord</th>\n",
       "      <th>vertex_id_x</th>\n",
       "      <th>v2_coord</th>\n",
       "      <th>vertex_id_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>0</td>\n",
       "      <td>chr2</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>80</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>chr2</td>\n",
       "      <td>80</td>\n",
       "      <td>1</td>\n",
       "      <td>75</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>2</td>\n",
       "      <td>chr2</td>\n",
       "      <td>75</td>\n",
       "      <td>4</td>\n",
       "      <td>65</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>intron</td>\n",
       "      <td>3</td>\n",
       "      <td>chr2</td>\n",
       "      <td>65</td>\n",
       "      <td>5</td>\n",
       "      <td>50</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>4</td>\n",
       "      <td>chr2</td>\n",
       "      <td>50</td>\n",
       "      <td>2</td>\n",
       "      <td>45</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>-</td>\n",
       "      <td>intron</td>\n",
       "      <td>5</td>\n",
       "      <td>chr2</td>\n",
       "      <td>80</td>\n",
       "      <td>1</td>\n",
       "      <td>60</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>6</td>\n",
       "      <td>chr2</td>\n",
       "      <td>60</td>\n",
       "      <td>6</td>\n",
       "      <td>50</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>7</td>\n",
       "      <td>chr2</td>\n",
       "      <td>75</td>\n",
       "      <td>4</td>\n",
       "      <td>60</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>intron</td>\n",
       "      <td>8</td>\n",
       "      <td>chr2</td>\n",
       "      <td>60</td>\n",
       "      <td>6</td>\n",
       "      <td>50</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>9</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>20</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>10</td>\n",
       "      <td>chr1</td>\n",
       "      <td>20</td>\n",
       "      <td>8</td>\n",
       "      <td>25</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>11</td>\n",
       "      <td>chr1</td>\n",
       "      <td>25</td>\n",
       "      <td>11</td>\n",
       "      <td>30</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>12</td>\n",
       "      <td>chr1</td>\n",
       "      <td>30</td>\n",
       "      <td>12</td>\n",
       "      <td>35</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>13</td>\n",
       "      <td>chr1</td>\n",
       "      <td>35</td>\n",
       "      <td>9</td>\n",
       "      <td>40</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>14</td>\n",
       "      <td>chr2</td>\n",
       "      <td>45</td>\n",
       "      <td>3</td>\n",
       "      <td>50</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    v1  v2 strand edge_type edge_id chrom v1_coord vertex_id_x v2_coord  \\\n",
       "0    0   1      -      exon       0  chr2      100           0       80   \n",
       "1    1   4      -    intron       1  chr2       80           1       75   \n",
       "2    4   5      -      exon       2  chr2       75           4       65   \n",
       "3    5   2      -    intron       3  chr2       65           5       50   \n",
       "4    2   3      -      exon       4  chr2       50           2       45   \n",
       "5    1   6      -    intron       5  chr2       80           1       60   \n",
       "6    6   2      -      exon       6  chr2       60           6       50   \n",
       "7    4   6      -      exon       7  chr2       75           4       60   \n",
       "8    6   2      -    intron       8  chr2       60           6       50   \n",
       "9    7   8      +      exon       9  chr1        1           7       20   \n",
       "10   8  11      +    intron      10  chr1       20           8       25   \n",
       "11  11  12      +      exon      11  chr1       25          11       30   \n",
       "12  12   9      +    intron      12  chr1       30          12       35   \n",
       "13   9  10      +      exon      13  chr1       35           9       40   \n",
       "14   3   2      +      exon      14  chr2       45           3       50   \n",
       "\n",
       "   vertex_id_y  \n",
       "0            1  \n",
       "1            4  \n",
       "2            5  \n",
       "3            2  \n",
       "4            3  \n",
       "5            6  \n",
       "6            2  \n",
       "7            6  \n",
       "8            2  \n",
       "9            8  \n",
       "10          11  \n",
       "11          12  \n",
       "12           9  \n",
       "13          10  \n",
       "14           2  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edge_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "north-lecture",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>coord</th>\n",
       "      <th>vertex_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>80</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>50</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>45</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2</td>\n",
       "      <td>75</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr2</td>\n",
       "      <td>65</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr2</td>\n",
       "      <td>60</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>20</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>35</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>40</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>25</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom coord vertex_id\n",
       "0   chr2   100         0\n",
       "1   chr2    80         1\n",
       "2   chr2    50         2\n",
       "3   chr2    45         3\n",
       "4   chr2    75         4\n",
       "5   chr2    65         5\n",
       "6   chr2    60         6\n",
       "7   chr1     1         7\n",
       "8   chr1    20         8\n",
       "9   chr1    35         9\n",
       "10  chr1    40        10\n",
       "11  chr1    25        11\n",
       "12  chr1    30        12"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.loc_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "meaningful-darkness",
   "metadata": {},
   "source": [
    "#### test - create_dfs - swangraph with data, duplicate transcript being added\n",
    "* also tests that old data (novelty in this case) is not overwritten"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "center-basketball",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           chrom  coord  vertex_id\n",
      "vertex_id                         \n",
      "0           True   True       True\n",
      "1           True   True       True\n",
      "2           True   True       True\n",
      "3           True   True       True\n",
      "4           True   True       True\n",
      "5           True   True       True\n",
      "6           True   True       True\n",
      "7           True   True       True\n",
      "8           True   True       True\n",
      "9           True   True       True\n",
      "10          True   True       True\n",
      "11          True   True       True\n",
      "12          True   True       True\n",
      "           v1    v2  strand  edge_type  edge_id\n",
      "edge_id                                        \n",
      "0        True  True    True       True     True\n",
      "1        True  True    True       True     True\n",
      "2        True  True    True       True     True\n",
      "3        True  True    True       True     True\n",
      "4        True  True    True       True     True\n",
      "5        True  True    True       True     True\n",
      "6        True  True    True       True     True\n",
      "7        True  True    True       True     True\n",
      "8        True  True    True       True     True\n",
      "9        True  True    True       True     True\n",
      "10       True  True    True       True     True\n",
      "11       True  True    True       True     True\n",
      "12       True  True    True       True     True\n",
      "13       True  True    True       True     True\n",
      "14       True  True    True       True     True\n",
      "       tname   gid  gname  path   tid  novelty\n",
      "tid                                           \n",
      "test1   True  True   True  True  True    False\n",
      "test2   True  True   True  True  True     True\n",
      "test3   True  True   True  True  True    False\n",
      "test4   True  True   True  True  True    False\n",
      "test5   True  True   True  True  True    False\n"
     ]
    }
   ],
   "source": [
    "transcripts, exons = get_test_transcript_exon_dicts()\n",
    "sg = swan.SwanGraph()\n",
    "\n",
    "# add dummy data\n",
    "\n",
    "# loc_df - format\n",
    "loc_df = pd.read_csv('data/test_preexisting_loc_df.tsv', sep='\\t')\n",
    "loc_df.set_index('vertex_id_index', inplace=True)\n",
    "loc_df.index.name = 'vertex_id'\n",
    "\n",
    "# edge_df - format and remove the columns that are there \n",
    "# just for debugging purposes\n",
    "edge_df = pd.read_csv('data/test_preexisting_edge_df.tsv', sep='\\t')\n",
    "edge_df.drop(['v2_coord', 'v1_coord'], axis=1, inplace=True)\n",
    "edge_df.set_index('edge_id_index', inplace=True)\n",
    "edge_df.index.name = 'edge_id'\n",
    "\n",
    "# t_df - remove and reformat columns that are there for debugging\n",
    "t_df = pd.read_csv('data/test_preexisting_t_df.tsv', sep='\\t')\n",
    "t_df.set_index('tid_index', inplace=True)\n",
    "t_df.index.name = 'tid'\n",
    "t_df.drop(['loc_path'], axis=1, inplace=True)\n",
    "t_df.rename({'edge_path': 'path'}, axis=1, inplace=True)\n",
    "t_df['path'] = t_df.apply(lambda x: [int(n) for n in x.path.split(',')], axis=1)\n",
    "t_df = t_df[t_df.columns]\n",
    "\n",
    "sg.loc_df = loc_df\n",
    "sg.edge_df = edge_df\n",
    "sg.t_df = t_df\n",
    "\n",
    "loc_df, edge_df, t_df = sg.create_dfs(transcripts, exons, False)\n",
    "\n",
    "\n",
    "# control data\n",
    "\n",
    "# loc_df - format\n",
    "ctrl_loc_df = pd.read_csv('data/test_preexisting_result_loc_df.tsv', sep='\\t')\n",
    "ctrl_loc_df.set_index('vertex_id_index', inplace=True)\n",
    "ctrl_loc_df.index.name = 'vertex_id'\n",
    "\n",
    "# edge_df - format and remove the columns that are there \n",
    "# just for debugging purposes\n",
    "ctrl_edge_df = pd.read_csv('data/test_preexisting_result_edge_df.tsv', sep='\\t')\n",
    "ctrl_edge_df.drop(['v2_coord', 'v1_coord'], axis=1, inplace=True)\n",
    "ctrl_edge_df.set_index('edge_id_index', inplace=True)\n",
    "ctrl_edge_df.index.name = 'edge_id'\n",
    "\n",
    "# t_df - remove and reformat columns that are there for debugging\n",
    "ctrl_t_df = pd.read_csv('data/test_preexisting_result_t_df.tsv', sep='\\t')\n",
    "ctrl_t_df.set_index('tid_index', inplace=True)\n",
    "ctrl_t_df.index.name = 'tid'\n",
    "ctrl_t_df.drop(['loc_path'], axis=1, inplace=True)\n",
    "ctrl_t_df.rename({'edge_path': 'path'}, axis=1, inplace=True)\n",
    "ctrl_t_df['path'] = ctrl_t_df.apply(lambda x: [int(n) for n in x.path.split(',')], axis=1)\n",
    "ctrl_t_df = ctrl_t_df[t_df.columns]\n",
    "\n",
    "# remove novelty for entries that are new \n",
    "new_tids = ['test1', 'test3', 'test4', 'test5']\n",
    "ctrl_t_df.loc[ctrl_t_df.tid.isin(new_tids), 'novelty'] = 'Undefined'\n",
    "\n",
    "# sort all values by their IDs\n",
    "loc_df.sort_index(inplace=True)\n",
    "edge_df.sort_index(inplace=True)\n",
    "t_df.sort_index(inplace=True)\n",
    "ctrl_loc_df.sort_index(inplace=True)\n",
    "ctrl_edge_df.sort_index(inplace=True)\n",
    "ctrl_t_df.sort_index(inplace=True)\n",
    "\n",
    "# bc I'm dumb, convert to integers\n",
    "loc_df.coord = loc_df.coord.astype('int')\n",
    "\n",
    "# print('test')\n",
    "# print(loc_df)\n",
    "# print('control')\n",
    "# print(ctrl_loc_df)\n",
    "print(ctrl_loc_df == loc_df)\n",
    "\n",
    "# print('test')\n",
    "# print(edge_df)\n",
    "# print('control')\n",
    "# print(ctrl_edge_df)\n",
    "print(ctrl_edge_df == edge_df)\n",
    "\n",
    "# print('test')\n",
    "# print(t_df)\n",
    "# print('control')\n",
    "# print(ctrl_t_df)\n",
    "print(ctrl_t_df == t_df)\n",
    "\n",
    "# test_paths = t_df.path.tolist()\n",
    "# ctrl_paths = ctrl_t_df.path.tolist()\n",
    "# for t, c in zip(test_paths, ctrl_paths):\n",
    "#     pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "legislative-stevens",
   "metadata": {},
   "source": [
    "#### test - create_dfs - swangraph with data, mixed novelty content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fifteen-revolution",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           chrom  coord  vertex_id\n",
      "vertex_id                         \n",
      "0           True   True       True\n",
      "1           True   True       True\n",
      "2           True   True       True\n",
      "3           True   True       True\n",
      "4           True   True       True\n",
      "5           True   True       True\n",
      "6           True   True       True\n",
      "7           True   True       True\n",
      "8           True   True       True\n",
      "9           True   True       True\n",
      "10          True   True       True\n",
      "11          True   True       True\n",
      "12          True   True       True\n",
      "           v1    v2  strand  edge_type  edge_id\n",
      "edge_id                                        \n",
      "0        True  True    True       True     True\n",
      "1        True  True    True       True     True\n",
      "2        True  True    True       True     True\n",
      "3        True  True    True       True     True\n",
      "4        True  True    True       True     True\n",
      "5        True  True    True       True     True\n",
      "6        True  True    True       True     True\n",
      "7        True  True    True       True     True\n",
      "8        True  True    True       True     True\n",
      "9        True  True    True       True     True\n",
      "10       True  True    True       True     True\n",
      "11       True  True    True       True     True\n",
      "12       True  True    True       True     True\n",
      "13       True  True    True       True     True\n",
      "14       True  True    True       True     True\n",
      "test\n",
      "             tname        gid        gname               path    tid  \\\n",
      "tid                                                                    \n",
      "test1  test1_tname  test1_gid  test1_gname    [5, 6, 7, 8, 9]  test1   \n",
      "test2  test2_tname  test2_gid  test2_gname    [0, 1, 2, 3, 4]  test2   \n",
      "test3  test3_tname  test2_gid  test2_gname  [0, 1, 10, 11, 4]  test3   \n",
      "test4  test4_tname  test4_gid  test4_gname               [12]  test4   \n",
      "test5  test5_tname  test2_gid  test2_gname        [0, 13, 14]  test5   \n",
      "\n",
      "         novelty  \n",
      "tid               \n",
      "test1  Undefined  \n",
      "test2      Known  \n",
      "test3  Undefined  \n",
      "test4  Undefined  \n",
      "test5  Undefined  \n",
      "control\n",
      "             tname        gid        gname               path    tid  \\\n",
      "tid                                                                    \n",
      "test1  test1_tname  test1_gid  test1_gname    [5, 6, 7, 8, 9]  test1   \n",
      "test2  test2_tname  test2_gid  test2_gname    [0, 1, 2, 3, 4]  test2   \n",
      "test3  test3_tname  test2_gid  test2_gname  [0, 1, 10, 11, 4]  test3   \n",
      "test4  test4_tname  test4_gid  test4_gname               [12]  test4   \n",
      "test5  test5_tname  test2_gid  test2_gname        [0, 13, 14]  test5   \n",
      "\n",
      "         novelty  \n",
      "tid               \n",
      "test1  Undefined  \n",
      "test2      Known  \n",
      "test3  Undefined  \n",
      "test4  Undefined  \n",
      "test5  Undefined  \n",
      "       tname   gid  gname  path   tid  novelty\n",
      "tid                                           \n",
      "test1   True  True   True  True  True     True\n",
      "test2   True  True   True  True  True     True\n",
      "test3   True  True   True  True  True     True\n",
      "test4   True  True   True  True  True     True\n",
      "test5   True  True   True  True  True     True\n"
     ]
    }
   ],
   "source": [
    "transcripts, exons = get_test_transcript_exon_dicts()\n",
    "\n",
    "# to do - remove transcript that's already there\n",
    "sg = swan.SwanGraph()\n",
    "\n",
    "# add dummy data\n",
    "\n",
    "# loc_df - format\n",
    "loc_df = pd.read_csv('data/test_preexisting_loc_df.tsv', sep='\\t')\n",
    "loc_df.set_index('vertex_id_index', inplace=True)\n",
    "loc_df.index.name = 'vertex_id'\n",
    "\n",
    "# edge_df - format and remove the columns that are there \n",
    "# just for debugging purposes\n",
    "edge_df = pd.read_csv('data/test_preexisting_edge_df.tsv', sep='\\t')\n",
    "edge_df.drop(['v2_coord', 'v1_coord'], axis=1, inplace=True)\n",
    "edge_df.set_index('edge_id_index', inplace=True)\n",
    "edge_df.index.name = 'edge_id'\n",
    "\n",
    "# t_df - remove and reformat columns that are there for debugging\n",
    "t_df = pd.read_csv('data/test_preexisting_t_df.tsv', sep='\\t')\n",
    "t_df.set_index('tid_index', inplace=True)\n",
    "t_df.index.name = 'tid'\n",
    "t_df.drop(['loc_path'], axis=1, inplace=True)\n",
    "t_df.rename({'edge_path': 'path'}, axis=1, inplace=True)\n",
    "t_df['path'] = t_df.apply(lambda x: [int(n) for n in x.path.split(',')], axis=1)\n",
    "t_df = t_df[t_df.columns]\n",
    "\n",
    "sg.loc_df = loc_df\n",
    "sg.edge_df = edge_df\n",
    "sg.t_df = t_df\n",
    "\n",
    "loc_df, edge_df, t_df = sg.create_dfs(transcripts, exons, False)\n",
    "\n",
    "\n",
    "# control data\n",
    "\n",
    "# loc_df - format\n",
    "ctrl_loc_df = pd.read_csv('data/test_preexisting_result_loc_df.tsv', sep='\\t')\n",
    "ctrl_loc_df.set_index('vertex_id_index', inplace=True)\n",
    "ctrl_loc_df.index.name = 'vertex_id'\n",
    "\n",
    "# edge_df - format and remove the columns that are there \n",
    "# just for debugging purposes\n",
    "ctrl_edge_df = pd.read_csv('data/test_preexisting_result_edge_df.tsv', sep='\\t')\n",
    "ctrl_edge_df.drop(['v2_coord', 'v1_coord'], axis=1, inplace=True)\n",
    "ctrl_edge_df.set_index('edge_id_index', inplace=True)\n",
    "ctrl_edge_df.index.name = 'edge_id'\n",
    "\n",
    "# t_df - remove and reformat columns that are there for debugging\n",
    "ctrl_t_df = pd.read_csv('data/test_preexisting_result_t_df.tsv', sep='\\t')\n",
    "ctrl_t_df.set_index('tid_index', inplace=True)\n",
    "ctrl_t_df.index.name = 'tid'\n",
    "ctrl_t_df.drop(['loc_path'], axis=1, inplace=True)\n",
    "ctrl_t_df.rename({'edge_path': 'path'}, axis=1, inplace=True)\n",
    "ctrl_t_df['path'] = ctrl_t_df.apply(lambda x: [int(n) for n in x.path.split(',')], axis=1)\n",
    "ctrl_t_df = ctrl_t_df[t_df.columns]\n",
    "\n",
    "# remove novelty for entries that are new \n",
    "new_tids = ['test1', 'test3', 'test4', 'test5']\n",
    "ctrl_t_df.loc[ctrl_t_df.tid.isin(new_tids), 'novelty'] = 'Undefined'\n",
    "\n",
    "# sort all values by their IDs\n",
    "loc_df.sort_index(inplace=True)\n",
    "edge_df.sort_index(inplace=True)\n",
    "t_df.sort_index(inplace=True)\n",
    "ctrl_loc_df.sort_index(inplace=True)\n",
    "ctrl_edge_df.sort_index(inplace=True)\n",
    "ctrl_t_df.sort_index(inplace=True)\n",
    "\n",
    "# bc I'm dumb, convert to integers\n",
    "loc_df.coord = loc_df.coord.astype('int')\n",
    "\n",
    "# print('test')\n",
    "# print(loc_df)\n",
    "# print('control')\n",
    "# print(ctrl_loc_df)\n",
    "print(ctrl_loc_df == loc_df)\n",
    "\n",
    "# print('test')\n",
    "# print(edge_df)\n",
    "# print('control')\n",
    "# print(ctrl_edge_df)\n",
    "print(ctrl_edge_df == edge_df)\n",
    "\n",
    "# print('test')\n",
    "# print(t_df)\n",
    "# print('control')\n",
    "# print(ctrl_t_df)\n",
    "print(ctrl_t_df == t_df)\n",
    "\n",
    "# test_paths = t_df.path.tolist()\n",
    "# ctrl_paths = ctrl_t_df.path.tolist()\n",
    "# for t, c in zip(test_paths, ctrl_paths):\n",
    "#     pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "entire-being",
   "metadata": {},
   "source": [
    "#### test - create_dfs - swangraph with data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "genetic-novel",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           chrom  coord  vertex_id\n",
      "vertex_id                         \n",
      "0           True   True       True\n",
      "1           True   True       True\n",
      "2           True   True       True\n",
      "3           True   True       True\n",
      "4           True   True       True\n",
      "5           True   True       True\n",
      "6           True   True       True\n",
      "7           True   True       True\n",
      "8           True   True       True\n",
      "9           True   True       True\n",
      "10          True   True       True\n",
      "11          True   True       True\n",
      "12          True   True       True\n",
      "           v1    v2  strand  edge_type  edge_id\n",
      "edge_id                                        \n",
      "0        True  True    True       True     True\n",
      "1        True  True    True       True     True\n",
      "2        True  True    True       True     True\n",
      "3        True  True    True       True     True\n",
      "4        True  True    True       True     True\n",
      "5        True  True    True       True     True\n",
      "6        True  True    True       True     True\n",
      "7        True  True    True       True     True\n",
      "8        True  True    True       True     True\n",
      "9        True  True    True       True     True\n",
      "10       True  True    True       True     True\n",
      "11       True  True    True       True     True\n",
      "12       True  True    True       True     True\n",
      "13       True  True    True       True     True\n",
      "14       True  True    True       True     True\n",
      "       tname   gid  gname  path   tid  novelty\n",
      "tid                                           \n",
      "test1   True  True   True  True  True     True\n",
      "test2   True  True   True  True  True     True\n",
      "test3   True  True   True  True  True     True\n",
      "test4   True  True   True  True  True     True\n",
      "test5   True  True   True  True  True     True\n"
     ]
    }
   ],
   "source": [
    "transcripts, exons = get_test_transcript_exon_dicts()\n",
    "\n",
    "# to do - remove transcript that's already there\n",
    "del transcripts['test2']\n",
    "sg = swan.SwanGraph()\n",
    "\n",
    "# add dummy data\n",
    "\n",
    "# loc_df - format\n",
    "loc_df = pd.read_csv('data/test_preexisting_loc_df.tsv', sep='\\t')\n",
    "loc_df.set_index('vertex_id_index', inplace=True)\n",
    "loc_df.index.name = 'vertex_id'\n",
    "\n",
    "# edge_df - format and remove the columns that are there \n",
    "# just for debugging purposes\n",
    "edge_df = pd.read_csv('data/test_preexisting_edge_df.tsv', sep='\\t')\n",
    "edge_df.drop(['v2_coord', 'v1_coord'], axis=1, inplace=True)\n",
    "edge_df.set_index('edge_id_index', inplace=True)\n",
    "edge_df.index.name = 'edge_id'\n",
    "\n",
    "# t_df - remove and reformat columns that are there for debugging\n",
    "t_df = pd.read_csv('data/test_preexisting_t_df.tsv', sep='\\t')\n",
    "t_df.set_index('tid_index', inplace=True)\n",
    "t_df.index.name = 'tid'\n",
    "t_df.drop(['loc_path'], axis=1, inplace=True)\n",
    "t_df.rename({'edge_path': 'path'}, axis=1, inplace=True)\n",
    "t_df['path'] = t_df.apply(lambda x: [int(n) for n in x.path.split(',')], axis=1)\n",
    "t_df = t_df[t_df.columns]\n",
    "\n",
    "sg.loc_df = loc_df\n",
    "sg.edge_df = edge_df\n",
    "sg.t_df = t_df\n",
    "\n",
    "loc_df, edge_df, t_df = sg.create_dfs(transcripts, exons, True)\n",
    "\n",
    "\n",
    "# control data\n",
    "\n",
    "# loc_df - format\n",
    "ctrl_loc_df = pd.read_csv('data/test_preexisting_result_loc_df.tsv', sep='\\t')\n",
    "ctrl_loc_df.set_index('vertex_id_index', inplace=True)\n",
    "ctrl_loc_df.index.name = 'vertex_id'\n",
    "\n",
    "# edge_df - format and remove the columns that are there \n",
    "# just for debugging purposes\n",
    "ctrl_edge_df = pd.read_csv('data/test_preexisting_result_edge_df.tsv', sep='\\t')\n",
    "ctrl_edge_df.drop(['v2_coord', 'v1_coord'], axis=1, inplace=True)\n",
    "ctrl_edge_df.set_index('edge_id_index', inplace=True)\n",
    "ctrl_edge_df.index.name = 'edge_id'\n",
    "\n",
    "# t_df - remove and reformat columns that are there for debugging\n",
    "ctrl_t_df = pd.read_csv('data/test_preexisting_result_t_df.tsv', sep='\\t')\n",
    "ctrl_t_df.set_index('tid_index', inplace=True)\n",
    "ctrl_t_df.index.name = 'tid'\n",
    "ctrl_t_df.drop(['loc_path'], axis=1, inplace=True)\n",
    "ctrl_t_df.rename({'edge_path': 'path'}, axis=1, inplace=True)\n",
    "ctrl_t_df['path'] = ctrl_t_df.apply(lambda x: [int(n) for n in x.path.split(',')], axis=1)\n",
    "ctrl_t_df = ctrl_t_df[t_df.columns]\n",
    "\n",
    "# sort all values by their IDs\n",
    "loc_df.sort_index(inplace=True)\n",
    "edge_df.sort_index(inplace=True)\n",
    "t_df.sort_index(inplace=True)\n",
    "ctrl_loc_df.sort_index(inplace=True)\n",
    "ctrl_edge_df.sort_index(inplace=True)\n",
    "ctrl_t_df.sort_index(inplace=True)\n",
    "\n",
    "# and order columns the same way\n",
    "ctrl_loc_df = ctrl_loc_df[loc_df.columns]\n",
    "ctrl_edge_df = ctrl_edge_df[edge_df.columns]\n",
    "ctrl_t_df = ctrl_t_df[t_df.columns]\n",
    "\n",
    "# bc I'm dumb, convert to integers\n",
    "loc_df.coord = loc_df.coord.astype('int')\n",
    "\n",
    "# print('test')\n",
    "# print(loc_df)\n",
    "# print('control')\n",
    "# print(ctrl_loc_df)\n",
    "print(ctrl_loc_df == loc_df)\n",
    "\n",
    "# print('test')\n",
    "# print(edge_df)\n",
    "# print('control')\n",
    "# print(ctrl_edge_df)\n",
    "print(ctrl_edge_df == edge_df)\n",
    "\n",
    "# print('test')\n",
    "# print(t_df)\n",
    "# print('control')\n",
    "# print(ctrl_t_df)\n",
    "print(ctrl_t_df == t_df)\n",
    "\n",
    "# test_paths = t_df.path.tolist()\n",
    "# ctrl_paths = ctrl_t_df.path.tolist()\n",
    "# for t, c in zip(test_paths, ctrl_paths):\n",
    "#     pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "automotive-desire",
   "metadata": {},
   "source": [
    "#### test - create_dfs - from_talon = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "needed-headset",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           chrom  coord  vertex_id\n",
      "vertex_id                         \n",
      "0           True   True       True\n",
      "1           True   True       True\n",
      "2           True   True       True\n",
      "3           True   True       True\n",
      "4           True   True       True\n",
      "5           True   True       True\n",
      "6           True   True       True\n",
      "7           True   True       True\n",
      "8           True   True       True\n",
      "9           True   True       True\n",
      "10          True   True       True\n",
      "11          True   True       True\n",
      "12          True   True       True\n",
      "           v1    v2  strand  edge_type  edge_id\n",
      "edge_id                                        \n",
      "0        True  True    True       True     True\n",
      "1        True  True    True       True     True\n",
      "2        True  True    True       True     True\n",
      "3        True  True    True       True     True\n",
      "4        True  True    True       True     True\n",
      "5        True  True    True       True     True\n",
      "6        True  True    True       True     True\n",
      "7        True  True    True       True     True\n",
      "8        True  True    True       True     True\n",
      "9        True  True    True       True     True\n",
      "10       True  True    True       True     True\n",
      "11       True  True    True       True     True\n",
      "12       True  True    True       True     True\n",
      "13       True  True    True       True     True\n",
      "14       True  True    True       True     True\n",
      "        tid  tname   gid  gname  path  novelty\n",
      "tid                                           \n",
      "test1  True   True  True   True  True     True\n",
      "test2  True   True  True   True  True     True\n",
      "test3  True   True  True   True  True     True\n",
      "test4  True   True  True   True  True     True\n",
      "test5  True   True  True   True  True     True\n"
     ]
    }
   ],
   "source": [
    "transcripts, exons = get_test_transcript_exon_dicts()\n",
    "sg = swan.SwanGraph()\n",
    "\n",
    "loc_df, edge_df, t_df = sg.create_dfs(transcripts, exons, True)\n",
    "\n",
    "ctrl_loc_df = pd.read_csv('data/test_loc_df.tsv', sep='\\t')\n",
    "ctrl_loc_df.set_index('vertex_id_index', inplace=True)\n",
    "ctrl_loc_df.index.name = 'vertex_id'\n",
    "\n",
    "# remove the columns that are there just for debugging purposes\n",
    "ctrl_edge_df = pd.read_csv('data/test_edge_df.tsv', sep='\\t')\n",
    "ctrl_edge_df.drop(['v2_coord', 'v1_coord'], axis=1, inplace=True)\n",
    "ctrl_edge_df.set_index('edge_id_index', inplace=True)\n",
    "ctrl_edge_df.index.name = 'edge_id'\n",
    "\n",
    "# again, remove and reformat columns that are there for debugging\n",
    "ctrl_t_df = pd.read_csv('data/test_t_df.tsv', sep='\\t')\n",
    "ctrl_t_df.set_index('tid_index', inplace=True)\n",
    "ctrl_t_df.index.name = 'tid'\n",
    "ctrl_t_df.drop(['loc_path'], axis=1, inplace=True)\n",
    "ctrl_t_df.rename({'edge_path': 'path'}, axis=1, inplace=True)\n",
    "ctrl_t_df['path'] = ctrl_t_df.apply(lambda x: [int(n) for n in x.path.split(',')], axis=1)\n",
    "ctrl_t_df = ctrl_t_df[t_df.columns]\n",
    "\n",
    "# sort all values by their IDs\n",
    "loc_df.sort_index(inplace=True)\n",
    "edge_df.sort_index(inplace=True)\n",
    "t_df.sort_index(inplace=True)\n",
    "ctrl_loc_df.sort_index(inplace=True)\n",
    "ctrl_edge_df.sort_index(inplace=True)\n",
    "ctrl_t_df.sort_index(inplace=True)\n",
    "\n",
    "# and order columns the same way\n",
    "ctrl_loc_df = ctrl_loc_df[loc_df.columns]\n",
    "ctrl_edge_df = ctrl_edge_df[edge_df.columns]\n",
    "ctrl_t_df = ctrl_t_df[t_df.columns]\n",
    "\n",
    "\n",
    "# bc I'm dumb, convert to integers\n",
    "loc_df.coord = loc_df.coord.astype('int')\n",
    "\n",
    "# print('test')\n",
    "# print(loc_df)\n",
    "# print('control')\n",
    "# print(ctrl_loc_df)\n",
    "print(ctrl_loc_df == loc_df)\n",
    "\n",
    "# print('test')\n",
    "# print(edge_df)\n",
    "# print('control')\n",
    "# print(ctrl_edge_df)\n",
    "print(ctrl_edge_df == edge_df)\n",
    "\n",
    "# print('test')\n",
    "# print(t_df)\n",
    "# print('control')\n",
    "# print(ctrl_t_df)\n",
    "print(ctrl_t_df == t_df)\n",
    "\n",
    "# test_paths = t_df.path.tolist()\n",
    "# ctrl_paths = ctrl_t_df.path.tolist()\n",
    "# for t, c in zip(test_paths, ctrl_paths):\n",
    "#     pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "careful-combination",
   "metadata": {},
   "source": [
    "#### test - create_dfs - empty swangraph\n",
    "* make sure that the emtpy dict -> df -> dict conversion doesn't screw up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "center-implement",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           chrom  coord  vertex_id\n",
      "vertex_id                         \n",
      "0           True   True       True\n",
      "1           True   True       True\n",
      "2           True   True       True\n",
      "3           True   True       True\n",
      "4           True   True       True\n",
      "5           True   True       True\n",
      "6           True   True       True\n",
      "7           True   True       True\n",
      "8           True   True       True\n",
      "9           True   True       True\n",
      "10          True   True       True\n",
      "11          True   True       True\n",
      "12          True   True       True\n",
      "           v1    v2  strand  edge_type  edge_id\n",
      "edge_id                                        \n",
      "0        True  True    True       True     True\n",
      "1        True  True    True       True     True\n",
      "2        True  True    True       True     True\n",
      "3        True  True    True       True     True\n",
      "4        True  True    True       True     True\n",
      "5        True  True    True       True     True\n",
      "6        True  True    True       True     True\n",
      "7        True  True    True       True     True\n",
      "8        True  True    True       True     True\n",
      "9        True  True    True       True     True\n",
      "10       True  True    True       True     True\n",
      "11       True  True    True       True     True\n",
      "12       True  True    True       True     True\n",
      "13       True  True    True       True     True\n",
      "14       True  True    True       True     True\n",
      "       tname   gid  gname  path   tid\n",
      "tid                                  \n",
      "test1   True  True   True  True  True\n",
      "test2   True  True   True  True  True\n",
      "test3   True  True   True  True  True\n",
      "test4   True  True   True  True  True\n",
      "test5   True  True   True  True  True\n"
     ]
    }
   ],
   "source": [
    "transcripts, exons = get_test_transcript_exon_dicts()\n",
    "sg = swan.SwanGraph()\n",
    "\n",
    "loc_df, edge_df, t_df = sg.create_dfs(transcripts, exons, False)\n",
    "\n",
    "ctrl_loc_df = pd.read_csv('data/test_loc_df.tsv', sep='\\t')\n",
    "ctrl_loc_df.set_index('vertex_id_index', inplace=True)\n",
    "ctrl_loc_df.index.name = 'vertex_id'\n",
    "\n",
    "# remove the columns that are there just for debugging purposes\n",
    "ctrl_edge_df = pd.read_csv('data/test_edge_df.tsv', sep='\\t')\n",
    "ctrl_edge_df.drop(['v2_coord', 'v1_coord'], axis=1, inplace=True)\n",
    "ctrl_edge_df.set_index('edge_id_index', inplace=True)\n",
    "ctrl_edge_df.index.name = 'edge_id'\n",
    "\n",
    "# again, remove and reformat columns that are there for debugging\n",
    "ctrl_t_df = pd.read_csv('data/test_t_df.tsv', sep='\\t')\n",
    "ctrl_t_df.set_index('tid_index', inplace=True)\n",
    "ctrl_t_df.index.name = 'tid'\n",
    "ctrl_t_df.drop(['loc_path', 'novelty'], axis=1, inplace=True)\n",
    "ctrl_t_df.rename({'edge_path': 'path'}, axis=1, inplace=True)\n",
    "ctrl_t_df['path'] = ctrl_t_df.apply(lambda x: [int(n) for n in x.path.split(',')], axis=1)\n",
    "\n",
    "# sort all values by their IDs\n",
    "loc_df.sort_index(inplace=True)\n",
    "edge_df.sort_index(inplace=True)\n",
    "t_df.sort_index(inplace=True)\n",
    "ctrl_loc_df.sort_index(inplace=True)\n",
    "ctrl_edge_df.sort_index(inplace=True)\n",
    "ctrl_t_df.sort_index(inplace=True)\n",
    "\n",
    "# and order columns the same way\n",
    "ctrl_loc_df = ctrl_loc_df[loc_df.columns]\n",
    "ctrl_edge_df = ctrl_edge_df[edge_df.columns]\n",
    "ctrl_t_df = ctrl_t_df[t_df.columns]\n",
    "\n",
    "# bc I'm dumb, convert to integers\n",
    "loc_df.coord = loc_df.coord.astype('int')\n",
    "\n",
    "# print('test')\n",
    "# print(loc_df)\n",
    "# print('control')\n",
    "# print(ctrl_loc_df)\n",
    "print(ctrl_loc_df == loc_df)\n",
    "\n",
    "# print('test')\n",
    "# print(edge_df)\n",
    "# print('control')\n",
    "# print(ctrl_edge_df)\n",
    "print(ctrl_edge_df == edge_df)\n",
    "\n",
    "# print('test')\n",
    "# print(t_df)\n",
    "# print('control')\n",
    "# print(ctrl_t_df)\n",
    "print(ctrl_t_df == t_df)\n",
    "\n",
    "# test_paths = t_df.path.tolist()\n",
    "# ctrl_paths = ctrl_t_df.path.tolist()\n",
    "# for t, c in zip(test_paths, ctrl_paths):\n",
    "#     pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bound-newspaper",
   "metadata": {},
   "source": [
    "#### test - create_transcript_edge_dict - transcripts already exist in swangraph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "incoming-worcester",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "honest-revision",
   "metadata": {},
   "source": [
    "#### test - create_transcript_edge_dict - edges already exist in swangraph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "sublime-malta",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "transcripts, exons = get_test_transcript_exon_dicts()\n",
    "\n",
    "# add some dummy data\n",
    "sg = swan.SwanGraph()\n",
    "\n",
    "data = [[0, 'chr1', 1],\n",
    "        [1, 'chr2', 20], \n",
    "        [2, 'chr2', 100],\n",
    "        [3, 'chr2', 80]]\n",
    "columns = ['vertex_id', 'chrom', 'coord']\n",
    "sg.loc_df = pd.DataFrame(data=data, columns=columns)\n",
    "locs = sg.create_loc_dict(exons)\n",
    "\n",
    "data = [[0, 0, 1, '+', 'exon'],\n",
    "        [1, 2, 3, '-', 'exon']]\n",
    "columns = ['edge_id', 'v1', 'v2', 'strand', 'edge_type']\n",
    "sg.edge_df = pd.DataFrame(data=data, columns=columns)\n",
    "\n",
    "transcripts, edges = sg.create_transcript_edge_dicts(transcripts, exons, locs)\n",
    "\n",
    "# just compare the paths for the transcripts, which is the only \n",
    "# part modified by this function\n",
    "transcripts = dict([(key, item['path']) for key, item in transcripts.items()])\n",
    "ctrl_transcript_paths = {\n",
    "    'test1': [0,2,3,4,5],\n",
    "    'test2': [1,6,7,8,9],\n",
    "    'test3': [1,6,10,11,9],\n",
    "    'test4': [12],\n",
    "    'test5': [1,13,14]\n",
    "}\n",
    "\n",
    "print(transcripts == ctrl_transcript_paths)\n",
    "\n",
    "ctrl_edges = {\n",
    "    ('chr1', 1, 20, '+', 'exon'): {\n",
    "        'edge_id': 0,\n",
    "        'edge_type': 'exon',\n",
    "        'v1': 0,\n",
    "        'v2': 1\n",
    "    },\n",
    "    ('chr1', 20, 25, '+', 'intron'): {\n",
    "        'edge_id': 2,\n",
    "        'edge_type': 'intron',\n",
    "        'v1': 4,\n",
    "        'v2': 5\n",
    "    },\n",
    "    ('chr1', 25, 30, '+', 'exon'): {\n",
    "        'edge_id': 3,\n",
    "        'edge_type': 'exon', \n",
    "        'v1': 5,\n",
    "        'v2': 6\n",
    "    },\n",
    "    ('chr1', 30, 35, '+', 'intron'): {\n",
    "        'edge_id': 4,\n",
    "        'edge_type': 'intron', \n",
    "        'v1': 6,\n",
    "        'v2': 7\n",
    "    },\n",
    "    ('chr1', 35, 40, '+', 'exon'): {\n",
    "        'edge_id': 5,\n",
    "        'edge_type': 'exon', \n",
    "        'v1': 7,\n",
    "        'v2': 8\n",
    "    },\n",
    "    ('chr2', 100, 80, '-', 'exon'): {\n",
    "        'edge_id': 1,\n",
    "        'edge_type': 'exon', \n",
    "        'v1': 2,\n",
    "        'v2': 3\n",
    "    },\n",
    "    ('chr2', 80, 75, '-', 'intron'): {\n",
    "        'edge_id': 6,\n",
    "        'edge_type': 'intron', \n",
    "        'v1': 3,\n",
    "        'v2': 9\n",
    "    },\n",
    "    ('chr2', 75, 60, '-', 'exon'): {\n",
    "        'edge_id': 7,\n",
    "        'edge_type': 'exon' , \n",
    "        'v1': 9,\n",
    "        'v2': 10\n",
    "    },\n",
    "    ('chr2', 60, 50, '-', 'intron'): {\n",
    "        'edge_id': 8,\n",
    "        'edge_type': 'intron', \n",
    "        'v1': 10,\n",
    "        'v2': 11\n",
    "    },\n",
    "    ('chr2', 50, 45, '-', 'exon'): {\n",
    "        'edge_id': 9,\n",
    "        'edge_type': 'exon', \n",
    "        'v1': 11,\n",
    "        'v2': 12\n",
    "    },\n",
    "    ('chr2', 75, 65, '-', 'exon'): {\n",
    "        'edge_id': 10,\n",
    "        'edge_type': 'exon', \n",
    "        'v1': 9,\n",
    "        'v2': 13\n",
    "    },\n",
    "    ('chr2', 65, 50, '-', 'intron'): {\n",
    "        'edge_id': 11,\n",
    "        'edge_type': 'intron', \n",
    "        'v1': 13,\n",
    "        'v2': 11\n",
    "    },\n",
    "    ('chr2', 45, 50, '+', 'exon'): {\n",
    "        'edge_id': 12,\n",
    "        'edge_type': 'exon', \n",
    "        'v1': 12,\n",
    "        'v2': 11\n",
    "    },\n",
    "    ('chr2', 80, 60, '-', 'intron'): {\n",
    "        'edge_id': 13,\n",
    "        'edge_type': 'intron', \n",
    "        'v1': 3,\n",
    "        'v2': 10\n",
    "    },\n",
    "    ('chr2', 60, 50, '-', 'exon'): {\n",
    "        'edge_id': 14,\n",
    "        'edge_type': 'exon', \n",
    "        'v1': 10,\n",
    "        'v2': 11\n",
    "    }\n",
    "}\n",
    "\n",
    "print(edges == ctrl_edges)\n",
    "\n",
    "# print(locs)\n",
    "# for key, _ in edges.items():\n",
    "#     print()\n",
    "#     print(key)\n",
    "#     print('test item')\n",
    "#     print(edges[key])\n",
    "#     print('ctrl item')\n",
    "#     print(ctrl_edges[key])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "committed-portland",
   "metadata": {},
   "source": [
    "#### test - create_transcript_edge_dict - empty swangraph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "substantial-manufacturer",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "transcripts, exons = get_test_transcript_exon_dicts()\n",
    "\n",
    "sg = swan.SwanGraph()\n",
    "locs = sg.create_loc_dict(exons)\n",
    "\n",
    "transcripts, edges = sg.create_transcript_edge_dicts(transcripts, exons, locs)\n",
    "\n",
    "# just compare the paths for the transcripts, which is the only \n",
    "# part modified by this function\n",
    "transcripts = dict([(key, item['path']) for key, item in transcripts.items()])\n",
    "ctrl_transcript_paths = {\n",
    "    'test1': [0,1,2,3,4],\n",
    "    'test2': [5,6,7,8,9],\n",
    "    'test3': [5,6,10,11,9],\n",
    "    'test4': [12],\n",
    "    'test5': [5,13,14]\n",
    "}\n",
    "\n",
    "print(transcripts == ctrl_transcript_paths)\n",
    "\n",
    "ctrl_edges = {\n",
    "    ('chr1', 1, 20, '+', 'exon'): {\n",
    "        'edge_id': 0,\n",
    "        'edge_type': 'exon',\n",
    "        'v1': 0,\n",
    "        'v2': 1\n",
    "    },\n",
    "    ('chr1', 20, 25, '+', 'intron'): {\n",
    "        'edge_id': 1,\n",
    "        'edge_type': 'intron',\n",
    "        'v1': 1,\n",
    "        'v2': 2\n",
    "    },\n",
    "    ('chr1', 25, 30, '+', 'exon'): {\n",
    "        'edge_id': 2,\n",
    "        'edge_type': 'exon', \n",
    "        'v1': 2,\n",
    "        'v2': 3\n",
    "    },\n",
    "    ('chr1', 30, 35, '+', 'intron'): {\n",
    "        'edge_id': 3,\n",
    "        'edge_type': 'intron', \n",
    "        'v1': 3,\n",
    "        'v2': 4\n",
    "    },\n",
    "    ('chr1', 35, 40, '+', 'exon'): {\n",
    "        'edge_id': 4,\n",
    "        'edge_type': 'exon', \n",
    "        'v1': 4,\n",
    "        'v2': 5\n",
    "    },\n",
    "    ('chr2', 100, 80, '-', 'exon'): {\n",
    "        'edge_id': 5,\n",
    "        'edge_type': 'exon', \n",
    "        'v1': 6,\n",
    "        'v2': 7\n",
    "    },\n",
    "    ('chr2', 80, 75, '-', 'intron'): {\n",
    "        'edge_id': 6,\n",
    "        'edge_type': 'intron', \n",
    "        'v1': 7,\n",
    "        'v2': 8\n",
    "    },\n",
    "    ('chr2', 75, 60, '-', 'exon'): {\n",
    "        'edge_id': 7,\n",
    "        'edge_type': 'exon' , \n",
    "        'v1': 8,\n",
    "        'v2': 9\n",
    "    },\n",
    "    ('chr2', 60, 50, '-', 'intron'): {\n",
    "        'edge_id': 8,\n",
    "        'edge_type': 'intron', \n",
    "        'v1': 9,\n",
    "        'v2': 10\n",
    "    },\n",
    "    ('chr2', 50, 45, '-', 'exon'): {\n",
    "        'edge_id': 9,\n",
    "        'edge_type': 'exon', \n",
    "        'v1': 10,\n",
    "        'v2': 11\n",
    "    },\n",
    "    ('chr2', 75, 65, '-', 'exon'): {\n",
    "        'edge_id': 10,\n",
    "        'edge_type': 'exon', \n",
    "        'v1': 8,\n",
    "        'v2': 12\n",
    "    },\n",
    "    ('chr2', 65, 50, '-', 'intron'): {\n",
    "        'edge_id': 11,\n",
    "        'edge_type': 'intron', \n",
    "        'v1': 12,\n",
    "        'v2': 10\n",
    "    },\n",
    "    ('chr2', 45, 50, '+', 'exon'): {\n",
    "        'edge_id': 12,\n",
    "        'edge_type': 'exon', \n",
    "        'v1': 11,\n",
    "        'v2': 10\n",
    "    },\n",
    "    ('chr2', 80, 60, '-', 'intron'): {\n",
    "        'edge_id': 13,\n",
    "        'edge_type': 'intron', \n",
    "        'v1': 7,\n",
    "        'v2': 9\n",
    "    },\n",
    "    ('chr2', 60, 50, '-', 'exon'): {\n",
    "        'edge_id': 14,\n",
    "        'edge_type': 'exon', \n",
    "        'v1': 9,\n",
    "        'v2': 10\n",
    "    }\n",
    "}\n",
    "\n",
    "print(edges == ctrl_edges)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "spoken-flood",
   "metadata": {},
   "source": [
    "#### test - create_loc_dict - locs already exist in swangraph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "prospective-sailing",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "_, exons = get_test_transcript_exon_dicts()\n",
    "\n",
    "# dummy preexisting data\n",
    "sg = swan.SwanGraph()\n",
    "data = [[0, 'chr1', 1], [1, 'chr2', 80]]\n",
    "columns = ['vertex_id', 'chrom', 'coord']\n",
    "sg.loc_df = pd.DataFrame(data=data, columns=columns)\n",
    "\n",
    "locs = sg.create_loc_dict(exons)\n",
    "\n",
    "ctrl_locs = {('chr1', 1):0,\n",
    "    ('chr2', 80): 1,\n",
    "    ('chr1', 20): 2,\n",
    "    ('chr1', 25): 3,\n",
    "    ('chr1', 30): 4,\n",
    "    ('chr1', 35): 5,\n",
    "    ('chr1', 40): 6,\n",
    "    ('chr2', 100): 7,\n",
    "    ('chr2', 75): 8,\n",
    "    ('chr2', 60): 9,\n",
    "    ('chr2', 50): 10,\n",
    "    ('chr2', 45): 11,\n",
    "    ('chr2', 65): 12\n",
    "     }\n",
    "\n",
    "print(ctrl_locs == locs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "pointed-plate",
   "metadata": {},
   "source": [
    "#### test - create_loc_dict - empty swangraph\n",
    "* also checks to make sure that exons with the same location as one of their splice sites aren't duplicated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "collaborative-label",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "_, exons = get_test_transcript_exon_dicts()\n",
    "\n",
    "sg = swan.SwanGraph()\n",
    "locs = sg.create_loc_dict(exons)\n",
    "ctrl_locs = {('chr1',1): 0, \n",
    "    ('chr1', 20): 1,\n",
    "    ('chr1', 25): 2,\n",
    "    ('chr1', 30): 3,\n",
    "    ('chr1', 35): 4,\n",
    "    ('chr1', 40): 5,\n",
    "    ('chr2', 100): 6,\n",
    "    ('chr2', 80): 7,\n",
    "    ('chr2', 75): 8,\n",
    "    ('chr2', 60): 9,\n",
    "    ('chr2', 50): 10,\n",
    "    ('chr2', 45): 11,\n",
    "    ('chr2', 65): 12\n",
    "     }\n",
    "print(ctrl_locs == locs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ecological-reservation",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_test_transcript_exon_dicts():\n",
    "\n",
    "    # features: \n",
    "    # exons that are \"out of order\"\n",
    "    # two transcripts from the same gene\n",
    "    # transcripts from - strand\n",
    "    # transcripts from + strand\n",
    "    transcripts = {\n",
    "        'test1': { # + strand\n",
    "            'gid': 'test1_gid',\n",
    "            'gname': 'test1_gname',\n",
    "            'tid': 'test1',\n",
    "            'tname': 'test1_tname',\n",
    "            'strand': '+',\n",
    "            'novelty': 'Known',\n",
    "            'exons': ['chr1_1_20_+_exon',\n",
    "                      'chr1_35_40_+_exon', # out of order exon (+)\n",
    "                      'chr1_25_30_+_exon',] # out of order exon (+)\n",
    "        },\n",
    "        'test2': { # - strand\n",
    "            'gid': 'test2_gid',\n",
    "            'gname': 'test2_gname',\n",
    "            'tid': 'test2',\n",
    "            'tname': 'test2_tname',\n",
    "            'strand': '-', \n",
    "            'novelty': 'Known',\n",
    "            'exons': ['chr2_100_80_-_exon', # duplicated exon/locations\n",
    "                      'chr2_75_60_-_exon', # same edge but intron vs. exon\n",
    "                      'chr2_50_45_-_exon'] # duplicated exon/locations, same exon different strand\n",
    "        },\n",
    "        'test3': { # - strand\n",
    "            'gid': 'test2_gid',\n",
    "            'gname': 'test2_gname',\n",
    "            'tid': 'test3',\n",
    "            'tname': 'test3_tname',\n",
    "            'strand': '-', \n",
    "            'novelty': 'NIC',\n",
    "            'exons': ['chr2_100_80_-_exon', # duplicated exon/locations\n",
    "                      'chr2_50_45_-_exon', # out of order exon (-), duplicated exon/locations\n",
    "                      'chr2_75_65_-_exon'] # out of order exon (-)\n",
    "        },\n",
    "        'test4': { # + strand\n",
    "            'gid': 'test4_gid',\n",
    "            'gname': 'test4_gname', \n",
    "            'tid': 'test4', \n",
    "            'tname': 'test4_tname',\n",
    "            'strand': '+',\n",
    "            'novelty': 'Known',\n",
    "            'exons': ['chr2_45_50_+_exon'] # same exon different strand\n",
    "        },\n",
    "        'test5': { # - strand\n",
    "            'gid': 'test2_gid',\n",
    "            'gname': 'test2_gname',\n",
    "            'tid': 'test5',\n",
    "            'tname': 'test5_tname',\n",
    "            'strand': '-', \n",
    "            'novelty': 'ISM',\n",
    "            'exons': ['chr2_100_80_-_exon', # duplicated exon/locations\n",
    "                      'chr2_60_50_-_exon'] # same edge but intron vs. exon\n",
    "        },\n",
    "    }\n",
    "\n",
    "    # features\n",
    "    # locations that are shared across exons\n",
    "    exons = {\n",
    "        'chr1_1_20_+_exon': {\n",
    "            'eid': 'chr1_1_20_+_exon',\n",
    "            'chrom': 'chr1',\n",
    "            'v1': 1,\n",
    "            'v2': 20,\n",
    "            'strand': '+', \n",
    "            'edge_type': 'exon'\n",
    "        },\n",
    "        'chr1_25_30_+_exon': {\n",
    "            'eid': 'chr1_25_30_+_exon',\n",
    "            'chrom': 'chr1',\n",
    "            'v1': 25,\n",
    "            'v2': 30,\n",
    "            'strand': '+', \n",
    "            'edge_type': 'exon'\n",
    "        },\n",
    "        'chr1_35_40_+_exon': {\n",
    "            'eid': 'chr1_35_40_+_exon',\n",
    "            'chrom': 'chr1',\n",
    "            'v1': 35,\n",
    "            'v2': 40,\n",
    "            'strand': '+', \n",
    "            'edge_type': 'exon'\n",
    "        },\n",
    "        'chr2_100_80_-_exon': {\n",
    "            'eid': 'chr2_100_80_-_exon',\n",
    "            'chrom': 'chr2',\n",
    "            'v1': 100,\n",
    "            'v2': 80,\n",
    "            'strand': '-', \n",
    "            'edge_type': 'exon'\n",
    "        },\n",
    "        'chr2_75_60_-_exon': {\n",
    "            'eid': 'chr2_75_60_-_exon',\n",
    "            'chrom': 'chr2',\n",
    "            'v1': 75,\n",
    "            'v2': 60,\n",
    "            'strand': '-', \n",
    "            'edge_type': 'exon'\n",
    "        },\n",
    "        'chr2_50_45_-_exon': {\n",
    "            'eid': 'chr2_50_45_-_exon',\n",
    "            'chrom': 'chr2',\n",
    "            'v1': 50,\n",
    "            'v2': 45,\n",
    "            'strand': '-', \n",
    "            'edge_type': 'exon'\n",
    "        },\n",
    "        'chr2_75_65_-_exon': {\n",
    "            'eid': 'chr2_75_60_-_exon',\n",
    "            'chrom': 'chr2',\n",
    "            'v1': 75,\n",
    "            'v2': 65,\n",
    "            'strand': '-', \n",
    "            'edge_type': 'exon'\n",
    "        }, \n",
    "        'chr2_45_50_+_exon': {\n",
    "            'eid': 'chr2_45_50_+_exon',\n",
    "            'chrom': 'chr2',\n",
    "            'v1': 45,\n",
    "            'v2': 50,\n",
    "            'strand': '+',\n",
    "            'edge_type': 'exon'\n",
    "            },\n",
    "        'chr2_100_80_-_exon': {\n",
    "            'eid': 'chr2_100_80_-_exon',\n",
    "            'chrom': 'chr2',\n",
    "            'v1': 100,\n",
    "            'v2': 80,\n",
    "            'strand': '-',\n",
    "            'edge_type': 'exon'\n",
    "        },\n",
    "        'chr2_60_50_-_exon': {\n",
    "            'eid': 'chr2_60_50_-_exon',\n",
    "            'chrom': 'chr2',\n",
    "            'v1': 60,\n",
    "            'v2': 50,\n",
    "            'strand': '-'\n",
    "        }\n",
    "    }\n",
    "\n",
    "    return transcripts, exons"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "unsigned-reverse",
   "metadata": {},
   "source": [
    "#### test - get_current_edges - swangraph with data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "living-franklin",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "\n",
    "cols = ['vertex_id', 'chrom', 'coord']\n",
    "data = [[0, 1, 2], [1, 1, 3], [2, 1, 50]]\n",
    "sg.loc_df = pd.DataFrame(data=data, columns=cols)\n",
    "\n",
    "cols = ['edge_id', 'v1', 'v2', 'strand', 'edge_type']\n",
    "data = [[0, 0, 1, '+', 'exon'],\n",
    "        [1, 1, 2, '+', 'intron']]\n",
    "sg.edge_df = pd.DataFrame(data=data, columns=cols)\n",
    "\n",
    "edges, n = sg.get_current_edges()\n",
    "ctrl = {(1,2,3,'+','exon'): {'edge_id': 0, \n",
    "                             'edge_type': 'exon',\n",
    "                             'v1': 0 ,\n",
    "                             'v2': 1},\n",
    "        (1,3,50,'+','intron'): {'edge_id': 1,\n",
    "                                'edge_type': 'intron',\n",
    "                                'v1': 1,\n",
    "                                'v2': 2}}\n",
    "print(edges == ctrl)\n",
    "print(n == 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "virgin-strain",
   "metadata": {},
   "source": [
    "#### test - get_current_edges - empty swangraph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "beginning-contrast",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "edges, n = sg.get_current_edges()\n",
    "\n",
    "print(edges == {})\n",
    "print(n == -1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "unknown-devices",
   "metadata": {},
   "source": [
    "#### test - get_current_locs - swangraph with data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "analyzed-oasis",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "cols = ['vertex_id', 'chrom', 'coord']\n",
    "data = [[0, 1, 2], [1, 1, 3], [2, 3, 50]]\n",
    "sg.loc_df = pd.DataFrame(data=data, columns=cols)\n",
    "locs, n = sg.get_current_locs()\n",
    "\n",
    "ctrl_locs = {(1,2):0, (1,3):1, (3,50):2}\n",
    "print(locs == ctrl_locs)\n",
    "print(n == 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "played-receiver",
   "metadata": {},
   "source": [
    "#### test - get_current_locs - empty swangraph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "featured-logan",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "locs, n = sg.get_current_locs()\n",
    "\n",
    "print(locs == {})\n",
    "print(n == -1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "latter-toner",
   "metadata": {},
   "source": [
    "#### test - parse_gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "published-reader",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gid</th>\n",
       "      <th>gname</th>\n",
       "      <th>tid</th>\n",
       "      <th>tname</th>\n",
       "      <th>strand</th>\n",
       "      <th>exons</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid_index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000020637.8</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000146979.1</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000153068.1</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000155801.1</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000179865.6</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       gid  gname   tid  tname  strand  exons\n",
       "tid_index                                                    \n",
       "ENSMUST00000020637.8  True   True  True   True    True   True\n",
       "ENSMUST00000146979.1  True   True  True   True    True   True\n",
       "ENSMUST00000153068.1  True   True  True   True    True   True\n",
       "ENSMUST00000155801.1  True   True  True   True    True   True\n",
       "ENSMUST00000179865.6  True   True  True   True    True   True"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# canx gtf - no gene names, from talon\n",
    "gtf_file = '../testing/input_files/Canx.gtf'\n",
    "t_df, exon_df, from_talon = swan.parse_gtf(gtf_file, False)\n",
    "\n",
    "t_df.index.name = 'tid_index'\n",
    "t_df = t_df.sort_values(by='tid_index')\n",
    "\n",
    "ctrl_t_df = pd.read_csv('../testing/input_files/Canx_transcript.tsv',sep='\\t')\n",
    "ctrl_t_df.set_index('tid_index', inplace=True)\n",
    "ctrl_t_df = ctrl_t_df.sort_values(by='tid_index')\n",
    "\n",
    "ctrl_exons = ctrl_t_df.exons.tolist()\n",
    "ctrl_exons = [exons.split(',') for exons in ctrl_exons]\n",
    "ctrl_t_df['exons'] = ctrl_exons\n",
    "\n",
    "t_df == ctrl_t_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "outside-monte",
   "metadata": {},
   "source": [
    "#### test - create_dfs_gtf - edge used as intron and exon\n",
    "* TODO - do I still need to do this now that I've chunked stuff down>? "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abstract-quantum",
   "metadata": {},
   "source": [
    "#### test - create_dfs_gtf - all of Canx.gtf\n",
    "* transcript that uses a preexisting location\n",
    "* TODO - do I stil lneed to do this now that I've chunked stuff down?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "substantial-midnight",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/5 [00:00<?, ?it/s]"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "gtf_file = '../testing/input_files/Canx.gtf'\n",
    "sg.create_dfs_gtf(gtf_file, verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "rental-conversion",
   "metadata": {},
   "source": [
    "#### test - 2x create_dfs_gtf, where second round contains edges/locations already in the SwanGraph\n",
    "* TODO - do I still need to do this now that I've chunked stuff down? "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "sensitive-template",
   "metadata": {},
   "source": [
    "### DIE tests"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "retained-lover",
   "metadata": {},
   "source": [
    "#### test get_die_gene_table\n",
    "    1. limit to isoforms with > 0 exp in at least one condition\n",
    "    2. limit to genes with more than 1 isoform\n",
    "    3. aggregate the last n - 11 isoforms\n",
    "    4. gene that doesn't have enough rc reads\n",
    "    5. gene with 11 > n isoforms > 1\n",
    "    6. gene with no expressed isoforms\n",
    "    7. gene that meets rc "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "photographic-azerbaijan",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     gid  cond1  cond2  total_counts  cond1_counts  cond2_counts  dpi\n",
      "tid                                                                  \n",
      "1      1    0.5    0.5            20            10            10  0.0\n",
      "2      1    0.5    0.5            20            10            10  0.0\n",
      "     gid  cond1  cond2  total_counts  cond1_counts  cond2_counts  dpi\n",
      "tid                                                                  \n",
      "1      1    0.5    0.5            20            10            10    0\n",
      "2      1    0.5    0.5            20            10            10    0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gid</th>\n",
       "      <th>cond1</th>\n",
       "      <th>cond2</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>cond1_counts</th>\n",
       "      <th>cond2_counts</th>\n",
       "      <th>dpi</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gid  cond1  cond2  total_counts  cond1_counts  cond2_counts   dpi\n",
       "tid                                                                    \n",
       "1    True   True   True          True          True          True  True\n",
       "2    True   True   True          True          True          True  True"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns = ['tid', 'gid', 'cond1', 'cond2', 'total_counts', 'cond1_counts', 'cond2_counts']\n",
    "data = [[1, 1, .5, .5, 20, 10, 10], \n",
    "        [2, 1, .5, .5, 20, 10, 10]]\n",
    "df = pd.DataFrame(data=data, columns=columns)\n",
    "conditions = ['cond1', 'cond2']\n",
    "df.set_index('tid', inplace=True)\n",
    "df\n",
    "\n",
    "df = swan.get_die_gene_table(df, conditions, rc=15)\n",
    "\n",
    "columns = ['tid', 'gid', 'cond1', 'cond2', 'total_counts', 'cond1_counts', 'cond2_counts', 'dpi']\n",
    "data = [[1, 1, .5, .5, 20, 10, 10, 0], \n",
    "        [2, 1, .5, .5, 20, 10, 10, 0]]\n",
    "ctrl = pd.DataFrame(data=data, columns=columns)\n",
    "ctrl.set_index('tid', inplace=True)\n",
    "print(df)\n",
    "print(ctrl)\n",
    "\n",
    "ctrl == df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "rocky-diagnosis",
   "metadata": {},
   "source": [
    "Tests 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "sudden-professional",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     gid  cond1  cond2  total_counts  cond1_counts  cond2_counts  dpi\n",
      "tid                                                                  \n",
      "1      1    0.5    0.5            20            10            10  0.0\n",
      "2      1    0.5    0.5            20            10            10  0.0\n",
      "     gid  cond1  cond2  total_counts  cond1_counts  cond2_counts  dpi\n",
      "tid                                                                  \n",
      "1      1    0.5    0.5            20            10            10    0\n",
      "2      1    0.5    0.5            20            10            10    0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gid</th>\n",
       "      <th>cond1</th>\n",
       "      <th>cond2</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>cond1_counts</th>\n",
       "      <th>cond2_counts</th>\n",
       "      <th>dpi</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gid  cond1  cond2  total_counts  cond1_counts  cond2_counts   dpi\n",
       "tid                                                                    \n",
       "1    True   True   True          True          True          True  True\n",
       "2    True   True   True          True          True          True  True"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns = ['tid', 'gid', 'cond1', 'cond2', 'total_counts', 'cond1_counts', 'cond2_counts']\n",
    "data = [[1, 1, .5, .5, 20, 10, 10], \n",
    "        [2, 1, .5, .5, 20, 10, 10]]\n",
    "df = pd.DataFrame(data=data, columns=columns)\n",
    "conditions = ['cond1', 'cond2']\n",
    "df.set_index('tid', inplace=True)\n",
    "df\n",
    "\n",
    "df = swan.get_die_gene_table(df, conditions, rc=0)\n",
    "\n",
    "columns = ['tid', 'gid', 'cond1', 'cond2', 'total_counts', 'cond1_counts', 'cond2_counts', 'dpi']\n",
    "data = [[1, 1, .5, .5, 20, 10, 10, 0], \n",
    "        [2, 1, .5, .5, 20, 10, 10, 0]]\n",
    "ctrl = pd.DataFrame(data=data, columns=columns)\n",
    "ctrl.set_index('tid', inplace=True)\n",
    "print(df)\n",
    "print(ctrl)\n",
    "\n",
    "ctrl == df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "sublime-casting",
   "metadata": {},
   "source": [
    "Tests 6 - this one is kinda handled upstream in multiple capacities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "steady-fishing",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "test - not enough isoforms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns = ['tid', 'gid', 'cond1', 'cond2', 'total_counts', 'cond1_counts', 'cond2_counts']\n",
    "data = [[1, 1, 0, 0, 0, 0, 0], \n",
    "        [2, 1, 0, 0, 0, 0, 0]]\n",
    "df = pd.DataFrame(data=data, columns=columns)\n",
    "conditions = ['cond1', 'cond2']\n",
    "df.set_index('tid', inplace=True)\n",
    "df\n",
    "\n",
    "df = swan.get_die_gene_table(df, conditions, rc=0)\n",
    "df == None "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "little-racing",
   "metadata": {},
   "source": [
    "Tests 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "buried-foster",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "test - not enough reads per condition\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns = ['tid', 'gid', 'cond1', 'cond2', 'total_counts', 'cond1_counts', 'cond2_counts']\n",
    "data = [[1, 1, .5, .5, 20, 10, 10], \n",
    "        [2, 1, .5, .5, 20, 10, 10]]\n",
    "df = pd.DataFrame(data=data, columns=columns)\n",
    "conditions = ['cond1', 'cond2']\n",
    "df.set_index('tid', inplace=True)\n",
    "df\n",
    "\n",
    "df = swan.get_die_gene_table(df, conditions, rc=30)\n",
    "df == None "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "extensive-sailing",
   "metadata": {},
   "source": [
    "Tests 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "waiting-proposal",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "test - not enough isoforms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns = ['tid', 'gid', 'cond1', 'cond2', 'total_counts', 'cond1_counts', 'cond2_counts', 'dpi']\n",
    "data = [1, 1, 1, 1, 30, 15, 15, 0]\n",
    "df = pd.DataFrame(data=[data], columns=columns)\n",
    "conditions = ['cond1', 'cond2']\n",
    "df.set_index('tid', inplace=True)\n",
    "df\n",
    "\n",
    "df = swan.get_die_gene_table(df, conditions, rc=0)\n",
    "df == None "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fewer-childhood",
   "metadata": {},
   "source": [
    "Tests 1. and 3. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "federal-ministry",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gid</th>\n",
       "      <th>cond1</th>\n",
       "      <th>cond2</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>cond1_counts</th>\n",
       "      <th>cond2_counts</th>\n",
       "      <th>dpi</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>all_other</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             gid  cond1  cond2  total_counts  cond1_counts  cond2_counts  \\\n",
       "tid                                                                        \n",
       "3           True   True   True          True          True          True   \n",
       "1           True   True   True          True          True          True   \n",
       "12          True   True   True          True          True          True   \n",
       "8           True   True   True          True          True          True   \n",
       "7           True   True   True          True          True          True   \n",
       "5           True   True   True          True          True          True   \n",
       "6           True   True   True          True          True          True   \n",
       "11          True   True   True          True          True          True   \n",
       "4           True   True   True          True          True          True   \n",
       "13          True   True   True          True          True          True   \n",
       "all_other  False   True   True          True          True          True   \n",
       "\n",
       "             dpi  \n",
       "tid               \n",
       "3           True  \n",
       "1           True  \n",
       "12          True  \n",
       "8          False  \n",
       "7          False  \n",
       "5           True  \n",
       "6           True  \n",
       "11          True  \n",
       "4          False  \n",
       "13          True  \n",
       "all_other  False  "
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conditions = ['cond1', 'cond2']\n",
    "df = pd.read_csv('data/test_pi_1.tsv', sep='\\t')\n",
    "df.set_index('tid', inplace=True)\n",
    "df = swan.get_die_gene_table(df, conditions, rc=0)\n",
    "df.tid = df.tid.astype('str')\n",
    "df.set_index('tid', inplace=True)\n",
    "# print(df.dtypes)\n",
    "# print(df)\n",
    "\n",
    "ctrl = pd.read_csv('data/test_pi_1_reference.tsv', sep='\\t')\n",
    "ctrl.tid = ctrl.tid.astype('str')\n",
    "ctrl.set_index('tid', inplace=True)\n",
    "# print(ctrl.dtypes)\n",
    "ctrl = ctrl[df.columns]\n",
    "\n",
    "# print(ctrl)\n",
    "\n",
    "\n",
    "# print(df.columns)\n",
    "# print(ctrl.columns)\n",
    "# print(df.columns == ctrl.columns)\n",
    "# print(df.index == ctrl.index)\n",
    "df == ctrl\n",
    "\n",
    "# still need better way to test this rip"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dietary-destination",
   "metadata": {},
   "source": [
    "#### test test_gene - all in pytest\n",
    "1. 2 up 2 down\n",
    "2. < 2 up, < 2 down\n",
    "3. all 0 "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "under-japanese",
   "metadata": {},
   "source": [
    "tests 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "disturbed-concentrate",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "columns = ['tid', 'gid', 'cond1', 'cond2', 'total_counts', 'cond1_counts', 'cond2_counts']\n",
    "data = [[1, 1, .5, .5, 20, 10, 10], \n",
    "        [2, 1, .5, .5, 20, 10, 10]]\n",
    "df = pd.DataFrame(data=data, columns=columns)\n",
    "conditions = ['cond1', 'cond2']\n",
    "df.set_index('tid', inplace=True)\n",
    "df = get_die_gene_table(df, conditions, rc=0)\n",
    "ctrl_gene_dpi = 0\n",
    "\n",
    "pval, gene_dpi = test_gene(df, conditions)\n",
    "\n",
    "print(gene_dpi)\n",
    "print(ctrl_gene_dpi)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "prescription-canadian",
   "metadata": {},
   "source": [
    "tests 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "formal-decision",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.22\n",
      "0.22\n"
     ]
    }
   ],
   "source": [
    "conditions = ['cond1', 'cond2']\n",
    "df = pd.read_csv('data/test_pi_1.tsv', sep='\\t')\n",
    "df.set_index('tid', inplace=True)\n",
    "df = swan.get_die_gene_table(df, conditions, rc=0)\n",
    "ctrl_gene_dpi = 0.22\n",
    "\n",
    "tids = [3, 1]\n",
    "df = df.loc[df.tid.isin(tids)]\n",
    "\n",
    "pval, gene_dpi = swan.test_gene(df, conditions)\n",
    "\n",
    "print(ctrl_gene_dpi)\n",
    "print(gene_dpi)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "framed-speaker",
   "metadata": {},
   "source": [
    "tests 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "oriental-editing",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.41000000000000003\n",
      "0.41\n"
     ]
    }
   ],
   "source": [
    "conditions = ['cond1', 'cond2']\n",
    "df = pd.read_csv('data/test_pi_1.tsv', sep='\\t')\n",
    "df.set_index('tid', inplace=True)\n",
    "df = swan.get_die_gene_table(df, conditions, rc=0)\n",
    "ctrl_gene_dpi = 0.41\n",
    "\n",
    "pval, gene_dpi = test_gene(df, conditions)\n",
    "# print(gene_dpi)\n",
    "# print(pval)\n",
    "\n",
    "print(gene_dpi)\n",
    "print(ctrl_gene_dpi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "danish-music",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding annotation to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing transcripts:  99%|█████████▉| 8400/8450 [00:02<00:00, 4170.51it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reindexing and sorting entries on genomic location...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reindexing vertices: 100%|█████████▉| 50000/50074 [00:00<00:00, 1057058.90it/s]\n",
      "Reindexing edges:  98%|█████████▊| 49000/50074 [00:00<00:00, 593869.17it/s]\n",
      "Reindexing transcripts:  99%|█████████▉| 8400/8450 [00:00<00:00, 161367.41it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Annotation to the SwanGraph\n",
      "\n",
      "Adding transcriptome to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "AnnData expects .obs.index to contain strings, but your first indices are: Int64Index([0, 1], dtype='int64'), …\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding abundance for datasets PB65_B017, PB65_B018 to SwanGraph.\n",
      "\n",
      "Adding abundance for datasets D12 to SwanGraph.\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "# just gencode vM21 chr 11 and tcf3\n",
    "gtf = 'data/chr11_and_Tcf3.gtf'\n",
    "sg.add_annotation(gtf, verbose=True)\n",
    "db = 'data/chr11_and_Tcf3_no_gname.db'\n",
    "sg.add_transcriptome(db, include_isms=True)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance_PB65.tsv'\n",
    "sg.add_abundance(ab)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance_D12.tsv'\n",
    "sg.add_abundance(ab)\n",
    "meta = 'data/chr11_and_Tcf3_metadata.tsv'\n",
    "sg.add_metadata(meta)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "headed-hampshire",
   "metadata": {},
   "source": [
    "#### test - get DIE, obs_col has more than 2 values and obs_conditions not given - in pytest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "geographic-karma",
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_col = 'dataset'\n",
    "id_col = 'tid'\n",
    "obs_conditions = ['D12']\n",
    "\n",
    "# commented out so I dont have to look at error for the time bean\n",
    "# test = sg.get_die_genes(obs_col=obs_col, obs_conditions=obs_conditions, rc_thresh=1)\n",
    "# should say \"obs_conditions must have exactly 2 values\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "composed-spelling",
   "metadata": {},
   "source": [
    "#### test - get DIE, obs_col w/o 2 elements given - in pytest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "yellow-ethics",
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_col = 'dataset'\n",
    "id_col = 'tid'\n",
    "\n",
    "# test = sg.get_die_genes(obs_col=obs_col, rc_thresh=1)\n",
    "# should say \"Must provide obs_conditions argument when obs_col has>2 unique values\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fuzzy-extreme",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tid\n",
       "ENCODE-mouseT000008451    ENCODE-mouseT000008451\n",
       "ENCODE-mouseT000008452    ENCODE-mouseT000008452\n",
       "ENCODE-mouseT000008453    ENCODE-mouseT000008453\n",
       "ENCODE-mouseT000008454    ENCODE-mouseT000008454\n",
       "ENCODE-mouseT000008455    ENCODE-mouseT000008455\n",
       "ENCODE-mouseT000008456    ENCODE-mouseT000008456\n",
       "ENCODE-mouseT000008457    ENCODE-mouseT000008457\n",
       "ENCODE-mouseT000008458    ENCODE-mouseT000008458\n",
       "ENCODE-mouseT000008459    ENCODE-mouseT000008459\n",
       "ENCODE-mouseT000008460    ENCODE-mouseT000008460\n",
       "ENCODE-mouseT000008461    ENCODE-mouseT000008461\n",
       "ENCODE-mouseT000008462    ENCODE-mouseT000008462\n",
       "ENSMUST00000020637.8        ENSMUST00000020637.8\n",
       "ENSMUST00000020741.10      ENSMUST00000020741.10\n",
       "ENSMUST00000105344.6        ENSMUST00000105344.6\n",
       "Name: tid, dtype: object"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.t_df.tid"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "neural-blake",
   "metadata": {},
   "source": [
    "#### test - get DIE between two clusters (non 'dataset' condition) - in pytest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "acute-housing",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                     gid        dpi\n",
      "0  ENSMUSG00000020167.13  33.333333\n",
      "1  ENSMUSG00000020368.14  66.666667\n",
      "                     gid        dpi\n",
      "0  ENSMUSG00000020167.13  33.333333\n",
      "1  ENSMUSG00000020368.14  66.666667\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gid</th>\n",
       "      <th>dpi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gid    dpi\n",
       "0  True  False\n",
       "1  True  False"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "obs_col = 'cluster'\n",
    "id_col = 'tid'\n",
    "\n",
    "test = sg.get_die_genes(obs_col=obs_col, rc_thresh=1)\n",
    "\n",
    "# don't test p vals cause that's tough \n",
    "test.drop(['p_val', 'adj_p_val'], axis=1, inplace=True)\n",
    "\n",
    "ctrl = pd.read_csv('data/chr11_and_Tcf3_cluster_dpi.tsv', sep='\\t')\n",
    "print(ctrl)\n",
    "print(test)\n",
    "ctrl == test"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "renewable-contrast",
   "metadata": {},
   "source": [
    "#### test - get DIE between two datasets  - in pytests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "purple-evolution",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0    ENSMUSG00000020167.13\n",
      "1    ENSMUSG00000020457.12\n",
      "2    ENSMUSG00000020368.14\n",
      "Name: gid, dtype: object\n"
     ]
    }
   ],
   "source": [
    "obs_col = 'dataset'\n",
    "obs_conditions = ['PB65_B017', 'PB65_B018']\n",
    "id_col = 'tid'\n",
    "\n",
    "test = sg.get_die_genes(obs_col=obs_col, obs_conditions=obs_conditions, rc_thresh=1)\n",
    "\n",
    "# don't test p vals cause that's tough \n",
    "test.drop(['p_val', 'adj_p_val'], axis=1, inplace=True)\n",
    "\n",
    "ctrl = pd.read_csv('data/chr11_and_Tcf3_PB65_B017_B018_dpi.tsv', sep='\\t')\n",
    "test == ctrl\n",
    "print(test.gid)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "stylish-locking",
   "metadata": {},
   "source": [
    "#### test - calc TPM for condition that is not \"dataset\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "worthy-message",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding annotation to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing transcripts:  99%|█████████▉| 8400/8450 [00:01<00:00, 4279.71it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reindexing and sorting entries on genomic location...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reindexing vertices: 100%|█████████▉| 50000/50074 [00:00<00:00, 996096.65it/s] \n",
      "Reindexing edges:  98%|█████████▊| 49000/50074 [00:00<00:00, 654539.74it/s]\n",
      "Reindexing transcripts:  99%|█████████▉| 8400/8450 [00:00<00:00, 157417.82it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Annotation to the SwanGraph\n",
      "\n",
      "Adding transcriptome to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "AnnData expects .obs.index to contain strings, but your first indices are: Int64Index([0, 1], dtype='int64'), …\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding abundance for datasets PB65_B017, PB65_B018 to SwanGraph.\n",
      "\n",
      "Adding abundance for datasets D12 to SwanGraph.\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "# just gencode vM21 chr 11 and tcf3\n",
    "gtf = 'data/chr11_and_Tcf3.gtf'\n",
    "sg.add_annotation(gtf, verbose=True)\n",
    "db = 'data/chr11_and_Tcf3_no_gname.db'\n",
    "sg.add_transcriptome(db, include_isms=True)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance_PB65.tsv'\n",
    "sg.add_abundance(ab)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance_D12.tsv'\n",
    "sg.add_abundance(ab)\n",
    "meta = 'data/chr11_and_Tcf3_metadata.tsv'\n",
    "sg.add_metadata(meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "activated-invalid",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                    2              1\n",
      "tid                                                 \n",
      "ENCODE-mouseT000008451  285714.285714  285714.285714\n",
      "ENCODE-mouseT000008452   71428.571429       0.000000\n",
      "ENCODE-mouseT000008453  142857.142857       0.000000\n",
      "ENCODE-mouseT000008454   71428.571429       0.000000\n",
      "ENCODE-mouseT000008455   71428.571429       0.000000\n",
      "ENCODE-mouseT000008456       0.000000  142857.142857\n",
      "ENCODE-mouseT000008457       0.000000  142857.142857\n",
      "ENCODE-mouseT000008458       0.000000  142857.142857\n",
      "ENCODE-mouseT000008459       0.000000  142857.142857\n",
      "ENCODE-mouseT000008460   71428.571429       0.000000\n",
      "ENCODE-mouseT000008461   71428.571429       0.000000\n",
      "ENCODE-mouseT000008462   71428.571429       0.000000\n",
      "ENSMUST00000020637.8         0.000000  142857.142857\n",
      "ENSMUST00000020741.10    71428.571429       0.000000\n",
      "ENSMUST00000105344.6     71428.571429       0.000000\n",
      "Index(['2', '1'], dtype='object')\n",
      "                                   2            1\n",
      "tid                                              \n",
      "ENCODE-mouseT000008451  285714.28570  285714.2857\n",
      "ENCODE-mouseT000008452   71428.57143       0.0000\n",
      "ENCODE-mouseT000008453  142857.14290       0.0000\n",
      "ENCODE-mouseT000008454   71428.57143       0.0000\n",
      "ENCODE-mouseT000008455   71428.57143       0.0000\n",
      "ENCODE-mouseT000008456       0.00000  142857.1429\n",
      "ENCODE-mouseT000008457       0.00000  142857.1429\n",
      "ENCODE-mouseT000008458       0.00000  142857.1429\n",
      "ENCODE-mouseT000008459       0.00000  142857.1429\n",
      "ENCODE-mouseT000008460   71428.57143       0.0000\n",
      "ENCODE-mouseT000008461   71428.57143       0.0000\n",
      "ENCODE-mouseT000008462   71428.57143       0.0000\n",
      "ENSMUST00000020637.8         0.00000  142857.1429\n",
      "ENSMUST00000020741.10    71428.57143       0.0000\n",
      "ENSMUST00000105344.6     71428.57143       0.0000\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>2</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008451</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008452</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008453</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008454</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008455</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008456</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008457</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008458</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008459</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008460</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008461</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008462</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000020637.8</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000020741.10</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000105344.6</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            2      1\n",
       "tid                                 \n",
       "ENCODE-mouseT000008451  False  False\n",
       "ENCODE-mouseT000008452  False   True\n",
       "ENCODE-mouseT000008453  False   True\n",
       "ENCODE-mouseT000008454  False   True\n",
       "ENCODE-mouseT000008455  False   True\n",
       "ENCODE-mouseT000008456   True  False\n",
       "ENCODE-mouseT000008457   True  False\n",
       "ENCODE-mouseT000008458   True  False\n",
       "ENCODE-mouseT000008459   True  False\n",
       "ENCODE-mouseT000008460  False   True\n",
       "ENCODE-mouseT000008461  False   True\n",
       "ENCODE-mouseT000008462  False   True\n",
       "ENSMUST00000020637.8     True  False\n",
       "ENSMUST00000020741.10   False   True\n",
       "ENSMUST00000105344.6    False   True"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = swan.calc_tpm(sg.adata, sg.t_df, obs_col='cluster')\n",
    "\n",
    "# test\n",
    "df = df.transpose()\n",
    "df.index.name = 'tid'\n",
    "df.sort_values(by='tid', inplace=True)\n",
    "print(df)\n",
    "print(df.columns)\n",
    "\n",
    "# ctrl\n",
    "ctrl = pd.read_csv('data/chr11_and_Tcf3_talon_abundance_clusters_tpm.tsv', sep='\\t')\n",
    "ctrl.set_index('tid', inplace=True)\n",
    "ctrl.sort_values(by='tid', inplace=True)\n",
    "print(ctrl)\n",
    "\n",
    "ctrl == df\n",
    "\n",
    "# need better way to test this bc of rounding, but it works.. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "better-macro",
   "metadata": {},
   "source": [
    "#### test - calc TPM from 2 abundance files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "accompanied-highway",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding annotation to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing transcripts:  99%|█████████▉| 8400/8450 [00:01<00:00, 4557.87it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reindexing and sorting entries on genomic location...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reindexing vertices: 100%|█████████▉| 50000/50074 [00:00<00:00, 1261353.77it/s]\n",
      "Reindexing edges:  98%|█████████▊| 49000/50074 [00:00<00:00, 708470.43it/s]\n",
      "Reindexing transcripts:  99%|█████████▉| 8400/8450 [00:00<00:00, 157610.77it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Annotation to the SwanGraph\n",
      "\n",
      "Adding transcriptome to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "AnnData expects .obs.index to contain strings, but your first indices are: Int64Index([0, 1], dtype='int64'), …\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding abundance for datasets PB65_B017, PB65_B018 to SwanGraph.\n",
      "\n",
      "Adding abundance for datasets D12 to SwanGraph.\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "# just gencode vM21 chr 11 and tcf3\n",
    "gtf = 'data/chr11_and_Tcf3.gtf'\n",
    "sg.add_annotation(gtf, verbose=True)\n",
    "db = 'data/chr11_and_Tcf3_no_gname.db'\n",
    "sg.add_transcriptome(db, include_isms=True)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance_PB65.tsv'\n",
    "sg.add_abundance(ab)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance_D12.tsv'\n",
    "sg.add_abundance(ab)\n",
    "meta = 'data/chr11_and_Tcf3_metadata.tsv'\n",
    "sg.add_metadata(meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "patient-locator",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                        PB65_B017  PB65_B018            D12\n",
      "tid                                                        \n",
      "ENCODE-mouseT000008451   300000.0   250000.0  285714.285714\n",
      "ENCODE-mouseT000008452   100000.0        0.0       0.000000\n",
      "ENCODE-mouseT000008453   100000.0   250000.0       0.000000\n",
      "ENCODE-mouseT000008454   100000.0        0.0       0.000000\n",
      "ENCODE-mouseT000008455   100000.0        0.0       0.000000\n",
      "ENCODE-mouseT000008456        0.0        0.0  142857.142857\n",
      "ENCODE-mouseT000008457        0.0        0.0  142857.142857\n",
      "ENCODE-mouseT000008458        0.0        0.0  142857.142857\n",
      "ENCODE-mouseT000008459        0.0        0.0  142857.142857\n",
      "ENCODE-mouseT000008460        0.0   250000.0       0.000000\n",
      "ENCODE-mouseT000008461   100000.0        0.0       0.000000\n",
      "ENCODE-mouseT000008462   100000.0        0.0       0.000000\n",
      "ENSMUST00000020637.8          0.0        0.0  142857.142857\n",
      "ENSMUST00000020741.10    100000.0        0.0       0.000000\n",
      "ENSMUST00000105344.6          0.0   250000.0       0.000000\n",
      "                        PB65_B017  PB65_B018          D12\n",
      "tid                                                      \n",
      "ENCODE-mouseT000008451     300000     250000  285714.2857\n",
      "ENCODE-mouseT000008452     100000          0       0.0000\n",
      "ENCODE-mouseT000008453     100000     250000       0.0000\n",
      "ENCODE-mouseT000008454     100000          0       0.0000\n",
      "ENCODE-mouseT000008455     100000          0       0.0000\n",
      "ENCODE-mouseT000008456          0          0  142857.1429\n",
      "ENCODE-mouseT000008457          0          0  142857.1429\n",
      "ENCODE-mouseT000008458          0          0  142857.1429\n",
      "ENCODE-mouseT000008459          0          0  142857.1429\n",
      "ENCODE-mouseT000008460          0     250000       0.0000\n",
      "ENCODE-mouseT000008461     100000          0       0.0000\n",
      "ENCODE-mouseT000008462     100000          0       0.0000\n",
      "ENSMUST00000020637.8            0          0  142857.1429\n",
      "ENSMUST00000020741.10      100000          0       0.0000\n",
      "ENSMUST00000105344.6            0     250000       0.0000\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PB65_B017</th>\n",
       "      <th>PB65_B018</th>\n",
       "      <th>D12</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008451</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008452</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008453</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008454</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008455</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008456</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008457</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008458</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008459</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008460</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008461</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008462</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000020637.8</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000020741.10</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000105344.6</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        PB65_B017  PB65_B018    D12\n",
       "tid                                                \n",
       "ENCODE-mouseT000008451       True       True  False\n",
       "ENCODE-mouseT000008452       True       True   True\n",
       "ENCODE-mouseT000008453       True       True   True\n",
       "ENCODE-mouseT000008454       True       True   True\n",
       "ENCODE-mouseT000008455       True       True   True\n",
       "ENCODE-mouseT000008456       True       True  False\n",
       "ENCODE-mouseT000008457       True       True  False\n",
       "ENCODE-mouseT000008458       True       True  False\n",
       "ENCODE-mouseT000008459       True       True  False\n",
       "ENCODE-mouseT000008460       True       True   True\n",
       "ENCODE-mouseT000008461       True       True   True\n",
       "ENCODE-mouseT000008462       True       True   True\n",
       "ENSMUST00000020637.8         True       True  False\n",
       "ENSMUST00000020741.10        True       True   True\n",
       "ENSMUST00000105344.6         True       True   True"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test\n",
    "df = pd.DataFrame(data=sg.adata.layers['tpm'], \\\n",
    "                  columns=sg.adata.var.index.tolist(), \\\n",
    "                  index=sg.adata.obs.index.tolist())\n",
    "df = df.transpose()\n",
    "df.index.name = 'tid'\n",
    "df.sort_values(by='tid', inplace=True)\n",
    "print(df)\n",
    "\n",
    "# ctrl\n",
    "ctrl = pd.read_csv('data/chr11_and_Tcf3_talon_abundance_tpm.tsv', sep='\\t')\n",
    "ctrl.set_index('tid', inplace=True)\n",
    "ctrl.sort_values(by='tid', inplace=True)\n",
    "print(ctrl)\n",
    "\n",
    "ctrl == df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "assigned-declaration",
   "metadata": {},
   "source": [
    "#### test - calc TPM from 1 abundance file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "arranged-davis",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding annotation to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing transcripts:  99%|█████████▉| 8400/8450 [00:01<00:00, 4478.18it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reindexing and sorting entries on genomic location...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reindexing vertices: 100%|█████████▉| 50000/50074 [00:00<00:00, 1219593.50it/s]\n",
      "Reindexing edges:  98%|█████████▊| 49000/50074 [00:00<00:00, 280590.31it/s]\n",
      "Reindexing transcripts:  99%|█████████▉| 8400/8450 [00:00<00:00, 140783.89it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Annotation to the SwanGraph\n",
      "\n",
      "Adding transcriptome to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "AnnData expects .obs.index to contain strings, but your first indices are: Int64Index([0, 1], dtype='int64'), …\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding abundance for datasets PB65_B017, PB65_B018, D12 to SwanGraph.\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "# just gencode vM21 chr 11 and tcf3\n",
    "gtf = 'data/chr11_and_Tcf3.gtf'\n",
    "sg.add_annotation(gtf, verbose=True)\n",
    "db = 'data/chr11_and_Tcf3_no_gname.db'\n",
    "sg.add_transcriptome(db, include_isms=True)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance.tsv'\n",
    "sg.add_abundance(ab)\n",
    "meta = 'data/chr11_and_Tcf3_metadata.tsv'\n",
    "sg.add_metadata(meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "elementary-framework",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                        PB65_B017  PB65_B018            D12\n",
      "tid                                                        \n",
      "ENCODE-mouseT000008451   300000.0   250000.0  285714.285714\n",
      "ENCODE-mouseT000008452   100000.0        0.0       0.000000\n",
      "ENCODE-mouseT000008453   100000.0   250000.0       0.000000\n",
      "ENCODE-mouseT000008454   100000.0        0.0       0.000000\n",
      "ENCODE-mouseT000008455   100000.0        0.0       0.000000\n",
      "ENCODE-mouseT000008456        0.0        0.0  142857.142857\n",
      "ENCODE-mouseT000008457        0.0        0.0  142857.142857\n",
      "ENCODE-mouseT000008458        0.0        0.0  142857.142857\n",
      "ENCODE-mouseT000008459        0.0        0.0  142857.142857\n",
      "ENCODE-mouseT000008460        0.0   250000.0       0.000000\n",
      "ENCODE-mouseT000008461   100000.0        0.0       0.000000\n",
      "ENCODE-mouseT000008462   100000.0        0.0       0.000000\n",
      "ENSMUST00000020637.8          0.0        0.0  142857.142857\n",
      "ENSMUST00000020741.10    100000.0        0.0       0.000000\n",
      "ENSMUST00000105344.6          0.0   250000.0       0.000000\n",
      "                        PB65_B017  PB65_B018          D12\n",
      "tid                                                      \n",
      "ENCODE-mouseT000008451     300000     250000  285714.2857\n",
      "ENCODE-mouseT000008452     100000          0       0.0000\n",
      "ENCODE-mouseT000008453     100000     250000       0.0000\n",
      "ENCODE-mouseT000008454     100000          0       0.0000\n",
      "ENCODE-mouseT000008455     100000          0       0.0000\n",
      "ENCODE-mouseT000008456          0          0  142857.1429\n",
      "ENCODE-mouseT000008457          0          0  142857.1429\n",
      "ENCODE-mouseT000008458          0          0  142857.1429\n",
      "ENCODE-mouseT000008459          0          0  142857.1429\n",
      "ENCODE-mouseT000008460          0     250000       0.0000\n",
      "ENCODE-mouseT000008461     100000          0       0.0000\n",
      "ENCODE-mouseT000008462     100000          0       0.0000\n",
      "ENSMUST00000020637.8            0          0  142857.1429\n",
      "ENSMUST00000020741.10      100000          0       0.0000\n",
      "ENSMUST00000105344.6            0     250000       0.0000\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PB65_B017</th>\n",
       "      <th>PB65_B018</th>\n",
       "      <th>D12</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008451</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008452</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008453</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008454</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008455</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008456</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008457</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008458</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008459</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008460</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008461</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008462</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000020637.8</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000020741.10</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000105344.6</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        PB65_B017  PB65_B018    D12\n",
       "tid                                                \n",
       "ENCODE-mouseT000008451       True       True  False\n",
       "ENCODE-mouseT000008452       True       True   True\n",
       "ENCODE-mouseT000008453       True       True   True\n",
       "ENCODE-mouseT000008454       True       True   True\n",
       "ENCODE-mouseT000008455       True       True   True\n",
       "ENCODE-mouseT000008456       True       True  False\n",
       "ENCODE-mouseT000008457       True       True  False\n",
       "ENCODE-mouseT000008458       True       True  False\n",
       "ENCODE-mouseT000008459       True       True  False\n",
       "ENCODE-mouseT000008460       True       True   True\n",
       "ENCODE-mouseT000008461       True       True   True\n",
       "ENCODE-mouseT000008462       True       True   True\n",
       "ENSMUST00000020637.8         True       True  False\n",
       "ENSMUST00000020741.10        True       True   True\n",
       "ENSMUST00000105344.6         True       True   True"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test\n",
    "df = pd.DataFrame(data=sg.adata.layers['tpm'], \\\n",
    "                  columns=sg.adata.var.index.tolist(), \\\n",
    "                  index=sg.adata.obs.index.tolist())\n",
    "df = df.transpose()\n",
    "df.index.name = 'tid'\n",
    "df.sort_values(by='tid', inplace=True)\n",
    "print(df)\n",
    "\n",
    "# ctrl\n",
    "ctrl = pd.read_csv('data/chr11_and_Tcf3_talon_abundance_tpm.tsv', sep='\\t')\n",
    "ctrl.set_index('tid', inplace=True)\n",
    "ctrl.sort_values(by='tid', inplace=True)\n",
    "print(ctrl)\n",
    "\n",
    "ctrl == df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "rotary-experiment",
   "metadata": {},
   "source": [
    "#### test - calc PI for a condition that is not 'dataset'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "excellent-montgomery",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/anndata/_core/anndata.py:21: FutureWarning: pandas.core.index is deprecated and will be removed in a future version.  The public classes are available in the top-level namespace.\n",
      "  from pandas.core.index import RangeIndex\n"
     ]
    }
   ],
   "source": [
    "import swan_vis as swan\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "settled-sussex",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding annotation to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing transcripts:  99%|█████████▉| 8400/8450 [00:01<00:00, 4330.09it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reindexing and sorting entries on genomic location...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reindexing vertices: 100%|█████████▉| 50000/50074 [00:00<00:00, 1006803.71it/s]\n",
      "Reindexing edges:  98%|█████████▊| 49000/50074 [00:00<00:00, 669436.09it/s]\n",
      "Reindexing transcripts:  99%|█████████▉| 8400/8450 [00:00<00:00, 157577.64it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Annotation to the SwanGraph\n",
      "\n",
      "Adding transcriptome to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "AnnData expects .obs.index to contain strings, but your first indices are: Int64Index([0, 1], dtype='int64'), …\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding abundance for datasets PB65_B017, PB65_B018, D12 to SwanGraph.\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "# just gencode vM21 chr 11 and tcf3\n",
    "gtf = 'data/chr11_and_Tcf3.gtf'\n",
    "sg.add_annotation(gtf, verbose=True)\n",
    "db = 'data/chr11_and_Tcf3_no_gname.db'\n",
    "sg.add_transcriptome(db, include_isms=True)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance.tsv'\n",
    "sg.add_abundance(ab)\n",
    "meta = 'data/chr11_and_Tcf3_metadata.tsv'\n",
    "sg.add_metadata(meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "blank-friday",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                 2      1\n",
      "tid                                      \n",
      "ENCODE-mouseT000008451   66.666672  100.0\n",
      "ENCODE-mouseT000008452   16.666668    0.0\n",
      "ENCODE-mouseT000008453   66.666672    0.0\n",
      "ENCODE-mouseT000008454  100.000000    0.0\n",
      "ENCODE-mouseT000008455  100.000000    0.0\n",
      "ENCODE-mouseT000008456    0.000000  100.0\n",
      "ENCODE-mouseT000008457    0.000000   25.0\n",
      "ENCODE-mouseT000008458    0.000000   25.0\n",
      "ENCODE-mouseT000008459    0.000000   25.0\n",
      "ENCODE-mouseT000008460   33.333336    0.0\n",
      "ENCODE-mouseT000008461   33.333336    0.0\n",
      "ENCODE-mouseT000008462   33.333336    0.0\n",
      "ENSMUST00000020637.8      0.000000   25.0\n",
      "ENSMUST00000020741.10    33.333336    0.0\n",
      "ENSMUST00000105344.6     16.666668    0.0\n",
      "Index(['2', '1'], dtype='object')\n",
      "                                 2    1\n",
      "tid                                    \n",
      "ENCODE-mouseT000008451   66.666672  100\n",
      "ENCODE-mouseT000008452   16.666668    0\n",
      "ENCODE-mouseT000008453   66.666672    0\n",
      "ENCODE-mouseT000008454  100.000000    0\n",
      "ENCODE-mouseT000008455  100.000000    0\n",
      "ENCODE-mouseT000008456    0.000000  100\n",
      "ENCODE-mouseT000008457    0.000000   25\n",
      "ENCODE-mouseT000008458    0.000000   25\n",
      "ENCODE-mouseT000008459    0.000000   25\n",
      "ENCODE-mouseT000008460   33.333336    0\n",
      "ENCODE-mouseT000008461   33.333336    0\n",
      "ENCODE-mouseT000008462   33.333336    0\n",
      "ENSMUST00000020637.8      0.000000   25\n",
      "ENSMUST00000020741.10    33.333336    0\n",
      "ENSMUST00000105344.6     16.666668    0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>2</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008451</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008452</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008453</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008454</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008455</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008456</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008457</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008458</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008459</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008460</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008461</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008462</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000020637.8</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000020741.10</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000105344.6</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            2     1\n",
       "tid                                \n",
       "ENCODE-mouseT000008451  False  True\n",
       "ENCODE-mouseT000008452  False  True\n",
       "ENCODE-mouseT000008453  False  True\n",
       "ENCODE-mouseT000008454   True  True\n",
       "ENCODE-mouseT000008455   True  True\n",
       "ENCODE-mouseT000008456   True  True\n",
       "ENCODE-mouseT000008457   True  True\n",
       "ENCODE-mouseT000008458   True  True\n",
       "ENCODE-mouseT000008459   True  True\n",
       "ENCODE-mouseT000008460  False  True\n",
       "ENCODE-mouseT000008461  False  True\n",
       "ENCODE-mouseT000008462  False  True\n",
       "ENSMUST00000020637.8     True  True\n",
       "ENSMUST00000020741.10   False  True\n",
       "ENSMUST00000105344.6    False  True"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = swan.calc_pi(sg.adata, sg.t_df, obs_col='cluster')\n",
    "\n",
    "# test\n",
    "df = df.transpose()\n",
    "df.index.name = 'tid'\n",
    "df.sort_values(by='tid', inplace=True)\n",
    "print(df)\n",
    "print(df.columns)\n",
    "\n",
    "# ctrl\n",
    "ctrl = pd.read_csv('data/chr11_and_Tcf3_talon_abundance_clusters_pi.tsv', sep='\\t')\n",
    "ctrl.set_index('tid', inplace=True)\n",
    "ctrl.sort_values(by='tid', inplace=True)\n",
    "print(ctrl)\n",
    "\n",
    "ctrl == df\n",
    "\n",
    "# need better way to test this bc of rounding, but it works.. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "coupled-positive",
   "metadata": {},
   "source": [
    "#### test - add abundance and then more abundance and calc pi (in add_abundance)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "political-prophet",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding annotation to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing transcripts:  99%|█████████▉| 8400/8450 [00:01<00:00, 4514.86it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reindexing and sorting entries on genomic location...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reindexing vertices: 100%|█████████▉| 50000/50074 [00:00<00:00, 1288659.76it/s]\n",
      "Reindexing edges:  98%|█████████▊| 49000/50074 [00:00<00:00, 704468.36it/s]\n",
      "Reindexing transcripts:  99%|█████████▉| 8400/8450 [00:00<00:00, 157235.16it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Annotation to the SwanGraph\n",
      "\n",
      "Adding transcriptome to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "AnnData expects .obs.index to contain strings, but your first indices are: Int64Index([0, 1], dtype='int64'), …\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding abundance for datasets PB65_B017, PB65_B018 to SwanGraph.\n",
      "\n",
      "Adding abundance for datasets D12 to SwanGraph.\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "# just gencode vM21 chr 11 and tcf3\n",
    "gtf = 'data/chr11_and_Tcf3.gtf'\n",
    "sg.add_annotation(gtf, verbose=True)\n",
    "db = 'data/chr11_and_Tcf3_no_gname.db'\n",
    "sg.add_transcriptome(db, include_isms=True)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance_PB65.tsv'\n",
    "sg.add_abundance(ab)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance_D12.tsv'\n",
    "sg.add_abundance(ab)\n",
    "meta = 'data/chr11_and_Tcf3_metadata.tsv'\n",
    "sg.add_metadata(meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "numerous-substitute",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                        PB65_B017  PB65_B018    D12\n",
      "tid                                                \n",
      "ENCODE-mouseT000008451       75.0       50.0  100.0\n",
      "ENCODE-mouseT000008452       25.0        0.0    0.0\n",
      "ENCODE-mouseT000008453       50.0      100.0    0.0\n",
      "ENCODE-mouseT000008454      100.0        0.0    0.0\n",
      "ENCODE-mouseT000008455      100.0        0.0    0.0\n",
      "ENCODE-mouseT000008456        0.0        0.0  100.0\n",
      "ENCODE-mouseT000008457        0.0        0.0   25.0\n",
      "ENCODE-mouseT000008458        0.0        0.0   25.0\n",
      "ENCODE-mouseT000008459        0.0        0.0   25.0\n",
      "ENCODE-mouseT000008460        0.0      100.0    0.0\n",
      "ENCODE-mouseT000008461       50.0        0.0    0.0\n",
      "ENCODE-mouseT000008462       50.0        0.0    0.0\n",
      "ENSMUST00000020637.8          0.0        0.0   25.0\n",
      "ENSMUST00000020741.10        50.0        0.0    0.0\n",
      "ENSMUST00000105344.6          0.0       50.0    0.0\n",
      "                        PB65_B017  PB65_B018  D12\n",
      "tid                                              \n",
      "ENCODE-mouseT000008451         75         50  100\n",
      "ENCODE-mouseT000008452         25          0    0\n",
      "ENCODE-mouseT000008453         50        100    0\n",
      "ENCODE-mouseT000008454        100          0    0\n",
      "ENCODE-mouseT000008455        100          0    0\n",
      "ENCODE-mouseT000008456          0          0  100\n",
      "ENCODE-mouseT000008457          0          0   25\n",
      "ENCODE-mouseT000008458          0          0   25\n",
      "ENCODE-mouseT000008459          0          0   25\n",
      "ENCODE-mouseT000008460          0        100    0\n",
      "ENCODE-mouseT000008461         50          0    0\n",
      "ENCODE-mouseT000008462         50          0    0\n",
      "ENSMUST00000020637.8            0          0   25\n",
      "ENSMUST00000020741.10          50          0    0\n",
      "ENSMUST00000105344.6            0         50    0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PB65_B017</th>\n",
       "      <th>PB65_B018</th>\n",
       "      <th>D12</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008451</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008452</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008453</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008454</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008455</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008456</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008457</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008458</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008459</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008460</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008461</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008462</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000020637.8</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000020741.10</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000105344.6</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        PB65_B017  PB65_B018   D12\n",
       "tid                                               \n",
       "ENCODE-mouseT000008451       True       True  True\n",
       "ENCODE-mouseT000008452       True       True  True\n",
       "ENCODE-mouseT000008453       True       True  True\n",
       "ENCODE-mouseT000008454       True       True  True\n",
       "ENCODE-mouseT000008455       True       True  True\n",
       "ENCODE-mouseT000008456       True       True  True\n",
       "ENCODE-mouseT000008457       True       True  True\n",
       "ENCODE-mouseT000008458       True       True  True\n",
       "ENCODE-mouseT000008459       True       True  True\n",
       "ENCODE-mouseT000008460       True       True  True\n",
       "ENCODE-mouseT000008461       True       True  True\n",
       "ENCODE-mouseT000008462       True       True  True\n",
       "ENSMUST00000020637.8         True       True  True\n",
       "ENSMUST00000020741.10        True       True  True\n",
       "ENSMUST00000105344.6         True       True  True"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test\n",
    "df = pd.DataFrame(data=sg.adata.layers['pi'], \\\n",
    "                  columns=sg.adata.var.index.tolist(), \\\n",
    "                  index=sg.adata.obs.index.tolist())\n",
    "df = df.transpose()\n",
    "df.index.name = 'tid'\n",
    "df.sort_values(by='tid', inplace=True)\n",
    "print(df)\n",
    "\n",
    "# ctrl\n",
    "ctrl = pd.read_csv('data/chr11_and_Tcf3_talon_abundance_pi.tsv', sep='\\t')\n",
    "ctrl.set_index('tid', inplace=True)\n",
    "ctrl.sort_values(by='tid', inplace=True)\n",
    "print(ctrl)\n",
    "\n",
    "ctrl == df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "sorted-editing",
   "metadata": {},
   "source": [
    "#### test - calculate percent iso expression per gene (in add_abundance)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "alternate-richmond",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding annotation to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing transcripts:  99%|█████████▉| 8400/8450 [00:01<00:00, 4535.96it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reindexing and sorting entries on genomic location...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reindexing vertices: 100%|█████████▉| 50000/50074 [00:00<00:00, 1227711.37it/s]\n",
      "Reindexing edges:  98%|█████████▊| 49000/50074 [00:00<00:00, 280962.39it/s]\n",
      "Reindexing transcripts:  99%|█████████▉| 8400/8450 [00:00<00:00, 146053.94it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Annotation to the SwanGraph\n",
      "\n",
      "Adding transcriptome to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "AnnData expects .obs.index to contain strings, but your first indices are: Int64Index([0, 1], dtype='int64'), …\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding abundance for datasets PB65_B017, PB65_B018, D12 to SwanGraph.\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "# just gencode vM21 chr 11 and tcf3\n",
    "gtf = 'data/chr11_and_Tcf3.gtf'\n",
    "sg.add_annotation(gtf, verbose=True)\n",
    "db = 'data/chr11_and_Tcf3_no_gname.db'\n",
    "sg.add_transcriptome(db, include_isms=True)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance.tsv'\n",
    "sg.add_abundance(ab)\n",
    "meta = 'data/chr11_and_Tcf3_metadata.tsv'\n",
    "sg.add_metadata(meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "peripheral-monte",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             dataset cluster   sample\n",
      "index                                \n",
      "PB65_B017  PB65_B017       2  GM12878\n",
      "PB65_B018  PB65_B018       2  GM12878\n",
      "D12              D12       1     K562\n",
      "[[ 75.  25.  50. 100. 100.   0.   0.   0.   0.   0.  50.  50.   0.  50.\n",
      "    0.]\n",
      " [ 50.   0. 100.   0.   0.   0.   0.   0.   0. 100.   0.   0.   0.   0.\n",
      "   50.]\n",
      " [100.   0.   0.   0.   0. 100.  25.  25.  25.   0.   0.   0.  25.   0.\n",
      "    0.]]\n"
     ]
    }
   ],
   "source": [
    "print(sg.adata.obs)\n",
    "print(sg.adata.layers['pi'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "unnecessary-kidney",
   "metadata": {},
   "source": [
    "The following two should be equivalent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "collected-drove",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PB65_B017</th>\n",
       "      <th>PB65_B018</th>\n",
       "      <th>D12</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008451</th>\n",
       "      <td>75.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008452</th>\n",
       "      <td>25.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008453</th>\n",
       "      <td>50.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008454</th>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008455</th>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008456</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008457</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008458</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008459</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008460</th>\n",
       "      <td>0.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008461</th>\n",
       "      <td>50.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008462</th>\n",
       "      <td>50.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000020637.8</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000020741.10</th>\n",
       "      <td>50.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000105344.6</th>\n",
       "      <td>0.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        PB65_B017  PB65_B018    D12\n",
       "tid                                                \n",
       "ENCODE-mouseT000008451       75.0       50.0  100.0\n",
       "ENCODE-mouseT000008452       25.0        0.0    0.0\n",
       "ENCODE-mouseT000008453       50.0      100.0    0.0\n",
       "ENCODE-mouseT000008454      100.0        0.0    0.0\n",
       "ENCODE-mouseT000008455      100.0        0.0    0.0\n",
       "ENCODE-mouseT000008456        0.0        0.0  100.0\n",
       "ENCODE-mouseT000008457        0.0        0.0   25.0\n",
       "ENCODE-mouseT000008458        0.0        0.0   25.0\n",
       "ENCODE-mouseT000008459        0.0        0.0   25.0\n",
       "ENCODE-mouseT000008460        0.0      100.0    0.0\n",
       "ENCODE-mouseT000008461       50.0        0.0    0.0\n",
       "ENCODE-mouseT000008462       50.0        0.0    0.0\n",
       "ENSMUST00000020637.8          0.0        0.0   25.0\n",
       "ENSMUST00000020741.10        50.0        0.0    0.0\n",
       "ENSMUST00000105344.6          0.0       50.0    0.0"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df = pd.read_csv('data/chr11_and_Tcf3_talon_abundance_pi.tsv', sep='\\t')\n",
    "df = pd.DataFrame(data=sg.adata.layers['pi'], \\\n",
    "                  columns=sg.adata.var.index.tolist(), \\\n",
    "                  index=sg.adata.obs.index.tolist())\n",
    "df = df.transpose()\n",
    "df.index.name = 'tid'\n",
    "df.sort_values(by='tid', inplace=True)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ruled-darwin",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PB65_B017</th>\n",
       "      <th>PB65_B018</th>\n",
       "      <th>D12</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008451</th>\n",
       "      <td>75</td>\n",
       "      <td>50</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008452</th>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008453</th>\n",
       "      <td>50</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008454</th>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008455</th>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008456</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008457</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008458</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008459</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008460</th>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008461</th>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008462</th>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000020637.8</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000020741.10</th>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000105344.6</th>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        PB65_B017  PB65_B018  D12\n",
       "tid                                              \n",
       "ENCODE-mouseT000008451         75         50  100\n",
       "ENCODE-mouseT000008452         25          0    0\n",
       "ENCODE-mouseT000008453         50        100    0\n",
       "ENCODE-mouseT000008454        100          0    0\n",
       "ENCODE-mouseT000008455        100          0    0\n",
       "ENCODE-mouseT000008456          0          0  100\n",
       "ENCODE-mouseT000008457          0          0   25\n",
       "ENCODE-mouseT000008458          0          0   25\n",
       "ENCODE-mouseT000008459          0          0   25\n",
       "ENCODE-mouseT000008460          0        100    0\n",
       "ENCODE-mouseT000008461         50          0    0\n",
       "ENCODE-mouseT000008462         50          0    0\n",
       "ENSMUST00000020637.8            0          0   25\n",
       "ENSMUST00000020741.10          50          0    0\n",
       "ENSMUST00000105344.6            0         50    0"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# tests\n",
    "ctrl = pd.read_csv('data/chr11_and_Tcf3_talon_abundance_pi.tsv', sep='\\t')\n",
    "ctrl.set_index('tid', inplace=True)\n",
    "ctrl.sort_values(by='tid', inplace=True)\n",
    "ctrl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "aggressive-abortion",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PB65_B017</th>\n",
       "      <th>PB65_B018</th>\n",
       "      <th>D12</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008451</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008452</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008453</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008454</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008455</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008456</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008457</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008458</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008459</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008460</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008461</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODE-mouseT000008462</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000020637.8</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000020741.10</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000105344.6</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        PB65_B017  PB65_B018   D12\n",
       "tid                                               \n",
       "ENCODE-mouseT000008451       True       True  True\n",
       "ENCODE-mouseT000008452       True       True  True\n",
       "ENCODE-mouseT000008453       True       True  True\n",
       "ENCODE-mouseT000008454       True       True  True\n",
       "ENCODE-mouseT000008455       True       True  True\n",
       "ENCODE-mouseT000008456       True       True  True\n",
       "ENCODE-mouseT000008457       True       True  True\n",
       "ENCODE-mouseT000008458       True       True  True\n",
       "ENCODE-mouseT000008459       True       True  True\n",
       "ENCODE-mouseT000008460       True       True  True\n",
       "ENCODE-mouseT000008461       True       True  True\n",
       "ENCODE-mouseT000008462       True       True  True\n",
       "ENSMUST00000020637.8         True       True  True\n",
       "ENSMUST00000020741.10        True       True  True\n",
       "ENSMUST00000105344.6         True       True  True"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctrl == df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "desirable-potter",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "KeysView(Layers with keys: counts, tpm, pi)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.adata.layers.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "integral-works",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[100.,   0.,   0.,   0., 100.,  25.,  25.,  25.,  25.,   0.,   0.],\n",
       "       [100.,  50., 100., 100.,   0.,   0.,   0.,   0.,   0.,  50.,   0.],\n",
       "       [ 50., 100.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,  50.]],\n",
       "      dtype=float32)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.adata.layers['pi']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "extreme-prince",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   ENCODE-mouseT000008451  ENSMUST00000105344.6\n",
      "1                     2.0                   0.0\n",
      "2                     4.0                   1.0\n",
      "                         1_pi  2_pi\n",
      "ENCODE-mouseT000008451  100.0  80.0\n",
      "ENSMUST00000105344.6      0.0  20.0\n",
      "\n",
      "\n",
      "   ENCODE-mouseT000008453  ENSMUST00000020741.10\n",
      "1                     0.0                    0.0\n",
      "2                     2.0                    1.0\n",
      "                        1_pi       2_pi\n",
      "ENCODE-mouseT000008453   0.0  66.666672\n",
      "ENSMUST00000020741.10    0.0  33.333336\n",
      "\n",
      "\n",
      "   ENCODE-mouseT000008454\n",
      "1                     0.0\n",
      "2                     1.0\n",
      "                        1_pi   2_pi\n",
      "ENCODE-mouseT000008454   0.0  100.0\n",
      "\n",
      "\n",
      "   ENCODE-mouseT000008455\n",
      "1                     0.0\n",
      "2                     1.0\n",
      "                        1_pi   2_pi\n",
      "ENCODE-mouseT000008455   0.0  100.0\n",
      "\n",
      "\n",
      "   ENCODE-mouseT000008456\n",
      "1                     1.0\n",
      "2                     0.0\n",
      "                         1_pi  2_pi\n",
      "ENCODE-mouseT000008456  100.0   0.0\n",
      "\n",
      "\n",
      "   ENCODE-mouseT000008457  ENCODE-mouseT000008458  ENCODE-mouseT000008459  \\\n",
      "1                     1.0                     1.0                     1.0   \n",
      "2                     0.0                     0.0                     0.0   \n",
      "\n",
      "   ENSMUST00000020637.8  \n",
      "1                   1.0  \n",
      "2                   0.0  \n",
      "                        1_pi  2_pi\n",
      "ENCODE-mouseT000008457  25.0   0.0\n",
      "ENCODE-mouseT000008458  25.0   0.0\n",
      "ENCODE-mouseT000008459  25.0   0.0\n",
      "ENSMUST00000020637.8    25.0   0.0\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/anndata/_core/anndata.py:1056: FutureWarning: is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead\n",
      "  if not is_categorical(df_full[k]):\n",
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/anndata/_core/anndata.py:1056: FutureWarning: is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead\n",
      "  if not is_categorical(df_full[k]):\n",
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/anndata/_core/anndata.py:1056: FutureWarning: is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead\n",
      "  if not is_categorical(df_full[k]):\n",
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/anndata/_core/anndata.py:1056: FutureWarning: is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead\n",
      "  if not is_categorical(df_full[k]):\n",
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/anndata/_core/anndata.py:1056: FutureWarning: is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead\n",
      "  if not is_categorical(df_full[k]):\n",
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/anndata/_core/anndata.py:1056: FutureWarning: is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead\n",
      "  if not is_categorical(df_full[k]):\n"
     ]
    }
   ],
   "source": [
    "# conditions = ['PB65_B017', 'PB65_B018', 'D12']\n",
    "# cond = 'dataset'\n",
    "cond = 'cluster'\n",
    "\n",
    "# first set default assay to counts\n",
    "sg.adata.X = sg.adata.layers['counts']\n",
    "\n",
    "for g in sg.t_df.gid.unique():\n",
    "    tids = sg.t_df.loc[sg.t_df.gid == g, 'tid'].tolist()\n",
    "    temp = sg.adata[:, tids]\n",
    "    temp = pd.DataFrame(data=temp.X, index=temp.obs[cond].tolist(), columns=temp.var.index.tolist())\n",
    "    \n",
    "    temp = temp.groupby(level=0).sum()\n",
    "    conditions = temp.index.tolist()\n",
    "    print(temp)\n",
    "    temp = temp.transpose()\n",
    "    \n",
    "    # aggregate on conditions\n",
    "    pi_cols = []\n",
    "    for c in conditions:\n",
    "        total_col = '{}_total'.format(c)\n",
    "        pi_col = '{}_pi'.format(c)\n",
    "        total_count = temp[c].sum()\n",
    "        temp[total_col] = total_count\n",
    "        temp[pi_col] = (temp[c]/temp[total_col])*100\n",
    "        pi_cols.append(pi_col)\n",
    "        \n",
    "    # fill nans (no expression) with 0\n",
    "    temp.fillna(0, inplace=True)\n",
    "    \n",
    "    # only pi columns\n",
    "    temp = temp[pi_cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "fifth-validity",
   "metadata": {},
   "outputs": [],
   "source": [
    "# calc pi\n",
    "\n",
    "cond = 'dataset'\n",
    "df = pd.DataFrame(data=sg.adata.X, index=sg.adata.obs[cond].tolist(), columns=sg.adata.var.index.tolist())\n",
    "\n",
    "# add up values on condition (row)\n",
    "# print(df.head())\n",
    "df = df.groupby(level=0).sum()\n",
    "# print(df.head())\n",
    "conditions = df.index.unique().tolist()\n",
    "# print(conditions)\n",
    "df = df.transpose()\n",
    "\n",
    "# add gid \n",
    "df = df.merge(sg.t_df['gid'], how='left', left_index=True, right_index=True)\n",
    "\n",
    "# calculate total number of reads per gene per condition\n",
    "# print(df.loc[df.gid == 'ENSMUSG00000020167.13'])\n",
    "temp = df.copy(deep=True)\n",
    "temp.reset_index(drop=True, inplace=True)\n",
    "totals = temp.groupby('gid').sum().reset_index()\n",
    "# print(totals.head())\n",
    "\n",
    "# merge back in \n",
    "df.reset_index(inplace=True)\n",
    "df.rename({'index':'tid'}, axis=1, inplace=True)\n",
    "df = df.merge(totals, on='gid', suffixes=(None, '_total'))\n",
    "# print(df.head())\n",
    "del totals\n",
    "\n",
    "# calculate percent iso exp for each gene / transcript / condition\n",
    "pi_cols = []\n",
    "for c in conditions:\n",
    "    cond_col = '{}_pi'.format(c)\n",
    "    total_col = '{}_total'.format(c)\n",
    "    df[cond_col] = (df[c]/df[total_col])*100\n",
    "    pi_cols.append(cond_col)\n",
    "    \n",
    "# fill nans with 0 \n",
    "df.fillna(0, inplace=True)\n",
    "# print(df.head())\n",
    "# \n",
    "# formatting\n",
    "df.set_index('tid', inplace=True)\n",
    "df = df[pi_cols]\n",
    "for col in pi_cols:\n",
    "    new_col = col[:-3]\n",
    "    df.rename({col: new_col}, axis=1, inplace=True)\n",
    "df = df.transpose()\n",
    "# df.set_index('tid', inplace=True)\n",
    "\n",
    "# reorder in sg.adata.var / sg.t_df order\n",
    "df = df[sg.t_df.tid.tolist()]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "generic-firmware",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>tid</th>\n",
       "      <th>ENCODE-mouseT000008451</th>\n",
       "      <th>ENCODE-mouseT000008452</th>\n",
       "      <th>ENCODE-mouseT000008453</th>\n",
       "      <th>ENCODE-mouseT000008454</th>\n",
       "      <th>ENCODE-mouseT000008455</th>\n",
       "      <th>ENCODE-mouseT000008456</th>\n",
       "      <th>ENCODE-mouseT000008457</th>\n",
       "      <th>ENCODE-mouseT000008458</th>\n",
       "      <th>ENCODE-mouseT000008459</th>\n",
       "      <th>ENCODE-mouseT000008460</th>\n",
       "      <th>ENCODE-mouseT000008461</th>\n",
       "      <th>ENCODE-mouseT000008462</th>\n",
       "      <th>ENSMUST00000020637.8</th>\n",
       "      <th>ENSMUST00000020741.10</th>\n",
       "      <th>ENSMUST00000105344.6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>D12</th>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PB65_B017</th>\n",
       "      <td>75.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PB65_B018</th>\n",
       "      <td>50.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "tid        ENCODE-mouseT000008451  ENCODE-mouseT000008452  \\\n",
       "D12                         100.0                     0.0   \n",
       "PB65_B017                    75.0                    25.0   \n",
       "PB65_B018                    50.0                     0.0   \n",
       "\n",
       "tid        ENCODE-mouseT000008453  ENCODE-mouseT000008454  \\\n",
       "D12                           0.0                     0.0   \n",
       "PB65_B017                    50.0                   100.0   \n",
       "PB65_B018                   100.0                     0.0   \n",
       "\n",
       "tid        ENCODE-mouseT000008455  ENCODE-mouseT000008456  \\\n",
       "D12                           0.0                   100.0   \n",
       "PB65_B017                   100.0                     0.0   \n",
       "PB65_B018                     0.0                     0.0   \n",
       "\n",
       "tid        ENCODE-mouseT000008457  ENCODE-mouseT000008458  \\\n",
       "D12                          25.0                    25.0   \n",
       "PB65_B017                     0.0                     0.0   \n",
       "PB65_B018                     0.0                     0.0   \n",
       "\n",
       "tid        ENCODE-mouseT000008459  ENCODE-mouseT000008460  \\\n",
       "D12                          25.0                     0.0   \n",
       "PB65_B017                     0.0                     0.0   \n",
       "PB65_B018                     0.0                   100.0   \n",
       "\n",
       "tid        ENCODE-mouseT000008461  ENCODE-mouseT000008462  \\\n",
       "D12                           0.0                     0.0   \n",
       "PB65_B017                    50.0                    50.0   \n",
       "PB65_B018                     0.0                     0.0   \n",
       "\n",
       "tid        ENSMUST00000020637.8  ENSMUST00000020741.10  ENSMUST00000105344.6  \n",
       "D12                        25.0                    0.0                   0.0  \n",
       "PB65_B017                   0.0                   50.0                   0.0  \n",
       "PB65_B018                   0.0                    0.0                  50.0  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "electronic-transcription",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make pi layer\n",
    "sg.adata.layers['pi'] = df.to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "loving-duration",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3, 11)\n",
      "(3, 11)\n"
     ]
    }
   ],
   "source": [
    "print(np.shape(df.to_numpy()))\n",
    "print(np.shape(sg.adata.X))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "behind-tonight",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11\n",
      "11\n",
      "True\n",
      "['ENCODE-mouseT000008451', 'ENCODE-mouseT000008453', 'ENCODE-mouseT000008454', 'ENCODE-mouseT000008455', 'ENCODE-mouseT000008456', 'ENCODE-mouseT000008457', 'ENCODE-mouseT000008458', 'ENCODE-mouseT000008459', 'ENSMUST00000020637.8', 'ENSMUST00000020741.10', 'ENSMUST00000105344.6']\n",
      "\n",
      "['ENCODE-mouseT000008451', 'ENCODE-mouseT000008453', 'ENCODE-mouseT000008454', 'ENCODE-mouseT000008455', 'ENCODE-mouseT000008456', 'ENCODE-mouseT000008457', 'ENCODE-mouseT000008458', 'ENCODE-mouseT000008459', 'ENSMUST00000020637.8', 'ENSMUST00000020741.10', 'ENSMUST00000105344.6']\n",
      "\n",
      "['ENCODE-mouseT000008451', 'ENCODE-mouseT000008453', 'ENCODE-mouseT000008454', 'ENCODE-mouseT000008455', 'ENCODE-mouseT000008456', 'ENCODE-mouseT000008457', 'ENCODE-mouseT000008458', 'ENCODE-mouseT000008459', 'ENSMUST00000020637.8', 'ENSMUST00000020741.10', 'ENSMUST00000105344.6']\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "# check that the orders are the same across t_df, new layer, and adata.obs\n",
    "print(len(df.columns))\n",
    "print(len(sg.adata.var.index.tolist()))\n",
    "print(df.columns.tolist() == sg.adata.var.index.tolist())\n",
    "print(df.columns.tolist())\n",
    "print()\n",
    "print(sg.adata.var.index.tolist())\n",
    "print()\n",
    "print(sg.t_df.index.tolist())\n",
    "print(sg.adata.var.index.tolist()==sg.t_df.index.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "experienced-container",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(11, 1)\n",
      "(3, 11)\n",
      "(3, 3)\n"
     ]
    }
   ],
   "source": [
    "print(np.shape(sg.adata.var))\n",
    "print(np.shape(sg.adata.X))\n",
    "print(np.shape(sg.adata.obs))"
   ]
  },
  {
   "cell_type": "raw",
   "id": "light-mouth",
   "metadata": {},
   "source": [
    "temp.X"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "floating-zimbabwe",
   "metadata": {},
   "source": [
    "#### test - add metadata one after another with dupe values, don't overwrite"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "korean-hotel",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding annotation to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing transcripts:  99%|█████████▉| 8400/8450 [00:02<00:00, 4166.37it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reindexing and sorting entries on genomic location...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reindexing vertices: 100%|█████████▉| 50000/50074 [00:00<00:00, 970872.24it/s] \n",
      "Reindexing edges:  98%|█████████▊| 49000/50074 [00:00<00:00, 628898.53it/s]\n",
      "Reindexing transcripts:  99%|█████████▉| 8400/8450 [00:00<00:00, 147347.67it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Annotation to the SwanGraph\n",
      "\n",
      "Adding transcriptome to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "AnnData expects .obs.index to contain strings, but your first indices are: Int64Index([0, 1], dtype='int64'), …\n",
      "AnnData expects .obs.index to contain strings, but your first indices are: Int64Index([0, 1], dtype='int64'), …\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding abundance for datasets PB65_B017, PB65_B018, D12 to SwanGraph.\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "# just gencode vM21 chr 11 and tcf3\n",
    "gtf = 'data/chr11_and_Tcf3.gtf'\n",
    "sg.add_annotation(gtf, verbose=True)\n",
    "db = 'data/chr11_and_Tcf3_no_gname.db'\n",
    "sg.add_transcriptome(db)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance.tsv'\n",
    "sg.add_abundance(ab)\n",
    "meta = 'data/chr11_and_Tcf3_metadata.tsv'\n",
    "sg.add_metadata(meta)\n",
    "meta = 'data/chr11_and_Tcf3_metadata_dupecols.tsv'\n",
    "sg.add_metadata(meta, overwrite=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "lightweight-command",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "      <th>cluster</th>\n",
       "      <th>sample</th>\n",
       "      <th>cell_state</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PB65_B017</th>\n",
       "      <td>PB65_B017</td>\n",
       "      <td>2</td>\n",
       "      <td>GM12878</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PB65_B018</th>\n",
       "      <td>PB65_B018</td>\n",
       "      <td>2</td>\n",
       "      <td>GM12878</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D12</th>\n",
       "      <td>D12</td>\n",
       "      <td>1</td>\n",
       "      <td>K562</td>\n",
       "      <td>G0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             dataset cluster   sample cell_state\n",
       "index                                           \n",
       "PB65_B017  PB65_B017       2  GM12878          M\n",
       "PB65_B018  PB65_B018       2  GM12878          S\n",
       "D12              D12       1     K562         G0"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.adata.obs # clusters should be 1 and 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "stunning-formula",
   "metadata": {},
   "source": [
    "#### test - add metadata one after another with dupe values, overwrite"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "external-serbia",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding annotation to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing transcripts:  99%|█████████▉| 8400/8450 [00:01<00:00, 4455.59it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reindexing and sorting entries on genomic location...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reindexing vertices: 100%|█████████▉| 50000/50074 [00:00<00:00, 1261475.16it/s]\n",
      "Reindexing edges:  98%|█████████▊| 49000/50074 [00:00<00:00, 685053.67it/s]\n",
      "Reindexing transcripts:  99%|█████████▉| 8400/8450 [00:00<00:00, 155110.70it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Annotation to the SwanGraph\n",
      "\n",
      "Adding transcriptome to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "AnnData expects .obs.index to contain strings, but your first indices are: Int64Index([0, 1], dtype='int64'), …\n",
      "AnnData expects .obs.index to contain strings, but your first indices are: Int64Index([0, 1], dtype='int64'), …\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding abundance for datasets PB65_B017, PB65_B018, D12 to SwanGraph.\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "# just gencode vM21 chr 11 and tcf3\n",
    "gtf = 'data/chr11_and_Tcf3.gtf'\n",
    "sg.add_annotation(gtf, verbose=True)\n",
    "db = 'data/chr11_and_Tcf3_no_gname.db'\n",
    "sg.add_transcriptome(db)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance.tsv'\n",
    "sg.add_abundance(ab)\n",
    "meta = 'data/chr11_and_Tcf3_metadata.tsv'\n",
    "sg.add_metadata(meta)\n",
    "meta = 'data/chr11_and_Tcf3_metadata_dupecols.tsv'\n",
    "sg.add_metadata(meta, overwrite=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ranging-genius",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "      <th>sample</th>\n",
       "      <th>cell_state</th>\n",
       "      <th>cluster</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PB65_B017</th>\n",
       "      <td>PB65_B017</td>\n",
       "      <td>GM12878</td>\n",
       "      <td>M</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PB65_B018</th>\n",
       "      <td>PB65_B018</td>\n",
       "      <td>GM12878</td>\n",
       "      <td>S</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D12</th>\n",
       "      <td>D12</td>\n",
       "      <td>K562</td>\n",
       "      <td>G0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             dataset   sample cell_state cluster\n",
       "index                                           \n",
       "PB65_B017  PB65_B017  GM12878          M       3\n",
       "PB65_B018  PB65_B018  GM12878          S       4\n",
       "D12              D12     K562         G0       3"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.adata.obs # clusters should be 3 and 4 now"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "broke-donor",
   "metadata": {},
   "source": [
    "#### test - add metadata one after another to obs table - PYTESTED"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "reported-tobago",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['beep']\n",
      "['bop', 'beep']\n",
      "['beep']\n"
     ]
    }
   ],
   "source": [
    "a = ['dataset', 'beep']\n",
    "b = ['beep', 'bop', 'dataset']\n",
    "sg_cols = list(set(a)-set(['dataset']))\n",
    "meta_cols = list(set(b)-set(['dataset']))\n",
    "print(sg_cols) # should be ['beep']\n",
    "print(meta_cols) # should be ['beep', 'bop']\n",
    "dupe_cols = [c for c in meta_cols if c in sg_cols]\n",
    "print(dupe_cols) # should be ['beep']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "detected-signature",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding annotation to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing transcripts:  99%|█████████▉| 8400/8450 [00:01<00:00, 4469.69it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reindexing and sorting entries on genomic location...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reindexing vertices: 100%|█████████▉| 50000/50074 [00:00<00:00, 1200657.25it/s]\n",
      "Reindexing edges:  98%|█████████▊| 49000/50074 [00:00<00:00, 286162.69it/s]\n",
      "Reindexing transcripts:  99%|█████████▉| 8400/8450 [00:00<00:00, 148394.65it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Annotation to the SwanGraph\n",
      "\n",
      "Adding transcriptome to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "AnnData expects .obs.index to contain strings, but your first indices are: Int64Index([0, 1], dtype='int64'), …\n",
      "AnnData expects .obs.index to contain strings, but your first indices are: Int64Index([0, 1], dtype='int64'), …\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding abundance for datasets PB65_B017, PB65_B018, D12 to SwanGraph.\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "# just gencode vM21 chr 11 and tcf3\n",
    "gtf = 'data/chr11_and_Tcf3.gtf'\n",
    "sg.add_annotation(gtf, verbose=True)\n",
    "db = 'data/chr11_and_Tcf3_no_gname.db'\n",
    "sg.add_transcriptome(db)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance.tsv'\n",
    "sg.add_abundance(ab)\n",
    "meta = 'data/chr11_and_Tcf3_metadata.tsv'\n",
    "sg.add_metadata(meta)\n",
    "meta = 'data/chr11_and_Tcf3_metadata_2.tsv'\n",
    "sg.add_metadata(meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "contemporary-entertainment",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tname</th>\n",
       "      <th>gid</th>\n",
       "      <th>gname</th>\n",
       "      <th>path</th>\n",
       "      <th>tid</th>\n",
       "      <th>loc_path</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>test1</th>\n",
       "      <td>test1_tname</td>\n",
       "      <td>test1_gid</td>\n",
       "      <td>test1_gname</td>\n",
       "      <td>[0, 1, 2, 3, 4]</td>\n",
       "      <td>test1</td>\n",
       "      <td>[0, 1, 2, 3, 4, 5]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test2</th>\n",
       "      <td>test2_tname</td>\n",
       "      <td>test2_gid</td>\n",
       "      <td>test2_gname</td>\n",
       "      <td>[5, 6, 7, 8, 9]</td>\n",
       "      <td>test2</td>\n",
       "      <td>[12, 11, 10, 8, 7, 6]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test3</th>\n",
       "      <td>test3_tname</td>\n",
       "      <td>test2_gid</td>\n",
       "      <td>test2_gname</td>\n",
       "      <td>[5, 6, 13, 14, 9]</td>\n",
       "      <td>test3</td>\n",
       "      <td>[12, 11, 10, 9, 7, 6]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test4</th>\n",
       "      <td>test4_tname</td>\n",
       "      <td>test4_gid</td>\n",
       "      <td>test4_gname</td>\n",
       "      <td>[10]</td>\n",
       "      <td>test4</td>\n",
       "      <td>[6, 7]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test5</th>\n",
       "      <td>test5_tname</td>\n",
       "      <td>test2_gid</td>\n",
       "      <td>test2_gname</td>\n",
       "      <td>[5, 11, 12]</td>\n",
       "      <td>test5</td>\n",
       "      <td>[12, 11, 8, 7]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             tname        gid        gname               path    tid  \\\n",
       "tid                                                                    \n",
       "test1  test1_tname  test1_gid  test1_gname    [0, 1, 2, 3, 4]  test1   \n",
       "test2  test2_tname  test2_gid  test2_gname    [5, 6, 7, 8, 9]  test2   \n",
       "test3  test3_tname  test2_gid  test2_gname  [5, 6, 13, 14, 9]  test3   \n",
       "test4  test4_tname  test4_gid  test4_gname               [10]  test4   \n",
       "test5  test5_tname  test2_gid  test2_gname        [5, 11, 12]  test5   \n",
       "\n",
       "                    loc_path  \n",
       "tid                           \n",
       "test1     [0, 1, 2, 3, 4, 5]  \n",
       "test2  [12, 11, 10, 8, 7, 6]  \n",
       "test3  [12, 11, 10, 9, 7, 6]  \n",
       "test4                 [6, 7]  \n",
       "test5         [12, 11, 8, 7]  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.t_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "secondary-canyon",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "      <th>cluster</th>\n",
       "      <th>sample</th>\n",
       "      <th>cell_state</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PB65_B017</th>\n",
       "      <td>PB65_B017</td>\n",
       "      <td>2</td>\n",
       "      <td>GM12878</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PB65_B018</th>\n",
       "      <td>PB65_B018</td>\n",
       "      <td>2</td>\n",
       "      <td>GM12878</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D12</th>\n",
       "      <td>D12</td>\n",
       "      <td>1</td>\n",
       "      <td>K562</td>\n",
       "      <td>G0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             dataset cluster   sample cell_state\n",
       "index                                           \n",
       "PB65_B017  PB65_B017       2  GM12878          M\n",
       "PB65_B018  PB65_B018       2  GM12878          S\n",
       "D12              D12       1     K562         G0"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "registered-motor",
   "metadata": {},
   "source": [
    "#### test - add metadata to obs table\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "criminal-breast",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "AnnData expects .obs.index to contain strings, but your first indices are: Int64Index([0, 1], dtype='int64'), …\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(meta, sep='\\t')\n",
    "sg.adata.obs = sg.adata.obs.merge(df, how='left', on='dataset')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "southern-response",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding annotation to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing transcripts:  99%|█████████▉| 8400/8450 [00:01<00:00, 4477.48it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reindexing and sorting entries on genomic location...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reindexing vertices: 100%|█████████▉| 50000/50074 [00:00<00:00, 1256133.16it/s]\n",
      "Reindexing edges:  98%|█████████▊| 49000/50074 [00:00<00:00, 676068.92it/s]\n",
      "Reindexing transcripts:  99%|█████████▉| 8400/8450 [00:00<00:00, 132285.60it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Annotation to the SwanGraph\n",
      "\n",
      "Adding transcriptome to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "AnnData expects .obs.index to contain strings, but your first indices are: Int64Index([0, 1], dtype='int64'), …\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding abundance for datasets PB65_B017, PB65_B018, D12 to SwanGraph.\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "# just gencode vM21 chr 11 and tcf3\n",
    "gtf = 'data/chr11_and_Tcf3.gtf'\n",
    "sg.add_annotation(gtf, verbose=True)\n",
    "db = 'data/chr11_and_Tcf3_no_gname.db'\n",
    "sg.add_transcriptome(db)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance.tsv'\n",
    "sg.add_abundance(ab)\n",
    "meta = 'data/chr11_and_Tcf3_metadata.tsv'\n",
    "sg.add_metadata(meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "egyptian-folder",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "      <th>cluster</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PB65_B017</th>\n",
       "      <td>PB65_B017</td>\n",
       "      <td>2</td>\n",
       "      <td>GM12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PB65_B018</th>\n",
       "      <td>PB65_B018</td>\n",
       "      <td>2</td>\n",
       "      <td>GM12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D12</th>\n",
       "      <td>D12</td>\n",
       "      <td>1</td>\n",
       "      <td>K562</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             dataset cluster   sample\n",
       "index                                \n",
       "PB65_B017  PB65_B017       2  GM12878\n",
       "PB65_B018  PB65_B018       2  GM12878\n",
       "D12              D12       1     K562"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "collectible-authentication",
   "metadata": {},
   "source": [
    "#### test - add abundance directly from TALON ab file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "tested-detector",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding annotation to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing transcripts:  99%|█████████▉| 8400/8450 [00:01<00:00, 4307.46it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reindexing and sorting entries on genomic location...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reindexing vertices: 100%|█████████▉| 50000/50074 [00:00<00:00, 969933.77it/s] \n",
      "Reindexing edges:  98%|█████████▊| 49000/50074 [00:00<00:00, 633928.53it/s]\n",
      "Reindexing transcripts:  99%|█████████▉| 8400/8450 [00:00<00:00, 151474.25it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Annotation to the SwanGraph\n",
      "\n",
      "Adding transcriptome to the SwanGraph\n",
      "['gene_ID', 'transcript_ID', 'annot_gene_id', 'annot_transcript_id', 'annot_gene_name', 'annot_transcript_name', 'n_exons', 'length', 'gene_novelty', 'transcript_novelty', 'ISM_subtype']\n",
      "['gene_ID', 'transcript_ID', 'annot_gene_id', 'annot_transcript_id', 'annot_gene_name', 'annot_transcript_name', 'n_exons', 'length', 'gene_novelty', 'transcript_novelty', 'ISM_subtype']\n",
      "hewwo?\n",
      "\n",
      "Adding abundance for datasets PB65_B017, PB65_B018, D12 to SwanGraph.\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "# just gencode vM21 chr 11 and tcf3\n",
    "gtf = 'data/chr11_and_Tcf3.gtf'\n",
    "sg.add_annotation(gtf, verbose=True)\n",
    "db = 'data/chr11_and_Tcf3_no_gname.db'\n",
    "sg.add_transcriptome(db)\n",
    "ab = 'data/chr11_and_Tcf3_talon_abundance.tsv'\n",
    "sg.add_abundance(ab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "thirty-facing",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1., 1., 1., 1., 0.],\n",
       "       [0., 1., 0., 0., 0.],\n",
       "       [0., 0., 0., 0., 1.]], dtype=float32)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.adata.layers['counts'][:5, :5]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "roman-escape",
   "metadata": {},
   "source": [
    "#### test - reformat talon ab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "difficult-ready",
   "metadata": {},
   "outputs": [],
   "source": [
    "ab = 'data/chr11_and_Tcf3_talon_abundance.tsv'\n",
    "new_ab = 'data/chr11_and_Tcf3_swan_abundance.tsv'\n",
    "swan.reformat_talon_abundance(ab, new_ab)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "recognized-bidder",
   "metadata": {},
   "source": [
    "#### test - add transcriptome from db and abundance from tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "conceptual-cycle",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding annotation to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing transcripts:  99%|█████████▉| 8400/8450 [00:01<00:00, 4438.75it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reindexing and sorting entries on genomic location...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reindexing vertices: 100%|█████████▉| 50000/50074 [00:00<00:00, 1183494.36it/s]\n",
      "Reindexing edges:  98%|█████████▊| 49000/50074 [00:00<00:00, 515513.75it/s]\n",
      "Reindexing transcripts:  99%|█████████▉| 8400/8450 [00:00<00:00, 112191.45it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Annotation to the SwanGraph\n",
      "\n",
      "Adding transcriptome to the SwanGraph\n",
      "\n",
      "Adding abundance for datasets PB65_B017, PB65_B018, D12 to SwanGraph.\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "# just gencode vM21 chr 11 and tcf3\n",
    "gtf = 'data/chr11_and_Tcf3.gtf'\n",
    "sg.add_annotation(gtf, verbose=True)\n",
    "db = 'data/chr11_and_Tcf3_no_gname.db'\n",
    "sg.add_transcriptome(db)\n",
    "new_ab = 'data/chr11_and_Tcf3_swan_abundance.tsv'\n",
    "sg.add_abundance(new_ab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "classified-wonder",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>tname</th>\n",
       "      <th>gid</th>\n",
       "      <th>gname</th>\n",
       "      <th>path</th>\n",
       "      <th>annotation</th>\n",
       "      <th>novelty</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENST00000023939.8</th>\n",
       "      <td>ENST00000023939.8</td>\n",
       "      <td>RTF2-201</td>\n",
       "      <td>ENSG00000022277.12</td>\n",
       "      <td>RTF2</td>\n",
       "      <td>[9022, 9029, 9035, 9036, 9037, 9039, 9040, 904...</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000045083.6</th>\n",
       "      <td>ENST00000045083.6</td>\n",
       "      <td>RIPOR3-201</td>\n",
       "      <td>ENSG00000042062.11</td>\n",
       "      <td>RIPOR3</td>\n",
       "      <td>[13836, 13837, 13838, 13839, 13840, 13841, 138...</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000062104.3</th>\n",
       "      <td>ENST00000062104.3</td>\n",
       "      <td>NNAT-201</td>\n",
       "      <td>ENSG00000053438.9</td>\n",
       "      <td>NNAT</td>\n",
       "      <td>[5907, 5912, 5913, 5916, 5918, 5923]</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000064571.2</th>\n",
       "      <td>ENST00000064571.2</td>\n",
       "      <td>CBLN4-201</td>\n",
       "      <td>ENSG00000054803.3</td>\n",
       "      <td>CBLN4</td>\n",
       "      <td>[13398, 13399, 13400, 13401, 13402, 13403]</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000170150.3</th>\n",
       "      <td>ENST00000170150.3</td>\n",
       "      <td>BPIFB2-201</td>\n",
       "      <td>ENSG00000078898.6</td>\n",
       "      <td>BPIFB2</td>\n",
       "      <td>[4157, 4159, 4163, 4166, 4168, 4170, 4172, 417...</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 tid       tname                 gid   gname  \\\n",
       "tid                                                                            \n",
       "ENST00000023939.8  ENST00000023939.8    RTF2-201  ENSG00000022277.12    RTF2   \n",
       "ENST00000045083.6  ENST00000045083.6  RIPOR3-201  ENSG00000042062.11  RIPOR3   \n",
       "ENST00000062104.3  ENST00000062104.3    NNAT-201   ENSG00000053438.9    NNAT   \n",
       "ENST00000064571.2  ENST00000064571.2   CBLN4-201   ENSG00000054803.3   CBLN4   \n",
       "ENST00000170150.3  ENST00000170150.3  BPIFB2-201   ENSG00000078898.6  BPIFB2   \n",
       "\n",
       "                                                                path  \\\n",
       "tid                                                                    \n",
       "ENST00000023939.8  [9022, 9029, 9035, 9036, 9037, 9039, 9040, 904...   \n",
       "ENST00000045083.6  [13836, 13837, 13838, 13839, 13840, 13841, 138...   \n",
       "ENST00000062104.3               [5907, 5912, 5913, 5916, 5918, 5923]   \n",
       "ENST00000064571.2         [13398, 13399, 13400, 13401, 13402, 13403]   \n",
       "ENST00000170150.3  [4157, 4159, 4163, 4166, 4168, 4170, 4172, 417...   \n",
       "\n",
       "                   annotation novelty  \n",
       "tid                                    \n",
       "ENST00000023939.8        True   Known  \n",
       "ENST00000045083.6        True   Known  \n",
       "ENST00000062104.3        True   Known  \n",
       "ENST00000064571.2        True   Known  \n",
       "ENST00000170150.3        True   Known  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.t_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "scientific-constitution",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PB65_B017</th>\n",
       "      <td>PB65_B017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PB65_B018</th>\n",
       "      <td>PB65_B018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D12</th>\n",
       "      <td>D12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             dataset\n",
       "PB65_B017  PB65_B017\n",
       "PB65_B018  PB65_B018\n",
       "D12              D12"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.adata.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "removed-batman",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1., 1., 1., 1., 0.],\n",
       "       [0., 1., 0., 0., 0.],\n",
       "       [0., 0., 0., 0., 1.]], dtype=float32)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.adata.X[:5, :5]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "overall-complexity",
   "metadata": {},
   "source": [
    "#### test - add transcriptome from db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "hungry-humor",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding annotation to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing transcripts:  99%|█████████▉| 8400/8450 [00:01<00:00, 4304.85it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reindexing and sorting entries on genomic location...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reindexing vertices: 100%|█████████▉| 50000/50074 [00:00<00:00, 968388.58it/s] \n",
      "Reindexing edges:  98%|█████████▊| 49000/50074 [00:00<00:00, 579532.86it/s]\n",
      "Reindexing transcripts:  99%|█████████▉| 8400/8450 [00:00<00:00, 146771.90it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Annotation to the SwanGraph\n",
      "\n",
      "Adding transcriptome to the SwanGraph\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "# just gencode vM21 chr 11 and tcf3\n",
    "gtf = 'data/chr11_and_Tcf3.gtf'\n",
    "sg.add_annotation(gtf, verbose=True)\n",
    "db = 'data/chr11_and_Tcf3_no_gname.db'\n",
    "sg.add_transcriptome(db)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "plastic-cattle",
   "metadata": {},
   "source": [
    "#### test - add some abundance and then more abundance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "horizontal-province",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding annotation to the SwanGraph\n",
      "\n",
      "Adding transcriptome to the SwanGraph\n",
      "\n",
      "Adding abundance for datasets hepg2_1_chr20, hepg2_2_chr20, hffc6_1_chr20, hffc6_2_chr20, hffc6_3_chr20 to SwanGraph.\n",
      "\n",
      "Adding abundance for datasets k562_1_chr20, k562_2_chr20 to SwanGraph.\n",
      "(2, 4555)\n",
      "(7, 4555)\n",
      "(2, 4555)\n",
      "(7, 4555)\n",
      "2\n",
      "7\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "# just chr 20\n",
    "gtf = 'data/gencode.v29.annotation.gtf'\n",
    "sg.add_annotation(gtf)\n",
    "# transcriptome\n",
    "gtf = 'data/hepg2_1_talon.gtf'\n",
    "sg.add_transcriptome(gtf)\n",
    "# abundance - hffc6 and hepg2\n",
    "ab = 'data/all_talon_abundance_filtered_counts_hffc6_hepg2.tsv'\n",
    "sg.add_abundance(ab)\n",
    "ab = 'data/all_talon_abundance_filtered_counts_k562.tsv'\n",
    "sg.add_abundance(ab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "committed-contamination",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "KeysView(Layers with keys: counts, tpm, pi)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.adata.layers.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "improving-saturday",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>hepg2_1_chr20</th>\n",
       "      <td>hepg2_1_chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hepg2_2_chr20</th>\n",
       "      <td>hepg2_2_chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hffc6_1_chr20</th>\n",
       "      <td>hffc6_1_chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hffc6_2_chr20</th>\n",
       "      <td>hffc6_2_chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hffc6_3_chr20</th>\n",
       "      <td>hffc6_3_chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k562_1_chr20</th>\n",
       "      <td>k562_1_chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k562_2_chr20</th>\n",
       "      <td>k562_2_chr20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     dataset\n",
       "hepg2_1_chr20  hepg2_1_chr20\n",
       "hepg2_2_chr20  hepg2_2_chr20\n",
       "hffc6_1_chr20  hffc6_1_chr20\n",
       "hffc6_2_chr20  hffc6_2_chr20\n",
       "hffc6_3_chr20  hffc6_3_chr20\n",
       "k562_1_chr20    k562_1_chr20\n",
       "k562_2_chr20    k562_2_chr20"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "excess-treasure",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENST00000023939.8</th>\n",
       "      <td>ENST00000023939.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000045083.6</th>\n",
       "      <td>ENST00000045083.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000062104.3</th>\n",
       "      <td>ENST00000062104.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000064571.2</th>\n",
       "      <td>ENST00000064571.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000170150.3</th>\n",
       "      <td>ENST00000170150.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 tid\n",
       "tid                                 \n",
       "ENST00000023939.8  ENST00000023939.8\n",
       "ENST00000045083.6  ENST00000045083.6\n",
       "ENST00000062104.3  ENST00000062104.3\n",
       "ENST00000064571.2  ENST00000064571.2\n",
       "ENST00000170150.3  ENST00000170150.3"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.adata.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "golden-ballet",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[  0.,   0.,   0.,   0., 847.],\n",
       "       [  0.,   0.,   0.,   0., 800.],\n",
       "       [  2.,   0.,   0.,   0.,   0.],\n",
       "       [  2.,   0.,   0.,   0.,   0.],\n",
       "       [  2.,   0.,   0.,   0.,   0.],\n",
       "       [  1.,   0.,   0.,   0.,   0.],\n",
       "       [  1.,   0.,   0.,   0.,   0.]], dtype=float32)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.adata.X[:, :5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "imported-duplicate",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'tpm'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-8-4f83231fd002>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0msg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlayers\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'tpm'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m:\u001b[0m\u001b[0;36m5\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/anndata/_core/aligned_mapping.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    145\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    146\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__getitem__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mV\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 147\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    148\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    149\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setitem__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mV\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'tpm'"
     ]
    }
   ],
   "source": [
    "sg.adata.layers['tpm'][:, :5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "sustained-penny",
   "metadata": {},
   "outputs": [],
   "source": [
    "sg.adata.obs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "stylish-tunisia",
   "metadata": {},
   "source": [
    "#### test - add abundance when trying to add more datasests that are already there"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "mexican-decrease",
   "metadata": {},
   "outputs": [],
   "source": [
    "sg = swan.SwanGraph()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "stainless-footwear",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding annotation to the SwanGraph\n"
     ]
    }
   ],
   "source": [
    "# just chr 20\n",
    "gtf = 'data/gencode.v29.annotation.gtf'\n",
    "sg.add_annotation(gtf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "quarterly-tribune",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding transcriptome to the SwanGraph\n"
     ]
    }
   ],
   "source": [
    "# transcriptome\n",
    "gtf = 'data/hepg2_1_talon.gtf'\n",
    "sg.add_transcriptome(gtf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "incorporate-rotation",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding abundance for datasets hepg2_1_chr20, hepg2_2_chr20, hffc6_1_chr20, hffc6_2_chr20, hffc6_3_chr20... (and 2 more) to SwanGraph\n"
     ]
    }
   ],
   "source": [
    "# abundance\n",
    "ab = 'data/all_talon_abundance_filtered_counts.tsv'\n",
    "sg.add_abundance(ab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "ready-affairs",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "counts\n",
      "tpm\n"
     ]
    }
   ],
   "source": [
    "for i in sg.adata.layers.keys():\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "seeing-listing",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transcripts absent from abundance file will be assigned 0 counts.\n",
      "Transcripts found in abundance matrix that are not in the SwanGraph will not have expression added.\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Dataset hepg2_1_chr20 already present in the SwanGraph.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-13-6de16e8aed29>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# abundance\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mab\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'data/all_talon_abundance_filtered_counts.tsv'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0msg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd_abundance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mab\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/swan_vis/swangraph.py\u001b[0m in \u001b[0;36madd_abundance\u001b[0;34m(self, counts_file)\u001b[0m\n\u001b[1;32m    321\u001b[0m                 \u001b[0;32mfor\u001b[0m \u001b[0md\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdatasets\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    322\u001b[0m                         \u001b[0;32mif\u001b[0m \u001b[0md\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdatasets\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 323\u001b[0;31m                                 \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Dataset {} already present in the SwanGraph.'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0md\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    324\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdatasets\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mextend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdatasets\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    325\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Dataset hepg2_1_chr20 already present in the SwanGraph."
     ]
    }
   ],
   "source": [
    "# abundance\n",
    "ab = 'data/all_talon_abundance_filtered_counts.tsv'\n",
    "sg.add_abundance(ab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "divided-doctrine",
   "metadata": {},
   "outputs": [],
   "source": [
    "sg.adata.X[:5,:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "taken-growing",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "buried-penguin",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "progressive-heaven",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>tname</th>\n",
       "      <th>gid</th>\n",
       "      <th>gname</th>\n",
       "      <th>path</th>\n",
       "      <th>annotation</th>\n",
       "      <th>novelty</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENST00000023939.8</th>\n",
       "      <td>ENST00000023939.8</td>\n",
       "      <td>RTF2-201</td>\n",
       "      <td>ENSG00000022277.12</td>\n",
       "      <td>RTF2</td>\n",
       "      <td>[9022, 9029, 9035, 9036, 9037, 9039, 9040, 904...</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000045083.6</th>\n",
       "      <td>ENST00000045083.6</td>\n",
       "      <td>RIPOR3-201</td>\n",
       "      <td>ENSG00000042062.11</td>\n",
       "      <td>RIPOR3</td>\n",
       "      <td>[13836, 13837, 13838, 13839, 13840, 13841, 138...</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000062104.3</th>\n",
       "      <td>ENST00000062104.3</td>\n",
       "      <td>NNAT-201</td>\n",
       "      <td>ENSG00000053438.9</td>\n",
       "      <td>NNAT</td>\n",
       "      <td>[5907, 5912, 5913, 5916, 5918, 5923]</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000064571.2</th>\n",
       "      <td>ENST00000064571.2</td>\n",
       "      <td>CBLN4-201</td>\n",
       "      <td>ENSG00000054803.3</td>\n",
       "      <td>CBLN4</td>\n",
       "      <td>[13398, 13399, 13400, 13401, 13402, 13403]</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000170150.3</th>\n",
       "      <td>ENST00000170150.3</td>\n",
       "      <td>BPIFB2-201</td>\n",
       "      <td>ENSG00000078898.6</td>\n",
       "      <td>BPIFB2</td>\n",
       "      <td>[4157, 4159, 4163, 4166, 4168, 4170, 4172, 417...</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 tid       tname                 gid   gname  \\\n",
       "tid                                                                            \n",
       "ENST00000023939.8  ENST00000023939.8    RTF2-201  ENSG00000022277.12    RTF2   \n",
       "ENST00000045083.6  ENST00000045083.6  RIPOR3-201  ENSG00000042062.11  RIPOR3   \n",
       "ENST00000062104.3  ENST00000062104.3    NNAT-201   ENSG00000053438.9    NNAT   \n",
       "ENST00000064571.2  ENST00000064571.2   CBLN4-201   ENSG00000054803.3   CBLN4   \n",
       "ENST00000170150.3  ENST00000170150.3  BPIFB2-201   ENSG00000078898.6  BPIFB2   \n",
       "\n",
       "                                                                path  \\\n",
       "tid                                                                    \n",
       "ENST00000023939.8  [9022, 9029, 9035, 9036, 9037, 9039, 9040, 904...   \n",
       "ENST00000045083.6  [13836, 13837, 13838, 13839, 13840, 13841, 138...   \n",
       "ENST00000062104.3               [5907, 5912, 5913, 5916, 5918, 5923]   \n",
       "ENST00000064571.2         [13398, 13399, 13400, 13401, 13402, 13403]   \n",
       "ENST00000170150.3  [4157, 4159, 4163, 4166, 4168, 4170, 4172, 417...   \n",
       "\n",
       "                   annotation novelty  \n",
       "tid                                    \n",
       "ENST00000023939.8        True   Known  \n",
       "ENST00000045083.6        True   Known  \n",
       "ENST00000062104.3        True   Known  \n",
       "ENST00000064571.2        True   Known  \n",
       "ENST00000170150.3        True   Known  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.t_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "respiratory-relation",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>tname</th>\n",
       "      <th>gid</th>\n",
       "      <th>gname</th>\n",
       "      <th>path</th>\n",
       "      <th>annotation</th>\n",
       "      <th>novelty</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TALONT000005280</th>\n",
       "      <td>TALONT000005280</td>\n",
       "      <td>TALONT000005280</td>\n",
       "      <td>ENSG00000198171.12</td>\n",
       "      <td>DDRGK1</td>\n",
       "      <td>[21055, 21057, 21059, 21061, 21063, 21065, 210...</td>\n",
       "      <td>False</td>\n",
       "      <td>NNC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TALONT000005555</th>\n",
       "      <td>TALONT000005555</td>\n",
       "      <td>TALONT000005555</td>\n",
       "      <td>ENSG00000125779.22</td>\n",
       "      <td>PANK2</td>\n",
       "      <td>[1020, 1022, 1038, 1040, 1045, 1048, 1050, 105...</td>\n",
       "      <td>False</td>\n",
       "      <td>NIC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TALONT000005727</th>\n",
       "      <td>TALONT000005727</td>\n",
       "      <td>TALONT000005727</td>\n",
       "      <td>ENSG00000089063.14</td>\n",
       "      <td>TMEM230</td>\n",
       "      <td>[20569, 20572, 20586, 20588, 20592, 20594, 205...</td>\n",
       "      <td>False</td>\n",
       "      <td>NNC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TALONT000006405</th>\n",
       "      <td>TALONT000006405</td>\n",
       "      <td>TALONT000006405</td>\n",
       "      <td>ENSG00000125868.15</td>\n",
       "      <td>DSTN</td>\n",
       "      <td>[2138, 2142, 2146, 2148, 2151, 2153, 2157, 2160]</td>\n",
       "      <td>False</td>\n",
       "      <td>NNC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TALONT000006566</th>\n",
       "      <td>TALONT000006566</td>\n",
       "      <td>TALONT000006566</td>\n",
       "      <td>ENSG00000089006.16</td>\n",
       "      <td>SNX5</td>\n",
       "      <td>[19444, 19447, 19456, 19458, 19472, 19477, 194...</td>\n",
       "      <td>False</td>\n",
       "      <td>NIC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             tid            tname                 gid  \\\n",
       "tid                                                                     \n",
       "TALONT000005280  TALONT000005280  TALONT000005280  ENSG00000198171.12   \n",
       "TALONT000005555  TALONT000005555  TALONT000005555  ENSG00000125779.22   \n",
       "TALONT000005727  TALONT000005727  TALONT000005727  ENSG00000089063.14   \n",
       "TALONT000006405  TALONT000006405  TALONT000006405  ENSG00000125868.15   \n",
       "TALONT000006566  TALONT000006566  TALONT000006566  ENSG00000089006.16   \n",
       "\n",
       "                   gname                                               path  \\\n",
       "tid                                                                           \n",
       "TALONT000005280   DDRGK1  [21055, 21057, 21059, 21061, 21063, 21065, 210...   \n",
       "TALONT000005555    PANK2  [1020, 1022, 1038, 1040, 1045, 1048, 1050, 105...   \n",
       "TALONT000005727  TMEM230  [20569, 20572, 20586, 20588, 20592, 20594, 205...   \n",
       "TALONT000006405     DSTN   [2138, 2142, 2146, 2148, 2151, 2153, 2157, 2160]   \n",
       "TALONT000006566     SNX5  [19444, 19447, 19456, 19458, 19472, 19477, 194...   \n",
       "\n",
       "                 annotation novelty  \n",
       "tid                                  \n",
       "TALONT000005280       False     NNC  \n",
       "TALONT000005555       False     NIC  \n",
       "TALONT000005727       False     NNC  \n",
       "TALONT000006405       False     NNC  \n",
       "TALONT000006566       False     NIC  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.t_df.loc[sg.t_df.annotation==False].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "unavailable-spanking",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>tname</th>\n",
       "      <th>gid</th>\n",
       "      <th>gname</th>\n",
       "      <th>path</th>\n",
       "      <th>annotation</th>\n",
       "      <th>novelty</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENST00000023939.8</th>\n",
       "      <td>ENST00000023939.8</td>\n",
       "      <td>RTF2-201</td>\n",
       "      <td>ENSG00000022277.12</td>\n",
       "      <td>RTF2</td>\n",
       "      <td>[9022, 9029, 9035, 9036, 9037, 9039, 9040, 904...</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000045083.6</th>\n",
       "      <td>ENST00000045083.6</td>\n",
       "      <td>RIPOR3-201</td>\n",
       "      <td>ENSG00000042062.11</td>\n",
       "      <td>RIPOR3</td>\n",
       "      <td>[13836, 13837, 13838, 13839, 13840, 13841, 138...</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000062104.3</th>\n",
       "      <td>ENST00000062104.3</td>\n",
       "      <td>NNAT-201</td>\n",
       "      <td>ENSG00000053438.9</td>\n",
       "      <td>NNAT</td>\n",
       "      <td>[5907, 5912, 5913, 5916, 5918, 5923]</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000064571.2</th>\n",
       "      <td>ENST00000064571.2</td>\n",
       "      <td>CBLN4-201</td>\n",
       "      <td>ENSG00000054803.3</td>\n",
       "      <td>CBLN4</td>\n",
       "      <td>[13398, 13399, 13400, 13401, 13402, 13403]</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000170150.3</th>\n",
       "      <td>ENST00000170150.3</td>\n",
       "      <td>BPIFB2-201</td>\n",
       "      <td>ENSG00000078898.6</td>\n",
       "      <td>BPIFB2</td>\n",
       "      <td>[4157, 4159, 4163, 4166, 4168, 4170, 4172, 417...</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TALONT000013657</th>\n",
       "      <td>TALONT000013657</td>\n",
       "      <td>TALONT000013657</td>\n",
       "      <td>ENSG00000101187.15</td>\n",
       "      <td>SLCO4A1</td>\n",
       "      <td>[10403, 10405, 10408, 10412, 10421, 10424, 104...</td>\n",
       "      <td>False</td>\n",
       "      <td>NIC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TALONT000013998</th>\n",
       "      <td>TALONT000013998</td>\n",
       "      <td>TALONT000013998</td>\n",
       "      <td>ENSG00000101246.19</td>\n",
       "      <td>ARFRP1</td>\n",
       "      <td>[11745, 11747, 11754, 11756, 11759, 11761, 117...</td>\n",
       "      <td>False</td>\n",
       "      <td>NIC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TALONT000014162</th>\n",
       "      <td>TALONT000014162</td>\n",
       "      <td>TALONT000014162</td>\n",
       "      <td>ENSG00000198276.15</td>\n",
       "      <td>UCKL1</td>\n",
       "      <td>[11635, 11639, 11645, 11649, 11651, 11653, 116...</td>\n",
       "      <td>False</td>\n",
       "      <td>NNC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TALONT000014163</th>\n",
       "      <td>TALONT000014163</td>\n",
       "      <td>TALONT000014163</td>\n",
       "      <td>ENSG00000198276.15</td>\n",
       "      <td>UCKL1</td>\n",
       "      <td>[11635, 11639, 11646, 11649, 11651, 11653, 116...</td>\n",
       "      <td>False</td>\n",
       "      <td>NNC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TALONT000014329</th>\n",
       "      <td>TALONT000014329</td>\n",
       "      <td>TALONT000014329</td>\n",
       "      <td>ENSG00000196421.8</td>\n",
       "      <td>C20orf204</td>\n",
       "      <td>[11342, 11344, 11346, 11352, 11355, 11357]</td>\n",
       "      <td>False</td>\n",
       "      <td>NIC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4555 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 tid            tname                 gid  \\\n",
       "tid                                                                         \n",
       "ENST00000023939.8  ENST00000023939.8         RTF2-201  ENSG00000022277.12   \n",
       "ENST00000045083.6  ENST00000045083.6       RIPOR3-201  ENSG00000042062.11   \n",
       "ENST00000062104.3  ENST00000062104.3         NNAT-201   ENSG00000053438.9   \n",
       "ENST00000064571.2  ENST00000064571.2        CBLN4-201   ENSG00000054803.3   \n",
       "ENST00000170150.3  ENST00000170150.3       BPIFB2-201   ENSG00000078898.6   \n",
       "...                              ...              ...                 ...   \n",
       "TALONT000013657      TALONT000013657  TALONT000013657  ENSG00000101187.15   \n",
       "TALONT000013998      TALONT000013998  TALONT000013998  ENSG00000101246.19   \n",
       "TALONT000014162      TALONT000014162  TALONT000014162  ENSG00000198276.15   \n",
       "TALONT000014163      TALONT000014163  TALONT000014163  ENSG00000198276.15   \n",
       "TALONT000014329      TALONT000014329  TALONT000014329   ENSG00000196421.8   \n",
       "\n",
       "                       gname  \\\n",
       "tid                            \n",
       "ENST00000023939.8       RTF2   \n",
       "ENST00000045083.6     RIPOR3   \n",
       "ENST00000062104.3       NNAT   \n",
       "ENST00000064571.2      CBLN4   \n",
       "ENST00000170150.3     BPIFB2   \n",
       "...                      ...   \n",
       "TALONT000013657      SLCO4A1   \n",
       "TALONT000013998       ARFRP1   \n",
       "TALONT000014162        UCKL1   \n",
       "TALONT000014163        UCKL1   \n",
       "TALONT000014329    C20orf204   \n",
       "\n",
       "                                                                path  \\\n",
       "tid                                                                    \n",
       "ENST00000023939.8  [9022, 9029, 9035, 9036, 9037, 9039, 9040, 904...   \n",
       "ENST00000045083.6  [13836, 13837, 13838, 13839, 13840, 13841, 138...   \n",
       "ENST00000062104.3               [5907, 5912, 5913, 5916, 5918, 5923]   \n",
       "ENST00000064571.2         [13398, 13399, 13400, 13401, 13402, 13403]   \n",
       "ENST00000170150.3  [4157, 4159, 4163, 4166, 4168, 4170, 4172, 417...   \n",
       "...                                                              ...   \n",
       "TALONT000013657    [10403, 10405, 10408, 10412, 10421, 10424, 104...   \n",
       "TALONT000013998    [11745, 11747, 11754, 11756, 11759, 11761, 117...   \n",
       "TALONT000014162    [11635, 11639, 11645, 11649, 11651, 11653, 116...   \n",
       "TALONT000014163    [11635, 11639, 11646, 11649, 11651, 11653, 116...   \n",
       "TALONT000014329           [11342, 11344, 11346, 11352, 11355, 11357]   \n",
       "\n",
       "                   annotation novelty  \n",
       "tid                                    \n",
       "ENST00000023939.8        True   Known  \n",
       "ENST00000045083.6        True   Known  \n",
       "ENST00000062104.3        True   Known  \n",
       "ENST00000064571.2        True   Known  \n",
       "ENST00000170150.3        True   Known  \n",
       "...                       ...     ...  \n",
       "TALONT000013657         False     NIC  \n",
       "TALONT000013998         False     NIC  \n",
       "TALONT000014162         False     NNC  \n",
       "TALONT000014163         False     NNC  \n",
       "TALONT000014329         False     NIC  \n",
       "\n",
       "[4555 rows x 7 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.t_df.loc[sg.t_df.novelty!=False]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "prerequisite-abuse",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>edge_type</th>\n",
       "      <th>strand</th>\n",
       "      <th>v1</th>\n",
       "      <th>v2</th>\n",
       "      <th>edge_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>edge_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>(0, 1)</th>\n",
       "      <td>exon</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>(0, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(1, 4)</th>\n",
       "      <td>intron</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>(1, 4)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(2, 3)</th>\n",
       "      <td>exon</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>(2, 3)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(3, 4)</th>\n",
       "      <td>intron</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>(3, 4)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(4, 5)</th>\n",
       "      <td>exon</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>(4, 5)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        edge_type strand  v1  v2 edge_id\n",
       "edge_id                                 \n",
       "(0, 1)       exon      +   0   1  (0, 1)\n",
       "(1, 4)     intron      +   1   4  (1, 4)\n",
       "(2, 3)       exon      +   2   3  (2, 3)\n",
       "(3, 4)     intron      +   3   4  (3, 4)\n",
       "(4, 5)       exon      +   4   5  (4, 5)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.edge_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "sonic-poultry",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>coord</th>\n",
       "      <th>strand</th>\n",
       "      <th>vertex_id</th>\n",
       "      <th>TSS</th>\n",
       "      <th>TES</th>\n",
       "      <th>internal</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vertex_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr20</td>\n",
       "      <td>87250</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr20</td>\n",
       "      <td>87359</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr20</td>\n",
       "      <td>87710</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr20</td>\n",
       "      <td>87767</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr20</td>\n",
       "      <td>96005</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           chrom  coord strand  vertex_id    TSS    TES  internal\n",
       "vertex_id                                                        \n",
       "0          chr20  87250      +          0   True  False     False\n",
       "1          chr20  87359      +          1  False  False      True\n",
       "2          chr20  87710      +          2   True  False     False\n",
       "3          chr20  87767      +          3  False  False      True\n",
       "4          chr20  96005      +          4  False  False      True"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.loc_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "scheduled-pasta",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "requested-ensemble",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>tname</th>\n",
       "      <th>gid</th>\n",
       "      <th>gname</th>\n",
       "      <th>path</th>\n",
       "      <th>annotation</th>\n",
       "      <th>novelty</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENST00000023939.8</th>\n",
       "      <td>ENST00000023939.8</td>\n",
       "      <td>RTF2-201</td>\n",
       "      <td>ENSG00000022277.12</td>\n",
       "      <td>RTF2</td>\n",
       "      <td>[9022, 9029, 9035, 9036, 9037, 9039, 9040, 904...</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000045083.6</th>\n",
       "      <td>ENST00000045083.6</td>\n",
       "      <td>RIPOR3-201</td>\n",
       "      <td>ENSG00000042062.11</td>\n",
       "      <td>RIPOR3</td>\n",
       "      <td>[13836, 13837, 13838, 13839, 13840, 13841, 138...</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000062104.3</th>\n",
       "      <td>ENST00000062104.3</td>\n",
       "      <td>NNAT-201</td>\n",
       "      <td>ENSG00000053438.9</td>\n",
       "      <td>NNAT</td>\n",
       "      <td>[5907, 5912, 5913, 5916, 5918, 5923]</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000064571.2</th>\n",
       "      <td>ENST00000064571.2</td>\n",
       "      <td>CBLN4-201</td>\n",
       "      <td>ENSG00000054803.3</td>\n",
       "      <td>CBLN4</td>\n",
       "      <td>[13398, 13399, 13400, 13401, 13402, 13403]</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000170150.3</th>\n",
       "      <td>ENST00000170150.3</td>\n",
       "      <td>BPIFB2-201</td>\n",
       "      <td>ENSG00000078898.6</td>\n",
       "      <td>BPIFB2</td>\n",
       "      <td>[4157, 4159, 4163, 4166, 4168, 4170, 4172, 417...</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 tid       tname                 gid   gname  \\\n",
       "tid                                                                            \n",
       "ENST00000023939.8  ENST00000023939.8    RTF2-201  ENSG00000022277.12    RTF2   \n",
       "ENST00000045083.6  ENST00000045083.6  RIPOR3-201  ENSG00000042062.11  RIPOR3   \n",
       "ENST00000062104.3  ENST00000062104.3    NNAT-201   ENSG00000053438.9    NNAT   \n",
       "ENST00000064571.2  ENST00000064571.2   CBLN4-201   ENSG00000054803.3   CBLN4   \n",
       "ENST00000170150.3  ENST00000170150.3  BPIFB2-201   ENSG00000078898.6  BPIFB2   \n",
       "\n",
       "                                                                path  \\\n",
       "tid                                                                    \n",
       "ENST00000023939.8  [9022, 9029, 9035, 9036, 9037, 9039, 9040, 904...   \n",
       "ENST00000045083.6  [13836, 13837, 13838, 13839, 13840, 13841, 138...   \n",
       "ENST00000062104.3               [5907, 5912, 5913, 5916, 5918, 5923]   \n",
       "ENST00000064571.2         [13398, 13399, 13400, 13401, 13402, 13403]   \n",
       "ENST00000170150.3  [4157, 4159, 4163, 4166, 4168, 4170, 4172, 417...   \n",
       "\n",
       "                   annotation novelty  \n",
       "tid                                    \n",
       "ENST00000023939.8        True   Known  \n",
       "ENST00000045083.6        True   Known  \n",
       "ENST00000062104.3        True   Known  \n",
       "ENST00000064571.2        True   Known  \n",
       "ENST00000170150.3        True   Known  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.t_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "permanent-vegetarian",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('data/all_talon_abundance_filtered_counts.tsv', sep='\\t')\n",
    "\n",
    "# rename first column \n",
    "col = df.columns[0]\n",
    "df.rename({col: 'tid'}, axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "certain-inspection",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "vulnerable-connection",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'ab_tids' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-8-fbb545520f63>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtid\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mab_tids\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msg_tids\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'ab_tids' is not defined"
     ]
    }
   ],
   "source": [
    "df.loc[df.tid.isin(list(set(ab_tids)-set(sg_tids)))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "suffering-singles",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transcripts absent from abundance file will be assigned 0 counts.\n",
      "Transcripts found in abundance matrix that are not in the SwanGraph will not have expression added.\n"
     ]
    }
   ],
   "source": [
    "# limit to just the transcripts already in the graph\n",
    "sg_tids = sg.t_df.tid.tolist()\n",
    "ab_tids = df.tid.tolist()\n",
    "\n",
    "if len(set(sg_tids)-set(ab_tids)) != 0:\n",
    "    print('Transcripts absent from abundance file will be assigned 0 counts.')\n",
    "if len(set(ab_tids)-set(sg_tids)) != 0:\n",
    "    print('Transcripts found in abundance matrix that are not in the SwanGraph will not have expression added.')\n",
    "\n",
    "    \n",
    "# right merge to keep everything in the t_df already\n",
    "df = df.merge(sg.t_df['tid'].to_frame(), how='right', left_on='tid', right_index=True)\n",
    "df.drop(['tid_x', 'tid_y'], axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "sudden-eagle",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.fillna(0, inplace=True)\n",
    "df.set_index('tid', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "caring-messenger",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(7, 4555)\n"
     ]
    }
   ],
   "source": [
    "tpm_df = df.index.to_frame()\n",
    "tpm_df.head()\n",
    "for c in df.columns.tolist():\n",
    "    total_counts = df[c].sum()    \n",
    "    tpm_df[c] = (df[c]*1000000)/total_counts\n",
    "tpm_df.drop('tid', axis=1, inplace=True)\n",
    "tpm_df = tpm_df.T\n",
    "tpm_X = tpm_df.to_numpy()\n",
    "print(np.shape(tpm_X))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "valued-stomach",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>tid</th>\n",
       "      <th>ENST00000023939.8</th>\n",
       "      <th>ENST00000045083.6</th>\n",
       "      <th>ENST00000062104.3</th>\n",
       "      <th>ENST00000064571.2</th>\n",
       "      <th>ENST00000170150.3</th>\n",
       "      <th>ENST00000190983.4</th>\n",
       "      <th>ENST00000191018.9</th>\n",
       "      <th>ENST00000201031.2</th>\n",
       "      <th>ENST00000201961.6</th>\n",
       "      <th>ENST00000201963.3</th>\n",
       "      <th>...</th>\n",
       "      <th>TALONT000013147</th>\n",
       "      <th>TALONT000013148</th>\n",
       "      <th>TALONT000013454</th>\n",
       "      <th>TALONT000013557</th>\n",
       "      <th>TALONT000013655</th>\n",
       "      <th>TALONT000013657</th>\n",
       "      <th>TALONT000013998</th>\n",
       "      <th>TALONT000014162</th>\n",
       "      <th>TALONT000014163</th>\n",
       "      <th>TALONT000014329</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>hepg2_1_chr20</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>53914.704010</td>\n",
       "      <td>0.0</td>\n",
       "      <td>254.614895</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>572.883514</td>\n",
       "      <td>381.922342</td>\n",
       "      <td>891.152132</td>\n",
       "      <td>700.190961</td>\n",
       "      <td>700.190961</td>\n",
       "      <td>381.922342</td>\n",
       "      <td>572.883514</td>\n",
       "      <td>1145.767027</td>\n",
       "      <td>1082.113304</td>\n",
       "      <td>318.268619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hepg2_2_chr20</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>29204.541306</td>\n",
       "      <td>0.0</td>\n",
       "      <td>255.539736</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>985.653269</td>\n",
       "      <td>292.045413</td>\n",
       "      <td>985.653269</td>\n",
       "      <td>511.079473</td>\n",
       "      <td>365.056766</td>\n",
       "      <td>401.562443</td>\n",
       "      <td>876.136239</td>\n",
       "      <td>2190.340598</td>\n",
       "      <td>876.136239</td>\n",
       "      <td>328.551090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hffc6_1_chr20</th>\n",
       "      <td>130.182907</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>325.457267</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>325.457267</td>\n",
       "      <td>130.182907</td>\n",
       "      <td>1236.737616</td>\n",
       "      <td>520.731628</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1497.103430</td>\n",
       "      <td>585.823081</td>\n",
       "      <td>65.091453</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hffc6_2_chr20</th>\n",
       "      <td>107.717994</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>592.448969</td>\n",
       "      <td>161.576991</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>107.717994</td>\n",
       "      <td>107.717994</td>\n",
       "      <td>969.461949</td>\n",
       "      <td>377.012980</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>969.461949</td>\n",
       "      <td>1023.320946</td>\n",
       "      <td>377.012980</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hffc6_3_chr20</th>\n",
       "      <td>138.504155</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>900.277008</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>415.512465</td>\n",
       "      <td>138.504155</td>\n",
       "      <td>1038.781163</td>\n",
       "      <td>277.008310</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1177.285319</td>\n",
       "      <td>900.277008</td>\n",
       "      <td>346.260388</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 4555 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "tid            ENST00000023939.8  ENST00000045083.6  ENST00000062104.3  \\\n",
       "hepg2_1_chr20           0.000000                0.0                0.0   \n",
       "hepg2_2_chr20           0.000000                0.0                0.0   \n",
       "hffc6_1_chr20         130.182907                0.0                0.0   \n",
       "hffc6_2_chr20         107.717994                0.0                0.0   \n",
       "hffc6_3_chr20         138.504155                0.0                0.0   \n",
       "\n",
       "tid            ENST00000064571.2  ENST00000170150.3  ENST00000190983.4  \\\n",
       "hepg2_1_chr20                0.0       53914.704010                0.0   \n",
       "hepg2_2_chr20                0.0       29204.541306                0.0   \n",
       "hffc6_1_chr20                0.0           0.000000                0.0   \n",
       "hffc6_2_chr20                0.0           0.000000                0.0   \n",
       "hffc6_3_chr20                0.0           0.000000                0.0   \n",
       "\n",
       "tid            ENST00000191018.9  ENST00000201031.2  ENST00000201961.6  \\\n",
       "hepg2_1_chr20         254.614895           0.000000                0.0   \n",
       "hepg2_2_chr20         255.539736           0.000000                0.0   \n",
       "hffc6_1_chr20         325.457267           0.000000                0.0   \n",
       "hffc6_2_chr20         592.448969         161.576991                0.0   \n",
       "hffc6_3_chr20         900.277008           0.000000                0.0   \n",
       "\n",
       "tid            ENST00000201963.3  ...  TALONT000013147  TALONT000013148  \\\n",
       "hepg2_1_chr20                0.0  ...       572.883514       381.922342   \n",
       "hepg2_2_chr20                0.0  ...       985.653269       292.045413   \n",
       "hffc6_1_chr20                0.0  ...       325.457267       130.182907   \n",
       "hffc6_2_chr20                0.0  ...       107.717994       107.717994   \n",
       "hffc6_3_chr20                0.0  ...       415.512465       138.504155   \n",
       "\n",
       "tid            TALONT000013454  TALONT000013557  TALONT000013655  \\\n",
       "hepg2_1_chr20       891.152132       700.190961       700.190961   \n",
       "hepg2_2_chr20       985.653269       511.079473       365.056766   \n",
       "hffc6_1_chr20      1236.737616       520.731628         0.000000   \n",
       "hffc6_2_chr20       969.461949       377.012980         0.000000   \n",
       "hffc6_3_chr20      1038.781163       277.008310         0.000000   \n",
       "\n",
       "tid            TALONT000013657  TALONT000013998  TALONT000014162  \\\n",
       "hepg2_1_chr20       381.922342       572.883514      1145.767027   \n",
       "hepg2_2_chr20       401.562443       876.136239      2190.340598   \n",
       "hffc6_1_chr20         0.000000      1497.103430       585.823081   \n",
       "hffc6_2_chr20         0.000000       969.461949      1023.320946   \n",
       "hffc6_3_chr20         0.000000      1177.285319       900.277008   \n",
       "\n",
       "tid            TALONT000014163  TALONT000014329  \n",
       "hepg2_1_chr20      1082.113304       318.268619  \n",
       "hepg2_2_chr20       876.136239       328.551090  \n",
       "hffc6_1_chr20        65.091453         0.000000  \n",
       "hffc6_2_chr20       377.012980         0.000000  \n",
       "hffc6_3_chr20       346.260388         0.000000  \n",
       "\n",
       "[5 rows x 4555 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tpm_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "chinese-yukon",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hepg2_1_chr20</th>\n",
       "      <th>hepg2_2_chr20</th>\n",
       "      <th>hffc6_1_chr20</th>\n",
       "      <th>hffc6_2_chr20</th>\n",
       "      <th>hffc6_3_chr20</th>\n",
       "      <th>k562_1_chr20</th>\n",
       "      <th>k562_2_chr20</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENST00000023939.8</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000045083.6</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000062104.3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000064571.2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000170150.3</th>\n",
       "      <td>847.0</td>\n",
       "      <td>800.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   hepg2_1_chr20  hepg2_2_chr20  hffc6_1_chr20  hffc6_2_chr20  \\\n",
       "tid                                                                             \n",
       "ENST00000023939.8            0.0            0.0            2.0            2.0   \n",
       "ENST00000045083.6            0.0            0.0            0.0            0.0   \n",
       "ENST00000062104.3            0.0            0.0            0.0            0.0   \n",
       "ENST00000064571.2            0.0            0.0            0.0            0.0   \n",
       "ENST00000170150.3          847.0          800.0            0.0            0.0   \n",
       "\n",
       "                   hffc6_3_chr20  k562_1_chr20  k562_2_chr20  \n",
       "tid                                                           \n",
       "ENST00000023939.8            2.0           1.0           1.0  \n",
       "ENST00000045083.6            0.0           0.0           0.0  \n",
       "ENST00000062104.3            0.0           0.0           0.0  \n",
       "ENST00000064571.2            0.0           0.0           0.0  \n",
       "ENST00000170150.3            0.0           0.0           0.0  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "superior-creativity",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "df = df.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "public-mailing",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>tid</th>\n",
       "      <th>ENST00000023939.8</th>\n",
       "      <th>ENST00000045083.6</th>\n",
       "      <th>ENST00000062104.3</th>\n",
       "      <th>ENST00000064571.2</th>\n",
       "      <th>ENST00000170150.3</th>\n",
       "      <th>ENST00000190983.4</th>\n",
       "      <th>ENST00000191018.9</th>\n",
       "      <th>ENST00000201031.2</th>\n",
       "      <th>ENST00000201961.6</th>\n",
       "      <th>ENST00000201963.3</th>\n",
       "      <th>...</th>\n",
       "      <th>TALONT000013147</th>\n",
       "      <th>TALONT000013148</th>\n",
       "      <th>TALONT000013454</th>\n",
       "      <th>TALONT000013557</th>\n",
       "      <th>TALONT000013655</th>\n",
       "      <th>TALONT000013657</th>\n",
       "      <th>TALONT000013998</th>\n",
       "      <th>TALONT000014162</th>\n",
       "      <th>TALONT000014163</th>\n",
       "      <th>TALONT000014329</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>hepg2_1_chr20</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>847.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hepg2_2_chr20</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>800.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>27.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hffc6_1_chr20</th>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hffc6_2_chr20</th>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hffc6_3_chr20</th>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 4555 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "tid            ENST00000023939.8  ENST00000045083.6  ENST00000062104.3  \\\n",
       "hepg2_1_chr20                0.0                0.0                0.0   \n",
       "hepg2_2_chr20                0.0                0.0                0.0   \n",
       "hffc6_1_chr20                2.0                0.0                0.0   \n",
       "hffc6_2_chr20                2.0                0.0                0.0   \n",
       "hffc6_3_chr20                2.0                0.0                0.0   \n",
       "\n",
       "tid            ENST00000064571.2  ENST00000170150.3  ENST00000190983.4  \\\n",
       "hepg2_1_chr20                0.0              847.0                0.0   \n",
       "hepg2_2_chr20                0.0              800.0                0.0   \n",
       "hffc6_1_chr20                0.0                0.0                0.0   \n",
       "hffc6_2_chr20                0.0                0.0                0.0   \n",
       "hffc6_3_chr20                0.0                0.0                0.0   \n",
       "\n",
       "tid            ENST00000191018.9  ENST00000201031.2  ENST00000201961.6  \\\n",
       "hepg2_1_chr20                4.0                0.0                0.0   \n",
       "hepg2_2_chr20                7.0                0.0                0.0   \n",
       "hffc6_1_chr20                5.0                0.0                0.0   \n",
       "hffc6_2_chr20               11.0                3.0                0.0   \n",
       "hffc6_3_chr20               13.0                0.0                0.0   \n",
       "\n",
       "tid            ENST00000201963.3  ...  TALONT000013147  TALONT000013148  \\\n",
       "hepg2_1_chr20                0.0  ...              9.0              6.0   \n",
       "hepg2_2_chr20                0.0  ...             27.0              8.0   \n",
       "hffc6_1_chr20                0.0  ...              5.0              2.0   \n",
       "hffc6_2_chr20                0.0  ...              2.0              2.0   \n",
       "hffc6_3_chr20                0.0  ...              6.0              2.0   \n",
       "\n",
       "tid            TALONT000013454  TALONT000013557  TALONT000013655  \\\n",
       "hepg2_1_chr20             14.0             11.0             11.0   \n",
       "hepg2_2_chr20             27.0             14.0             10.0   \n",
       "hffc6_1_chr20             19.0              8.0              0.0   \n",
       "hffc6_2_chr20             18.0              7.0              0.0   \n",
       "hffc6_3_chr20             15.0              4.0              0.0   \n",
       "\n",
       "tid            TALONT000013657  TALONT000013998  TALONT000014162  \\\n",
       "hepg2_1_chr20              6.0              9.0             18.0   \n",
       "hepg2_2_chr20             11.0             24.0             60.0   \n",
       "hffc6_1_chr20              0.0             23.0              9.0   \n",
       "hffc6_2_chr20              0.0             18.0             19.0   \n",
       "hffc6_3_chr20              0.0             17.0             13.0   \n",
       "\n",
       "tid            TALONT000014163  TALONT000014329  \n",
       "hepg2_1_chr20             17.0              5.0  \n",
       "hepg2_2_chr20             24.0              9.0  \n",
       "hffc6_1_chr20              1.0              0.0  \n",
       "hffc6_2_chr20              7.0              0.0  \n",
       "hffc6_3_chr20              5.0              0.0  \n",
       "\n",
       "[5 rows x 4555 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "postal-broadcast",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENST00000023939.8</th>\n",
       "      <td>ENST00000023939.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000045083.6</th>\n",
       "      <td>ENST00000045083.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000062104.3</th>\n",
       "      <td>ENST00000062104.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000064571.2</th>\n",
       "      <td>ENST00000064571.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000170150.3</th>\n",
       "      <td>ENST00000170150.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 tid\n",
       "tid                                 \n",
       "ENST00000023939.8  ENST00000023939.8\n",
       "ENST00000045083.6  ENST00000045083.6\n",
       "ENST00000062104.3  ENST00000062104.3\n",
       "ENST00000064571.2  ENST00000064571.2\n",
       "ENST00000170150.3  ENST00000170150.3"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var = df.columns.to_frame()\n",
    "var.columns = ['tid']\n",
    "var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "union-mechanics",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENST00000023939.8\n",
      "ENST00000045083.6\n",
      "ENST00000062104.3\n",
      "ENST00000064571.2\n",
      "ENST00000170150.3\n"
     ]
    }
   ],
   "source": [
    "for v in var.tid.tolist()[:5]:\n",
    "    print(v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "characteristic-liver",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>hepg2_1_chr20</th>\n",
       "      <td>hepg2_1_chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hepg2_2_chr20</th>\n",
       "      <td>hepg2_2_chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hffc6_1_chr20</th>\n",
       "      <td>hffc6_1_chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hffc6_2_chr20</th>\n",
       "      <td>hffc6_2_chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hffc6_3_chr20</th>\n",
       "      <td>hffc6_3_chr20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     dataset\n",
       "hepg2_1_chr20  hepg2_1_chr20\n",
       "hepg2_2_chr20  hepg2_2_chr20\n",
       "hffc6_1_chr20  hffc6_1_chr20\n",
       "hffc6_2_chr20  hffc6_2_chr20\n",
       "hffc6_3_chr20  hffc6_3_chr20"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "obs = df.index.to_frame()\n",
    "obs.columns = ['dataset']\n",
    "obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "spatial-prompt",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hepg2_1_chr20\n",
      "hepg2_2_chr20\n",
      "hffc6_1_chr20\n",
      "hffc6_2_chr20\n",
      "hffc6_3_chr20\n"
     ]
    }
   ],
   "source": [
    "for o in obs.dataset.tolist()[:5]:\n",
    "    print(o)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "anticipated-hepatitis",
   "metadata": {},
   "outputs": [],
   "source": [
    "sg.datasets.extend(obs.dataset.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "secondary-london",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['hepg2_1_chr20',\n",
       " 'hepg2_2_chr20',\n",
       " 'hffc6_1_chr20',\n",
       " 'hffc6_2_chr20',\n",
       " 'hffc6_3_chr20',\n",
       " 'k562_1_chr20',\n",
       " 'k562_2_chr20',\n",
       " 'hepg2_1_chr20',\n",
       " 'hepg2_2_chr20',\n",
       " 'hffc6_1_chr20',\n",
       " 'hffc6_2_chr20',\n",
       " 'hffc6_3_chr20',\n",
       " 'k562_1_chr20',\n",
       " 'k562_2_chr20']"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "rapid-scout",
   "metadata": {},
   "outputs": [],
   "source": [
    "X = df.to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "regional-celtic",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = anndata.AnnData(obs=obs, var=var, X=X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "asian-richards",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>hepg2_1_chr20</th>\n",
       "      <td>hepg2_1_chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hepg2_2_chr20</th>\n",
       "      <td>hepg2_2_chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hffc6_1_chr20</th>\n",
       "      <td>hffc6_1_chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hffc6_2_chr20</th>\n",
       "      <td>hffc6_2_chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hffc6_3_chr20</th>\n",
       "      <td>hffc6_3_chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k562_1_chr20</th>\n",
       "      <td>k562_1_chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k562_2_chr20</th>\n",
       "      <td>k562_2_chr20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     dataset\n",
       "hepg2_1_chr20  hepg2_1_chr20\n",
       "hepg2_2_chr20  hepg2_2_chr20\n",
       "hffc6_1_chr20  hffc6_1_chr20\n",
       "hffc6_2_chr20  hffc6_2_chr20\n",
       "hffc6_3_chr20  hffc6_3_chr20\n",
       "k562_1_chr20    k562_1_chr20\n",
       "k562_2_chr20    k562_2_chr20"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "remarkable-brunei",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENST00000023939.8</th>\n",
       "      <td>ENST00000023939.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000045083.6</th>\n",
       "      <td>ENST00000045083.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000062104.3</th>\n",
       "      <td>ENST00000062104.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000064571.2</th>\n",
       "      <td>ENST00000064571.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000170150.3</th>\n",
       "      <td>ENST00000170150.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 tid\n",
       "tid                                 \n",
       "ENST00000023939.8  ENST00000023939.8\n",
       "ENST00000045083.6  ENST00000045083.6\n",
       "ENST00000062104.3  ENST00000062104.3\n",
       "ENST00000064571.2  ENST00000064571.2\n",
       "ENST00000170150.3  ENST00000170150.3"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "appointed-phenomenon",
   "metadata": {},
   "outputs": [],
   "source": [
    "sg.adata = adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "attached-remedy",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[  0.,   0.,   0.,   0., 847.],\n",
       "       [  0.,   0.,   0.,   0., 800.],\n",
       "       [  2.,   0.,   0.,   0.,   0.],\n",
       "       [  2.,   0.,   0.,   0.,   0.],\n",
       "       [  2.,   0.,   0.,   0.,   0.]])"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X[:5, :5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "institutional-necklace",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENST00000023939.8</th>\n",
       "      <td>ENST00000023939.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000045083.6</th>\n",
       "      <td>ENST00000045083.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000062104.3</th>\n",
       "      <td>ENST00000062104.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000064571.2</th>\n",
       "      <td>ENST00000064571.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000170150.3</th>\n",
       "      <td>ENST00000170150.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 tid\n",
       "tid                                 \n",
       "ENST00000023939.8  ENST00000023939.8\n",
       "ENST00000045083.6  ENST00000045083.6\n",
       "ENST00000062104.3  ENST00000062104.3\n",
       "ENST00000064571.2  ENST00000064571.2\n",
       "ENST00000170150.3  ENST00000170150.3"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "billion-state",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>hepg2_1_chr20</th>\n",
       "      <td>hepg2_1_chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hepg2_2_chr20</th>\n",
       "      <td>hepg2_2_chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hffc6_1_chr20</th>\n",
       "      <td>hffc6_1_chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hffc6_2_chr20</th>\n",
       "      <td>hffc6_2_chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hffc6_3_chr20</th>\n",
       "      <td>hffc6_3_chr20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     dataset\n",
       "hepg2_1_chr20  hepg2_1_chr20\n",
       "hepg2_2_chr20  hepg2_2_chr20\n",
       "hffc6_1_chr20  hffc6_1_chr20\n",
       "hffc6_2_chr20  hffc6_2_chr20\n",
       "hffc6_3_chr20  hffc6_3_chr20"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "obs[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "designed-kruger",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>tid</th>\n",
       "      <th>ENST00000023939.8</th>\n",
       "      <th>ENST00000045083.6</th>\n",
       "      <th>ENST00000062104.3</th>\n",
       "      <th>ENST00000064571.2</th>\n",
       "      <th>ENST00000170150.3</th>\n",
       "      <th>ENST00000190983.4</th>\n",
       "      <th>ENST00000191018.9</th>\n",
       "      <th>ENST00000201031.2</th>\n",
       "      <th>ENST00000201961.6</th>\n",
       "      <th>ENST00000201963.3</th>\n",
       "      <th>...</th>\n",
       "      <th>TALONT000013147</th>\n",
       "      <th>TALONT000013148</th>\n",
       "      <th>TALONT000013454</th>\n",
       "      <th>TALONT000013557</th>\n",
       "      <th>TALONT000013655</th>\n",
       "      <th>TALONT000013657</th>\n",
       "      <th>TALONT000013998</th>\n",
       "      <th>TALONT000014162</th>\n",
       "      <th>TALONT000014163</th>\n",
       "      <th>TALONT000014329</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>hepg2_1_chr20</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>847.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hepg2_2_chr20</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>800.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>27.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hffc6_1_chr20</th>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hffc6_2_chr20</th>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hffc6_3_chr20</th>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 4555 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "tid            ENST00000023939.8  ENST00000045083.6  ENST00000062104.3  \\\n",
       "hepg2_1_chr20                0.0                0.0                0.0   \n",
       "hepg2_2_chr20                0.0                0.0                0.0   \n",
       "hffc6_1_chr20                2.0                0.0                0.0   \n",
       "hffc6_2_chr20                2.0                0.0                0.0   \n",
       "hffc6_3_chr20                2.0                0.0                0.0   \n",
       "\n",
       "tid            ENST00000064571.2  ENST00000170150.3  ENST00000190983.4  \\\n",
       "hepg2_1_chr20                0.0              847.0                0.0   \n",
       "hepg2_2_chr20                0.0              800.0                0.0   \n",
       "hffc6_1_chr20                0.0                0.0                0.0   \n",
       "hffc6_2_chr20                0.0                0.0                0.0   \n",
       "hffc6_3_chr20                0.0                0.0                0.0   \n",
       "\n",
       "tid            ENST00000191018.9  ENST00000201031.2  ENST00000201961.6  \\\n",
       "hepg2_1_chr20                4.0                0.0                0.0   \n",
       "hepg2_2_chr20                7.0                0.0                0.0   \n",
       "hffc6_1_chr20                5.0                0.0                0.0   \n",
       "hffc6_2_chr20               11.0                3.0                0.0   \n",
       "hffc6_3_chr20               13.0                0.0                0.0   \n",
       "\n",
       "tid            ENST00000201963.3  ...  TALONT000013147  TALONT000013148  \\\n",
       "hepg2_1_chr20                0.0  ...              9.0              6.0   \n",
       "hepg2_2_chr20                0.0  ...             27.0              8.0   \n",
       "hffc6_1_chr20                0.0  ...              5.0              2.0   \n",
       "hffc6_2_chr20                0.0  ...              2.0              2.0   \n",
       "hffc6_3_chr20                0.0  ...              6.0              2.0   \n",
       "\n",
       "tid            TALONT000013454  TALONT000013557  TALONT000013655  \\\n",
       "hepg2_1_chr20             14.0             11.0             11.0   \n",
       "hepg2_2_chr20             27.0             14.0             10.0   \n",
       "hffc6_1_chr20             19.0              8.0              0.0   \n",
       "hffc6_2_chr20             18.0              7.0              0.0   \n",
       "hffc6_3_chr20             15.0              4.0              0.0   \n",
       "\n",
       "tid            TALONT000013657  TALONT000013998  TALONT000014162  \\\n",
       "hepg2_1_chr20              6.0              9.0             18.0   \n",
       "hepg2_2_chr20             11.0             24.0             60.0   \n",
       "hffc6_1_chr20              0.0             23.0              9.0   \n",
       "hffc6_2_chr20              0.0             18.0             19.0   \n",
       "hffc6_3_chr20              0.0             17.0             13.0   \n",
       "\n",
       "tid            TALONT000014163  TALONT000014329  \n",
       "hepg2_1_chr20             17.0              5.0  \n",
       "hepg2_2_chr20             24.0              9.0  \n",
       "hffc6_1_chr20              1.0              0.0  \n",
       "hffc6_2_chr20              7.0              0.0  \n",
       "hffc6_3_chr20              5.0              0.0  \n",
       "\n",
       "[5 rows x 4555 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "sweet-problem",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[  0.,   0.,   0.,   0., 847.],\n",
       "       [  0.,   0.,   0.,   0., 800.],\n",
       "       [  2.,   0.,   0.,   0.,   0.],\n",
       "       [  2.,   0.,   0.,   0.,   0.],\n",
       "       [  2.,   0.,   0.,   0.,   0.]], dtype=float32)"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.adata.X[:5,:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "assured-collectible",
   "metadata": {},
   "outputs": [],
   "source": [
    "sg.adata.layers['tpm'] = tpm_X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "compatible-kansas",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[    0.        ,     0.        ,     0.        ,     0.        ,\n",
       "        53914.70401018],\n",
       "       [    0.        ,     0.        ,     0.        ,     0.        ,\n",
       "        29204.54130617],\n",
       "       [  130.18290698,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ],\n",
       "       [  107.71799429,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ],\n",
       "       [  138.50415512,     0.        ,     0.        ,     0.        ,\n",
       "            0.        ]])"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.adata.X = sg.adata.layers['tpm']\n",
    "sg.adata.X[:5,:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "hydraulic-church",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
